--- psi.orig/src/chatdlg.cpp
+++ psi/src/chatdlg.cpp
@@ -187,8 +187,9 @@ void ChatDlg::doTrackBar()
 
 void ChatDlg::initActions()
 {
-	act_send_ = new QAction(this);
-	addAction(act_send_);
+	act_send_ = new QAction(chatEdit());
+	act_send_->setShortcutContext(Qt::WidgetShortcut);
+	chatEdit()->addAction(act_send_);
 	connect(act_send_, SIGNAL(triggered()), SLOT(doSend()));
 
 	act_close_ = new QAction(this);
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -832,8 +832,9 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	ui_.toolbar->setSizePolicy(QSizePolicy::Minimum, QSizePolicy::Maximum);
 
 	// Common actions
-	d->act_send = new QAction(this);
-	addAction(d->act_send);
+	d->act_send = new QAction(ui_.mle->chatEdit());
+	d->act_send->setShortcutContext(Qt::WidgetShortcut);
+	ui_.mle->chatEdit()->addAction(d->act_send);
 	connect(d->act_send,SIGNAL(triggered()), SLOT(mle_returnPressed()));
  	ui_.pb_send->setIcon(IconsetFactory::icon("psi/action_button_send").icon());
 	connect(ui_.pb_send, SIGNAL(clicked()), SLOT(mle_returnPressed()));
