--- psi.orig/src/eventdb.cpp
+++ psi/src/eventdb.cpp
@@ -28,6 +28,9 @@
 #include <QTimer>
 #include <QTextStream>
 #include <QDateTime>
+#include <QSqlError>
+#include <QSqlQuery>
+#include <QSqlDriver>
 
 #include "common.h"
 #include "applicationinfo.h"
@@ -72,6 +75,15 @@ const QString & EDBItem::prevId() const
 	return v_prevId;
 }
 
+void EDBItem::setNextId(const QString &id)
+{
+	v_nextId = id;
+}
+
+void EDBItem::setPrevId(const QString &id)
+{
+	v_prevId = id;
+}
 
 //----------------------------------------------------------------------------
 // EDBHandle
@@ -436,6 +448,7 @@ int EDBFlatFile::append(const Jid &j, Ps
 	d->rlist.append(r);
 
 	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+
 	return r->id;
 }
 
@@ -968,3 +981,689 @@ QString EDBFlatFile::File::eventToLine(P
 	
 	return "";
 }
+
+//----------------------------------------------------------------------------
+// EDBSqLite
+//----------------------------------------------------------------------------
+
+EDBSqLite::EDBSqLite() : EDB(),
+	transactionsCounter(0),
+	lastCommitTime(QDateTime::currentDateTime()),
+	commitTimer(NULL),
+	mirror_(NULL)
+{
+	status = NotActive;
+	QString path = ApplicationInfo::historyDir() + "/history.db";
+	QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE", "history");
+	db.setDatabaseName(path);
+	if (!db.open()) {
+		qWarning("EDBSqLite::EDBSqLite(): Can't open base.\n" + db.lastError().text().toAscii());
+		return;
+	}
+	setInsertingMode(Normal);
+	if (db.tables(QSql::Tables).size() == 0) {
+		// no tables found.
+		if (db.transaction()) {
+			QSqlQuery query(db);
+			query.exec("CREATE TABLE `jids` (`id` INTEGER PRIMARY KEY ASC, `type` INTEGER, `jid` TEXT);");
+			query.exec("CREATE TABLE `events` (`id` INTEGER PRIMARY KEY ASC AUTOINCREMENT, `jid_id` INTEGER, `date` TEXT, `type` INTEGER, `direction` INTEGER,  `flags` INTEGER, `subject` TEXT, `m_text` TEXT, `lang` TEXT);");
+			query.exec("CREATE TABLE `urls` (`event_id` INTEGER, `url` TEXT, `desc` TEXT);");
+			query.exec("CREATE INDEX `jid` ON `jids` (`jid`);");
+			query.exec("CREATE INDEX `jid_id` ON `events` (`jid_id`);");
+			query.exec("CREATE INDEX `date` ON `events` (`date`);");
+			query.exec("CREATE INDEX `event_id` ON `urls` (`event_id`);");
+			if (db.commit()) {
+				status = Commited;
+			}
+		}
+	} else {
+		status = Commited;
+	}
+}
+
+EDBSqLite::~EDBSqLite()
+{
+	commit();
+	{
+		QSqlDatabase db = QSqlDatabase::database("history", false);
+		if (db.isOpen())
+			db.close();
+	}
+	QSqlDatabase::removeDatabase("history");
+}
+
+int EDBSqLite::getLatest(const XMPP::Jid & jid, int len)
+{
+	item_query_req *r = new item_query_req;
+	r->j = jid;
+	r->type = item_query_req::Type_getLatest;
+	r->len = len < 1 ? 1 : len;
+	r->id = genUniqueId();
+	rlist.append(r);
+
+	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+	return r->id;
+}
+
+int EDBSqLite::getOldest(const XMPP::Jid &jid, int len)
+{
+	item_query_req *r = new item_query_req;
+	r->j = jid;
+	r->type = item_query_req::Type_getOldest;
+	r->len = len < 1 ? 1 : len;
+	r->id = genUniqueId();
+	rlist.append(r);
+
+	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+	return r->id;
+}
+
+int EDBSqLite::get(const XMPP::Jid &jid, const QString &id, int direction, int len)
+{
+	item_query_req *r = new item_query_req;
+	r->j = jid;
+	r->type = item_query_req::Type_get;
+	r->len = len < 1 ? 1 : len;
+	r->dir = direction;
+	r->eventId = id;
+	r->id = genUniqueId();
+	rlist.append(r);
+
+	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+	return r->id;
+}
+
+int EDBSqLite::getByDate(const XMPP::Jid &jid, QDateTime first, QDateTime last)
+{
+	item_query_req *r = new item_query_req;
+	r->j = jid;
+	r->type = item_query_req::Type_getByDate;
+	r->len = 1;
+	r->id = genUniqueId();
+	r->first = first;
+	r->last = last;
+	rlist.append(r);
+
+	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+	return r->id;
+}
+
+int EDBSqLite::find(const QString &str, const XMPP::Jid &jid, const QString &id, int direction)
+{
+	item_query_req *r = new item_query_req;
+	r->j = jid;
+	r->type = item_query_req::Type_find;
+	r->len = 1;
+	r->dir = direction;
+	r->findStr = str;
+	r->eventId = id;
+	r->id = genUniqueId();
+	rlist.append(r);
+
+	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+	return r->id;
+}
+
+int EDBSqLite::append(const XMPP::Jid &jid, PsiEvent *e)
+{
+	item_query_req *r = new item_query_req;
+	r->j = jid;
+	r->type = item_file_req::Type_append;
+	r->event = e->copy();
+	if ( !r->event ) {
+		qWarning("EDBSqLite::append(): Attempted to append incompatible type.");
+		delete r;
+		return 0;
+	}
+	r->id = genUniqueId();
+	rlist.append(r);
+
+	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+
+	if (mirror_)
+		mirror_->append(jid, e);
+
+	return r->id;
+}
+
+int EDBSqLite::erase(const XMPP::Jid &jid)
+{
+	item_query_req *r = new item_query_req;
+	r->j = jid;
+	r->type = item_query_req::Type_erase;
+	r->id = genUniqueId();
+	rlist.append(r);
+
+	QTimer::singleShot(FAKEDELAY, this, SLOT(performRequests()));
+	return r->id;
+}
+
+void EDBSqLite::setInsertingMode(InsertMode mode)
+{
+	// in the case of a flow of new records
+	if (mode == Import) {
+		// Commit after 10000 inserts and every 5 seconds
+		maxUncommitedRecs = 10000;
+		maxUncommitedSecs = 5;
+	} else {
+		// Commit after 3 inserts and every 1 second
+		maxUncommitedRecs = 3;
+		maxUncommitedSecs = 1;
+	}
+	// Commit if there were no new additions for 1 second
+	commitByTimeoutSecs = 1;
+	//--
+	commit();
+}
+
+void EDBSqLite::setMirror(EDBFlatFile *mirr)
+{
+	if (mirr != mirror_) {
+		if (mirror_)
+			delete mirror_;
+		mirror_ = mirr;
+	}
+}
+
+EDBFlatFile *EDBSqLite::mirror() const
+{
+	return mirror_;
+}
+
+void EDBSqLite::performRequests()
+{
+	if (rlist.isEmpty())
+		return;
+	item_query_req *r = rlist.takeFirst();
+	const int type = r->type;
+	if (type == item_query_req::Type_append) {
+		bool b = appendEvent(r->j, r->event);
+		writeFinished(r->id, b);
+		delete r->event;
+	}
+	else if (type >= item_query_req::Type_getLatest && type <= item_query_req::Type_get) {
+		commit();
+		PreparedQueryes::QueryType queryType;
+		if (type == item_query_req::Type_getLatest) {
+			queryType = PreparedQueryes::QueryLatest;
+		} else if (type == item_query_req::Type_getOldest) {
+			queryType = PreparedQueryes::QueryOldest;
+		} else if (type == item_query_req::Type_get) {
+			if (r->dir == Backward)
+				queryType = PreparedQueryes::QueryIdBackward;
+			else
+				queryType = PreparedQueryes::QueryIdForward;
+		}
+		EDBSqLite::PreparedQuery *query = queryes.getPreparedQuery(queryType);
+		QString sJid = r->j.bare();
+		const int len = r->len + 1;
+		query->bindValue(":jid", sJid);
+		if (type == item_query_req::Type_get)
+			query->bindValue(":id", r->eventId);
+		query->bindValue(":cnt", len);
+		EDBResult result;
+		if (query->exec()) {
+			EDBItem *ei = NULL;
+			int i = 0;
+			const bool fForward = (queryType == PreparedQueryes::QueryOldest
+					       || queryType == PreparedQueryes::QueryIdForward);
+			while (query->next()) {
+				++i;
+				QString id = query->value(0).toString();
+				if (i == len) {
+					if (ei) {
+						if (fForward)
+							ei->setNextId(id);
+						else
+							ei->setPrevId(id);
+					}
+					break;
+
+				}
+				PsiEvent *e = getEvent(query);
+				if (e) {
+					QString prevId, nextId;
+					if (ei) {
+						if (fForward) {
+							ei->setNextId(id);
+							prevId = ei->id();
+						} else {
+							ei->setPrevId(id);
+							nextId = ei->id();
+						}
+					}
+					ei = new EDBItem(e, id, prevId, nextId);
+					result.append(EDBItemPtr(ei));
+				}
+			}
+			query->freeResult();
+			if (type == item_query_req::Type_get && !result.isEmpty()) {
+				query = queryes.getPreparedQuery(fForward ? PreparedQueryes::QueryId2Forward : PreparedQueryes::QueryId2Backward);
+				query->bindValue(":jid", sJid);
+				query->bindValue(":id", r->eventId);
+				if (query->exec() && query->next()) {
+					if (fForward)
+						result.first()->setPrevId(query->value(0).toString());
+					else
+						result.first()->setNextId(query->value(0).toString());
+					query->freeResult();
+				}
+			}
+		}
+		resultReady(r->id, result);
+	} else if(type == item_file_req::Type_getByDate ) {
+		commit();
+		EDBSqLite::PreparedQuery *query = queryes.getPreparedQuery(PreparedQueryes::QueryDate1);
+		QString jidStr = r->j.bare();
+		query->bindValue(":jid", jidStr);
+		query->bindValue(":date1", r->first);
+		query->bindValue(":date2", r->last);
+		EDBResult result;
+		if (query->exec()) {
+			EDBItem *ei = NULL;
+			while (query->next()) {
+				QString id = query->value(0).toString();
+				PsiEvent *e = getEvent(query);
+				if (e) {
+					QString prevId;
+					if (ei) {
+						prevId = ei->id();
+						ei->setNextId(id);
+					}
+					ei = new EDBItem(e, id, prevId, QString());
+					EDBItemPtr eip = EDBItemPtr(ei);
+					result.append(eip);
+				}
+			}
+			query->freeResult();
+			if (!result.isEmpty()) {
+				query = queryes.getPreparedQuery(PreparedQueryes::QueryDate2);
+				query->bindValue(":jid", jidStr);
+				query->bindValue(":date", r->first);
+				if (query->exec()) {
+					if (query->next()) {
+						result.first()->setPrevId(query->value(0).toString());
+					}
+				}
+				query->freeResult();
+				query = queryes.getPreparedQuery(PreparedQueryes::QueryDate3);
+				query->bindValue(":jid", jidStr);
+				query->bindValue(":date", r->last);
+				if (query->exec() && query->next()) {
+					result.last()->setNextId(query->value(0).toString());
+				}
+				query->freeResult();
+			}
+		}
+		resultReady(r->id, result);
+	} else if(type == item_file_req::Type_find) {
+		commit();
+		EDBSqLite::PreparedQuery *query = queryes.getPreparedQuery(PreparedQueryes::QueryFindText);
+		query->bindValue(":jid", r->j.bare());
+		EDBResult result;
+		if (query->exec()) {
+			EDBItem *ei = NULL;
+			QString str = r->findStr.toLower();
+			while (query->next()) {
+				QString id = query->value(0).toString();
+				if (!query->value(7).toString().toLower().contains(str, Qt::CaseSensitive))
+					continue;
+				PsiEvent *e = getEvent(query);
+				if (e) {
+					QString prevId;
+					if (ei) {
+						prevId = ei->id();
+						ei->setNextId(id);
+					}
+					ei = new EDBItem(e, id, prevId, QString());
+					EDBItemPtr eip = EDBItemPtr(ei);
+					result.append(eip);
+				}
+			}
+			query->freeResult();
+		}
+		resultReady(r->id, result);
+
+	} else if(type == item_file_req::Type_erase) {
+		writeFinished(r->id, eraseHistory(r->j));
+	}
+	delete r;
+}
+
+bool EDBSqLite::appendEvent(const XMPP::Jid &jid, PsiEvent *e)
+{
+	QSqlDatabase db = QSqlDatabase::database("history");
+	const qint64 jidId = ensureJidRowId(jid);
+	if (jidId == 0)
+		return false;
+	QDateTime dTime;
+	int nType = 0;
+	int nFlags = 0;
+	if (e->type() == PsiEvent::Message) {
+		MessageEvent *me = (MessageEvent *)e;
+		const Message &m = me->message();
+		dTime = m.timeStamp();
+		if (!m.subject().isEmpty() || !m.body().isEmpty())
+			nFlags |= 1;
+		if (!m.urlList().isEmpty())
+			nFlags |= 2;
+		if (m.type() == "chat")
+			nType = 1;
+		else if(m.type() == "error")
+			nType = 4;
+		else if(m.type() == "headline")
+			nType = 5;
+	} else if (e->type() == PsiEvent::Auth) {
+		AuthEvent *ae = (AuthEvent *)e;
+		dTime = ae->timeStamp();
+		QString subType = ae->authType();
+		if(subType == "subscribe")
+			nType = 3;
+		else if(subType == "subscribed")
+			nType = 6;
+		else if(subType == "unsubscribe")
+			nType = 7;
+		else if(subType == "unsubscribed")
+			nType = 8;
+	} else {
+		return false;
+	}
+	int nDirection = e->originLocal() ? 1 : 2;
+	if (!transaction(false))
+		return false;
+	QSqlQuery query(db);
+	query.prepare("INSERT INTO `events` (`jid_id`, `date`, `type`, `direction`, `flags`, `subject`, `m_text`, `lang`) VALUES (?, ?, ?, ?, ?, ?, ?, ?);");
+	query.bindValue(0, jidId);
+	query.bindValue(1, dTime);
+	query.bindValue(2, nType);
+	query.bindValue(3, nDirection);
+	query.bindValue(4, nFlags);
+	if (nFlags == 0) {
+		query.bindValue(5, QVariant(QVariant::String));
+		query.bindValue(6, QVariant(QVariant::String));
+		query.bindValue(7, QVariant(QVariant::String));
+	} else {
+		MessageEvent *me = (MessageEvent *)e;
+		const Message &m = me->message();
+		if (nFlags & 1) {
+			QString lang = m.lang();
+			query.bindValue(5, m.subject(lang));
+			query.bindValue(6, m.body(lang));
+			query.bindValue(7, lang);
+		}
+		if (!query.exec())
+			return false;
+		if (nFlags & 2) {
+			// Getting last record id
+			qint64 eventId = query.lastInsertId().toLongLong();
+			// insert into urls table
+			const UrlList &urls = m.urlList();
+			foreach (const Url &url, urls) {
+				query.prepare("INSERT INTO `urls` (`event_id`, `url`, `desc`) VALUES (?, ?, ?);");
+				query.addBindValue(eventId);
+				query.addBindValue(url.url());
+				query.addBindValue(url.desc());
+				if (!query.exec())
+					return false;
+			}
+		}
+	}
+	return true;
+}
+
+PsiEvent *EDBSqLite::getEvent(PreparedQuery *q)
+{
+	// `events`.`id`       - 0
+	// `jid`               - 1
+	// `data`              - 2
+	// `type`              - 3
+	// `direction`         - 4
+	// `flags`             - 5
+	// `subject`           - 6
+	// `m_text`            - 7
+	// `lang`              - 8
+	int type = q->value(3).toInt();
+	if(type == 0 || type == 1 || type == 4 || type == 5) {
+		Message m;
+		m.setTimeStamp(q->value(2).toDateTime());
+		if(type == 1)
+			m.setType("chat");
+		else if(type == 4)
+			m.setType("error");
+		else if(type == 5)
+			m.setType("headline");
+		else
+			m.setType("");
+		m.setFrom(Jid(q->value(1).toString()));
+		QVariant text = q->value(7);
+		if (!text.isNull()) {
+			m.setBody(text.toString());
+			m.setLang(q->value(8).toString());
+			m.setSubject(q->value(6).toString());
+		}
+		m.setSpooled(true);
+		if (q->value(5).toInt() & 2) {
+			EDBSqLite::PreparedQuery *urlQuery = queryes.getPreparedQuery(PreparedQueryes::QueryUrls);
+			urlQuery->bindValue(":id", q->value(0));
+			if (urlQuery->exec()) {
+				while (urlQuery->next()) {
+					m.urlAdd(Url(urlQuery->value(0).toString(), urlQuery->value(1).toString()));
+				}
+				urlQuery->freeResult();
+			}
+		}
+		MessageEvent *me = new MessageEvent(m, 0);
+		me->setOriginLocal((q->value(4).toInt() == 1));
+		return me;
+	}
+	if(type == 2 || type == 3 || type == 6 || type == 7 || type == 8) {
+		QString subType = "subscribe";
+		// if(type == 2) { // Not used (stupid "system message" from Psi <= 0.8.6)
+		if(type == 3)
+			subType = "subscribe";
+		else if(type == 6)
+			subType = "subscribed";
+		else if(type == 7)
+			subType = "unsubscribe";
+		else if(type == 8)
+			subType = "unsubscribed";
+
+		AuthEvent *ae = new AuthEvent(Jid(q->value(1).toString()), subType, 0);
+		ae->setTimeStamp(q->value(2).toDateTime());
+		return ae;
+	}
+	return NULL;
+}
+
+qint64 EDBSqLite::ensureJidRowId(const XMPP::Jid &jid)
+{
+	QString sJid = jid.bare();
+	if (sJid.isEmpty())
+		return 0;
+	qint64 id = jidsHash.value(sJid, 0);
+	if (id != 0)
+		return id;
+
+	EDBSqLite::PreparedQuery *query = queryes.getPreparedQuery(PreparedQueryes::QueryJidRowId);
+	query->bindValue(":jid", sJid);
+	if (query->exec()) {
+		if (query->first()) {
+			id = query->value(0).toLongLong();
+		} else {
+			QSqlQuery queryIns(QSqlDatabase::database("history"));
+			queryIns.prepare("INSERT INTO `jids` (`type`, `jid`) VALUES (:type, :jid);");
+			queryIns.bindValue(":type", 1);  // contact jid
+			queryIns.bindValue(":jid", sJid);
+			if (queryIns.exec()) {
+				id  = queryIns.lastInsertId().toLongLong();
+			}
+		}
+		query->freeResult();
+	}
+	return id;
+}
+
+bool EDBSqLite::eraseHistory(const XMPP::Jid &jid)
+{
+	bool res = false;
+	if (!transaction(true))
+		return false;
+	QSqlQuery query(QSqlDatabase::database("history"));
+	query.prepare("SELECT `id` FROM `jids` WHERE `jid` = :jid;");
+	query.bindValue(":jid", jid.bare());
+	if (query.exec() && query.first()) {
+		const qint64 id = query.value(0).toLongLong();
+		QString queryStr = QString("DELETE FROM `urls` WHERE `event_id` IN (SELECT `id` FROM `events` WHERE `jid_id` = :id);");
+		query.prepare(queryStr);
+		query.bindValue(":id", id);
+		if (query.exec()) {
+			query.prepare("DELETE FROM `events` WHERE `jid_id` = :id;");
+			query.bindValue(":id", id);
+			if (query.exec()) {
+				res = true;
+			}
+		}
+	}
+	if (res)
+		res = commit();
+	else
+		rollback();
+	return res;
+}
+
+bool EDBSqLite::transaction(bool now)
+{
+	if (status == NotActive)
+		return false;
+	if (now || transactionsCounter >= maxUncommitedRecs || lastCommitTime.secsTo(QDateTime::currentDateTime()) >= maxUncommitedSecs) {
+		if (!commit())
+			return false;
+	}
+
+	if (status == Commited) {
+		if (!QSqlDatabase::database("history").transaction())
+			return false;
+		status = NotCommited;
+	}
+	++transactionsCounter;
+
+	startAutocommitTimer();
+
+	return true;
+}
+
+bool EDBSqLite::commit()
+{
+	if (status != NotActive) {
+		if (status == Commited || QSqlDatabase::database("history").commit()) {
+			transactionsCounter = 0;
+			lastCommitTime = QDateTime::currentDateTime();
+			status = Commited;
+			stopAutocommitTimer();
+			return true;
+		}
+	}
+	return false;
+}
+
+bool EDBSqLite::rollback()
+{
+	if (status == NotCommited && QSqlDatabase::database("history").rollback()) {
+		transactionsCounter = 0;
+		lastCommitTime = QDateTime::currentDateTime();
+		status = Commited;
+		stopAutocommitTimer();
+		return true;
+	}
+	return false;
+}
+
+void EDBSqLite::startAutocommitTimer()
+{
+	if (!commitTimer) {
+		commitTimer = new QTimer(this);
+		connect(commitTimer, SIGNAL(timeout()), this, SLOT(commit()));
+		commitTimer->setSingleShot(true);
+		commitTimer->setInterval(commitByTimeoutSecs * 1000);
+	}
+	commitTimer->start();
+}
+
+void EDBSqLite::stopAutocommitTimer()
+{
+	if (commitTimer && commitTimer->isActive())
+		commitTimer->stop();
+}
+
+// ****************** class PreparedQueryes ********************
+
+EDBSqLite::PreparedQueryes::PreparedQueryes()
+{
+}
+
+EDBSqLite::PreparedQueryes::~PreparedQueryes()
+{
+	foreach (EDBSqLite::PreparedQuery *q, queryList.values()) {
+		if (q)
+			delete q;
+	}
+}
+
+EDBSqLite::PreparedQuery *EDBSqLite::PreparedQueryes::getPreparedQuery(QueryType type)
+{
+	EDBSqLite::PreparedQuery *q = queryList.value(type, NULL);
+	if (q != NULL)
+		return q;
+
+	q = new EDBSqLite::PreparedQuery(QSqlDatabase::database("history"));
+	q->setForwardOnly(true);
+	QString queryStr;
+	if (type == QueryDate1) {
+		queryStr = "SELECT `events`.`id`, `jid`, `date`, `events`.`type`, `direction`, `flags`, `subject`, `m_text`, `lang`"
+			   " FROM `events`, `jids`"
+			   " WHERE `jid` = :jid  AND `jids`.`id` = `jid_id` AND `date` >= :date1 AND `date` <= :date2 ORDER BY `date` ASC;";
+	} else if (type == QueryDate2) {
+		queryStr = "SELECT `events`.`id` FROM `events`, `jids`"
+			   " WHERE `jid` = :jid AND `jids`.`id` = `jid_id` AND `date` < :date ORDER BY `date` DESC LIMIT 1;";
+	} else if (type == QueryDate3) {
+		queryStr = "SELECT `events`.`id` FROM `events`, `jids`"
+			   " WHERE `jid` = :jid AND `jids`.`id` = `jid_id`  AND `date` > :date ORDER BY `date` ASC LIMIT 1;";
+	} else if (type == QueryFindText) {
+		queryStr = "SELECT `events`.`id`, `jid`, `date`, `events`.`type`, `direction`, `flags`, `subject`, `m_text`, `lang`"
+			   " FROM `events`, `jids`"
+			   " WHERE `jid` = :jid  AND `jids`.`id` = `jid_id` AND `m_text` IS NOT NULL ORDER BY `date`;";
+	} else if (type == QueryLatest || type == QueryOldest || type == QueryIdBackward || type == QueryIdForward) {
+		queryStr = "SELECT `events`.`id`, `jid`, `date`, `events`.`type`, `direction`, `flags`, `subject`, `m_text`, `lang`"
+			   " FROM `events`, `jids`"
+			   " WHERE `jid` = :jid AND `jids`.`id` = `jid_id`";
+		if (type == QueryIdBackward) {
+			queryStr.append(" AND `events`.`id` <= :id");
+		} else if (type == QueryIdForward) {
+			queryStr.append(" AND `events`.`id` >= :id");
+		}
+		if (type == QueryLatest || type == QueryIdBackward) {
+			queryStr.append(" ORDER BY `events`.`id` DESC");
+		} else {
+			queryStr.append(" ORDER BY `events`.`id` ASC");
+		}
+		queryStr.append(" LIMIT :cnt;");
+	} else if (type == QueryId2Backward || type == QueryId2Forward) {
+		queryStr = "SELECT `events`.`id` FROM `events`, `jids`"
+			   " WHERE `jid` = :jid AND `jids`.`id` = `jid_id`";
+		queryStr.append(QString(" AND `events`.`id` %1 :id").arg((type == QueryId2Forward) ? "<" : ">"));
+		queryStr.append(QString(" ORDER BY `events`.`id` %1").arg((type == QueryId2Forward) ? "DESC" : "ASC"));
+		queryStr.append(" LIMIT 1;");
+	} else if (type == QueryUrls) {
+		queryStr = "SELECT `url`, `desc` FROM `urls` WHERE `event_id` = :id;";
+	} else if (type == QueryJidRowId) {
+		queryStr = "SELECT `id` FROM `jids` WHERE `jid` = :jid;";
+	}
+	q->prepare(queryStr);
+
+	queryList[type] = q;
+	return q;
+}
+
+EDBSqLite::PreparedQuery::PreparedQuery(QSqlDatabase db) : QSqlQuery(db)
+{
+}
--- psi.orig/src/eventdb.h
+++ psi/src/eventdb.h
@@ -26,6 +26,10 @@
 #include <QFile>
 #include <QSharedPointer>
 #include <QDateTime>
+#include <QSqlDatabase>
+#include <QSqlQuery>
+#include <QHash>
+#include <QVariant>
 
 #include "xmpp_jid.h"
 
@@ -34,13 +38,15 @@ class PsiEvent;
 class EDBItem
 {
 public:
-	EDBItem(PsiEvent *, const QString &id, const QString &nextId, const QString &prevId);
+	EDBItem(PsiEvent *, const QString &id, const QString &prevId, const QString &nextId);
 	~EDBItem();
 
 	PsiEvent *event() const;
 	const QString & id() const;
 	const QString & nextId() const;
 	const QString & prevId() const;
+	void setNextId(const QString &id);
+	void setPrevId(const QString &id);
 
 private:
 	QString v_id, v_prevId, v_nextId;
@@ -189,4 +195,114 @@ private:
 	void ensureIndex();
 };
 
+class EDBSqLite : public EDB
+{
+	Q_OBJECT
+
+	class PreparedQueryes;
+	//--------
+	class PreparedQuery : private QSqlQuery
+	{
+	public:
+		void bindValue(const QString &placeholder, const QVariant &val) {QSqlQuery::bindValue(placeholder, val);};
+		bool exec() {return QSqlQuery::exec();};
+		bool first() {return QSqlQuery::first();};
+		bool next() {return QSqlQuery::next();};
+		QVariant value(int i) const {return QSqlQuery::value(i);};
+		void freeResult() {QSqlQuery::finish();};
+	private:
+		friend class PreparedQueryes;
+		PreparedQuery(QSqlDatabase db);
+		~PreparedQuery() {};
+
+	};
+	//--------
+	class PreparedQueryes
+	{
+	public:
+		enum QueryType {
+			QueryLatest, QueryOldest, QueryIdForward, QueryIdBackward,
+			QueryId2Forward, QueryId2Backward,
+			QueryDate1, QueryDate2, QueryDate3,
+			QueryFindText,
+			QueryUrls,
+			QueryJidRowId
+		};
+		PreparedQueryes();
+		~PreparedQueryes();
+		PreparedQuery *getPreparedQuery(QueryType);
+	private:
+		QHash<QueryType, PreparedQuery *> queryList;
+	};
+	//--------
+
+public:
+	enum InsertMode {Normal, Import};
+
+	EDBSqLite();
+	~EDBSqLite();
+
+	int getLatest(const XMPP::Jid &, int len);
+	int getOldest(const XMPP::Jid &, int len);
+	int get(const XMPP::Jid &jid, const QString &id, int direction, int len);
+	int getByDate(const XMPP::Jid &jid, QDateTime first, QDateTime last);
+	int find(const QString &str, const XMPP::Jid &jid, const QString &id, int direction);
+	int append(const XMPP::Jid &jid, PsiEvent *e);
+	int erase(const XMPP::Jid &jid);
+	void setInsertingMode(InsertMode mode);
+	void setMirror(EDBFlatFile *mirr);
+	EDBFlatFile *mirror() const;
+
+private:
+	enum {NotActive, NotCommited, Commited};
+	struct item_query_req
+	{
+		XMPP::Jid j;
+		int type; // 0 = latest, 1 = oldest, 2 = random, 3 = write
+		int len;
+		int dir;
+		int id;
+		QString eventId;
+		QString findStr;
+		PsiEvent *event;
+
+		QDateTime first, last;
+		enum Type {
+			Type_getLatest = 0,
+			Type_getOldest,
+			Type_get,
+			Type_append,
+			Type_find,
+			Type_getByDate,
+			Type_erase
+		};
+	};
+	int  status;
+	unsigned int transactionsCounter;
+	QDateTime lastCommitTime;
+	unsigned int maxUncommitedRecs;
+	int maxUncommitedSecs;
+	unsigned int commitByTimeoutSecs;
+	QTimer *commitTimer;
+	EDBFlatFile *mirror_;
+	QList<item_query_req*> rlist;
+	QHash<QString, qint64>jidsHash;
+	PreparedQueryes queryes;
+
+private:
+	bool appendEvent(const XMPP::Jid &, PsiEvent *);
+	PsiEvent *getEvent(PreparedQuery *q);
+	qint64 ensureJidRowId(const XMPP::Jid &);
+	bool eraseHistory(const XMPP::Jid &);
+	bool transaction(bool now);
+	bool rollback();
+	void startAutocommitTimer();
+	void stopAutocommitTimer();
+
+private slots:
+	void performRequests();
+	bool commit();
+
+};
+
 #endif
--- psi.orig/src/historydlg.cpp
+++ psi/src/historydlg.cpp
@@ -36,6 +36,7 @@
 #include "fileutil.h"
 #include "userlist.h"
 #include "common.h"
+#include "historyimp.h"
 
 static const QString geometryOption = "options.ui.history.size";
 
@@ -400,6 +401,8 @@ void HistoryDlg::doMenu()
 	m->addAction(IconsetFactory::icon("psi/chat").icon(), tr("&Open chat"), this, SLOT(openChat()));
 	m->addAction(IconsetFactory::icon("psi/save").icon(), tr("&Export history"), this, SLOT(exportHistory()));
 	m->addAction(IconsetFactory::icon("psi/clearChat").icon(), tr("&Delete history"), this, SLOT(removeHistory()));
+	m->addSeparator();
+	m->addAction(tr("&Import all history"), this, SLOT(importHistory()));
 	m->exec(QCursor::pos());
 }
 
@@ -626,3 +629,25 @@ EDBHandle* HistoryDlg::getEDBHandle()
 	connect(h, SIGNAL(finished()), SLOT(edb_finished()));
 	return h;
 }
+
+void HistoryDlg::importHistory()
+{
+	startRequest();
+	HistoryImp *imp = new HistoryImp(d->psi);
+	connect(imp, SIGNAL(finished(int)), this, SLOT(importFinished(int)));
+	imp->start();
+}
+
+void HistoryDlg::importFinished(int result)
+{
+	stopRequest();
+	HistoryImp *imp = (HistoryImp *)sender();
+	if (result == HistoryImp::Normal) {
+		int sec = imp->importDuration();
+		int min = sec / 60;
+		sec = sec % 60;
+		QMessageBox::information(this, tr("Import"), tr("Imports is finished.\nDuration is %1 min. %2 sec.").arg(min).arg(sec));
+	} else if (result == HistoryImp::Error)
+		QMessageBox::critical(this, tr("Import"), tr("Import error."));
+	imp->deleteLater();
+}
--- psi.orig/src/historydlg.h
+++ psi/src/historydlg.h
@@ -69,6 +69,8 @@ private slots:
 	void openChat();
 	void doMenu();
 	void removedContact(PsiContact*);
+	void importHistory();
+	void importFinished(int);
 
 protected:
 	bool eventFilter(QObject *, QEvent *);
--- psi.orig/src/historyimp.cpp
+++ psi/src/historyimp.cpp
@@ -0,0 +1,163 @@
+/*
+ * historyimp.cpp
+ * Copyright (C) 20011   Aleksey Andreev
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#include <QDir>
+#include <QTimer>
+#include <QMessageBox>
+
+#include "historyimp.h"
+#include "applicationinfo.h"
+
+HistoryImp::HistoryImp(PsiCon *psi) : QObject(),
+	psi_(psi),
+	prompt(true),
+	hRead(NULL),
+	hWrite(NULL),
+	active(false)
+{
+}
+
+HistoryImp::~HistoryImp()
+{
+	clear();
+}
+
+void HistoryImp::clear()
+{
+	if (desEdb) {
+		((EDBSqLite *)desEdb)->setInsertingMode(EDBSqLite::Normal);
+		((EDBSqLite *)desEdb)->setMirror((EDBFlatFile *)souEdb);
+	}
+	if (hRead) {
+		delete hRead;
+		hRead = NULL;
+	}
+	if (hWrite) {
+		delete hWrite;
+		hWrite = NULL;
+	}
+}
+
+void HistoryImp::start()
+{
+	active = true;
+	startTime = QDateTime::currentDateTime();
+	QString impDirStr = ApplicationInfo::historyDir() + "/";
+	QDir impDir(impDirStr);
+	QFileInfoList fl = impDir.entryInfoList(QStringList("*.history"), QDir::Files);
+	impSize = 0;
+	foreach (const QFileInfo &fi, fl) {
+		XMPP::Jid jid(JIDUtil::decode(fi.completeBaseName()));
+		if (jid.isValid()) {
+			importList.append(QPair<XMPP::Jid, QString>(jid, QString()));
+			impSize += fi.size();
+		}
+	}
+	if (importList.isEmpty()) {
+		stop(Normal);
+		return;
+	}
+	desEdb = psi_->edb();
+	((EDBSqLite *)desEdb)->setMirror(NULL);
+	((EDBSqLite *)desEdb)->setInsertingMode(EDBSqLite::Import);
+
+	souEdb = new EDBFlatFile;
+
+	QTimer::singleShot(0, this, SLOT(readFromFiles()));
+}
+
+void HistoryImp::stop(int reason)
+{
+	active = false;
+	stopTime = QDateTime::currentDateTime();
+	clear();
+	emit finished(reason);
+}
+
+int HistoryImp::importDuration()
+{
+	int duration = 0;
+	if (startTime.isValid() && stopTime.isValid()) {
+		duration = startTime.secsTo(stopTime);
+		if (duration < 0)
+			duration = 0;
+	}
+	return duration;
+}
+
+void HistoryImp::readFromFiles()
+{
+	if (!active)
+		return;
+	if (prompt) {
+		prompt = false;
+		QString text = tr("Found %1 files. Common size is %2 bytes.\nContinue?").arg(importList.size()).arg(impSize);
+		if (QMessageBox::question(NULL, tr("Import history"), text, QMessageBox::Yes, QMessageBox::No) != QMessageBox::Yes) {
+			stop(Cancel);
+			return;
+		}
+	}
+	if (importList.isEmpty()) {
+		stop(Normal);
+		return;
+	}
+	if (hWrite != NULL && !hWrite->writeSuccess()) {
+		stop(Error); // Write error
+		return;
+	}
+	if (hRead == NULL) {
+		hRead = new EDBHandle(souEdb);
+		connect(hRead, SIGNAL(finished()), this, SLOT(writeToSqlite()));
+	}
+	QPair<XMPP::Jid, QString> &rec = importList.first();
+	QString nextId = rec.second;
+	if (nextId.isEmpty()) {
+		qWarning(QString("Importing %1").arg(JIDUtil::toString(rec.first, true)).toAscii());
+		hRead->getOldest(rec.first, 1);
+	} else {
+		hRead->get(rec.first, nextId, EDB::Forward, 1);
+	}
+}
+
+void HistoryImp::writeToSqlite()
+{
+	if (!active)
+		return;
+	const EDBResult r = hRead->result();
+	if (hRead->lastRequestType() != EDBHandle::Read || r.size() > 1) {
+		stop(Error);
+		return;
+	}
+	if (r.isEmpty()) {
+		importList.removeFirst();
+		QTimer::singleShot(0, this, SLOT(readFromFiles()));
+		return;
+	}
+	if (hWrite == NULL) {
+		hWrite = new EDBHandle(desEdb);
+		connect(hWrite, SIGNAL(finished()), this, SLOT(readFromFiles()));
+	}
+	EDBItemPtr it = r.first();
+	hWrite->append(importList.first().first, it->event());
+	if (!it->nextId().isEmpty())
+		importList.first().second = it->nextId();
+	else
+		importList.removeFirst();
+}
--- psi.orig/src/historyimp.h
+++ psi/src/historyimp.h
@@ -0,0 +1,69 @@
+/*
+ * historyimp.h
+ * Copyright (C) 20011   Aleksey Andreev
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ */
+
+#ifndef HISTORYIMP_H
+#define HISTORYIMP_H
+
+#include <QObject>
+#include <QPair>
+
+#include "xmpp/jid/jid.h"
+#include "jidutil.h"
+#include "psicon.h"
+#include "eventdb.h"
+
+class HistoryImp : public QObject
+{
+	Q_OBJECT
+
+public:
+	enum {Normal, Cancel, Error};
+	HistoryImp(PsiCon *psi);
+	~HistoryImp();
+	void start();
+	void stop(int reason = Cancel);
+	int importDuration();
+
+private:
+	PsiCon *psi_;
+	bool prompt;
+	qint64 impSize;
+	QList< QPair<XMPP::Jid, QString> > importList;
+	EDB *souEdb;
+	EDB *desEdb;
+	EDBHandle *hRead;
+	EDBHandle *hWrite;
+	QDateTime startTime;
+	QDateTime stopTime;
+	bool active;
+
+private:
+	void clear();
+
+private slots:
+	void readFromFiles();
+	void writeToSqlite();
+
+signals:
+	void finished(int);
+
+};
+
+#endif
--- psi.orig/src/psicon.cpp
+++ psi/src/psicon.cpp
@@ -324,7 +324,10 @@ PsiCon::PsiCon()
 	d->ftwin = 0;
 #endif
 
-	d->edb = new EDBFlatFile;
+	//d->edb = new EDBFlatFile;
+	EDBSqLite *esl = new EDBSqLite;
+	esl->setMirror(new EDBFlatFile);
+	d->edb = esl;
 
 	d->s5bServer = 0;
 	d->tuneManager = 0;
--- psi.orig/src/src.pri
+++ psi/src/src.pri
@@ -1,4 +1,4 @@
-QT += xml network
+QT += xml network sql
 
 # modules
 include($$PWD/protocol/protocol.pri)
@@ -155,6 +155,7 @@ HEADERS += \
 	$$PWD/translationmanager.h \
 	$$PWD/eventdb.h \
 	$$PWD/historydlg.h \
+	$$PWD/historyimp.h \
 	$$PWD/tipdlg.h \
 	$$PWD/searchdlg.h \
 	$$PWD/registrationdlg.h \
@@ -303,6 +304,7 @@ SOURCES += \
 	$$PWD/translationmanager.cpp \
 	$$PWD/eventdb.cpp \
 	$$PWD/historydlg.cpp \
+	$$PWD/historyimp.cpp \
 	$$PWD/searchdlg.cpp \
 	$$PWD/registrationdlg.cpp \
 	$$PWD/psitoolbar.cpp \
