Index: trunk/src/alerticon.cpp
===================================================================
--- trunk.orig/src/alerticon.cpp	2008-09-05 21:01:41.000000000 +0300
+++ trunk/src/alerticon.cpp	2008-09-05 21:03:48.000000000 +0300
@@ -62,7 +62,7 @@
 {
 	animTimer = new QTimer(this);
 	connect(animTimer, SIGNAL(timeout()), SLOT(animTimeout()));
-	animTimer->start(120 * 5);
+	// animTimer->start(120 * 5);
 	frame = 0;
 
 	// blank icon
Index: trunk/src/contactview.cpp
===================================================================
--- trunk.orig/src/contactview.cpp	2008-09-05 21:11:02.000000000 +0300
+++ trunk/src/contactview.cpp	2008-09-05 21:11:31.000000000 +0300
@@ -2016,7 +2016,7 @@
 
 	// animation timer
 	d->animTimer = new QTimer(this);
-	d->animTimer->start(120 * 5);
+	// d->animTimer->start(120 * 5);
 
 	d->recalculateSizeTimer = new QTimer(this);
 	connect(d->recalculateSizeTimer, SIGNAL(timeout()), d, SLOT(recalculateSize()));
Index: trunk/src/tools/iconset/anim.cpp
===================================================================
--- trunk.orig/src/tools/iconset/anim.cpp	2008-09-05 21:05:11.000000000 +0300
+++ trunk/src/tools/iconset/anim.cpp	2008-09-05 21:05:26.000000000 +0300
@@ -184,6 +184,7 @@
 
 	void restartTimer()
 	{
+		return;
 		if ( !paused && speed > 0 ) {
 			int frameperiod = frames[frame].period;
 			int i = frameperiod >= 0 ? frameperiod * 100/speed : 0;
Index: trunk/src/libpsi/tools/idle/idle.cpp
===================================================================
--- trunk.orig/src/libpsi/tools/idle/idle.cpp	2008-09-05 21:09:16.000000000 +0300
+++ trunk/src/libpsi/tools/idle/idle.cpp	2008-09-05 21:10:04.000000000 +0300
@@ -93,8 +93,8 @@
 		d->idleSince = QDateTime::currentDateTime();
 	}
 
-	// poll every second (use a lower value if you need more accuracy)
-	d->checkTimer.start(1000);
+	// don't poll every second (use a lower value if you need more accuracy)
+	d->checkTimer.start(300000);
 }
 
 void Idle::stop()
Index: trunk/src/tools/tunecontroller/pollingtunecontroller.cpp
===================================================================
--- trunk.orig/src/tools/tunecontroller/pollingtunecontroller.cpp	2008-09-05 21:02:36.000000000 +0300
+++ trunk/src/tools/tunecontroller/pollingtunecontroller.cpp	2008-09-05 21:05:13.000000000 +0300
@@ -38,7 +38,7 @@
 PollingTuneController::PollingTuneController()
 {
 	connect(&timer_, SIGNAL(timeout()), SLOT(check()));
-	timer_.start(3000);
+	// timer_.start(3000);
 }
 
 
@@ -48,6 +48,7 @@
  */
 void PollingTuneController::setInterval(int interval)
 {
+	return;
 	if (interval > 0)
 		timer_.start(interval);
 	else
