diff --git a/src/networkaccessmanager.cpp b/src/networkaccessmanager.cpp
index 0ef88554..60f061c1 100644
--- a/src/networkaccessmanager.cpp
+++ b/src/networkaccessmanager.cpp
@@ -31,10 +31,12 @@ NetworkAccessManager::NetworkAccessManager(QObject *parent)
 }
 
 QNetworkReply *NetworkAccessManager::createRequest(Operation op, const QNetworkRequest & req,
-                                                   QIODevice * outgoingData = 0)
+                                                   QIODevice * outgoingData = nullptr)
 {
     if (req.url().host() != QLatin1String("psi")) {
-        return QNetworkAccessManager::createRequest(op, req, outgoingData);
+        QNetworkRequest request(req);
+        request.setRawHeader("User-Agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/72.0.3626.53 Safari/537.36");
+        return QNetworkAccessManager::createRequest(op, request, outgoingData);
     }
 
     QNetworkReply *reply = nullptr;
