diff --git a/psi-plus-snapshots/src/mainwin.cpp b/psi-plus-snapshots/src/mainwin.cpp
index ecd20cc..5c0ce47 100644
--- a/psi-plus-snapshots/src/mainwin.cpp
+++ b/psi-plus-snapshots/src/mainwin.cpp
@@ -1616,7 +1616,7 @@ void MainWin::trayClicked(const QPoint &, int button)
 		return;
 	}
 
-	if(button == Qt::MidButton) {
+	if(button == Qt::MidButton || d->nextAmount > 0) {
 		doRecvNextEvent();
 		return;
 	}
