diff --git a/src/accountmodify.ui b/src/accountmodify.ui
index d6350eb..3fdfd5d 100644
--- a/src/accountmodify.ui
+++ b/src/accountmodify.ui
@@ -900,10 +900,13 @@
         </widget>
        </item>
        <item>
-        <widget class="QGroupBox" name="groupBox_2">
+        <widget class="QGroupBox" name="gb_stun">
          <property name="title">
           <string>STUN/TURN</string>
          </property>
+         <property name="checkable">
+          <bool>true</bool>
+         </property>
          <layout class="QVBoxLayout" name="verticalLayout">
           <item>
            <layout class="QHBoxLayout">
@@ -915,7 +918,32 @@
              </widget>
             </item>
             <item>
-             <widget class="QLineEdit" name="le_stunHost"/>
+             <widget class="QComboBox" name="cb_stunHost">
+              <property name="sizePolicy">
+               <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
+                <horstretch>0</horstretch>
+                <verstretch>0</verstretch>
+               </sizepolicy>
+              </property>
+              <property name="editable">
+               <bool>true</bool>
+              </property>
+              <item>
+               <property name="text">
+                <string>stun.jabber.ru:5249</string>
+               </property>
+              </item>
+              <item>
+               <property name="text">
+                <string>stun.habahaba.im:3478</string>
+               </property>
+              </item>
+              <item>
+               <property name="text">
+                <string>stun.ekiga.net:3478</string>
+               </property>
+              </item>
+             </widget>
             </item>
             <item>
              <widget class="QLabel" name="lb_stunPort">
@@ -1118,7 +1146,6 @@
   <tabstop>ck_req_mutual</tabstop>
   <tabstop>cb_security_level</tabstop>
   <tabstop>le_dtProxy</tabstop>
-  <tabstop>le_stunHost</tabstop>
   <tabstop>le_stunPort</tabstop>
   <tabstop>le_stunUser</tabstop>
   <tabstop>le_stunPass</tabstop>
@@ -1140,10 +1167,29 @@
      <y>439</y>
     </hint>
     <hint type="destinationlabel">
-     <x>523</x>
+     <x>499</x>
      <y>386</y>
     </hint>
    </hints>
   </connection>
+  <connection>
+   <sender>cb_stunHost</sender>
+   <signal>activated(QString)</signal>
+   <receiver>AccountModify</receiver>
+   <slot>setStunServer(QString)</slot>
+   <hints>
+    <hint type="sourcelabel">
+     <x>322</x>
+     <y>291</y>
+    </hint>
+    <hint type="destinationlabel">
+     <x>499</x>
+     <y>264</y>
+    </hint>
+   </hints>
+  </connection>
  </connections>
+ <slots>
+  <slot>setStunServer(QString)</slot>
+ </slots>
 </ui>
diff --git a/src/accountmodifydlg.cpp b/src/accountmodifydlg.cpp
index 7b76e6f..2dbd55b 100644
--- a/src/accountmodifydlg.cpp
+++ b/src/accountmodifydlg.cpp
@@ -152,7 +152,8 @@ void AccountModifyDlg::init()
 	ck_keepAlive->setChecked(acc.opt_keepAlive);
 	ck_ibbOnly->setChecked(acc.ibbOnly);
 	le_dtProxy->setText(acc.dtProxy.full());
-	le_stunHost->setText(acc.stunHost);
+	gb_stun->setChecked(acc.useStun);
+	cb_stunHost->setEditText(acc.stunHost);
 	le_stunPort->setText(QString::number(acc.stunPort));
 	le_stunUser->setText(acc.stunUser);
 	le_stunPass->setText(acc.stunPass);
@@ -472,6 +473,18 @@ void AccountModifyDlg::detailsChangePW()
 	}
 }
 
+void AccountModifyDlg::setStunServer(const QString &server)
+{
+	QString host = server.section(':', 0, 0);
+	QString port = server.section(':', 1, 1);
+	if (port.isEmpty()) {
+		port = "3478";
+	}
+
+	cb_stunHost->setEditText(host);
+	le_stunPort->setText(port);
+}
+
 void AccountModifyDlg::save()
 {
 	/*if(pa && le_name->text().isEmpty()) {
@@ -539,7 +552,8 @@ void AccountModifyDlg::save()
 	acc.opt_keepAlive = ck_keepAlive->isChecked();
 	acc.ibbOnly = ck_ibbOnly->isChecked();
 	acc.dtProxy = le_dtProxy->text();
-	acc.stunHost = le_stunHost->text();
+	acc.useStun = gb_stun->isChecked();
+	acc.stunHost = cb_stunHost->currentText();
 	acc.stunPort = le_stunPort->text().toInt();
 	acc.stunUser = le_stunUser->text();
 	acc.stunPass = le_stunPass->text();
diff --git a/src/accountmodifydlg.h b/src/accountmodifydlg.h
index f838c9c..f0e85dc 100644
--- a/src/accountmodifydlg.h
+++ b/src/accountmodifydlg.h
@@ -56,22 +56,24 @@ private slots:
 	void detailsVCard();
 	void detailsChangePW();
 
+	void setStunServer(const QString &server);
+
 	void save();
 
 	//void pgpToggled(bool);
 	void chooseKey();
 	void clearKey();
- 	void tabChanged(int);
+	void tabChanged(int);
 
 	// Privacy
- 	void privacyClicked();
- 	void updatePrivacyTab();
- 	void setPrivacyTabEnabled(bool b);
- 	void addBlockClicked();
- 	void removeBlockClicked();
- 	void updateBlockedContacts(const PrivacyList&);
- 	void getDefaultList_error();
- 	void changeList_error();
+	void privacyClicked();
+	void updatePrivacyTab();
+	void setPrivacyTabEnabled(bool b);
+	void addBlockClicked();
+	void removeBlockClicked();
+	void updateBlockedContacts(const PrivacyList&);
+	void getDefaultList_error();
+	void changeList_error();
 
 	void resourceCbChanged(int index);
 	void priorityCbChanged(int index);
@@ -85,9 +87,9 @@ private:
 	UserAccount acc;
 	
 	// Privacy
- 	PrivacyListModel privacyModel;
- 	PrivacyListBlockedModel privacyBlockedModel;
- 	bool privacyInitialized;
+	PrivacyListModel privacyModel;
+	PrivacyListBlockedModel privacyBlockedModel;
+	bool privacyInitialized;
 
 	void updateUserID();
 	void setKeyID(bool b, const QString &s="");
diff --git a/src/profiles.h b/src/profiles.h
index ee1b986..3600406 100644
--- a/src/profiles.h
+++ b/src/profiles.h
@@ -89,6 +89,7 @@ public:
 
 	QString optionsBase;
 
+	bool useStun;
 	QString stunHost;
 	int stunPort;
 	QString stunUser;
diff --git a/src/psi_profiles.cpp b/src/psi_profiles.cpp
index dfe9a1f..f9dbf39 100644
--- a/src/psi_profiles.cpp
+++ b/src/psi_profiles.cpp
@@ -154,7 +154,9 @@ void UserAccount::reset()
 	proxy_user = "";
 	proxy_pass = "";
 
-	stunPort = 3478;
+	useStun = true;
+	stunHost = "stun.jabber.ru";
+	stunPort = 5249;
 
 	keybind.clear();
 
@@ -307,6 +309,10 @@ void UserAccount::fromOptions(OptionsTree *o, QString base)
 	ibbOnly = o->getOption(base + ".ibb-only").toBool();
 
 
+	if (allSetOptions.contains(base + ".use-stun")) {
+		useStun = o->getOption(base + ".use-stun").toBool();
+	}
+
 	if (allSetOptions.contains(base + ".stun-host")) {
 		stunHost = o->getOption(base + ".stun-host").toString();
 	}
@@ -460,6 +466,7 @@ void UserAccount::toOptions(OptionsTree *o, QString base)
 	o->setOption(base + ".bytestreams-proxy", dtProxy.full());
 	o->setOption(base + ".ibb-only", ibbOnly);
 
+	o->setOption(base + ".use-stun", useStun);
 	o->setOption(base + ".stun-host", stunHost);
 	o->setOption(base + ".stun-port", QString::number(stunPort));
 	o->setOption(base + ".stun-username", stunUser);
diff --git a/src/psiaccount.cpp b/src/psiaccount.cpp
index b185d1b..8b36f30 100644
--- a/src/psiaccount.cpp
+++ b/src/psiaccount.cpp
@@ -949,9 +949,10 @@ public:
 
 	void updateAvCallSettings(const UserAccount &acc)
 	{
-		avCallManager->setStunBindService(acc.stunHost, acc.stunPort);
-		avCallManager->setStunRelayUdpService(acc.stunHost, acc.stunPort, acc.stunUser, acc.stunPass);
-		avCallManager->setStunRelayTcpService(acc.stunHost, acc.stunPort, convert_proxy(acc, jid), acc.stunUser, acc.stunPass);
+		QString stunHost = acc.useStun ? acc.stunHost : "";
+		avCallManager->setStunBindService(stunHost, acc.stunPort);
+		avCallManager->setStunRelayUdpService(stunHost, acc.stunPort, acc.stunUser, acc.stunPass);
+		avCallManager->setStunRelayTcpService(stunHost, acc.stunPort, convert_proxy(acc, jid), acc.stunUser, acc.stunPass);
 	}
 
 	bool isAlwaysVisibleContact(const Jid& jid) const
