--- git.orig/iris/tools/tools.pro
+++ git/iris/tools/tools.pro
@@ -1,2 +1,2 @@
 TEMPLATE = subdirs
-SUBDIRS = nettool icetunnel xmpptest
+#SUBDIRS = nettool icetunnel xmpptest
--- git.orig/src/main.cpp
+++ git/src/main.cpp
@@ -512,6 +512,9 @@ int main(int argc, char *argv[])
 	// version.
 	// see http://www.mozilla.org/security/announce/2007/mfsa2007-23.html
 	// for how this problem affected firefox on windows.
+#ifdef HAVE_QT5
+	const QString appPath = QFileInfo(QString::fromLocal8Bit(argv[0])).absoluteDir().absolutePath();
+#endif
 
 	PsiCli cli;
 
@@ -548,6 +551,7 @@ int main(int argc, char *argv[])
 	// it must be initialized first in order for ApplicationInfo::resourcesDir() to work
 #ifdef HAVE_QT5
 	qInstallMessageHandler(psiMessageOutput);
+	QCoreApplication::addLibraryPath(appPath);
 #else
 	qInstallMsgHandler(psiMessageOutput);
 #endif
