diff --git a/iris/src/xmpp/CMakeLists.txt b/iris/src/xmpp/CMakeLists.txt
index 593b5dc1..3d72b436 100644
--- a/iris/src/xmpp/CMakeLists.txt
+++ b/iris/src/xmpp/CMakeLists.txt
@@ -123,8 +123,9 @@ set(SOURCES
     xmpp-im/xmpp_vcard.cpp
     xmpp-im/xmpp_xdata.cpp
     xmpp-im/xmpp_xmlcommon.cpp
-    xmpp-im/jingle-s5b.cpp
+    xmpp-im/jingle.cpp
     xmpp-im/jingle-ft.cpp
+    xmpp-im/jingle-s5b.cpp
 
     base/randomnumbergenerator.cpp
     base/timezone.cpp
@@ -150,7 +151,6 @@ set(SOURCES
 
     xmpp-im/s5b.cpp
     xmpp-im/xmpp_features.cpp
-    xmpp-im/jingle.cpp
     )
 
 add_library(iris
diff --git a/iris/src/xmpp/xmpp-im/jingle-s5b.cpp b/iris/src/xmpp/xmpp-im/jingle-s5b.cpp
index 718d0b0d..ee417d07 100644
--- a/iris/src/xmpp/xmpp-im/jingle-s5b.cpp
+++ b/iris/src/xmpp/xmpp-im/jingle-s5b.cpp
@@ -16,7 +16,6 @@
  * along with this library.  If not, see <https://www.gnu.org/licenses/>.
  *
  */
-
 #include "jingle-s5b.h"
 #include "s5b.h"
 #include "xmpp/jid/jid.h"
@@ -1193,3 +1192,5 @@ void Pad::registerSid(const QString &sid)
 } // namespace S5B
 } // namespace Jingle
 } // namespace XMPP
+
+#include "jingle-s5b.moc"
