diff --git a/src/contactlistdragview.cpp b/src/contactlistdragview.cpp
index 6c36094..dd9b44d 100644
--- a/src/contactlistdragview.cpp
+++ b/src/contactlistdragview.cpp
@@ -740,6 +740,10 @@ void ContactListDragView::mousePressEvent(QMouseEvent* e)
 		itemActivated(index);
 	}
 	ContactListView::mousePressEvent(e);
+    
+	if (activateItemsOnSingleClick()) {
+		ContactListView::mouseDoubleClickEvent(e);
+	}
 }
 
 void ContactListDragView::mouseMoveEvent(QMouseEvent* e)
diff --git a/src/contactlistview.cpp b/src/contactlistview.cpp
index 0496809..e8d9757 100644
--- a/src/contactlistview.cpp
+++ b/src/contactlistview.cpp
@@ -99,7 +99,7 @@ ContactListView::ContactListView(QWidget* parent)
 	setCornerWidget(new ContactListViewCorner());
 #endif
 
-	connect(this, SIGNAL(activated(const QModelIndex&)), SLOT(itemActivated(const QModelIndex&)));
+	connect(this, SIGNAL(doubleClicked(const QModelIndex&)), SLOT(itemActivated(const QModelIndex&)));
 	// showStatus_ = PsiOptions::instance()->getOption("options.ui.contactlist.status-messages.show").toBool();
 }
 
@@ -279,7 +279,7 @@ void ContactListView::keyPressEvent(QKeyEvent* event)
 	case Qt::Key_Return:
 		if (state() != EditingState || hasFocus()) {
 			if (currentIndex().isValid())
-				emit activated(currentIndex());
+				emit doubleClicked(currentIndex());
 		}
 		else {
 			event->ignore();
