--- psi.orig/iris/src/xmpp/xmpp-im/filetransfer.cpp
+++ psi/iris/src/xmpp/xmpp-im/filetransfer.cpp
@@ -438,11 +438,14 @@ void FileTransferManager::setDisabled(co
 void FileTransferManager::pft_incoming(const FTRequest &req)
 {
 	QString streamType;
-	foreach (const QString &ns, req.streamTypes) {
-		BytestreamManager *manager = streamManager(ns);
-		if (manager && manager->isAcceptableSID(req.from, req.id)) {
-			streamType = ns;
-			break;
+	for(int i = 0; i < d->streamPriority.size(); i++) {
+		const QString ns = d->streamPriority.at(i);
+		if(req.streamTypes.contains(ns)) {
+			BytestreamManager *manager = streamManager(ns);
+			if (manager && manager->isAcceptableSID(req.from, req.id)) {
+				streamType = ns;
+				break;
+			}
 		}
 	}
 
