--- git.orig/src/main.cpp
+++ git/src/main.cpp
@@ -360,7 +360,7 @@ static QByteArray encodeUri(const QByteA
 	return out;
 }
 
-static QString decodeUri(const QString &in)
+static QByteArray decodeUri(const QByteArray &in)
 {
 	// note that the input is an 8-bit text encoding which is then escaped,
 	//   and the result is a latin1-compatible string.  to decode this, we
@@ -368,17 +368,17 @@ static QString decodeUri(const QString &
 	//   into unicode
 
 	QByteArray dec;
-	for(int n = 0; n < in.length(); ++n) {
-		if(in[n] == '\\') {
-			if(n + 1 < in.length()) {
+	for(int n = 0; n < in.size(); ++n) {
+		if(in.at(n) == '\\') {
+			if(n + 1 < in.size()) {
 				++n;
-				if(in[n] == '\\') {
+				if(in.at(n) == '\\') {
 					dec += '\\';
 				}
-				else if(in[n] == 'x') {
-					if(n + 2 < in.length()) {
+				else if(in.at(n) == 'x') {
+					if(n + 2 < in.size()) {
 						++n;
-						QString xs = in.mid(n, 2);
+						QByteArray xs = in.mid(n, 2);
 						++n;
 						bool ok = false;
 						int x = xs.toInt(&ok, 16);
@@ -391,10 +391,10 @@ static QString decodeUri(const QString &
 			}
 		}
 		else
-			dec += in[n].toLatin1();
+			dec += in.at(n);
 	}
 
-	return QString::fromLocal8Bit(dec);
+	return dec;
 }
 
 // NOTE: this function is called without qapp existing
