--- psi.orig/src/contactlistdragmodel.cpp
+++ psi/src/contactlistdragmodel.cpp
@@ -528,7 +528,7 @@ QList<PsiContact*> ContactListDragModel:
 		QStringList groups = psiContact->groups();
 
 		foreach(ContactListModelOperationList::Operation op, contactOperation.operations) {
-			groups.removeAll(op.groupFrom);
+			groups.removeAll(processContactSetGroupName(op.groupFrom));
 		}
 
 		if (!groupsEnabled()) {
@@ -631,7 +631,8 @@ QString ContactListDragModel::sourceOper
 	if (specialGroup) {
 		return specialGroup->sourceOperationsForSpecialGroupContact(contact);
 	}
-	return processContactSetGroupName(groupName);
+	return groupName;
+	//return processContactSetGroupName(groupName);
 }
 
 QString ContactListDragModel::destinationOperationsForContactGroup(const QString& groupName, PsiContact* contact) const
