diff --git a/src/libpsi/dialogs/CMakeLists.txt b/src/libpsi/dialogs/CMakeLists.txt
index 701fee7..7d604c2 100644
--- a/src/libpsi/dialogs/CMakeLists.txt
+++ b/src/libpsi/dialogs/CMakeLists.txt
@@ -1,9 +1,6 @@
-unset(MOC_SOURCES)
 unset(HEADERS)
-unset(PLAIN_HEADERS)
 unset(FORMS)
 unset(SOURCES)
-unset(PLAIN_SOURCES)
 unset(UI_FORMS)
 unset(EXTRA_LDFLAGS)
 
@@ -13,7 +10,7 @@ list(APPEND HEADERS
     grepshortcutkeydialog.h
     )
 
-list(APPEND PLAIN_SOURCES
+list(APPEND SOURCES
     grepshortcutkeydialog.cpp
     )
 
@@ -22,7 +19,6 @@ list(APPEND FORMS
     )
 
 qt5_wrap_ui(UI_FORMS ${FORMS})
-qt_wrap_cpp(MOC_SOURCES ${HEADERS})
-add_library(libpsi_dialogs STATIC ${HEADERS} ${MOC_SOURCES} ${UI_FORMS} ${PLAIN_SOURCES} ${PLAIN_HEADERS})
+add_library(libpsi_dialogs STATIC ${HEADERS} ${SOURCES} ${UI_FORMS})
 target_link_libraries(libpsi_dialogs ${QT_LIBRARIES})
 target_include_directories(libpsi_dialogs PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
diff --git a/src/libpsi/tools/CMakeLists.txt b/src/libpsi/tools/CMakeLists.txt
index ed489bb..29802d6 100644
--- a/src/libpsi/tools/CMakeLists.txt
+++ b/src/libpsi/tools/CMakeLists.txt
@@ -1,24 +1,36 @@
 set(CMAKE_CXX_STANDARD 14)
 
-unset(MOC_SOURCES)
 unset(HEADERS)
-unset(PLAIN_HEADERS)
 unset(FORMS)
 unset(SOURCES)
-unset(PLAIN_SOURCES)
 unset(UI_FORMS)
 unset(EXTRA_LDFLAGS)
 
 if(APPLE)
-    # mac {
-    #     # Growl
-    #     contains(DEFINES, HAVE_GROWL) {
-    #         include($$PWD/growlnotifier/growlnotifier.pri)
-    #     }
-
-    #     # Mac dock
-    #     include($$PWD/mac_dock/mac_dock.pri)
-    # }
+    if(USE_GROWL)
+        list(APPEND EXTRA_LDFLAGS
+            "-framework Growl -framework Cocoa"
+            )
+        list(APPEND HEADERS
+            growlnotifier/growlnotifier.h
+            )
+        list(APPEND SOURCES
+            growlnotifier/growlnotifier.mm
+            )
+    endif()
+    if(USE_MAC_DOC)
+        list(APPEND EXTRA_LDFLAGS
+            "-framework Carbon"
+            )
+        list(APPEND HEADERS
+            mac_dock/mac_dock.h
+            mac_dock/privateqt_mac.h
+            )
+        list(APPEND SOURCES
+            mac_dock/mac_dock.mm
+            mac_dock/privateqt_mac.mm
+            )
+    endif()
 endif()
 
 list(APPEND HEADERS
@@ -33,9 +45,7 @@ list(APPEND HEADERS
 
     # globalshortcut
     globalshortcut/globalshortcuttrigger.h
-    )
 
-list(APPEND PLAIN_HEADERS
     # tools
     maybe.h
     iodeviceopener.h
@@ -55,7 +65,7 @@ list(APPEND PLAIN_HEADERS
     spellchecker/spellhighlighter.h
     )
 
-list(APPEND PLAIN_SOURCES
+list(APPEND SOURCES
     # tools
     priorityvalidator.cpp
     languagemanager.cpp
@@ -84,12 +94,12 @@ list(APPEND PLAIN_SOURCES
     )
 
 if(APPLE)
-    list(APPEND PLAIN_HEADERS
+    list(APPEND HEADERS
         mac_dock/mac_dock.h
         mac_dock/privateqt_mac.h
         )
 
-    list(APPEND PLAIN_SOURCES
+    list(APPEND SOURCES
         # globalshortcut
         globalshortcut/globalshortcutmanager_mac.mm
         globalshortcut/NDKeyboardLayout.m
@@ -114,12 +124,12 @@ if(APPLE)
         systemwatch/systemwatch_mac.cpp
         )
 elseif(WIN32)
-    list(APPEND PLAIN_HEADERS
+    list(APPEND HEADERS
         # spellchecker
         spellchecker/hunspellchecker.h
         )
 
-    list(APPEND PLAIN_SOURCES
+    list(APPEND SOURCES
         #idle
         idle/idle_win.cpp
 
@@ -138,7 +148,7 @@ elseif(HAIKU)
         systemwatch/systemwatch_unix.h
         )
 
-        list(APPEND PLAIN_SOURCES
+        list(APPEND SOURCES
         #idle
         idle/idle_x11.cpp
 
@@ -154,14 +164,13 @@ else()
         systemwatch/systemwatch_unix.h
         )
 
-    list(APPEND PLAIN_SOURCES
+    list(APPEND SOURCES
         #idle
         idle/idle_x11.cpp
 
         # systemwatch
         systemwatch/systemwatch_unix.cpp
 
-
         # globalshortcut
         globalshortcut/globalshortcutmanager_x11.cpp
         )
@@ -188,11 +197,11 @@ if(USE_ENCHANT)
         ${Enchant_LIBRARY}
         )
 
-    list(APPEND PLAIN_HEADERS
+    list(APPEND HEADERS
         spellchecker/enchantchecker.h
         )
 
-    list(APPEND PLAIN_SOURCES
+    list(APPEND SOURCES
         spellchecker/enchantchecker.cpp
         )
 elseif(USE_HUNSPELL)
@@ -210,15 +219,15 @@ elseif(USE_HUNSPELL)
         ${HUNSPELL_LIBRARY}
         )
 
-    list(APPEND PLAIN_HEADERS
+    list(APPEND HEADERS
         spellchecker/hunspellchecker.h
         )
 
-    list(APPEND PLAIN_SOURCES
+    list(APPEND SOURCES
         spellchecker/hunspellchecker.cpp
         )
 elseif(APPLE)
-    list(APPEND PLAIN_SOURCES
+    list(APPEND SOURCES
         spellchecker/macspellchecker.mm
         )
 
@@ -228,8 +237,7 @@ elseif(APPLE)
 endif()
 
 qt5_wrap_ui(UI_FORMS ${FORMS})
-qt_wrap_cpp(MOC_SOURCES ${SOURCES} ${HEADERS})
-add_library(libpsi_tools STATIC  ${SOURCES} ${HEADERS} ${MOC_SOURCES} ${UI_FORMS} ${PLAIN_SOURCES} ${PLAIN_HEADERS})
+add_library(libpsi_tools STATIC  ${SOURCES} ${HEADERS} ${UI_FORMS})
 target_link_libraries(libpsi_tools ${QT_LIBRARIES} tools ${EXTRA_LDFLAGS})
 target_include_directories(libpsi_tools PUBLIC ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
 
diff --git a/src/libpsi/tools/zip/CMakeLists.txt b/src/libpsi/tools/zip/CMakeLists.txt
index 565f693..10b8a21 100644
--- a/src/libpsi/tools/zip/CMakeLists.txt
+++ b/src/libpsi/tools/zip/CMakeLists.txt
@@ -1,9 +1,6 @@
-unset(MOC_SOURCES)
 unset(HEADERS)
-unset(PLAIN_HEADERS)
 unset(FORMS)
 unset(SOURCES)
-unset(PLAIN_SOURCES)
 unset(UI_FORMS)
 unset(EXTRA_LDFLAGS)
 
