--- psi.orig/src/contactlistviewdelegate.cpp
+++ psi/src/contactlistviewdelegate.cpp
@@ -373,11 +373,14 @@ bool ContactListViewDelegate::eventFilte
 	QWidget* editor = ::qobject_cast<QWidget*>(object);
 	if (event->type() == QEvent::KeyPress) {
 		QKeyEvent* keyEvent = static_cast<QKeyEvent*>(event);
-		if (keyEvent->key() == Qt::Key_Up || keyEvent->key() == Qt::Key_Home) {
+		if (keyEvent->key() == Qt::Key_Up ||
+		   (keyEvent->key() == Qt::Key_Home && keyEvent->modifiers() != Qt::ShiftModifier))
+		{
 			setEditorCursorPosition(editor, 0);
 			return true;
 		}
-		else if (keyEvent->key() == Qt::Key_Down || keyEvent->key() == Qt::Key_End) {
+		else if (keyEvent->key() == Qt::Key_Down ||
+			(keyEvent->key() == Qt::Key_End && keyEvent->modifiers() != Qt::ShiftModifier)) {
 			setEditorCursorPosition(editor, -1);
 			return true;
 		}
