--- psi.orig/src/contactlistmodel.h
+++ psi/src/contactlistmodel.h
@@ -97,6 +97,7 @@ public:
 		TotalContactsRole = Qt::UserRole + 16,
 		InternalGroupNameRole = Qt::UserRole + 17,
 		SpecialGroupTypeRole = Qt::UserRole + 18,
+		DisplayGroupRole = Qt::UserRole + 33,
 
 		// accounts
 		UsingSSLRole = Qt::UserRole + 19,
--- psi.orig/src/psicontactlistmodel.cpp
+++ psi/src/psicontactlistmodel.cpp
@@ -19,6 +19,7 @@
  */
 
 #include <QTimer>
+#include <QTextDocument> // for Qt::escape()
 
 #include "psicontactlistmodel.h"
 
@@ -57,8 +58,10 @@ QVariant PsiContactListModel::contactDat
 
 QVariant PsiContactListModel::contactGroupData(const ContactListGroup* group, int role) const
 {
-	if (role == Qt::ToolTipRole) {
+	if (role == DisplayGroupRole || role == Qt::ToolTipRole) {
 		QString text = itemData(group, Qt::DisplayRole).toString();
+		if (role == Qt::ToolTipRole)
+			text = Qt::escape(text);
 		text += QString(" (%1/%2)")
 		        .arg(itemData(group, ContactListModel::OnlineContactsRole).toInt())
 		        .arg(itemData(group, ContactListModel::TotalContactsRole).toInt());
--- psi.orig/src/psicontactlistviewdelegate.cpp
+++ psi/src/psicontactlistviewdelegate.cpp
@@ -422,7 +422,7 @@ void PsiContactListViewDelegate::drawGro
 
 	r.setLeft(pixmapRect.right() + 3);
 
-	QString text = index.data(Qt::ToolTipRole).toString();
+	QString text = index.data(ContactListModel::DisplayGroupRole).toString();
 	drawText(painter, o, r, text, index);
 
 	if(slimGroup_ && !(option.state & QStyle::State_Selected)) {
