--- psi.orig/iris/src/xmpp/sasl/scramsha1response.cpp
+++ psi/iris/src/xmpp/sasl/scramsha1response.cpp
@@ -38,7 +38,7 @@ namespace XMPP {
 	SCRAMSHA1Response::SCRAMSHA1Response(const QByteArray& server_first_message, const QByteArray& password_in, const QByteArray& client_first_message, const QString &salted_password_base64, const RandomNumberGenerator& rand)
 {
 	Q_UNUSED(rand);
-	QString pass_in(password_in);
+	QString pass_in = QString::fromUtf8(password_in);
 	QString pass_out;
 
 	QRegExp pattern("r=(.*),s=(.+),i=(\\d+)");
