diff --git a/iris/src/xmpp/xmpp-core/protocol.h b/iris/src/xmpp/xmpp-core/protocol.h
--- a/iris/src/xmpp/xmpp-core/protocol.h
+++ b/iris/src/xmpp/xmpp-core/protocol.h
@@ -43,7 +43,7 @@
 #define NS_COMPRESS_FEATURE "http://jabber.org/features/compress"
 #define NS_COMPRESS_PROTOCOL "http://jabber.org/protocol/compress"
 #define NS_HOSTS    "http://barracuda.com/xmppextensions/hosts"
-#define NS_STREAM_MANAGEMENT  "urn:xmpp:sm:3"
+#define NS_STREAM_MANAGEMENT  "urn:xmpp:sm:xxx"
 
 namespace XMPP
 {
