--- psi.orig/qcm/plugins.qcm
+++ psi/qcm/plugins.qcm
@@ -1,8 +1,6 @@
 /*
 -----BEGIN QCMOD-----
-name: Plugins
-arg: enable-plugins,Enable Psi Plugin support
-
+name: Psi Plugin support
 -----END QCMOD-----
 */
 
@@ -13,17 +11,10 @@ class qc_plugins : public ConfObj
 {
 public:
 	qc_plugins(Conf *c) : ConfObj(c) {}
-	QString name() const { return "Plugins"; }
-	QString shortname() const { return "Plugins"; }
+	QString name() const { return "Psi Plugin support"; }
+	QString shortname() const { return "plugins"; }
 	bool exec()
 	{
-		QString s;
-
-		// Check if Jingle was enabled explicitly
-		s = conf->getenv("QC_ENABLE_PLUGINS");
-		if(s.isEmpty())
-			return false;
-
 		conf->addDefine("PSI_PLUGINS");
 
 		// Finish
--- psi.orig/psi.qc
+++ psi/psi.qc
@@ -45,7 +45,7 @@
   <dep type='spell'>
     <required/>
   </dep>
-  <!--<dep type='plugins'/>-->
+  <dep type='plugins'><disabled/></dep>
   <dep type='conf'>
     <required/>
   </dep>
