--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -345,6 +345,10 @@
 						<usertext type="QColor" comment="Color used to indicate additional text for informational messages.">#606060</usertext>
 						<highlighting type="QColor">#FF0000</highlighting>
 					</messages>
+					<chat>
+						<link-color type="QColor">#0000ff</link-color>
+						<mailto-color type="QColor">#0000ff</mailto-color>
+					</chat>
 					<passive-popup>
 						<border type="QColor">#5297f9</border>
 					</passive-popup>
--- psi.orig/src/textutil.cpp
+++ psi/src/textutil.cpp
@@ -411,7 +411,8 @@ QString TextUtil::linkify(const QString
 			href = escape(href);
 			href = linkify_htmlsafe(href);
 			//printf("link: [%s], href=[%s]\n", link.latin1(), href.latin1());
-			linked = QString("<a href=\"%1\">").arg(href) + escape(link) + "</a>" + escape(pre.mid(cutoff));
+			QColor lcolor=PsiOptions::instance()->getOption("options.ui.look.colors.chat.link-color").toString();
+			linked = QString("<a href=\"%1\" style=\"color:"+lcolor.name()+"\">").arg(href) + escape(link) + "</a>" + escape(pre.mid(cutoff));
 			out.replace(x1, len, linked);
 			n = x1 + linked.length() - 1;
 		}
@@ -444,7 +445,8 @@ QString TextUtil::linkify(const QString
 
 			href += link;
 			//printf("link: [%s], href=[%s]\n", link.latin1(), href.latin1());
-			linked = QString("<a href=\"%1\">").arg(href) + link + "</a>";
+			QColor mcolor=PsiOptions::instance()->getOption("options.ui.look.colors.chat.mailto-color").toString();
+			linked = QString("<a href=\"%1\" style=\"color:"+mcolor.name()+"\">").arg(href) + link + "</a>";
 			out.replace(x1, len, linked);
 			n = x1 + linked.length() - 1;
 		}
