--- psi.orig/src/mainwin.cpp
+++ psi/src/mainwin.cpp
@@ -133,7 +133,7 @@ public:
 	bool isLeftRoster;
 
 	PopupAction* optionsButton, *statusButton;
-	IconActionGroup* statusGroup;
+	IconActionGroup* statusGroup, *viewGroups;
 	IconAction* statusSmallerAlt;
 	EventNotifierAction* eventNotifier;
 	PsiCon* psi;
@@ -177,6 +177,7 @@ MainWin::Private::Private(PsiCon* _psi,
 {
 
 	statusGroup   = (IconActionGroup *)getAction("status_group");
+	viewGroups    = (IconActionGroup *)getAction("view_groups");
 	eventNotifier = (EventNotifierAction *)getAction("event_notifier");
 
 	optionsButton = (PopupAction *)getAction("button_options");
@@ -946,6 +947,7 @@ void MainWin::buildOptionsMenu()
 {
 	buildGeneralMenu( d->optionsMenu );
 	d->optionsMenu->addSeparator();
+	d->optionsMenu->addAction(d->viewGroups);
 
 	// help menu
 	QMenu* helpMenu = new QMenu(tr("&Help"), d->optionsMenu);
