--- git.orig/src/psiaccount.cpp
+++ git/src/psiaccount.cpp
@@ -25,6 +25,7 @@
 
 
 #include <QFileDialog>
+#include <QSet>
 #include <QInputDialog>
 #include <QTimer>
 #include <QMessageBox>
@@ -2274,9 +2275,19 @@ void PsiAccount::bookmarksAvailabilityCh
 	}
 
 #ifdef GROUPCHAT
+	QSet<QString> joinSkip;
+	QFile file(pathToProfile(activeProfile, ApplicationInfo::ConfigLocation)+"/mucskipautojoin.txt");
+	if (file.exists() && file.open(QIODevice::ReadOnly)) {
+		QTextStream stream(&file);
+		while (!stream.atEnd()) {
+			joinSkip << stream.readLine().toUpper();
+		}
+	}
+
 	foreach(ConferenceBookmark c, d->bookmarkManager->conferences()) {
 		if (!findDialog<GCMainDlg*>(Jid(c.jid().bare())) && c.autoJoin()) {
-			actionJoin(c, true);
+			if  (!joinSkip.contains(c.jid().full().toUpper()))
+				actionJoin(c, true);
 		}
 	}
 #endif
