--- psi.orig/src/gcuserview.cpp
+++ psi/src/gcuserview.cpp
@@ -557,8 +557,8 @@ bool GCUserView::maybeTip(const QPoint &
 	UserResource ur;
 	ur.setName(nick);
 	ur.setStatus(s);
-	//ur.setClient(client_name,client_version,"");
-	ur.setClient(QString(),QString(),"");
+	ur.setClient(client_name,client_version,"");
+	//ur.setClient(QString(),QString(),"");
 	u.userResourceList().append(ur);
 	u.setPrivate(true);
 	u.setAvatarFactory(dlg->account()->avatarFactory());
--- psi.orig/src/psiaccount.cpp
+++ psi/src/psiaccount.cpp
@@ -2538,6 +2538,8 @@ void PsiAccount::client_resourceAvailabl
 			//ur.setSecurityEnabled(true);
 			if(local)
 				ur.setClient(ApplicationInfo::name(),ApplicationInfo::version(),SystemInfo::instance()->os());
+			else
+				ur.setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
 
 			u->userResourceList().append(ur);
 			rp = &u->userResourceList().last();
@@ -2607,8 +2609,8 @@ void PsiAccount::client_resourceAvailabl
 		foreach(UserListItem* u, findRelevant(j)) {
 			UserResourceList::Iterator rit = u->userResourceList().find(j.resource());
 			if (rit != u->userResourceList().end()) {
-				//(*rit).setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
-				(*rit).setClient(QString(),QString(),"");
+				(*rit).setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
+				//(*rit).setClient(QString(),QString(),"");
 				cpUpdate(*u,(*rit).name());
 			}
 		}
@@ -3174,8 +3176,8 @@ void PsiAccount::capsChanged(const Jid&
 		bool found = (rit == u->userResourceList().end()) ? false: true;
 		if(!found)
 			continue;
-		//(*rit).setClient(name,version,"");
-		(*rit).setClient(QString(),QString(),"");
+		(*rit).setClient(name,version,"");
+		//(*rit).setClient(QString(),QString(),"");
 		cpUpdate(*u);
 	}
 }
