--- psi.orig/src/gcuserview.cpp
+++ psi/src/gcuserview.cpp
@@ -557,8 +557,8 @@ bool GCUserView::maybeTip(const QPoint &
 	UserResource ur;
 	ur.setName(nick);
 	ur.setStatus(s);
-	//ur.setClient(client_name,client_version,"");
-	ur.setClient(QString(),QString(),"");
+	ur.setClient(client_name,client_version,"");
+	//ur.setClient(QString(),QString(),"");
 	u.userResourceList().append(ur);
 	u.setPrivate(true);
 	u.setAvatarFactory(dlg->account()->avatarFactory());

