--- psi.orig/src/gcuserview.cpp
+++ psi/src/gcuserview.cpp
@@ -484,7 +484,6 @@ void GCUserView::contextMenuRequested(co
 	act->setEnabled( (!self || lvi->s.mucItem().role() == MUCItem::Moderator) && MUCManager::canSetRole(c->s.mucItem(),lvi->s.mucItem(),MUCItem::Moderator));
 	pm->addMenu(rm);
 
-	/*
 	QMenu* am = new QMenu(tr("Change Affiliation"), pm);
 	act = am->addAction(tr("Unaffiliated"));
 	act->setData(15);
@@ -511,7 +510,6 @@ void GCUserView::contextMenuRequested(co
 	act->setEnabled((!self || lvi->s.mucItem().affiliation() == MUCItem::Owner) && MUCManager::canSetAffiliation(c->s.mucItem(),lvi->s.mucItem(),MUCItem::Owner));
 
 	pm->addMenu(am);
-	*/
 	pm->addSeparator();
 	//pm->insertItem(tr("Send &File"), 4);
 	//pm->insertSeparator();
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -1660,7 +1660,7 @@ void GCMainDlg::lv_action(const QString 
 		}
 
 	}
-	/*else if(x == 15) {
+	else if(x == 15) {
 		GCUserViewItem *contact = (GCUserViewItem*) ui_.lv_users->findEntry(nick);
 		if (contact->s.mucItem().affiliation() != MUCItem::NoAffiliation)
 			d->mucManager->setAffiliation(contact->s.mucItem().jid(), MUCItem::NoAffiliation);
@@ -1679,7 +1679,7 @@ void GCMainDlg::lv_action(const QString 
 		GCUserViewItem *contact = (GCUserViewItem*) ui_.lv_users->findEntry(nick);
 		if (contact->s.mucItem().affiliation() != MUCItem::Owner)
 			d->mucManager->setAffiliation(contact->s.mucItem().jid(), MUCItem::Owner);
-	}*/
+	}
 }
 
 void GCMainDlg::contextMenuEvent(QContextMenuEvent *)
