--- a/src/mainwin.cpp
+++ b/src/mainwin.cpp
@@ -339,6 +339,8 @@ MainWin::MainWin(bool _onTop, bool _asTool, PsiCon* psi)
 	} else
 		setCentralWidget(rosterBar);
 
+	connect(PsiOptions::instance(), SIGNAL(optionChanged(const QString&)), SLOT(optionChanged(const QString&)));
+
 	d->vb_roster = new QVBoxLayout(rosterBar);
 	d->rosterWidget_ = new PsiRosterWidget(rosterBar);
 	d->rosterWidget_->setContactList(psi->contactList());
@@ -544,6 +546,14 @@ void MainWin::splitterMoved()
 	d->tabsSize = d->isLeftRoster ? list.last() : list.first();
 }
 
+void MainWin::optionChanged(const QString& option)
+{
+	if (option == toolbarsStateOptionPath) { //for compatibility with skins plugin
+		loadToolbarsState();
+		//buildToolbars(); //FIXME!!! not a very good solution
+	}
+}
+
 void MainWin::registerAction( IconAction* action )
 {
 	const char *activated  = SIGNAL( triggered() );
--- a/src/mainwin.h
+++ b/src/mainwin.h
@@ -166,6 +166,8 @@ private slots:
 
 	void hideTimerTimeout();
 
+	void optionChanged(const QString&);
+
 public slots:
 	void setWindowIcon(const QPixmap&);
 	void showNoFocus();
