--- psi.orig/src/mainwin.cpp
+++ psi/src/mainwin.cpp
@@ -358,6 +358,8 @@ MainWin::MainWin(bool _onTop, bool _asTo
 	} else
 		setCentralWidget(rosterBar);
 
+	connect(PsiOptions::instance(), SIGNAL(optionChanged(const QString&)), SLOT(optionChanged(const QString&)));
+
 	d->vb_roster = new QVBoxLayout(rosterBar);
 #ifndef NEWCONTACTLIST
 	cvlist = new ContactView(rosterBar);
@@ -600,6 +602,14 @@ void MainWin::splitterMoved()
 	d->tabsSize = d->isLeftRoster ? list.last() : list.first();
 }
 
+void MainWin::optionChanged(const QString& option)
+{
+	if (option == toolbarsStateOptionPath) { //for compatibility with skins plugin
+		loadToolbarsState();
+		//buildToolbars(); //FIXME!!! not a very good solution
+	}
+}
+
 void MainWin::registerAction( IconAction* action )
 {
 	const char *activated  = SIGNAL( triggered() );
--- psi.orig/src/mainwin.h
+++ psi/src/mainwin.h
@@ -168,6 +168,8 @@ private slots:
 
 	void hideTimerTimeout();
 
+	void optionChanged(const QString&);
+
 public slots:
 	void setWindowIcon(const QPixmap&);
 	void showNoFocus();
