Index: psi/src/gcuserview.cpp
===================================================================
--- psi.orig/src/gcuserview.cpp	2013-01-15 14:04:00.268825973 +0400
+++ psi/src/gcuserview.cpp	2013-01-15 14:04:00.264825973 +0400
@@ -487,7 +487,6 @@
 	act->setEnabled( (!self || lvi->s.mucItem().role() == MUCItem::Moderator) && MUCManager::canSetRole(c->s.mucItem(),lvi->s.mucItem(),MUCItem::Moderator));
 	pm->addMenu(rm);
 
-	/*
 	QMenu* am = new QMenu(tr("Change Affiliation"), pm);
 	act = am->addAction(tr("Unaffiliated"));
 	act->setData(15);
@@ -514,7 +513,6 @@
 	act->setEnabled((!self || lvi->s.mucItem().affiliation() == MUCItem::Owner) && MUCManager::canSetAffiliation(c->s.mucItem(),lvi->s.mucItem(),MUCItem::Owner));
 
 	pm->addMenu(am);
-	*/
 	pm->addSeparator();
 	//pm->insertItem(tr("Send &File"), 4);
 	//pm->insertSeparator();
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:00.268825973 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:00.264825973 +0400
@@ -1662,7 +1662,7 @@
 		}
 
 	}
-	/*else if(x == 15) {
+	else if(x == 15) {
 		GCUserViewItem *contact = (GCUserViewItem*) ui_.lv_users->findEntry(nick);
 		if (contact->s.mucItem().affiliation() != MUCItem::NoAffiliation)
 			d->mucManager->setAffiliation(contact->s.mucItem().jid(), MUCItem::NoAffiliation);
@@ -1681,7 +1681,7 @@
 		GCUserViewItem *contact = (GCUserViewItem*) ui_.lv_users->findEntry(nick);
 		if (contact->s.mucItem().affiliation() != MUCItem::Owner)
 			d->mucManager->setAffiliation(contact->s.mucItem().jid(), MUCItem::Owner);
-	}*/
+	}
 }
 
 void GCMainDlg::contextMenuEvent(QContextMenuEvent *)
