--- a/src/mainwin.cpp
+++ b/src/mainwin.cpp
@@ -127,7 +127,7 @@ public:
 	bool isLeftRoster;
 
 	PopupAction* optionsButton, *statusButton;
-	IconActionGroup* statusGroup;
+	IconActionGroup* statusGroup, *viewGroups;
 	IconAction* statusSmallerAlt;
 	EventNotifierAction* eventNotifier;
 	PsiCon* psi;
@@ -171,6 +171,7 @@ MainWin::Private::Private(PsiCon* _psi, MainWin* _mainWin) : splitter(0), mainTa
 {
 
 	statusGroup   = (IconActionGroup *)getAction("status_group");
+	viewGroups    = (IconActionGroup *)getAction("view_groups");
 	eventNotifier = (EventNotifierAction *)getAction("event_notifier");
 
 	optionsButton = (PopupAction *)getAction("button_options");
@@ -906,6 +907,7 @@ void MainWin::buildOptionsMenu()
 {
 	buildGeneralMenu( d->optionsMenu );
 	d->optionsMenu->addSeparator();
+	d->optionsMenu->addAction(d->viewGroups);
 
 	// help menu
 	QMenu* helpMenu = new QMenu(tr("&Help"), d->optionsMenu);
