--- a/options/default.xml
+++ b/options/default.xml
@@ -113,6 +113,7 @@
                 <auto-popup type="bool">false</auto-popup>
                 <auto-capitalize type="bool">false</auto-capitalize>
                 <auto-scroll-to-bottom type="bool">false</auto-scroll-to-bottom>
+                <caption type="QString"></caption>
                 <delete-contents-after type="QString">hour</delete-contents-after>
                 <raise-chat-windows-on-new-messages type="bool">false</raise-chat-windows-on-new-messages>
                 <switch-tab-on-new-messages type="bool">false</switch-tab-on-new-messages>
--- a/src/chatdlg.cpp
+++ b/src/chatdlg.cpp
@@ -699,7 +699,10 @@ void ChatDlg::setSelfDestruct(int minutes)
 
 QString ChatDlg::desiredCaption() const
 {
-    QString cap = "";
+    QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+    if(!cap.isEmpty()) {
+        return cap;
+    }
 
     if (pending_ > 0) {
         cap += "* ";
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -2259,7 +2259,10 @@ const QString &GCMainDlg::getDisplayName() const
 
 QString GCMainDlg::desiredCaption() const
 {
-    QString cap = "";
+    QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+    if(!cap.isEmpty()) {
+        return cap;
+    }
 
     if (d->pending > 0) {
         cap += "* ";
--- a/src/tabs/tabdlg.cpp
+++ b/src/tabs/tabdlg.cpp
@@ -541,7 +541,10 @@ void TabDlg::activated()
 
 QString TabDlg::desiredCaption() const
 {
-    QString cap = "";
+    QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+    if(!cap.isEmpty()) {
+        return cap;
+    }
     uint pending = 0;
     foreach(TabbableWidget* tab, tabs_) {
         pending += tab->unreadMessageCount();
