--- a/psi.desktop
+++ b/psi.desktop
@@ -2,7 +2,7 @@
 # This is the spec version, *not* the application version
 Version=1.0
 Type=Application
-Name=Psi
+Name=Psi+
 GenericName=XMPP Client
 Comment=Communicate over the XMPP network
 Icon=psi
--- a/src/src.pro
+++ b/src/src.pro
@@ -47,15 +47,15 @@ unix {
 	ad.path=$$PREFIX/share/appdata/
 	ad.file = ../psi.appdata.xml
 	icon1.path=$$PREFIX/share/icons/hicolor/16x16/apps
-	icon1.extra = cp -f $$top_srcdir/iconsets/system/default/logo_16.png $(INSTALL_ROOT)$$icon1.path/psi.png
+	icon1.extra = cp -f $$top_srcdir/iconsets/system/default/psiplus/logo_16.png $(INSTALL_ROOT)$$icon1.path/psi.png
 	icon2.path=$$PREFIX/share/icons/hicolor/32x32/apps
-	icon2.extra = cp -f $$top_srcdir/iconsets/system/default/logo_32.png $(INSTALL_ROOT)$$icon2.path/psi.png
+	icon2.extra = cp -f $$top_srcdir/iconsets/system/default/psiplus/logo_32.png $(INSTALL_ROOT)$$icon2.path/psi.png
 	icon3.path=$$PREFIX/share/icons/hicolor/48x48/apps
-	icon3.extra = cp -f $$top_srcdir/iconsets/system/default/logo_48.png $(INSTALL_ROOT)$$icon3.path/psi.png
+	icon3.extra = cp -f $$top_srcdir/iconsets/system/default/psiplus/logo_48.png $(INSTALL_ROOT)$$icon3.path/psi.png
 	icon4.path=$$PREFIX/share/icons/hicolor/64x64/apps
-	icon4.extra = cp -f $$top_srcdir/iconsets/system/default/logo_64.png $(INSTALL_ROOT)$$icon4.path/psi.png
+	icon4.extra = cp -f $$top_srcdir/iconsets/system/default/psiplus/logo_64.png $(INSTALL_ROOT)$$icon4.path/psi.png
 	icon5.path=$$PREFIX/share/icons/hicolor/128x128/apps
-	icon5.extra = cp -f $$top_srcdir/iconsets/system/default/logo_128.png $(INSTALL_ROOT)$$icon5.path/psi.png
+	icon5.extra = cp -f $$top_srcdir/iconsets/system/default/psiplus/logo_128.png $(INSTALL_ROOT)$$icon5.path/psi.png
 	INSTALLS += dt ad icon1 icon2 icon3 icon4 icon5
 
 	psi_plugins {
