--- git.temp.orig/options/default.xml
+++ git.temp/options/default.xml
@@ -213,6 +213,7 @@
 			<spell-check comment="Options related to the spell checker">
 				<enabled comment="Whether the spell checker is enabled" type="bool">true</enabled>
 			</spell-check>
+			<disable-send-button type="bool">false</disable-send-button>
 			<systemtray comment="Options related to the system tray">
 				<use-old comment="Use the old system tray code (deprecated)" type="bool">false</use-old>
 				<enable type="bool">true</enable>
--- git.temp.orig/src/chatdlg.ui
+++ git.temp/src/chatdlg.ui
@@ -285,6 +285,31 @@
           </layout>
          </item>
          <item>
+          <widget class="QPushButton" name="pb_send">
+           <property name="sizePolicy">
+            <sizepolicy vsizetype="Expanding" hsizetype="Minimum">
+             <horstretch>0</horstretch>
+             <verstretch>0</verstretch>
+            </sizepolicy>
+           </property>
+           <property name="minimumSize">
+            <size>
+             <width>26</width>
+             <height>0</height>
+            </size>
+           </property>
+           <property name="maximumSize" >
+            <size>
+             <width>26</width>
+             <height>16777215</height>
+            </size>
+           </property>
+           <property name="toolTip">
+            <string>Send</string>
+           </property>
+          </widget>
+         </item>
+         <item>
           <widget class="QLabel" name="avatar" >
            <property name="sizePolicy" >
             <sizepolicy>
--- git.temp.orig/src/groupchatdlg.cpp
+++ git.temp/src/groupchatdlg.cpp
@@ -680,6 +680,8 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	d->act_send = new QAction(this);
 	addAction(d->act_send);
 	connect(d->act_send,SIGNAL(triggered()), SLOT(mle_returnPressed()));
+ 	ui_.pb_send->setIcon(IconsetFactory::icon("psi/action_button_send").icon());
+	connect(ui_.pb_send, SIGNAL(clicked()), SLOT(mle_returnPressed()));
 	d->act_close = new QAction(this);
 	addAction(d->act_close);
 	connect(d->act_close,SIGNAL(triggered()), SLOT(close()));
@@ -1600,6 +1602,13 @@ void GCMainDlg::setLooks()
 		ui_.tb_actions->show();
 	}
 
+	if (PsiOptions::instance()->getOption("options.ui.disable-send-button").toBool()) {
+		ui_.pb_send->hide();
+	}
+	else {
+		ui_.pb_send->show();
+	}
+
 	setWindowOpacity(double(qMax(MINIMUM_OPACITY,PsiOptions::instance()->getOption("options.ui.chat.opacity").toInt()))/100);
 
 	// update the widget icon
--- git.temp.orig/src/groupchatdlg.ui
+++ git.temp/src/groupchatdlg.ui
@@ -263,6 +263,31 @@
            </property>
           </widget>
          </item>
+         <item>
+          <widget class="QPushButton" name="pb_send">
+           <property name="sizePolicy">
+            <sizepolicy vsizetype="Expanding" hsizetype="Minimum">
+             <horstretch>0</horstretch>
+             <verstretch>0</verstretch>
+            </sizepolicy>
+           </property>
+           <property name="minimumSize">
+            <size>
+             <width>26</width>
+             <height>0</height>
+            </size>
+           </property>
+           <property name="maximumSize">
+            <size>
+             <width>26</width>
+             <height>16777215</height>
+            </size>
+           </property>
+           <property name="toolTip">
+            <string>Send</string>
+           </property>
+          </widget>
+         </item>
         </layout>
        </item>
       </layout>
--- git.temp.orig/src/psichatdlg.cpp
+++ git.temp/src/psichatdlg.cpp
@@ -216,6 +216,8 @@ void PsiChatDlg::initUi()
 	ui_.splitter->setSizes(list);
 
 	smallChat_ = PsiOptions::instance()->getOption("options.ui.chat.use-small-chats").toBool();
+ 	ui_.pb_send->setIcon(IconsetFactory::icon("psi/action_button_send").icon());
+	connect(ui_.pb_send, SIGNAL(clicked()), this, SLOT(doSend()));
 
 	act_mini_cmd_ = new QAction(this);
 	act_mini_cmd_->setText(tr("Input command..."));
@@ -298,6 +300,13 @@ void PsiChatDlg::updateIdentityVisibilit
 	else {
 		ui_.lb_ident->setVisible(false);
 	}
+
+	if (PsiOptions::instance()->getOption("options.ui.disable-send-button").toBool()) {
+		ui_.pb_send->hide();
+	}
+	else {
+		ui_.pb_send->show();
+	}
 }
 
 void PsiChatDlg::initToolButtons()
