--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -293,6 +293,7 @@
 					<incoming-file-transfer type="bool">true</incoming-file-transfer>
 					<incoming-headline type="bool">true</incoming-headline>
 					<incoming-message type="bool">true</incoming-message>
+					<notify-every-muc-message type="bool">false</notify-every-muc-message>
 					<maximum-jid-length type="int">25</maximum-jid-length>
 					<maximum-text-length type="int">300</maximum-text-length>
 					<maximum-status-length type="int">-1</maximum-status-length>
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -90,6 +90,8 @@
 #include <windows.h>
 #endif
 
+#include "psipopup.h"
+#include "psievent.h"
 
 #define MCMDMUC		"http://psi-im.org/ids/mcmd#mucmain"
 #define MCMDMUCNICK	"http://psi-im.org/ids/mcmd#mucnick"
@@ -1445,9 +1447,17 @@ void GCMainDlg::message(const Message &_
 		if(!m.spooled())
 			account()->playSound(PsiAccount::eSend);
 	}
-	else {
+	else if(account()->status().type() != Status::DND) {
 		if(alert || (PsiOptions::instance()->getOption("options.ui.notifications.sounds.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) )
 			account()->playSound(PsiAccount::eChat2);
+		if(alert || (PsiOptions::instance()->getOption("options.ui.notifications.passive-popups.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) )
+			if (!m.spooled() && !isActiveTab() && !m.from().resource().isEmpty()) {
+				PsiPopup *popup = new PsiPopup(PsiPopup::AlertGcHighlight, account());
+				XMPP::Jid jid = m.from().withDomain("");
+				const UserListItem u;
+				const MessageEvent *e = new MessageEvent(m, account());
+				popup->setData(jid, m.from().resource(), &u, (PsiEvent *)e);
+			}
 	}
 
 	if(from.isEmpty())
--- psi.orig/src/psipopup.cpp
+++ psi/src/psipopup.cpp
@@ -291,6 +291,11 @@ PsiPopup::PsiPopup(PopupType type, PsiAc
 		icon= (PsiIcon *)IconsetFactory::iconPtr("psi/call");
 		doAlertIcon = true;
 		break;
+	case AlertGcHighlight:
+		text += PsiPopup::tr("Groupchat highlight");
+		icon = (PsiIcon *)IconsetFactory::iconPtr("psi/headline");
+		doAlertIcon = true;
+		break;
 	default:
 		break;
 	}
--- psi.orig/src/psipopup.h
+++ psi/src/psipopup.h
@@ -55,7 +55,8 @@ public:
 		AlertChat,
 		AlertHeadline,
 		AlertFile,
-		AlertAvCall
+		AlertAvCall,
+		AlertGcHighlight
 	};
 	PsiPopup(PopupType type, PsiAccount *acc);
 
