--- git.orig/options/default.xml
+++ git/options/default.xml
@@ -606,6 +606,7 @@ QLineEdit#le_status_text {
 				<show-messageNext type="QKeySequence" comment="Show next message" >Ctrl+Down</show-messageNext>
 				<show-messagePrev type="QKeySequence" comment="Show prev message" >Ctrl+Up</show-messagePrev>
 				<change-case type="QKeySequence" comment="Change case of letters in the chat edit" >Ctrl+Alt+x</change-case>
+				<minimize type="QKeySequence" comment="Minimize active chat window"/>
 			</chat>
 			<contactlist comment="Shortcuts in the contactlist">
 				<assign-custom-avatar type="QKeySequence" comment="Assign a custom avatar to the selected contact"/>
--- git.orig/src/groupchatdlg.cpp
+++ git/src/groupchatdlg.cpp
@@ -227,6 +227,7 @@ public:
 //#endif
 	QAction *act_send, *act_scrollup, *act_scrolldown, *act_close;
 	QAction *act_mini_cmd, *act_nick, *act_hide, *act_copy_muc_jid;
+	QAction *act_minimize;
 
 	MCmdSimpleSite mCmdSite;
 	MCmdManager mCmdManager;
@@ -795,6 +796,10 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	connect(bm, SIGNAL(conferencesChanged(QList<ConferenceBookmark>)), SLOT(updateBookmarkIcon()));
 	connect(bm, SIGNAL(bookmarksSaved()), SLOT(updateBookmarkIcon()));
 
+	d->act_minimize = new QAction(this);
+	connect(d->act_minimize, SIGNAL(triggered()), SLOT(doMinimize()));
+	addAction(d->act_minimize);
+
 	int s = PsiIconset::instance()->system().iconSize();
 	ui_.toolbar->setIconSize(QSize(s,s));
 
@@ -912,6 +917,11 @@ void GCMainDlg::horizSplitterMoved()
 	PsiOptions::instance()->setOption("options.ui.muc.roster-width", d->rosterSize);
 }
 
+void GCMainDlg::doMinimize()
+{
+	window()->showMinimized();
+}
+
 void GCMainDlg::ensureTabbedCorrectly()
 {
 	TabbableWidget::ensureTabbedCorrectly();
@@ -955,6 +965,7 @@ void GCMainDlg::setShortcuts()
 	d->act_scrollup->setShortcuts(ShortcutManager::instance()->shortcuts("common.scroll-up"));
 	d->act_scrolldown->setShortcuts(ShortcutManager::instance()->shortcuts("common.scroll-down"));
 	d->act_mini_cmd->setShortcuts(ShortcutManager::instance()->shortcuts("chat.quick-command"));
+	d->act_minimize->setShortcuts(ShortcutManager::instance()->shortcuts("chat.minimize"));
 }
 
 void GCMainDlg::scrollUp()
--- git.orig/src/groupchatdlg.h
+++ git/src/groupchatdlg.h
@@ -146,6 +146,7 @@ private slots:
 #endif
 	void chatEditCreated();
 	void horizSplitterMoved();
+	void doMinimize();
 	void avatarUpdated(const Jid& jid);
 
 public:
--- git.orig/src/psichatdlg.cpp
+++ git/src/psichatdlg.cpp
@@ -354,6 +354,11 @@ void PsiChatDlg::setShortcuts()
 
 	act_mini_cmd_->setShortcuts(ShortcutManager::instance()->shortcuts("chat.quick-command"));
 
+	act_minimize_ = new QAction(this);
+
+	connect(act_minimize_, SIGNAL(triggered()), SLOT(doMinimize()));
+	addAction(act_minimize_);
+	act_minimize_->setShortcuts(ShortcutManager::instance()->shortcuts("chat.minimize"));
 }
 
 void PsiChatDlg::updateIdentityVisibility()
@@ -900,6 +905,10 @@ void PsiChatDlg::chatEditCreated()
 	tabCompletion.setTextEdit(chatEdit());
 }
 
+void PsiChatDlg::doMinimize()
+{
+	window()->showMinimized();
+}
 
 void PsiChatDlg::doSend() {
 	tabCompletion.reset();
--- git.orig/src/psichatdlg.h
+++ git/src/psichatdlg.h
@@ -35,6 +35,7 @@ private slots:
 	void toggleSmallChat();
 	void doClearButton();
 	void doMiniCmd();
+	void doMinimize();
 	void addContact();
 	void buildMenu();
 	void updateCounter();
@@ -84,6 +85,7 @@ private:
 
 	ActionList* actions_;
 	QAction *act_mini_cmd_;
+	QAction *act_minimize_;
 	TypeAheadFindBar *typeahead_;
 
 	ActionLineEdit *le_autojid;
