--- psi.orig/src/mucjoin.ui
+++ psi/src/mucjoin.ui
@@ -27,7 +27,7 @@
    <item row="0" column="1">
     <widget class="AccountsComboBox" name="cb_ident"/>
    </item>
-   <item row="1" column="0">
+   <item row="2" column="0">
     <widget class="QLabel" name="TextLabel1_2">
      <property name="text">
       <string>Recent:</string>
@@ -37,7 +37,7 @@
      </property>
     </widget>
    </item>
-   <item row="1" column="1">
+   <item row="2" column="1">
     <widget class="QComboBox" name="cb_recent">
      <property name="sizePolicy">
       <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
@@ -47,7 +47,7 @@
      </property>
     </widget>
    </item>
-   <item row="2" column="0" colspan="2">
+   <item row="4" column="0" colspan="2">
     <widget class="QGroupBox" name="gb_info">
      <property name="title">
       <string>Room information</string>
@@ -119,7 +119,7 @@
      </layout>
     </widget>
    </item>
-   <item row="3" column="0" colspan="2">
+   <item row="5" column="0" colspan="2">
     <spacer>
      <property name="orientation">
       <enum>Qt::Vertical</enum>
@@ -135,7 +135,7 @@
      </property>
     </spacer>
    </item>
-   <item row="4" column="0" colspan="2">
+   <item row="6" column="0" colspan="2">
     <widget class="Line" name="Line1">
      <property name="frameShape">
       <enum>QFrame::HLine</enum>
@@ -148,16 +148,29 @@
      </property>
     </widget>
    </item>
-   <item row="5" column="0">
+   <item row="7" column="0">
     <widget class="BusyWidget" name="busy" native="true"/>
    </item>
-   <item row="5" column="1">
+   <item row="7" column="1">
     <widget class="QDialogButtonBox" name="buttonBox">
      <property name="standardButtons">
       <set>QDialogButtonBox::Cancel</set>
      </property>
     </widget>
    </item>
+   <item row="1" column="0">
+    <widget class="QLabel" name="label">
+     <property name="text">
+      <string>Bookmarks:</string>
+     </property>
+     <property name="alignment">
+      <set>Qt::AlignRight|Qt::AlignTrailing|Qt::AlignVCenter</set>
+     </property>
+    </widget>
+   </item>
+   <item row="1" column="1">
+    <widget class="QComboBox" name="cb_bookmarks"/>
+   </item>
   </layout>
  </widget>
  <layoutdefault spacing="6" margin="11"/>
--- psi.orig/src/mucjoindlg.cpp
+++ psi/src/mucjoindlg.cpp
@@ -29,6 +29,7 @@
 #include "mucjoindlg.h"
 #include "psicontactlist.h"
 #include "groupchatdlg.h"
+#include "bookmarkmanager.h"
 
 static const int nickConflictCode = 409;
 static const QString additionalSymbol = "_";
@@ -70,6 +71,7 @@ MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiA
 
 	setWindowTitle(CAP(windowTitle()));
 	connect(ui_.cb_recent, SIGNAL(activated(int)), SLOT(recent_activated(int)));
+	connect(ui_.cb_bookmarks, SIGNAL(activated(int)), SLOT(bookmarksActivated(int)));
 	if (!ui_.cb_recent->count()) {
 		ui_.cb_recent->setEnabled(false);
 		ui_.le_host->setFocus();
@@ -77,6 +79,13 @@ MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiA
 	else {
 		recent_activated(0);
 	}
+	if (!ui_.cb_bookmarks->count()) {
+		ui_.cb_bookmarks->setEnabled(false);
+		ui_.le_host->setFocus();
+	}
+	else {
+		bookmarksActivated(0);
+	}
 
 	setWidgetsEnabled(true);
 	adjustSize();
@@ -114,6 +123,8 @@ void MUCJoinDlg::updateIdentity(PsiAccou
 		return;
 	}
 
+	updateBookmarks(account_);
+
 	connect(account_, SIGNAL(disconnected()), this, SLOT(pa_disconnected()));
 }
 
@@ -142,6 +153,33 @@ void MUCJoinDlg::recent_activated(int x)
 	ui_.le_nick->setText(jid.resource());
 }
 
+void MUCJoinDlg::bookmarksActivated(int x)
+{
+	Jid jid(ui_.cb_bookmarks->itemData(x).toString());
+	if (jid.full().isEmpty())
+		return;
+
+	ui_.le_host->setText(jid.domain());
+	ui_.le_room->setText(jid.node());
+	ui_.le_nick->setText(jid.resource());
+}
+
+void MUCJoinDlg::updateBookmarks(PsiAccount *pa)
+{
+	ui_.cb_bookmarks->clear();
+	if (pa && pa->bookmarkManager()->isAvailable()) {
+		ui_.cb_bookmarks->setEnabled(true);
+		foreach(ConferenceBookmark c, pa->bookmarkManager()->conferences()) {
+			Jid jid = c.jid().withResource(c.nick());
+			ui_.cb_bookmarks->addItem(JIDUtil::toString(jid, false));
+			ui_.cb_bookmarks->setItemData(ui_.cb_bookmarks->count()-1, QVariant(jid.full()));
+		}
+	}
+	else {
+		ui_.cb_bookmarks->setEnabled(false);
+	}
+}
+
 void MUCJoinDlg::doJoin(MucJoinReason r)
 {
 	if (!account_ || !account_->checkConnected(this))
@@ -197,6 +235,7 @@ void MUCJoinDlg::setWidgetsEnabled(bool
 {
 	ui_.cb_ident->setEnabled(enabled);
 	ui_.cb_recent->setEnabled(enabled && ui_.cb_recent->count() > 0);
+	ui_.cb_bookmarks->setEnabled(enabled && ui_.cb_bookmarks->count() > 0);
 	ui_.gb_info->setEnabled(enabled);
 	joinButton_->setEnabled(enabled);
 }
--- psi.orig/src/mucjoindlg.h
+++ psi/src/mucjoindlg.h
@@ -68,6 +68,8 @@ private slots:
 	void updateIdentityVisibility();
 	void pa_disconnected();
 	void recent_activated(int);
+	void bookmarksActivated(int);
+	void updateBookmarks(PsiAccount *pa);
 
 private:
 	Ui::MUCJoin ui_;
