diff --git a/src/psiaccount.cpp b/src/psiaccount.cpp
index 68de96e6..b6c27247 100644
--- a/src/psiaccount.cpp
+++ b/src/psiaccount.cpp
@@ -404,7 +404,7 @@ public:
                 SLOT(updateOnlineContactsCountTimeout()));
 
         logoutTimer = new QTimer(this);
-        logoutTimer->setInterval(500);
+        logoutTimer->setInterval(1000);
         logoutTimer->setSingleShot(true);
         connect(logoutTimer, SIGNAL(timeout()), SLOT(finishLogout()));
     }
@@ -1822,7 +1822,9 @@ void PsiAccount::logout(bool fast, const Status &s)
     stateChanged();
 
     d->client->close(fast);
-    if (fast) {
+    // if (fast) {
+    if (false) { // Temporary workaround
+        // See: https://github.com/psi-im/psi/issues/451
         d->finishLogout();
     } else { // let iris queue write all the stuff
         d->logoutTimer->start();
