--- psi.orig/src/chatdlg.cpp
+++ psi/src/chatdlg.cpp
@@ -1027,6 +1027,12 @@ void ChatDlg::addEmoticon(QString text)
 	PsiRichText::addEmoticon(chatEdit(), text);
 }
 
+void ChatDlg::setText(const QString& text)
+{
+	chatEdit()->clear();
+	chatEdit()->insertHtml(text + ' ');
+}
+
 /**
  * Records that the user is composing
  */
--- psi.orig/src/chatdlg.h
+++ psi/src/chatdlg.h
@@ -74,6 +74,7 @@ public:
 
 public:
 	PsiAccount* account() const;
+	void setText(const QString& );
 
 signals:
 	void aInfo(const Jid &);
--- psi.orig/src/common.cpp
+++ psi/src/common.cpp
@@ -25,6 +25,7 @@
 #include "psiiconset.h"
 #include "applicationinfo.h"
 #include "psioptions.h"
+#include "tabdlg.h"
 
 #include <QUrl>
 #include <QProcess>
@@ -377,6 +378,32 @@ void closeDialogs(QWidget *w)
 	}
 }
 
+TabbableWidget* findActiveTab()
+{
+	QWidget* chat = QApplication::activeWindow();
+	TabbableWidget* tw = 0;
+	if(chat) {
+		TabDlg* td = qobject_cast<TabDlg*>(chat);
+		if(td) {
+			tw = td->getCurrentTab();
+		}
+		else {
+			tw = qobject_cast<TabbableWidget*>(chat);
+			if (!tw) {
+				QList<TabDlg*> tmp = chat->findChildren<TabDlg*>(); // all-in-one
+				while(!tmp.isEmpty()) {
+					TabDlg* td = tmp.takeFirst();
+					tw = td->getCurrentTab();
+					if(tw) {
+						break;
+					}
+				}
+			}
+		}
+	}
+	return tw;
+}
+
 #ifdef Q_WS_X11
 #include <X11/Xlib.h>
 #include <X11/Xutil.h> // needed for WM_CLASS hinting
--- psi.orig/src/common.h
+++ psi/src/common.h
@@ -29,6 +29,7 @@
 #include <QColor>
 
 class QMenu;
+class TabbableWidget;
 
 #include "statuspreset.h"
 
@@ -125,6 +126,7 @@ void clearMenu(QMenu *m); // deletes all
 void bringToFront(QWidget *w, bool grabFocus = true);
 void replaceWidget(QWidget *, QWidget *);
 void closeDialogs(QWidget *);
+TabbableWidget* findActiveTab();
 #ifdef Q_WS_X11
 #include <QWidget>
 #include <QX11Info>
--- psi.orig/src/psiaccount.cpp
+++ psi/src/psiaccount.cpp
@@ -3870,7 +3870,7 @@ void PsiAccount::actionHistoryBox(PsiEve
 	w->show();
 }
 
-void PsiAccount::actionOpenChat(const Jid &j)
+void PsiAccount::actionOpenChat(const Jid &j, const QString & body)
 {
 	UserListItem *u = find(j);
 	if(!u) {
@@ -3904,9 +3904,9 @@ void PsiAccount::actionOpenChat(const Ji
 	}
 
 	if(!res.isEmpty())
-		openChat(j.withResource(res), UserAction);
+		openChat(j.withResource(res), UserAction, body);
 	else
-		openChat(j, UserAction);
+		openChat(j, UserAction, body);
 }
 
 // unlike actionOpenChat(), this function will work on jids that aren't
@@ -4206,11 +4206,11 @@ void PsiAccount::openUri(const QUrl &uri
 		QString subject = uri.queryItemValue("subject");
 		QString body = uri.queryItemValue("body");
 		QString type = uri.queryItemValue("type");
-		if (type == "chat" && subject.isEmpty() && body.isEmpty()) {
+		if (type == "chat" && subject.isEmpty()) {
 			if (!find(entity)) {
 				addUserListItem(entity);
 			}
-			actionOpenChat(entity);
+			actionOpenChat(entity, body);
 		} else {
 			dj_newMessage(entity, body, subject, "");
 		}
@@ -5181,7 +5181,7 @@ void PsiAccount::processChats(const Jid 
 	processChatsHelper(j, true);
 }
 
-void PsiAccount::openChat(const Jid& j, ActivationType activationType)
+void PsiAccount::openChat(const Jid& j, ActivationType activationType, const QString& body)
 {
 	ChatDlg* chat = ensureChatDlg(j);
 	chat->ensureTabbedCorrectly();
@@ -5197,6 +5197,9 @@ void PsiAccount::openChat(const Jid& j, 
 			QTimer::singleShot(1000, dlg, SLOT(showWithoutActivation()));
 		}
 	}
+	if (!body.isEmpty()) {
+		chat->setText(body);
+	}
 }
 
 void PsiAccount::chatMessagesRead(const Jid &j)
--- psi.orig/src/psiaccount.h
+++ psi/src/psiaccount.h
@@ -318,7 +318,7 @@ public slots:
 	void actionRename(const Jid &, const QString &);
 	void actionGroupRename(const QString &, const QString &);
 	void actionHistory(const Jid &);
-	void actionOpenChat(const Jid &);
+	void actionOpenChat(const Jid &,const QString & body = QString());
 	void actionOpenSavedChat(const Jid &);
 	void actionOpenChat2(const Jid &);
 	void actionOpenChatSpecific(const Jid &);
@@ -493,7 +493,7 @@ private:
 
 	void processChatsHelper(const Jid& jid, bool removeEvents);
 	void processChats(const Jid &);
-	void openChat(const Jid &, ActivationType activationType);
+	void openChat(const Jid &, ActivationType activationType, const QString & body = QString());
 	EventDlg *ensureEventDlg(const Jid &);
 	friend class PsiCon;
 
--- psi.orig/src/psicon.cpp
+++ psi/src/psicon.cpp
@@ -1178,7 +1178,14 @@ void PsiCon::openUri(const QUrl &uri)
 	// authority
 	PsiAccount *pa = 0;
 	//if (uri.authority().isEmpty()) {
-		pa = d->contactList->defaultAccount();
+		TabbableWidget* tw = findActiveTab();
+		if (tw) {
+			pa = tw->account();
+		}
+		if (!pa) {
+			pa = d->contactList->defaultAccount();
+		}
+
 		if (!pa) {
 			QMessageBox::critical(0, tr("Error"), QString("You need to have an account configured and enabled to open URIs (links)."));
 			return;
--- psi.orig/src/tabs/tabdlg.cpp
+++ psi/src/tabs/tabdlg.cpp
@@ -770,3 +770,8 @@ void TabDlg::setSimplifiedCaptionEnabled
 	simplifiedCaption_ = enabled;
 	updateCaption();
 }
+
+TabbableWidget* TabDlg::getCurrentTab() const
+{
+	return dynamic_cast<TabbableWidget*>(tabWidget_->currentPage());
+}
--- psi.orig/src/tabs/tabdlg.h
+++ psi/src/tabs/tabdlg.h
@@ -84,6 +84,7 @@ public:
 	void setUserManagementEnabled(bool enabled); // default enabled
 	void setTabBarShownForSingles(bool enabled); // default enabled
 	void setSimplifiedCaptionEnabled(bool enabled); // default disabled
+	TabbableWidget* getCurrentTab() const;
 
 protected:
 	void setShortcuts();
