Entering 'iris'
Entering 'src/libpsi'
diff --git a/src/chatdlg.cpp b/src/chatdlg.cpp
index a21537d..c05dea1 100644
--- a/src/chatdlg.cpp
+++ b/src/chatdlg.cpp
@@ -1006,6 +1006,12 @@ void ChatDlg::addEmoticon(QString text)
 	PsiRichText::addEmoticon(chatEdit(), text);
 }
 
+void ChatDlg::setText(const QString& text)
+{
+	chatEdit()->clear();
+	chatEdit()->insertHtml(text + ' ');
+}
+
 /**
  * Records that the user is composing
  */
diff --git a/src/chatdlg.h b/src/chatdlg.h
index d434520..95320fa 100644
--- a/src/chatdlg.h
+++ b/src/chatdlg.h
@@ -72,6 +72,7 @@ public:
 
 public:
 	PsiAccount* account() const;
+	void setText(const QString& );
 
 signals:
 	void aInfo(const Jid &);
diff --git a/src/common.cpp b/src/common.cpp
index 8e4a5e4..cad6f16 100644
--- a/src/common.cpp
+++ b/src/common.cpp
@@ -25,6 +25,7 @@
 #include "psiiconset.h"
 #include "applicationinfo.h"
 #include "psioptions.h"
+#include "tabdlg.h"
 
 #include <QUrl>
 #include <QProcess>
@@ -380,6 +381,32 @@ void closeDialogs(QWidget *w)
 	}
 }
 
+TabbableWidget* findActiveTab()
+{
+	QWidget* chat = QApplication::activeWindow();
+	TabbableWidget* tw = 0;
+	if(chat) {
+		TabDlg* td = qobject_cast<TabDlg*>(chat);
+		if(td) {
+			tw = td->getCurrentTab();
+		}
+		else {
+			tw = qobject_cast<TabbableWidget*>(chat);
+			if (!tw) {
+				QList<TabDlg*> tmp = chat->findChildren<TabDlg*>(); // all-in-one
+				while(!tmp.isEmpty()) {
+					TabDlg* td = tmp.takeFirst();
+					tw = td->getCurrentTab();
+					if(tw) {
+						break;
+					}
+				}
+			}
+		}
+	}
+	return tw;
+}
+
 #ifdef HAVE_X11
 #include <X11/Xlib.h>
 #include <X11/Xutil.h> // needed for WM_CLASS hinting
diff --git a/src/common.h b/src/common.h
index c8e0092..8a4208a 100644
--- a/src/common.h
+++ b/src/common.h
@@ -29,6 +29,7 @@
 #include <QColor>
 
 class QMenu;
+class TabbableWidget;
 
 #include "statuspreset.h"
 
@@ -147,6 +148,7 @@ void clearMenu(QMenu *m); // deletes all items, including submenus, from given Q
 void bringToFront(QWidget *w, bool grabFocus = true);
 void replaceWidget(QWidget *, QWidget *);
 void closeDialogs(QWidget *);
+TabbableWidget* findActiveTab();
 #ifdef HAVE_X11
 #include <QWidget>
 #include "x11info.h"
diff --git a/src/psiaccount.cpp b/src/psiaccount.cpp
index 20bd698..8e25780 100644
--- a/src/psiaccount.cpp
+++ b/src/psiaccount.cpp
@@ -3907,7 +3907,7 @@ void PsiAccount::actionHistoryBox(PsiEvent *e)
 	w->show();
 }
 
-void PsiAccount::actionOpenChat(const Jid &j)
+void PsiAccount::actionOpenChat(const Jid &j, const QString & body)
 {
 	UserListItem *u = find(j);
 	if(!u) {
@@ -3941,9 +3941,9 @@ void PsiAccount::actionOpenChat(const Jid &j)
 	}
 
 	if(!res.isEmpty())
-		openChat(j.withResource(res), UserAction);
+		openChat(j.withResource(res), UserAction, body);
 	else
-		openChat(j, UserAction);
+		openChat(j, UserAction, body);
 }
 
 // unlike actionOpenChat(), this function will work on jids that aren't
@@ -4247,11 +4247,11 @@ void PsiAccount::openUri(const QUrl &uriToOpen)
 		QString subject = uri.queryItemValue("subject");
 		QString body = uri.queryItemValue("body");
 		QString type = uri.queryItemValue("type");
-		if (type == "chat" && subject.isEmpty() && body.isEmpty()) {
+		if (type == "chat" && subject.isEmpty()) {
 			if (!find(entity)) {
 				addUserListItem(entity);
 			}
-			actionOpenChat(entity);
+			actionOpenChat(entity, body);
 		} else {
 			dj_newMessage(entity, body, subject, "");
 		}
@@ -5226,7 +5226,7 @@ void PsiAccount::processChats(const Jid &j)
 	processChatsHelper(j, true);
 }
 
-void PsiAccount::openChat(const Jid& j, ActivationType activationType)
+void PsiAccount::openChat(const Jid& j, ActivationType activationType, const QString& body)
 {
 	ChatDlg* chat = ensureChatDlg(j);
 	chat->ensureTabbedCorrectly();
@@ -5242,6 +5242,9 @@ void PsiAccount::openChat(const Jid& j, ActivationType activationType)
 			QTimer::singleShot(1000, dlg, SLOT(showTabWithoutActivation()));
 		}
 	}
+	if (!body.isEmpty()) {
+		chat->setText(body);
+	}
 }
 
 void PsiAccount::chatMessagesRead(const Jid &j)
diff --git a/src/psiaccount.h b/src/psiaccount.h
index 49cf66e..c8fcbcf 100644
--- a/src/psiaccount.h
+++ b/src/psiaccount.h
@@ -329,7 +329,7 @@ public slots:
 	void actionRename(const Jid &, const QString &);
 	void actionGroupRename(const QString &, const QString &);
 	void actionHistory(const Jid &);
-	void actionOpenChat(const Jid &);
+	void actionOpenChat(const Jid &,const QString & body = QString());
 	void actionOpenSavedChat(const Jid &);
 	void actionOpenChat2(const Jid &);
 	void actionOpenChatSpecific(const Jid &);
@@ -504,7 +504,7 @@ private:
 
 	void processChatsHelper(const Jid& jid, bool removeEvents);
 	void processChats(const Jid &);
-	void openChat(const Jid &, ActivationType activationType);
+	void openChat(const Jid &, ActivationType activationType, const QString & body = QString());
 	EventDlg *ensureEventDlg(const Jid &);
 	friend class PsiCon;
 
diff --git a/src/psicon.cpp b/src/psicon.cpp
index db35cd8..0db9e0a 100644
--- a/src/psicon.cpp
+++ b/src/psicon.cpp
@@ -1227,7 +1227,14 @@ void PsiCon::openUri(const QUrl &uri)
 	// authority
 	PsiAccount *pa = 0;
 	//if (uri.authority().isEmpty()) {
-		pa = d->contactList->defaultAccount();
+		TabbableWidget* tw = findActiveTab();
+		if (tw) {
+			pa = tw->account();
+		}
+		if (!pa) {
+			pa = d->contactList->defaultAccount();
+		}
+
 		if (!pa) {
 			QMessageBox::critical(0, tr("Error"), QString("You need to have an account configured and enabled to open URIs (links)."));
 			return;
diff --git a/src/tabs/tabdlg.cpp b/src/tabs/tabdlg.cpp
index 5c5d2d7..cde7dec 100644
--- a/src/tabs/tabdlg.cpp
+++ b/src/tabs/tabdlg.cpp
@@ -812,3 +812,8 @@ void TabDlg::setSimplifiedCaptionEnabled(bool enabled)
 	simplifiedCaption_ = enabled;
 	updateCaption();
 }
+
+TabbableWidget* TabDlg::getCurrentTab() const
+{
+	return dynamic_cast<TabbableWidget*>(tabWidget_->currentPage());
+}
diff --git a/src/tabs/tabdlg.h b/src/tabs/tabdlg.h
index 1704efa..a1e67c7 100644
--- a/src/tabs/tabdlg.h
+++ b/src/tabs/tabdlg.h
@@ -86,6 +86,7 @@ public:
 	void setUserManagementEnabled(bool enabled); // default enabled
 	void setTabBarShownForSingles(bool enabled); // default enabled
 	void setSimplifiedCaptionEnabled(bool enabled); // default disabled
+	TabbableWidget* getCurrentTab() const;
 
 protected:
 	void setShortcuts();
