Entering 'iris'
Entering 'src/libpsi'
Index: psi/src/chatdlg.cpp
===================================================================
--- psi.orig/src/chatdlg.cpp	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/chatdlg.cpp	2013-01-15 14:04:01.460826022 +0400
@@ -1006,6 +1006,12 @@
 	PsiRichText::addEmoticon(chatEdit(), text);
 }

+void ChatDlg::setText(const QString& text)
+{
+	chatEdit()->clear();
+	chatEdit()->insertHtml(text + ' ');
+}
+
 /**
  * Records that the user is composing
  */
Index: psi/src/chatdlg.h
===================================================================
--- psi.orig/src/chatdlg.h	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/chatdlg.h	2013-01-15 14:04:01.464826022 +0400
@@ -72,6 +72,7 @@

 public:
 	PsiAccount* account() const;
+	void setText(const QString& );

 signals:
 	void aInfo(const Jid &);
Index: psi/src/common.cpp
===================================================================
--- psi.orig/src/common.cpp	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/common.cpp	2013-01-15 14:04:01.464826022 +0400
@@ -25,6 +25,7 @@
 #include "psiiconset.h"
 #include "applicationinfo.h"
 #include "psioptions.h"
+#include "tabdlg.h"

 #include <QUrl>
 #include <QProcess>
@@ -380,6 +381,32 @@
 	}
 }

+TabbableWidget* findActiveTab()
+{
+	QWidget* chat = QApplication::activeWindow();
+	TabbableWidget* tw = 0;
+	if(chat) {
+		TabDlg* td = qobject_cast<TabDlg*>(chat);
+		if(td) {
+			tw = td->getCurrentTab();
+		}
+		else {
+			tw = qobject_cast<TabbableWidget*>(chat);
+			if (!tw) {
+				QList<TabDlg*> tmp = chat->findChildren<TabDlg*>(); // all-in-one
+				while(!tmp.isEmpty()) {
+					TabDlg* td = tmp.takeFirst();
+					tw = td->getCurrentTab();
+					if(tw) {
+						break;
+					}
+				}
+			}
+		}
+	}
+	return tw;
+}
+
 #ifdef HAVE_X11
 #include <X11/Xlib.h>
 #include <X11/Xutil.h> // needed for WM_CLASS hinting
Index: psi/src/common.h
===================================================================
--- psi.orig/src/common.h	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/common.h	2013-01-15 14:04:01.464826022 +0400
@@ -29,6 +29,7 @@
 #include <QColor>

 class QMenu;
+class TabbableWidget;

 #include "statuspreset.h"

@@ -147,6 +148,7 @@
 void bringToFront(QWidget *w, bool grabFocus = true);
 void replaceWidget(QWidget *, QWidget *);
 void closeDialogs(QWidget *);
+TabbableWidget* findActiveTab();
 #ifdef HAVE_X11
 #include <QWidget>
 #include "x11info.h"
Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:01.464826022 +0400
@@ -3907,7 +3907,7 @@
 	w->show();
 }

-void PsiAccount::actionOpenChat(const Jid &j)
+void PsiAccount::actionOpenChat(const Jid &j, const QString & body)
 {
 	UserListItem *u = find(j);
 	if(!u) {
@@ -3941,9 +3941,9 @@
 	}

 	if(!res.isEmpty())
-		openChat(j.withResource(res), UserAction);
+		openChat(j.withResource(res), UserAction, body);
 	else
-		openChat(j, UserAction);
+		openChat(j, UserAction, body);
 }

 // unlike actionOpenChat(), this function will work on jids that aren't
@@ -4247,11 +4247,11 @@
 		QString subject = uri.queryItemValue("subject");
 		QString body = uri.queryItemValue("body");
 		QString type = uri.queryItemValue("type");
-		if (type == "chat" && subject.isEmpty() && body.isEmpty()) {
+		if (type == "chat" && subject.isEmpty()) {
 			if (!find(entity)) {
 				addUserListItem(entity);
 			}
-			actionOpenChat(entity);
+			actionOpenChat(entity, body);
 		} else {
 			dj_newMessage(entity, body, subject, "");
 		}
@@ -5226,7 +5226,7 @@
 	processChatsHelper(j, true);
 }

-void PsiAccount::openChat(const Jid& j, ActivationType activationType)
+void PsiAccount::openChat(const Jid& j, ActivationType activationType, const QString& body)
 {
 	ChatDlg* chat = ensureChatDlg(j);
 	chat->ensureTabbedCorrectly();
@@ -5242,6 +5242,9 @@
 			QTimer::singleShot(1000, dlg, SLOT(showTabWithoutActivation()));
 		}
 	}
+	if (!body.isEmpty()) {
+		chat->setText(body);
+	}
 }

 void PsiAccount::chatMessagesRead(const Jid &j)
Index: psi/src/psiaccount.h
===================================================================
--- psi.orig/src/psiaccount.h	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/psiaccount.h	2013-01-15 14:04:01.464826022 +0400
@@ -329,7 +329,7 @@
 	void actionRename(const Jid &, const QString &);
 	void actionGroupRename(const QString &, const QString &);
 	void actionHistory(const Jid &);
-	void actionOpenChat(const Jid &);
+	void actionOpenChat(const Jid &,const QString & body = QString());
 	void actionOpenSavedChat(const Jid &);
 	void actionOpenChat2(const Jid &);
 	void actionOpenChatSpecific(const Jid &);
@@ -504,7 +504,7 @@

 	void processChatsHelper(const Jid& jid, bool removeEvents);
 	void processChats(const Jid &);
-	void openChat(const Jid &, ActivationType activationType);
+	void openChat(const Jid &, ActivationType activationType, const QString & body = QString());
 	EventDlg *ensureEventDlg(const Jid &);
 	friend class PsiCon;

Index: psi/src/psicon.cpp
===================================================================
--- psi.orig/src/psicon.cpp	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/psicon.cpp	2013-01-15 14:04:01.464826022 +0400
@@ -1227,7 +1227,14 @@
 	// authority
 	PsiAccount *pa = 0;
 	//if (uri.authority().isEmpty()) {
-		pa = d->contactList->defaultAccount();
+		TabbableWidget* tw = findActiveTab();
+		if (tw) {
+			pa = tw->account();
+		}
+		if (!pa) {
+			pa = d->contactList->defaultAccount();
+		}
+
 		if (!pa) {
 			QMessageBox::critical(0, tr("Error"), QString("You need to have an account configured and enabled to open URIs (links)."));
 			return;
Index: psi/src/tabs/tabdlg.cpp
===================================================================
--- psi.orig/src/tabs/tabdlg.cpp	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/tabs/tabdlg.cpp	2013-01-15 14:04:01.464826022 +0400
@@ -812,3 +812,8 @@
 	simplifiedCaption_ = enabled;
 	updateCaption();
 }
+
+TabbableWidget* TabDlg::getCurrentTab() const
+{
+	return dynamic_cast<TabbableWidget*>(tabWidget_->currentPage());
+}
Index: psi/src/tabs/tabdlg.h
===================================================================
--- psi.orig/src/tabs/tabdlg.h	2013-01-15 14:04:01.468826022 +0400
+++ psi/src/tabs/tabdlg.h	2013-01-15 14:04:01.464826022 +0400
@@ -86,6 +86,7 @@
 	void setUserManagementEnabled(bool enabled); // default enabled
 	void setTabBarShownForSingles(bool enabled); // default enabled
 	void setSimplifiedCaptionEnabled(bool enabled); // default disabled
+	TabbableWidget* getCurrentTab() const;

 protected:
 	void setShortcuts();
