--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -403,6 +403,8 @@ QLineEdit#le_status_text {
 					<chat>
 						<link-color type="QColor">#0000ff</link-color>
 						<mailto-color type="QColor">#0000ff</mailto-color>
+						<composing-color type="QColor">darkGreen</composing-color>
+						<unread-message-color type="QColor">red</unread-message-color>
 					</chat>
 					<passive-popup>
 						<border type="QColor">#5297f9</border>
--- psi.orig/src/tabs/tabdlg.cpp
+++ psi/src/tabs/tabdlg.cpp
@@ -653,15 +653,15 @@ void TabDlg::updateTab(TabbableWidget* c
 	//now set text colour based upon whether there are new messages/composing etc
 
 	if (chat->state() == TabbableWidget::StateComposing) {
-		tabWidget_->setTabTextColor(chat, Qt::darkGreen);
+		tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.composing-color").value<QColor>());
 		tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/typing")->icon());
 	}
 	else if (chat->unreadMessageCount()) {
-		tabWidget_->setTabTextColor(chat, Qt::red);
+		tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.unread-message-color").value<QColor>());
 		tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/chat")->icon());
 	}
 	else {
-		tabWidget_->setTabTextColor(chat, palette().windowText().color());
+		tabWidget_->setTabTextColor(chat, palette().color(QPalette::Text));
 		tabWidget_->setTabIcon(chat, chat->icon());
 	}
 	updateCaption();
