--- git.orig/src/widgets/psitextview.cpp
+++ git/src/widgets/psitextview.cpp
@@ -70,6 +70,13 @@ public:
 PsiTextView::PsiTextView(QWidget *parent)
 : QTextEdit(parent)
 {
+	QPalette pal = palette(); // copy widget's palette to non const QPalette
+	QColor colorText = pal.color(QPalette::Active, QPalette::HighlightedText);
+	pal.setColor(QPalette::Inactive, QPalette::HighlightedText, colorText);
+	QColor color = pal.color(QPalette::Active, QPalette::Highlight);
+	pal.setColor(QPalette::Inactive, QPalette::Highlight, color);
+	setPalette( pal );        // set the widget's palette
+    
 	d = new Private(this);
 
 	setReadOnly(true);
