Index: psi/options/default.xml
===================================================================
--- psi.orig/options/default.xml	2013-01-15 14:04:02.112826048 +0400
+++ psi/options/default.xml	2013-01-15 14:04:02.108826048 +0400
@@ -352,6 +352,7 @@
 				<recent-jids type="QStringList" />
 			</service-discovery>
 			<tabs>
+				<show-tab-icons type="bool">true</show-tab-icons>
 				<put-tabs-at-bottom type="bool">false</put-tabs-at-bottom>
 				<use-tabs type="bool">false</use-tabs>
 				<size type="QString"></size> <!-- will be invalid when converted to QSize so we can detect first load -->
Index: psi/src/chatdlg.cpp
===================================================================
--- psi.orig/src/chatdlg.cpp	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/chatdlg.cpp	2013-01-15 14:04:02.108826048 +0400
@@ -191,16 +191,6 @@
 	connect(act_scrolldown_, SIGNAL(triggered()), SLOT(scrollDown()));
 }

-void ChatDlg::ensureTabbedCorrectly()
-{
-	TabbableWidget::ensureTabbedCorrectly();
-	setShortcuts();
-	if(!isTabbed() && geometryOptionPath().isEmpty()) {
-		setGeometryOptionPath(geometryOption);
-	}
-}
-
-
 void ChatDlg::setShortcuts()
 {
 	act_send_->setShortcuts(ShortcutManager::instance()->shortcuts("chat.send"));
@@ -436,6 +426,18 @@
 	return u;
 }

+void ChatDlg::ensureTabbedCorrectly()
+{
+	TabbableWidget::ensureTabbedCorrectly();
+	setShortcuts();
+	QList<UserListItem*> ul = account()->findRelevant(jid());
+	UserStatus userStatus = userStatusFor(jid(), ul, false);
+	setTabIcon(PsiIconset::instance()->statusPtr(jid(), userStatus.statusType)->icon());
+	if(!isTabbed() && geometryOptionPath().isEmpty()) {
+		setGeometryOptionPath(geometryOption);
+	}
+}
+
 void ChatDlg::updateContact(const Jid &j, bool fromPresence)
 {
 	// if groupchat, only update if the resource matches
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:02.108826048 +0400
@@ -762,6 +762,13 @@
 	QList<int> tmp = ui_.hsplitter->sizes();
 	ui_.hsplitter->setSizes(QList<int>() << 0);
 	ui_.hsplitter->setSizes(tmp);
+	UserListItem* u = account()->find(jid().bare());
+	if(u) {
+		setTabIcon(PsiIconset::instance()->statusPtr(u)->icon());
+	}
+	else {
+		setStatusTabIcon((d->state == Private::Connected ? STATUS_ONLINE : STATUS_OFFLINE));
+	}
 	if(!isTabbed() && geometryOptionPath().isEmpty()) {
 		setGeometryOptionPath(geometryOption);
 	}
@@ -1056,6 +1063,7 @@
 	if(d->state != Private::Idle && d->state != Private::ForcedLeave) {
 		d->state = Private::Idle;
 		ui_.pb_topic->setEnabled(false);
+		setStatusTabIcon(STATUS_OFFLINE);
 		appendSysMsg(tr("Disconnected."), true);
 		ui_.mle->chatEdit()->setEnabled(false);
 	}
@@ -1143,6 +1151,7 @@
 void GCMainDlg::error(int, const QString &str)
 {
 	ui_.pb_topic->setEnabled(false);
+	setStatusTabIcon(STATUS_ERROR);

 	if(d->state == Private::Connecting)
 		appendSysMsg(tr("Unable to join groupchat.	Reason: %1").arg(str), true);
@@ -1196,6 +1205,8 @@
 	}

 	if (nick == d->self) {
+		if(s.isAvailable())
+			setStatusTabIcon(s.type());
 		// Update configuration dialog
 		if (d->configDlg) {
 			d->configDlg->setRoleAffiliation(s.mucItem().role(),s.mucItem().affiliation());
@@ -1469,6 +1480,7 @@
 		ui_.lv_users->clear();
 		d->state = Private::Connected;
 		ui_.pb_topic->setEnabled(true);
+		setStatusTabIcon(STATUS_ONLINE);
 		ui_.mle->chatEdit()->setEnabled(true);
 		setConnecting();
 		appendSysMsg(tr("Connected."), true);
@@ -1772,6 +1784,11 @@
 	return d->pending;
 }

+void GCMainDlg::setStatusTabIcon(int status)
+{
+	setTabIcon(PsiIconset::instance()->statusPtr(jid(), status)->icon());
+}
+
 //----------------------------------------------------------------------------
 // GCFindDlg
 //----------------------------------------------------------------------------
Index: psi/src/groupchatdlg.h
===================================================================
--- psi.orig/src/groupchatdlg.h	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/groupchatdlg.h	2013-01-15 14:04:02.108826048 +0400
@@ -92,6 +92,7 @@
 	void dropEvent(QDropEvent *);
 	void closeEvent(QCloseEvent *);
 	void mucInfoDialog(const QString& title, const QString& message, const Jid& actor, const QString& reason);
+	void setStatusTabIcon(int status);

 signals:
 	void aSend(const Message &);
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:02.108826048 +0400
@@ -180,6 +180,7 @@

 	PsiToolTip::install(ui_.lb_status);
 	ui_.lb_status->setPsiIcon(IconsetFactory::iconPtr("status/noauth"));
+	setTabIcon(IconsetFactory::iconPtr("status/noauth")->icon());

 	ui_.tb_emoticons->setIcon(IconsetFactory::icon("psi/smile").icon());

@@ -393,9 +394,11 @@

 	if (status == -1 || !u) {
 		ui_.lb_status->setPsiIcon(IconsetFactory::iconPtr("status/noauth"));
+		setTabIcon(IconsetFactory::iconPtr("status/noauth")->icon());
 	}
 	else {
 		ui_.lb_status->setPsiIcon(PsiIconset::instance()->statusPtr(jid(), status));
+		setTabIcon(PsiIconset::instance()->statusPtr(jid(), status)->icon());
 	}

 	if (u) {
Index: psi/src/tabs/tabbablewidget.cpp
===================================================================
--- psi.orig/src/tabs/tabbablewidget.cpp	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/tabs/tabbablewidget.cpp	2013-01-15 14:04:02.108826048 +0400
@@ -209,3 +209,20 @@
 		}
 	}
 }
+
+/**
+ * Set the icon of the tab.
+ */
+void TabbableWidget::setTabIcon(const QIcon &icon)
+{
+	icon_ = icon;
+	TabDlg* tabDlg = tabManager_->getManagingTabs(this);
+	if (tabDlg) {
+		tabDlg->setTabIcon(this, icon);
+	}
+}
+
+const QIcon &TabbableWidget::icon() const
+{
+	return icon_;
+}
Index: psi/src/tabs/tabbablewidget.h
===================================================================
--- psi.orig/src/tabs/tabbablewidget.h	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/tabs/tabbablewidget.h	2013-01-15 14:04:02.108826048 +0400
@@ -21,6 +21,7 @@
 #ifndef TABBABLE_H
 #define TABBABLE_H

+#include <QIcon>
 #include "advwidget.h"
 #include "im.h" // ChatState

@@ -42,6 +43,8 @@
 	~TabbableWidget();

 	PsiAccount* account() const;
+	void setTabIcon(const QIcon &);
+	const QIcon &icon() const;

 	virtual Jid jid() const;
 	virtual const QString & getDisplayName();
@@ -86,6 +89,7 @@
 	Jid jid_;
 	PsiAccount *pa_;
 	TabManager *tabManager_;
+	QIcon icon_;
 };

 #endif
Index: psi/src/tabs/tabdlg.cpp
===================================================================
--- psi.orig/src/tabs/tabdlg.cpp	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/tabs/tabdlg.cpp	2013-01-15 14:04:02.108826048 +0400
@@ -398,7 +398,7 @@
 {
 	setUpdatesEnabled(false);
 	tabs_.append(tab);
-	tabWidget_->addTab(tab, captionForTab(tab));
+	tabWidget_->addTab(tab, captionForTab(tab), tab->icon());

 	connect(tab, SIGNAL(invalidateTabInfo()), SLOT(updateTab()));
 	connect(tab, SIGNAL(updateFlashState()), SLOT(updateFlashState()));
@@ -593,12 +593,15 @@

 	if (chat->state() == TabbableWidget::StateComposing) {
 		tabWidget_->setTabTextColor(chat, Qt::darkGreen);
+		tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/typing")->icon());
 	}
 	else if (chat->unreadMessageCount()) {
 		tabWidget_->setTabTextColor(chat, Qt::red);
+		tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/chat")->icon());
 	}
 	else {
 		tabWidget_->setTabTextColor(chat, palette().windowText().color());
+		tabWidget_->setTabIcon(chat, chat->icon());
 	}
 	updateCaption();
 }
@@ -817,3 +820,11 @@
 {
 	return dynamic_cast<TabbableWidget*>(tabWidget_->currentPage());
 }
+
+/**
+ * Set the icon of the tab.
+ */
+void TabDlg::setTabIcon(QWidget *widget,const QIcon &icon)
+{
+	tabWidget_->setTabIcon(widget, icon);
+}
Index: psi/src/tabs/tabdlg.h
===================================================================
--- psi.orig/src/tabs/tabdlg.h	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/tabs/tabdlg.h	2013-01-15 14:04:02.108826048 +0400
@@ -87,6 +87,7 @@
 	void setTabBarShownForSingles(bool enabled); // default enabled
 	void setSimplifiedCaptionEnabled(bool enabled); // default disabled
 	TabbableWidget* getCurrentTab() const;
+	void setTabIcon(QWidget *,const QIcon &);

 protected:
 	void setShortcuts();
Index: psi/src/widgets/psitabwidget.cpp
===================================================================
--- psi.orig/src/widgets/psitabwidget.cpp	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/widgets/psitabwidget.cpp	2013-01-15 14:04:02.108826048 +0400
@@ -21,6 +21,7 @@
 #include "psitabwidget.h"
 #include "psitabbar.h"
 #include "common.h"
+#include "psioptions.h"
 #include <QWidget>
 #include <QVBoxLayout>
 #include <QHBoxLayout>
@@ -155,14 +156,18 @@
 /**
  * Add the Widget to the tab stack.
  */
-void PsiTabWidget::addTab(QWidget *widget, QString name) {
+void PsiTabWidget::addTab(QWidget *widget, QString name, const QIcon &icon)
+{
 	Q_ASSERT(widget);
 	if (widgets_.contains(widget)) {
 		return;
 	}
 	widgets_.append(widget);
 	stacked_->addWidget(widget);
-	tabBar_->addTab(name);
+	if (PsiOptions::instance()->getOption("options.ui.tabs.show-tab-icons").toBool())
+		tabBar_->addTab(icon, name);
+	else
+		tabBar_->addTab(name);
 	showPage(currentPage());
 }

@@ -230,6 +235,18 @@
 	tabBar_->setTabText(index, label);
 }

+/**
+ * Set the icon of the tab.
+ */
+void PsiTabWidget::setTabIcon(QWidget *widget, const QIcon &icon)
+{
+	int index = getIndex(widget);
+	if (index == -1 || !PsiOptions::instance()->getOption("options.ui.tabs.show-tab-icons").toBool()) {
+		return;
+	}
+	tabBar_->setTabIcon(index, icon);
+}
+
 void PsiTabWidget::setCurrentPage(int index) {
 	if (index >= 0 && index < count()) {
 		showPage(widgets_[index]);
Index: psi/src/widgets/psitabwidget.h
===================================================================
--- psi.orig/src/widgets/psitabwidget.h	2013-01-15 14:04:02.112826048 +0400
+++ psi/src/widgets/psitabwidget.h	2013-01-15 14:04:02.112826048 +0400
@@ -48,7 +48,7 @@
 	QWidget *currentPage();
 	int currentPageIndex();
 	QWidget *widget(int index);
-	void addTab(QWidget *, QString);
+	void addTab(QWidget *, QString, const QIcon &icon = QIcon());
 	void showPage(QWidget *);
 	void showPageDirectly(QWidget *);

@@ -56,6 +56,7 @@
 	QWidget* page(int index);
 	int getIndex(QWidget *);
 	void setTabText(QWidget *, const QString &);
+	void setTabIcon(QWidget *, const QIcon &);
 	void setTabPosition(QTabWidget::TabPosition pos);
 	void setCloseIcon(const QIcon &);

