--- git.temp.orig/options/default.xml
+++ git.temp/options/default.xml
@@ -138,6 +138,7 @@
 				<always-on-top type="bool">false</always-on-top>
 				<automatically-resize-roster type="bool">false</automatically-resize-roster>
 				<grow-roster-upwards type="bool">true</grow-roster-upwards>
+				<disable-scrollbar type="bool">true</disable-scrollbar>
 				<contact-sort-style type="QString">status</contact-sort-style>
 				<disable-service-discovery type="bool">false</disable-service-discovery>
 				<ensure-contact-visible-on-event type="bool">true</ensure-contact-visible-on-event>
--- git.temp.orig/src/groupchatdlg.cpp
+++ git.temp/src/groupchatdlg.cpp
@@ -638,6 +638,9 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 
 	ui_.lv_users->setMainDlg(this);
 	ui_.lv_users->model()->setSupportedDragActions(Qt::CopyAction);
+	if ( PsiOptions::instance()->getOption("options.ui.contactlist.disable-scrollbar").toBool() ) {
+		ui_.lv_users->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+	}
 	connect(ui_.lv_users, SIGNAL(action(const QString &, const Status &, int)), SLOT(lv_action(const QString &, const Status &, int)));
 	connect(ui_.lv_users, SIGNAL(insertNick(const QString&)), d, SLOT(insertNick(const QString&)));
 
--- git.temp.orig/src/psirosterwidget.cpp
+++ git.temp/src/psirosterwidget.cpp
@@ -48,6 +48,7 @@ static const QString showAgentsOptionPat
 static const QString showSelfOptionPath = "options.ui.contactlist.show.self-contact";
 static const QString showStatusMessagesOptionPath = "options.ui.contactlist.status-messages.show";
 static const QString allowAutoResizeOptionPath = "options.ui.contactlist.automatically-resize-roster";
+static const QString showScrollBarOptionPath = "options.ui.contactlist.disable-scrollbar";
 
 //----------------------------------------------------------------------------
 // PsiRosterFilterProxyModel
@@ -131,6 +132,9 @@ PsiRosterWidget::PsiRosterWidget(QWidget
 	contactListPageView_ = new PsiContactListView(contactListPage_);
 	contactListPageView_->installEventFilter(this);
 	contactListPageView_->setObjectName("contactListView");
+	if ( PsiOptions::instance()->getOption(showScrollBarOptionPath).toBool() ) {
+		contactListPageView_->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+	}
 	contactListPageLayout->addWidget(contactListPageView_);
 
 	// filterPage_
@@ -235,6 +239,11 @@ void PsiRosterWidget::optionChanged(cons
 	else if (option == allowAutoResizeOptionPath) {
 		contactListPageView_->setAutoResizeEnabled(PsiOptions::instance()->getOption(allowAutoResizeOptionPath).toBool());
 	}
+	else if (option == showScrollBarOptionPath) {
+		contactListPageView_->setVerticalScrollBarPolicy(
+			PsiOptions::instance()->getOption(showScrollBarOptionPath).toBool()?
+			Qt::ScrollBarAlwaysOff : Qt::ScrollBarAsNeeded);
+	}
 }
 
 void PsiRosterWidget::removeSelection(QMimeData* selection)
