--- a/src/chatdlg.cpp
+++ b/src/chatdlg.cpp
@@ -1228,15 +1228,17 @@ void ChatDlg::chatEditCreated()
 
 TabbableWidget::State ChatDlg::state() const
 {
-    switch (contactChatState_) {
-        case XMPP::StateComposing:
-            return TabbableWidget::State::Composing;
-        case XMPP::StateInactive:
-        case XMPP::StateGone:
-            return TabbableWidget::State::Inactive;
-        default:
-            return TabbableWidget::State::None;
+    TabbableWidget::State state = TabbableWidget::State::None;
+    if(contactChatState_ == XMPP::StateComposing) {
+        state = TabbableWidget::State::Composing;
     }
+    else if(unreadMessageCount()) {
+        state = TabbableWidget::State::Highlighted;
+    } else if (contactChatState_ == XMPP::StateInactive || contactChatState_ == XMPP::StateGone) {
+        state = TabbableWidget::State::Inactive;
+    }
+
+    return state;
 }
 
 int ChatDlg::unreadMessageCount() const
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -2552,7 +2552,7 @@ void GCMainDlg::chatEditCreated()
 
 TabbableWidget::State GCMainDlg::state() const
 {
-    return TabbableWidget::State::None;
+    return d->hPending ? TabbableWidget::State::Highlighted : TabbableWidget::State::None;
 }
 
 int GCMainDlg::unreadMessageCount() const
--- a/src/tabs/tabbablewidget.h
+++ b/src/tabs/tabbablewidget.h
@@ -65,7 +65,8 @@ public:
     enum class State : char {
         None = 0,
         Composing,
-        Inactive
+        Inactive,
+        Highlighted
     };
     virtual State state() const = 0;
     virtual int unreadMessageCount() const = 0;
--- a/src/tabs/tabdlg.cpp
+++ b/src/tabs/tabdlg.cpp
@@ -656,21 +656,27 @@ void TabDlg::updateTab(TabbableWidget* chat)
     tabWidget_->setTabText(chat, captionForTab(chat));
     //now set text colour based upon whether there are new messages/composing etc
 
-    if (chat->state() == TabbableWidget::State::Composing) {
+    TabbableWidget::State state = chat->state();
+    if (state == TabbableWidget::State::Composing) {
         tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.composing-color").value<QColor>());
         tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/typing")->icon());
     }
-    else if (chat->unreadMessageCount()) {
-        tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.unread-message-color").value<QColor>());
-        tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/chat")->icon());
-    }
-    else if (chat->state() == TabbableWidget::State::Inactive) {
-        tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.inactive-color").value<QColor>());
-        tabWidget_->setTabIcon(chat, chat->icon());
-    }
     else {
-        tabWidget_->setTabTextColor(chat, palette().color(QPalette::Text));
-        tabWidget_->setTabIcon(chat, chat->icon());
+        if (state == TabbableWidget::State::Highlighted) {
+            tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.unread-message-color").value<QColor>());
+        }
+        else if (state == TabbableWidget::State::Inactive) {
+            tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.inactive-color").value<QColor>());
+        } else {
+            tabWidget_->setTabTextColor(chat, palette().color(QPalette::Text));
+        }
+
+        if(chat->unreadMessageCount()) {
+            tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/chat")->icon());
+        }
+        else {
+            tabWidget_->setTabIcon(chat, chat->icon());
+        }
     }
     updateCaption();
 }
