Index: psi/src/mainwin_p.cpp
===================================================================
--- psi.orig/src/mainwin_p.cpp	2013-01-15 14:04:02.956826083 +0400
+++ psi/src/mainwin_p.cpp	2013-01-15 14:04:02.952826083 +0400
@@ -127,9 +127,9 @@
 	if(qVersionInt() >= 0x040300)
 	{
 		if((showText && !label.isEmpty()) && styleSheet().isEmpty()) {
-			setStyleSheet("text-align: left");
+			setStyleSheet("text-align: center");
 		}
-		else if((!showText || label.isEmpty()) && styleSheet() == "text-align: left") {
+		else if((!showText || label.isEmpty()) && styleSheet() == "text-align: center") {
 			setStyleSheet(QString());
 		}
 	}
