Entering 'iris'
Entering 'src/libpsi'
Index: psi/options/default.xml
===================================================================
--- psi.orig/options/default.xml	2013-01-15 14:04:04.604826150 +0400
+++ psi/options/default.xml	2013-01-15 14:04:04.600826150 +0400
@@ -219,7 +219,7 @@
 					<width>580</width>
 					<height>420</height>
 				</size>
-				<use-highlighting type="bool">true</use-highlighting><use-nick-coloring type="bool">true</use-nick-coloring>
+				<use-highlighting type="bool">true</use-highlighting><use-nick-coloring type="bool">true</use-nick-coloring><use-hash-nick-coloring type="bool">true</use-hash-nick-coloring><colored-history type="bool">true</colored-history>
 			</muc>
 			<show-deprecated comment="Deprecated functionality or protocols">
 				<service-discovery comment="Service discovery dialog">
Index: psi/src/chatview_te.cpp
===================================================================
--- psi.orig/src/chatview_te.cpp	2013-01-15 14:04:04.604826150 +0400
+++ psi/src/chatview_te.cpp	2013-01-15 14:04:04.600826150 +0400
@@ -290,7 +290,7 @@
 		alerttagsc = "</b>";
 	}

-	if(mv.isSpooled()) {
+	if(mv.isSpooled() && !PsiOptions::instance()->getOption("options.ui.muc.colored-history").toBool()) {
 		nickcolor = ColorOpt::instance()->color(infomrationalColorOpt).name();
 	} else {
 		nickcolor = getMucNickColor(mv.nick(), mv.isLocal());
Index: psi/src/chatviewcommon.cpp
===================================================================
--- psi.orig/src/chatviewcommon.cpp	2013-01-15 14:04:04.604826150 +0400
+++ psi/src/chatviewcommon.cpp	2013-01-15 14:04:04.600826150 +0400
@@ -19,6 +19,9 @@
  */
 
 #include <QWidget>
+#include <QColor>
+#include <QRegExp>
+#include <QApplication>
 
 #include "chatviewcommon.h"
 #include "psioptions.h"
@@ -42,30 +45,60 @@
 
 QString ChatViewCommon::getMucNickColor(const QString &nick, bool isSelf, QStringList validList)
 {
+	const char* rgbBlack = "#000000";
+	QString nickwoun = nick; // nick without underscores
+	nickwoun.replace(QRegExp("^_*"), "");
+	nickwoun.replace(QRegExp("_*$"), "");
+
 	int sender;
-	if(isSelf || nick.isEmpty()) {
+	if(isSelf || nickwoun.isEmpty()) {
 		sender = -1;
 	} else {
-		if (!_nicks.contains(nick)) {
+		if (!_nicks.contains(nickwoun)) {
 			//not found in map
-			_nicks.insert(nick,_nickNumber);
+			_nicks.insert(nickwoun,_nickNumber);
 			_nickNumber++;
 		}
-		sender=_nicks[nick];
+		sender=_nicks[nickwoun];
 	}
 
 	QStringList nickColors = validList.isEmpty()
 		? PsiOptions::instance()->getOption("options.ui.look.colors.muc.nick-colors").toStringList()
 		: validList;
 
-	if(!PsiOptions::instance()->getOption("options.ui.muc.use-nick-coloring").toBool() || nickColors.empty()) {
-		return "#000000";
-	}
-	else if(sender == -1 || nickColors.size() == 1) {
-		return nickColors[nickColors.size()-1];
-	}
-	else {
-		int n = sender % (nickColors.size()-1);
-		return nickColors[n];
+	if(!PsiOptions::instance()->getOption("options.ui.muc.use-nick-coloring").toBool()) {
+		return rgbBlack;
+	} else {
+		if (PsiOptions::instance()->getOption("options.ui.muc.use-hash-nick-coloring").toBool()) {
+			/* Hash-driven colors */
+			Q_ASSERT(nickwoun.size());
+			quint32 hash = qHash(nickwoun); // main hash
+			quint32 num = hash + nickwoun.size(); // additional hash which differs from main hash
+			QColor bg = QApplication::palette().color(QPalette::Base); // background color
+			int bgH = bg.hue() >= 0? bg.hue() : 60; // 60 == yellow
+			int bgV = bg.lightness() >= 0? bg.lightness() : 255; // 255 == white
+			int bgA = bg.alpha() >= 0? bg.alpha() : 255; // 255 == fully opaque color
+			int t = 300*(hash%30); // scale to 0-300 range with limited palette
+			int h = bgH + t + 30 ; // do not use colors close to background color
+			while (h > 359) h -= 360; // use only 0-359 range
+			int v = bgV > 127?
+						25  + 25*(num%4): // for bright themes
+						230 - 25*(num%4); // for dark themes
+			int a = bgA; // use the same transparency as in background color
+			int s = 255; // use only clear colors
+			QColor precolor = QColor::fromHsv(h, s, v, a);
+			return precolor.name();
+		} else {
+			/* Colors from list */
+			if (nickColors.empty()) {
+				return rgbBlack;
+			}
+			if(sender == -1 || nickColors.size() == 1) {
+				return nickColors[nickColors.size()-1];
+			} else {
+				int n = sender % (nickColors.size()-1);
+				return nickColors[n];
+			}
+		}
 	}
 }
Index: psi/src/options/opt_general_groupchat.ui
===================================================================
--- psi.orig/src/options/opt_general_groupchat.ui	2013-01-15 14:04:04.604826150 +0400
+++ psi/src/options/opt_general_groupchat.ui	2013-01-15 14:04:04.604826150 +0400
@@ -157,6 +157,13 @@
        </property>
       </widget>
      </item>
+      <item>
+      <widget class="QCheckBox" name="ck_gcHashNickColoring" >
+       <property name="text" >
+        <string>Enable hash nick coloring</string>
+       </property>
+      </widget>
+     </item>
      <item>
       <widget class="QGroupBox" name="ncGroupBox" >
        <property name="title" >
@@ -296,11 +303,30 @@
   <tabstop>pb_removeHighlightWord</tabstop>
   <tabstop>pb_addHighlightWord</tabstop>
   <tabstop>ck_gcNickColoring</tabstop>
+  <tabstop>ck_gcHashNickColoring</tabstop>
   <tabstop>le_newNickColor</tabstop>
   <tabstop>pb_nickColor</tabstop>
   <tabstop>pb_removeNickColor</tabstop>
   <tabstop>pb_addNickColor</tabstop>
  </tabstops>
  <resources/>
- <connections/>
-</ui>
+ <connections>
+  <connection>
+   <sender>ck_gcNickColoring</sender>
+   <signal>toggled(bool)</signal>
+   <receiver>ck_gcHashNickColoring</receiver>
+   <slot>setEnabled(bool)</slot>
+   <hints>
+    <hint type="sourcelabel">
+     <x>326</x>
+     <y>18</y>
+    </hint>
+    <hint type="destinationlabel">
+     <x>326</x>
+     <y>42</y>
+    </hint>
+   </hints>
+  </connection>
+ </connections>
+ </ui>
+
Index: psi/src/options/opt_groupchat.cpp
===================================================================
--- psi.orig/src/options/opt_groupchat.cpp	2013-01-15 14:04:04.604826150 +0400
+++ psi/src/options/opt_groupchat.cpp	2013-01-15 14:04:04.604826150 +0400
@@ -67,6 +67,7 @@
 	GeneralGroupchatUI *d = (GeneralGroupchatUI *)w;
 	PsiOptions::instance()->setOption("options.ui.muc.use-highlighting", d->ck_gcHighlights->isChecked());
 	PsiOptions::instance()->setOption("options.ui.muc.use-nick-coloring", d->ck_gcNickColoring->isChecked());
+	PsiOptions::instance()->setOption("options.ui.muc.use-hash-nick-coloring", d->ck_gcHashNickColoring->isChecked());

 	QStringList highlight;
 	int i;
@@ -96,6 +97,8 @@
 	d->ck_gcHighlights->setChecked( PsiOptions::instance()->getOption("options.ui.muc.use-highlighting").toBool() );
 	d->ck_gcNickColoring->setChecked( true );
 	d->ck_gcNickColoring->setChecked( PsiOptions::instance()->getOption("options.ui.muc.use-nick-coloring").toBool() );
+	d->ck_gcHashNickColoring->setChecked( true );
+	d->ck_gcHashNickColoring->setChecked( PsiOptions::instance()->getOption("options.ui.muc.use-hash-nick-coloring").toBool() );
 	d->lw_highlightWords->clear();
 	d->lw_highlightWords->addItems( PsiOptions::instance()->getOption("options.ui.muc.highlight-words").toStringList() );
 	d->lw_nickColors->clear();
