--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -112,6 +112,7 @@
 				<auto-popup type="bool">false</auto-popup>
 				<auto-capitalize type="bool">false</auto-capitalize>
 				<auto-scroll-to-bottom type="bool">true</auto-scroll-to-bottom>
+				<caption type="QString"></caption>
 				<delete-contents-after type="QString">never</delete-contents-after>
 				<raise-chat-windows-on-new-messages type="bool">false</raise-chat-windows-on-new-messages>
 				<switch-tab-on-new-messages type="bool">false</switch-tab-on-new-messages>
--- psi.orig/src/chatdlg.cpp
+++ psi/src/chatdlg.cpp
@@ -637,7 +637,10 @@ void ChatDlg::setSelfDestruct(int minute
 
 QString ChatDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 
 	if (pending_ > 0) {
 		cap += "* ";
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -1941,7 +1941,10 @@ void GCMainDlg::doAlert()
 
 QString GCMainDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 
 	if (d->pending > 0) {
 		cap += "* ";
--- psi.orig/src/tabs/tabdlg.cpp
+++ psi/src/tabs/tabdlg.cpp
@@ -558,7 +558,10 @@ void TabDlg::activated()
 
 QString TabDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 	uint pending = 0;
 	foreach(TabbableWidget* tab, tabs_) {
 		pending += tab->unreadMessageCount();
