--- git.orig/options/default.xml
+++ git/options/default.xml
@@ -288,6 +288,7 @@
 						<sent type="QColor" comment="Color used to indicate sent messages.">#ff0000</sent>
 						<informational type="QColor" comment="Color used to indicate informational (status change, spooled) messages.">#008000</informational>
 						<usertext type="QColor" comment="Color used to indicate additional text for informational messages.">#606060</usertext>
+						<highlighting type="QColor">#FF0000</highlighting>
 					</messages>
 					<passive-popup>
 						<border type="QColor">#5297f9</border>
--- git.orig/src/chatview_te.cpp
+++ git/src/chatview_te.cpp
@@ -285,7 +285,7 @@ void ChatView::renderMucMessage(const Me
 	QString icon = useMessageIcons_?
 					(QString("<img src=\"%1\" />").arg(mv.isLocal()?"icon:log_icon_delivered":"icon:log_icon_receive")):"";
 	if(mv.isAlert()) {
-		textcolor = "#FF0000";
+		textcolor = PsiOptions::instance()->getOption("options.ui.look.colors.messages.highlighting").value<QColor>().name();
 		alerttagso = "<b>";
 		alerttagsc = "</b>";
 	}
--- git.orig/themes/chatview/psi/classic/index.html
+++ git/themes/chatview/psi/classic/index.html
@@ -15,7 +15,7 @@ body > div > img:first-child { vertical-
 .received {}
 .infmsg {}
 .usertext {}
-.alert {font-weight:bold; color:red;}
+.alert {font-weight:bold;}
 .trackbar { height:1px; border:1px solid black; border-color:#bbf #66f #66f #bbf }
 .nicklink { text-decoration:none; color:inherit; }
   </style>
@@ -35,6 +35,7 @@ window[chatServer.jsNamespace()].theme =
 	var cssInfMsg = chat.util.findStyleSheet(document.styleSheets[0], ".infmsg").style;
 	var cssUserText = chat.util.findStyleSheet(document.styleSheets[0], ".usertext").style;
 	var cssChatSays = chat.util.findStyleSheet(document.styleSheets[0], ".msg>span:first").style;
+	var cssAlertMsg = chat.util.findStyleSheet(document.styleSheets[0], ".alert").style;
 	var useMessageIcons, messageIconsSize = 0;
 
 	var applyPsiSettings = function() {
@@ -45,6 +46,7 @@ window[chatServer.jsNamespace()].theme =
 		cssReceivedMsg.color = shared.colorOption("options.ui.look.colors.messages.received");
 		cssInfMsg.color = shared.colorOption("options.ui.look.colors.messages.informational");
 		cssUserText.color = shared.colorOption("options.ui.look.colors.messages.usertext");
+		cssAlertMsg.color = shared.psiOption("options.ui.look.colors.messages.highlighting");
 		useMessageIcons = shared.psiOption("options.ui.chat.use-message-icons");
 		if (shared.psiOption("options.ui.chat.scaled-message-icons")) {
 			var d = document.createElement("div");
--- a/src/options/opt_appearance.cpp
+++ b/src/options/opt_appearance.cpp
@@ -248,8 +248,8 @@ QWidget *OptionsTabAppearanceGeneral::widget()
 		{d->ck_cMessageSent,     d->pb_cMessageSent,     "messages.sent", ""},
 		{d->ck_cMessageReceived, d->pb_cMessageReceived, "messages.received", ""},
 		{d->ck_cSysMsg,          d->pb_cSysMsg,          "messages.informational", ""},
-		{d->ck_cUserText,        d->pb_cUserText,        "messages.usertext", ""}
-
+		{d->ck_cUserText,        d->pb_cUserText,        "messages.usertext", ""},
+		{d->ck_highlight,        d->pb_highlight,        "messages.highlighting", ""}
 	};
 
 	bg_color = new QButtonGroup(this);
--- a/src/options/opt_appearance.ui
+++ b/src/options/opt_appearance.ui
@@ -455,6 +455,32 @@
             </property>
            </widget>
           </item>
+          <item row="16" column="0">
+           <widget class="QCheckBox" name="ck_highlight">
+            <property name="text">
+             <string>Highlight:</string>
+            </property>
+           </widget>
+          </item>
+          <item row="16" column="1">
+           <widget class="QToolButton" name="pb_highlight">
+            <property name="sizePolicy">
+             <sizepolicy hsizetype="Fixed" vsizetype="Fixed">
+              <horstretch>0</horstretch>
+              <verstretch>0</verstretch>
+             </sizepolicy>
+            </property>
+            <property name="maximumSize">
+             <size>
+              <width>20</width>
+              <height>20</height>
+             </size>
+            </property>
+            <property name="text">
+             <string/>
+            </property>
+           </widget>
+          </item>
           <item row="9" column="0">
            <widget class="QCheckBox" name="ck_cUserText">
             <property name="whatsThis">
