Entering 'iris'
Entering 'src/libpsi'
diff --git a/options/default.xml b/options/default.xml
index 389d7c7..1f0a50b 100644
--- a/options/default.xml
+++ b/options/default.xml
@@ -376,6 +376,7 @@
 				<group-state comment="Saved state data of the tabsets defined by options.ui.tabs.grouping"/>
 				<tab-singles type="QString" comment="Tab types that would have been untabbed are given their own tabset. 'C' for chat and 'M' for mucs"/>
 				<use-tab-shortcuts type="bool">true</use-tab-shortcuts>
+				<show-tab-buttons type="bool">true</show-tab-buttons>
 			</tabs>
 		</ui>
 		<shortcuts comment="Shortcuts">
diff --git a/src/options/opt_chat.cpp b/src/options/opt_chat.cpp
index bbad4b1..e791b28 100644
--- a/src/options/opt_chat.cpp
+++ b/src/options/opt_chat.cpp
@@ -103,6 +103,7 @@ void OptionsTabChat::applyOptions()
 	PsiOptions::instance()->setOption("options.ui.chat.alert-for-already-open-chats", d->ck_alertOpenChats->isChecked());
 	PsiOptions::instance()->setOption("options.ui.chat.raise-chat-windows-on-new-messages", d->ck_raiseChatWindow->isChecked());
 	PsiOptions::instance()->setOption("options.ui.chat.use-small-chats", d->ck_smallChats->isChecked());
+	PsiOptions::instance()->setOption("options.ui.tabs.show-tab-buttons", d->ck_showTabButtons->isChecked());
 	
 	QString delafter;
 	switch (bg_delChats->buttons().indexOf( bg_delChats->checkedButton() )) {
@@ -179,6 +180,7 @@ void OptionsTabChat::restoreOptions()
 	d->ck_alertOpenChats->setChecked( PsiOptions::instance()->getOption("options.ui.chat.alert-for-already-open-chats").toBool() );
 	d->ck_raiseChatWindow->setChecked( PsiOptions::instance()->getOption("options.ui.chat.raise-chat-windows-on-new-messages").toBool() );
 	d->ck_smallChats->setChecked( PsiOptions::instance()->getOption("options.ui.chat.use-small-chats").toBool() );
+	d->ck_showTabButtons->setChecked( PsiOptions::instance()->getOption("options.ui.tabs.show-tab-buttons").toBool() );
 	d->ck_tabChats->setChecked( PsiOptions::instance()->getOption("options.ui.tabs.use-tabs").toBool() );
 	d->cb_tabGrouping->setEnabled(PsiOptions::instance()->getOption("options.ui.tabs.use-tabs").toBool());
 	QString tabGrouping = PsiOptions::instance()->getOption("options.ui.tabs.grouping").toString();
diff --git a/src/options/opt_chat.ui b/src/options/opt_chat.ui
index 571be2d..36e7ab8 100644
--- a/src/options/opt_chat.ui
+++ b/src/options/opt_chat.ui
@@ -120,6 +120,13 @@
     </layout>
    </item>
    <item>
+    <widget class="QCheckBox" name="ck_showTabButtons" >
+     <property name="text" >
+      <string>Show tab buttons</string>
+     </property>
+    </widget>
+   </item>
+   <item>
     <widget class="QCheckBox" name="ck_autoResize" >
      <property name="text" >
       <string>Auto-resize text input field</string>
diff --git a/src/psicon.cpp b/src/psicon.cpp
index e2688e4..b24db10 100644
--- a/src/psicon.cpp
+++ b/src/psicon.cpp
@@ -1318,7 +1318,8 @@ void PsiCon::optionChanged(const QString& option)
 	}
 
 	if (option == "options.ui.tabs.use-tabs" || 
-		option == "options.ui.tabs.grouping") {
+		option == "options.ui.tabs.grouping" || 
+		option == "options.ui.tabs.show-tab-buttons") {
 		QMessageBox::information(0, tr("Information"), tr("Some of the options you changed will only have full effect upon restart."));
 		//notifyRestart = false;
 	}
diff --git a/src/tabs/tabdlg.cpp b/src/tabs/tabdlg.cpp
index a4d1b3f..9b7d071 100644
--- a/src/tabs/tabdlg.cpp
+++ b/src/tabs/tabdlg.cpp
@@ -781,7 +781,7 @@ void TabDlg::setUserManagementEnabled(bool enabled)
 	}
 
 	userManagement_ = enabled;
-	tabWidget_->setTabButtonsShown(enabled);
+	tabWidget_->setTabButtonsShown(PsiOptions::instance()->getOption("options.ui.tabs.show-tab-buttons").toBool());
 	tabWidget_->setDragsEnabled(enabled);
 }
 
diff --git a/src/widgets/psitabwidget.cpp b/src/widgets/psitabwidget.cpp
index adb627d..cdb42a5 100644
--- a/src/widgets/psitabwidget.cpp
+++ b/src/widgets/psitabwidget.cpp
@@ -75,6 +75,10 @@ PsiTabWidget::PsiTabWidget(QWidget *parent)
 
 	setTabPosition(QTabWidget::North);
 
+	if (!PsiOptions::instance()->getOption("options.ui.tabs.show-tab-buttons").toBool()){
+		closeButton_->hide();
+		downButton_->hide();
+	}
 	connect( tabBar_, SIGNAL(mouseDoubleClickTab(int)), SLOT(mouseDoubleClickTab(int)));
 	connect( tabBar_, SIGNAL(mouseMiddleClickTab(int)), SLOT(mouseMiddleClickTab(int)));
 	connect( tabBar_, SIGNAL( currentChanged(int)), SLOT(tab_currentChanged(int)));
