Index: psi/src/chatdlg.ui
===================================================================
--- psi.orig/src/chatdlg.ui	2013-01-15 14:04:05.188826174 +0400
+++ psi/src/chatdlg.ui	2013-01-15 14:04:05.184826173 +0400
@@ -184,7 +184,7 @@
             <enum>QToolButton::InstantPopup</enum>
            </property>
            <property name="arrowType" >
-            <enum>Qt::DownArrow</enum>
+            <enum>Qt::NoArrow</enum>
            </property>
           </widget>
          </item>
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:05.188826174 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:05.184826173 +0400
@@ -705,6 +705,7 @@
 	d->pm_settings = new QMenu(this);
 	connect(d->pm_settings, SIGNAL(aboutToShow()), SLOT(buildMenu()));
 	ui_.tb_actions->setMenu(d->pm_settings);
+	ui_.tb_actions->setIcon(IconsetFactory::icon("psi/select").icon());
 	ui_.tb_actions->setStyleSheet(" QToolButton::menu-indicator { image:none } ");
 
 	// resize the horizontal splitter
Index: psi/src/groupchatdlg.ui
===================================================================
--- psi.orig/src/groupchatdlg.ui	2013-01-15 14:04:05.188826174 +0400
+++ psi/src/groupchatdlg.ui	2013-01-15 14:04:05.184826173 +0400
@@ -148,7 +148,7 @@
              <enum>QToolButton::InstantPopup</enum>
             </property>
             <property name="arrowType" >
-             <enum>Qt::DownArrow</enum>
+             <enum>Qt::NoArrow</enum>
             </property>
            </widget>
           </item>
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:05.188826174 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:05.184826173 +0400
@@ -207,6 +207,7 @@
 	pm_settings_ = new QMenu(this);
 	connect(pm_settings_, SIGNAL(aboutToShow()), SLOT(buildMenu()));
 	ui_.tb_actions->setMenu(pm_settings_);
+	ui_.tb_actions->setIcon(IconsetFactory::icon("psi/select").icon());
 	ui_.tb_actions->setStyleSheet(" QToolButton::menu-indicator { image:none } ");
 
 	connect(account()->capsManager(), SIGNAL(capsChanged(const Jid&)), SLOT(capsChanged(const Jid&)));
Index: psi/src/psitoolbar.cpp
===================================================================
--- psi.orig/src/psitoolbar.cpp	2013-01-15 14:04:05.188826174 +0400
+++ psi/src/psitoolbar.cpp	2013-01-15 14:04:05.184826173 +0400
@@ -26,6 +26,7 @@
 #include <QContextMenuEvent>
 #include <QList>
 #include <QMainWindow>
+#include <QToolButton>
 
 #include "iconset.h"
 #include "iconaction.h"
@@ -97,6 +98,9 @@
 	setWindowTitle(o->getOption(base_ + ".name").toString());
 
 	ActionList actions = actionList_->suitableActions(PsiActionList::Actions_MainWin | PsiActionList::Actions_Common);
+	QList<QString> skipList;
+	skipList << "button_options" << "button_status" << "event_notifier" << "spacer";
+
 	foreach(QString actionName, o->getOption(base_ + ".actions").toStringList()) {
 		IconAction* action = actions.action(actionName);
 
@@ -104,6 +108,12 @@
 			if (action->isSeparator()) {
 				addSeparator();
 			}
+			else if (!skipList.contains(actionName)) {
+				QToolButton *button = new QToolButton;
+				button->setDefaultAction(action);
+				button->setPopupMode(QToolButton::InstantPopup);
+				addWidget(button);
+			}
 			else {
 				action->addTo(this);
 			}
