--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -328,6 +328,7 @@ QLineEdit#le_status_text {
 					<height>420</height>
 				</size>
 				<use-highlighting type="bool">true</use-highlighting><use-nick-coloring type="bool">true</use-nick-coloring><use-hash-nick-coloring type="bool">true</use-hash-nick-coloring><colored-history type="bool">true</colored-history>
+				<allow-highlight-events type="bool">false</allow-highlight-events>
 			</muc>
 			<show-deprecated comment="Deprecated functionality or protocols">
 				<service-discovery comment="Service discovery dialog">
--- psi.orig/src/chatdlg.cpp
+++ psi/src/chatdlg.cpp
@@ -449,6 +449,8 @@ void ChatDlg::ensureTabbedCorrectly()
 
 void ChatDlg::updateContact(const Jid &j, bool fromPresence)
 {
+	if (account()->groupchats().contains(j.full()))
+		return;
 	// if groupchat, only update if the resource matches
 	if (account()->findGCContact(j) && !jid().compare(j)) {
 		return;
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -199,6 +199,7 @@ public:
 		dlg = d;
 		nickSeparator = ":";
 		nonAnonymous = false;
+		alert = false;
 
 		trackBar = false;
 		mCmdManager.registerProvider(this);
@@ -236,6 +237,7 @@ public:
 	int pending;
 	int hPending; // highlight pending
 	bool connecting;
+	bool alert;
 
 	QStringList hist;
 	int histAt;
@@ -1025,6 +1027,7 @@ void GCMainDlg::activated()
 			u->setPending(d->pending, d->hPending);
 			account()->updateEntry(*u);
 		}
+		messagesRead(jid());
 		invalidateTab();
 	}
 	doFlash(false);
@@ -1733,11 +1736,11 @@ void GCMainDlg::avatarUpdated(const Jid
 	}
 }
 
-void GCMainDlg::message(const Message &_m)
+void GCMainDlg::message(const Message &_m, const PsiEvent::Ptr &e)
 {
 	Message m = _m;
 	QString from = m.from().resource();
-	bool alert = false;
+	d->alert = false;
 
 	if (m.getMUCStatuses().contains(100)) {
 		d->nonAnonymous = true;
@@ -1787,7 +1790,7 @@ void GCMainDlg::message(const Message &_
 
 	// code to determine if the speaker was addressing this client in chat
 	if(m.body().contains(d->self))
-		alert = true;
+		d->alert = true;
 
 	if (m.body().left(d->self.length()) == d->self)
 		d->lastReferrer = m.from().resource();
@@ -1796,7 +1799,7 @@ void GCMainDlg::message(const Message &_
 		QStringList highlightWords = options->getOption("options.ui.muc.highlight-words").toStringList();
 		foreach (QString word, highlightWords) {
 			if(m.body().contains((word), Qt::CaseInsensitive)) {
-				alert = true;
+				d->alert = true;
 			}
 		}
 	}
@@ -1807,13 +1810,12 @@ void GCMainDlg::message(const Message &_
 			account()->playSound(PsiAccount::eSend);
 	}
 	else {
-		if(alert || (options->getOption("options.ui.notifications.sounds.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) )
+		if(d->alert || (options->getOption("options.ui.notifications.sounds.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) )
 			account()->playSound(PsiAccount::eGroupChat);
 
-		if(alert || (options->getOption("options.ui.notifications.passive-popups.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) ) {
+		if(d->alert || (options->getOption("options.ui.notifications.passive-popups.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) ) {
 			if (!m.spooled() && !isActiveTab() && !m.from().resource().isEmpty()) {
 				XMPP::Jid jid = m.from()/*.withDomain("")*/;
-				MessageEvent::Ptr e(new MessageEvent(m, account()));
 				UserListItem i;
 				i.setPrivate(true);
 				account()->psi()->popupManager()->doPopup(account(), PopupManager::AlertGcHighlight, jid, m.from().resource(), &i, e);
@@ -1822,9 +1824,9 @@ void GCMainDlg::message(const Message &_
 	}
 
 	if(from.isEmpty())
-		appendSysMsg(m.body(), alert, m.timeStamp());
+		appendSysMsg(m.body(), d->alert, m.timeStamp());
 	else
-		appendMessage(m, alert);
+		appendMessage(m, d->alert);
 }
 
 void GCMainDlg::joined()
@@ -1861,6 +1863,11 @@ const QDateTime & GCMainDlg::lastMsgTime
 	return d->te_log()->lastMsgTime();
 }
 
+bool GCMainDlg::isLastMessageAlert() const
+{
+	return d->alert;
+}
+
 void GCMainDlg::appendSysMsg(const QString &str, bool alert, const QDateTime &ts)
 {
 	MessageView mv = MessageView::fromPlainText(str, MessageView::System);
--- psi.orig/src/groupchatdlg.h
+++ psi/src/groupchatdlg.h
@@ -31,6 +31,7 @@
 #include "ui_groupchatdlg.h"
 #include "mucmanager.h"
 #include "advwidget.h"
+#include "psievent.h"
 
 using namespace XMPP;
 
@@ -70,12 +71,13 @@ public:
 
 	void error(int, const QString &);
 	void presence(const QString &, const Status &);
-	void message(const Message &);
+	void message(const Message &, const PsiEvent::Ptr &e = PsiEvent::Ptr());
 	void joined();
 	void setPassword(const QString&);
 	const QString& nick() const;
 	const QString& topic() const;
 	const QDateTime& lastMsgTime() const;
+	bool isLastMessageAlert() const;
 
 	bool isInactive() const;
 	void reactivate();
@@ -100,6 +102,7 @@ protected:
 
 signals:
 	void aSend(const Message &);
+	void messagesRead(const Jid &);
 
 public slots:
 	// reimplemented
--- psi.orig/src/psiaccount.cpp
+++ psi/src/psiaccount.cpp
@@ -2789,9 +2789,9 @@ void PsiAccount::processIncomingMessage(
 
 #ifdef GROUPCHAT
 	if(_m.type() == "groupchat") {
-		GCMainDlg *w = findDialog<GCMainDlg*>(Jid(_m.from().bare()));
-		if(w)
-			w->message(_m);
+		MessageEvent::Ptr me(new MessageEvent(_m, this));
+		me->setOriginLocal(false);
+		handleEvent(me, IncomingStanza);
 		return;
 	}
 #endif
@@ -5076,7 +5076,17 @@ void PsiAccount::handleEvent(const PsiEv
 				!(e->type() == PsiEvent::Message &&
 				 e.staticCast<MessageEvent>()->message().body().isEmpty()))
 			{
-				logEvent(e->from(), e);
+				bool isMuc = false;
+#ifdef GROUPCHAT
+				if(e->type() == PsiEvent::Message) {
+					MessageEvent::Ptr me = e.staticCast<MessageEvent>();
+					if (me->message().type() == "groupchat")
+						isMuc = true;
+				}
+#endif
+				if (!isMuc) {
+					logEvent(e->from(), e);
+				}
 			}
 		}
 	}
@@ -5116,7 +5126,7 @@ void PsiAccount::handleEvent(const PsiEv
 		}
 
 		// Pass message events to chat window
-		if ((m.containsEvents() || m.chatState() != StateNone) && m.body().isEmpty()) {
+		if ((m.containsEvents() || m.chatState() != StateNone) && m.body().isEmpty() && m.type() != "groupchat") {
 			if (o->getOption("options.messages.send-composing-events").toBool()) {
 				ChatDlg *c = findChatDialogEx(e->from());
 				if (c) {
@@ -5180,6 +5190,22 @@ void PsiAccount::handleEvent(const PsiEv
 			doPopup = true;
 			popupType = PopupManager::AlertHeadline;
 		} // /headline
+#ifdef GROUPCHAT
+		else if (m.type() == "groupchat") {
+			putToQueue = false;
+			bool allowMucEvents = o->getOption("options.ui.muc.allow-highlight-events").toBool();
+			if (activationType != FromXml) {
+				GCMainDlg *c = findDialog<GCMainDlg*>(e->from());
+				if (c) {
+					c->message(m, e);
+					if (!c->isActiveTab() && c->isLastMessageAlert() && !m.spooled() && allowMucEvents)
+						putToQueue = true;
+				}
+			}
+			else if (allowMucEvents)
+				putToQueue = true;
+		} // /groupchat
+#endif
 		else if (m.type().isEmpty()) {
 			soundType = eMessage;
 			doPopup = true;
@@ -5577,11 +5603,18 @@ void PsiAccount::processReadNext(const U
 		return;
 
 	bool isChat = false;
+#ifdef GROUPCHAT
+	bool isMuc = false;
+#endif
 	if(e->type() == PsiEvent::Message) {
 		MessageEvent::Ptr me = e.staticCast<MessageEvent>();
 		const Message &m = me->message();
 		if(m.type() == "chat" && m.getForm().fields().empty())
 			isChat = true;
+#ifdef GROUPCHAT
+		else if (m.type() == "groupchat")
+			isMuc = true;
+#endif
 	}
 
 	// if it's a chat message, just open the chat window.  there is no need to do
@@ -5592,6 +5625,17 @@ void PsiAccount::processReadNext(const U
 		return;
 	}
 
+#ifdef GROUPCHAT
+	if (isMuc) {
+		GCMainDlg *c = findDialog<GCMainDlg*>(e->from());
+		if (c)
+		{
+			c->bringToFront(true);
+			return;
+		}
+	}
+#endif
+
 	// remove from queue
 	e = d->eventQueue->dequeue(u.jid());
 
@@ -5696,6 +5740,19 @@ void PsiAccount::chatMessagesRead(const
 //	}
 }
 
+#ifdef GROUPCHAT
+void PsiAccount::groupChatMessagesRead(const Jid &j)
+{
+	d->eventQueue->clear(j);
+
+	QList<UserListItem*> ul = findRelevant(j);
+	if(!ul.isEmpty()) {
+		UserListItem *u = ul.first();
+		cpUpdate(*u);
+	}
+}
+#endif
+
 void PsiAccount::logEvent(const Jid &j, const PsiEvent::Ptr &e)
 {
 	if (!d->acc.opt_log)
@@ -5718,6 +5775,7 @@ void PsiAccount::openGroupChat(const Jid
 	GCMainDlg *w = new GCMainDlg(this, j, d->tabManager);
 	w->setPassword(d->client->groupChatPassword(j.domain(), j.node()));
 	connect(w, SIGNAL(aSend(const Message &)), SLOT(dj_sendMessage(const Message &)));
+	connect(w, SIGNAL(messagesRead(const Jid &)), SLOT(groupChatMessagesRead(const Jid &)));
 	connect(d->psi, SIGNAL(emitOptionsUpdate()), w, SLOT(optionsUpdate()));
 	if(reason != MUCJoinDlg::MucAutoJoin || !PsiOptions::instance()->getOption("options.ui.muc.hide-on-autojoin").toBool()) {
 		w->ensureTabbedCorrectly();
--- psi.orig/src/psiaccount.h
+++ psi/src/psiaccount.h
@@ -451,6 +451,9 @@ private slots:
 	void itemRetracted(const Jid&, const QString&, const PubSubRetraction&);
 
 	void chatMessagesRead(const Jid &);
+#ifdef GROUPCHAT
+	void groupChatMessagesRead(const Jid &);
+#endif
 
 	void slotCheckVCard();
 	void edb_finished();
--- psi.orig/src/psicon.cpp
+++ psi/src/psicon.cpp
@@ -81,6 +81,9 @@
 #include "accountscombobox.h"
 #include "tabdlg.h"
 #include "chatdlg.h"
+#ifdef GROUPCHAT
+#include "groupchatdlg.h"
+#endif
 #include "spellchecker/aspellchecker.h"
 #ifdef WEBKIT
 #include "avatars.h"
@@ -1735,16 +1738,28 @@ void PsiCon::processEvent(const PsiEvent
 	}
 
 	bool isChat = false;
+#ifdef GROUPCHAT
+	bool isMuc = false;
+#endif
 	bool sentToChatWindow = false;
 	if ( e->type() == PsiEvent::Message ) {
 		MessageEvent::Ptr me = e.staticCast<MessageEvent>();
 		const Message &m = me->message();
-		bool emptyForm = m.getForm().fields().empty();
-		// FIXME: Refactor this, PsiAccount and PsiEvent out
-		if (m.type() == "chat" && emptyForm) {
-			isChat = true;
-			sentToChatWindow = me->sentToChatWindow();
+#ifdef GROUPCHAT
+		if (m.type() == "groupchat") {
+			isMuc = true;
 		}
+		else {
+#endif
+			bool emptyForm = m.getForm().fields().empty();
+			// FIXME: Refactor this, PsiAccount and PsiEvent out
+			if (m.type() == "chat" && emptyForm) {
+				isChat = true;
+				sentToChatWindow = me->sentToChatWindow();
+			}
+#ifdef GROUPCHAT
+		}
+#endif
 	}
 
 	if ( isChat ) {
@@ -1776,6 +1791,17 @@ void PsiCon::processEvent(const PsiEvent
 	}
 #endif
 	else {
+#ifdef GROUPCHAT
+		if (isMuc) {
+			PsiAccount *account = e->account();
+			GCMainDlg *c = account->findDialog<GCMainDlg*>(e->from());
+			if (c) {
+				c->ensureTabbedCorrectly();
+				c->bringToFront(true);
+				return;
+			}
+		}
+#endif
 		// search for an already opened eventdlg
 		EventDlg *w = e->account()->findDialog<EventDlg*>(u->jid());
 
--- psi.orig/src/psievent.cpp
+++ psi/src/psievent.cpp
@@ -241,6 +241,10 @@ int MessageEvent::type() const
 
 Jid MessageEvent::from() const
 {
+#ifdef GROUPCHAT
+	if (v_m.type() == "groupchat")
+		return v_m.from().bare();
+#endif
 	return v_m.from();
 }
 
--- psi.orig/src/psiiconset.cpp
+++ psi/src/psiiconset.cpp
@@ -686,7 +686,7 @@ PsiIcon *PsiIconset::event2icon(const Ps
 		const Message &m = me->message();
 		if(m.type() == "headline")
 			icon = "psi/headline";
-		else if(m.type() == "chat")
+		else if(m.type() == "chat" || m.type() == "groupchat")
 			icon = "psi/chat";
 		else if(m.type() == "error")
 			icon = "psi/system";
