Index: psi/src/chatdlg.ui
===================================================================
--- psi.orig/src/chatdlg.ui	2013-01-15 14:04:05.688826194 +0400
+++ psi/src/chatdlg.ui	2013-01-15 14:04:05.688826194 +0400
@@ -132,6 +132,22 @@
           </widget>
          </item>
          <item>
+          <widget class="QToolButton" name="tb_voice" >
+           <property name="toolTip" >
+            <string>Voice Call</string>
+           </property>
+           <property name="text" >
+            <string/>
+           </property>
+           <property name="iconSize" >
+            <size>
+             <width>16</width>
+             <height>16</height>
+            </size>
+           </property>
+          </widget>
+         </item>
+         <item>
           <widget class="QToolButton" name="tb_pgp" >
            <property name="toolTip" >
             <string>Toggle encryption</string>
Index: psi/src/gcuserview.cpp
===================================================================
--- psi.orig/src/gcuserview.cpp	2013-01-15 14:04:05.688826194 +0400
+++ psi/src/gcuserview.cpp	2013-01-15 14:04:05.688826194 +0400
@@ -36,6 +36,7 @@
 #include "common.h"
 #include "psioptions.h"
 #include "coloropt.h"
+#include "avcall/avcall.h"
 
 static bool caseInsensitiveLessThan(const QString &s1, const QString &s2)
 {
@@ -414,6 +415,11 @@
 	act = new QAction(IconsetFactory::icon("psi/start-chat").icon(), tr("Open &Chat Window"), pm);
 	pm->addAction(act);
 	act->setData(1);
+	if (AvCallManager::isSupported()) {
+		act = new QAction(IconsetFactory::icon("psi/avcall").icon(), tr("Voice Call"), pm);
+		pm->addAction(act);
+		act->setData(5);
+	}
 	pm->addSeparator();
 
 	// Kick and Ban submenus
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:05.688826194 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:05.688826194 +0400
@@ -1675,6 +1675,9 @@
 	else if(x == 4) {
 		account()->invokeGCFile(jid().withResource(nick));
 	}
+	else if(x == 5) {
+		account()->actionVoice(jid().withResource(nick));
+	}
 	else if(x == 10) {
 		d->mucManager->kick(nick);
 	}
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:05.688826194 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:05.688826194 +0400
@@ -43,6 +43,7 @@
 #include "jidutil.h"
 #include "xmpp_tasks.h"
 #include "lastactivitytask.h"
+#include "avcall/avcall.h"
 
 
 #define MCMDCHAT		"http://psi-im.org/ids/mcmd#chatmain"
@@ -254,6 +255,7 @@
 		ui_.tb_actions->hide();
 		ui_.tb_emoticons->hide();
 		ui_.toolbar->hide();
+		ui_.tb_voice->hide();
 	}
 	else {
 		ui_.lb_status->show();
@@ -262,11 +264,13 @@
 			ui_.toolbar->show();
 			ui_.tb_actions->hide();
 			ui_.tb_emoticons->hide();
+			ui_.tb_voice->hide();
 		}
 		else {
 			ui_.toolbar->hide();
 			ui_.tb_emoticons->setVisible(PsiOptions::instance()->getOption("options.ui.emoticons.use-emoticons").toBool());
 			ui_.tb_actions->show();
+			ui_.tb_voice->setVisible(AvCallManager::isSupported());
 		}
 	}
 
@@ -323,7 +327,8 @@
 
 	act_voice_ = new IconAction(tr("Voice Call"), "psi/voice", tr("Voice Call"), 0, this);
 	connect(act_voice_, SIGNAL(triggered()), SLOT(doVoice()));
-	act_voice_->setEnabled(false);
+	//act_voice_->setEnabled(false);
+	ui_.tb_voice->setDefaultAction(act_voice_);
 
 	act_file_ = new IconAction(tr("Send File"), "psi/upload", tr("Send File"), 0, this);
 	connect(act_file_, SIGNAL(triggered()), SLOT(doFile()));
@@ -354,7 +359,7 @@
 	}
 	ui_.toolbar->addAction(act_info_);
 	ui_.toolbar->addAction(act_history_);
-	if (account()->voiceCaller()) {
+	if (AvCallManager::isSupported()) {
 		ui_.toolbar->addAction(act_voice_);
 	}
 }
@@ -373,7 +378,7 @@
 	if (resource.isEmpty() && ul && !ul->userResourceList().isEmpty()) {
 		resource = (*(ul->userResourceList().priority())).name();
 	}
-	act_voice_->setEnabled(!account()->capsManager()->isEnabled() || (ul && ul->isAvailable() && account()->capsManager()->features(jid().withResource(resource)).canVoice()));
+	//act_voice_->setEnabled(!account()->capsManager()->isEnabled() || (ul && ul->isAvailable() && account()->capsManager()->features(jid().withResource(resource)).canVoice()));
 }
 
 void PsiChatDlg::activated()
@@ -526,8 +531,9 @@
 
 	pm_settings_->addAction(act_icon_);
 	pm_settings_->addAction(act_file_);
-	if (account()->voiceCaller())
-		act_voice_->addTo(pm_settings_);
+	if (AvCallManager::isSupported()) {
+		pm_settings_->addAction(act_voice_);
+	}
 	pm_settings_->addAction(act_pgp_);
 	pm_settings_->addSeparator();
 
Index: psi/src/psicontactmenu.cpp
===================================================================
--- psi.orig/src/psicontactmenu.cpp	2013-01-15 14:04:05.688826194 +0400
+++ psi/src/psicontactmenu.cpp	2013-01-15 14:04:05.688826194 +0400
@@ -38,6 +38,7 @@
 #include "userlist.h"
 #include "statusdlg.h"
 #include "xmpp_tasks.h"
+#include "avcall/avcall.h"
 #ifdef HAVE_PGPUTIL
 #include "pgputil.h"
 #endif
@@ -413,7 +414,9 @@
 		menu_->addMenu(openWhiteboardToMenu_);
 		menu_->addMenu(executeCommandMenu_);
 		menu_->addMenu(activeChatsMenu_);
-		menu_->addAction(voiceCallAction_);
+		if(AvCallManager::isSupported()) {
+			menu_->addAction(voiceCallAction_);
+		}
 		menu_->addSeparator();
 		menu_->addAction(sendFileAction_);
 		menu_->addMenu(inviteToGroupchatMenu_);
