--- git.orig/src/chatdlg.ui
+++ git/src/chatdlg.ui
@@ -132,6 +132,22 @@
           </widget>
          </item>
          <item>
+          <widget class="QToolButton" name="tb_voice" >
+           <property name="toolTip" >
+            <string>Voice Call</string>
+           </property>
+           <property name="text" >
+            <string/>
+           </property>
+           <property name="iconSize" >
+            <size>
+             <width>16</width>
+             <height>16</height>
+            </size>
+           </property>
+          </widget>
+         </item>
+         <item>
           <widget class="QToolButton" name="tb_pgp" >
            <property name="toolTip" >
             <string>Toggle encryption</string>
--- git.orig/src/gcuserview.cpp
+++ git/src/gcuserview.cpp
@@ -35,6 +35,7 @@
 #include "common.h"
 #include "psioptions.h"
 #include "coloropt.h"
+#include "avcall/avcall.h"
 
 static bool caseInsensitiveLessThan(const QString &s1, const QString &s2)
 {
@@ -410,6 +411,11 @@ void GCUserView::doContextMenu(QTreeWidg
 	act = new QAction(IconsetFactory::icon("psi/start-chat").icon(), tr("Open &Chat Window"), pm);
 	pm->addAction(act);
 	act->setData(1);
+	if (AvCallManager::isSupported()) {
+		act = new QAction(IconsetFactory::icon("psi/avcall").icon(), tr("Voice Call"), pm);
+		pm->addAction(act);
+		act->setData(5);
+	}
 	pm->addSeparator();
 
 	// Kick and Ban submenus
--- git.orig/src/groupchatdlg.cpp
+++ git/src/groupchatdlg.cpp
@@ -1677,6 +1677,9 @@ void GCMainDlg::lv_action(const QString 
 	else if(x == 4) {
 		account()->invokeGCFile(jid().withResource(nick));
 	}
+	else if(x == 5) {
+		account()->actionVoice(jid().withResource(nick));
+	}
 	else if(x == 10) {
 		d->mucManager->kick(nick);
 	}
--- git.orig/src/psichatdlg.cpp
+++ git/src/psichatdlg.cpp
@@ -42,6 +42,7 @@
 #include "jidutil.h"
 #include "xmpp_tasks.h"
 #include "lastactivitytask.h"
+#include "avcall/avcall.h"
 
 
 #define MCMDCHAT		"http://psi-im.org/ids/mcmd#chatmain"
@@ -253,6 +254,7 @@ void PsiChatDlg::setLooks()
 		ui_.tb_actions->hide();
 		ui_.tb_emoticons->hide();
 		ui_.toolbar->hide();
+		ui_.tb_voice->hide();
 	}
 	else {
 		ui_.lb_status->show();
@@ -261,11 +263,13 @@ void PsiChatDlg::setLooks()
 			ui_.toolbar->show();
 			ui_.tb_actions->hide();
 			ui_.tb_emoticons->hide();
+			ui_.tb_voice->hide();
 		}
 		else {
 			ui_.toolbar->hide();
 			ui_.tb_emoticons->setVisible(PsiOptions::instance()->getOption("options.ui.emoticons.use-emoticons").toBool());
 			ui_.tb_actions->show();
+			ui_.tb_voice->setVisible(AvCallManager::isSupported());
 		}
 	}
 
@@ -322,7 +326,8 @@ void PsiChatDlg::initToolButtons()
 
 	act_voice_ = new IconAction(tr("Voice Call"), "psi/voice", tr("Voice Call"), 0, this);
 	connect(act_voice_, SIGNAL(triggered()), SLOT(doVoice()));
-	act_voice_->setEnabled(false);
+	//act_voice_->setEnabled(false);
+	ui_.tb_voice->setDefaultAction(act_voice_);
 
 	act_file_ = new IconAction(tr("Send File"), "psi/upload", tr("Send File"), 0, this);
 	connect(act_file_, SIGNAL(triggered()), SLOT(doFile()));
@@ -353,7 +358,7 @@ void PsiChatDlg::initToolBar()
 	}
 	ui_.toolbar->addAction(act_info_);
 	ui_.toolbar->addAction(act_history_);
-	if (account()->voiceCaller()) {
+	if (AvCallManager::isSupported()) {
 		ui_.toolbar->addAction(act_voice_);
 	}
 }
@@ -372,7 +377,7 @@ void PsiChatDlg::capsChanged()
 	if (resource.isEmpty() && ul && !ul->userResourceList().isEmpty()) {
 		resource = (*(ul->userResourceList().priority())).name();
 	}
-	act_voice_->setEnabled(!account()->capsManager()->isEnabled() || (ul && ul->isAvailable() && account()->capsManager()->features(jid().withResource(resource)).canVoice()));
+	//act_voice_->setEnabled(!account()->capsManager()->isEnabled() || (ul && ul->isAvailable() && account()->capsManager()->features(jid().withResource(resource)).canVoice()));
 }
 
 void PsiChatDlg::activated()
@@ -525,8 +530,9 @@ void PsiChatDlg::buildMenu()
 
 	pm_settings_->addAction(act_icon_);
 	pm_settings_->addAction(act_file_);
-	if (account()->voiceCaller())
-		act_voice_->addTo(pm_settings_);
+	if (AvCallManager::isSupported()) {
+		pm_settings_->addAction(act_voice_);
+	}
 	pm_settings_->addAction(act_pgp_);
 	pm_settings_->addSeparator();
 
--- git.orig/src/psicontactmenu.cpp
+++ git/src/psicontactmenu.cpp
@@ -38,6 +38,7 @@
 #include "userlist.h"
 #include "statusdlg.h"
 #include "xmpp_tasks.h"
+#include "avcall/avcall.h"
 #ifdef HAVE_PGPUTIL
 #include "pgputil.h"
 #endif
@@ -413,7 +414,9 @@ public:
 		menu_->addMenu(openWhiteboardToMenu_);
 		menu_->addMenu(executeCommandMenu_);
 		menu_->addMenu(activeChatsMenu_);
-		menu_->addAction(voiceCallAction_);
+		if(AvCallManager::isSupported()) {
+			menu_->addAction(voiceCallAction_);
+		}
 		menu_->addSeparator();
 		menu_->addAction(sendFileAction_);
 		menu_->addMenu(inviteToGroupchatMenu_);
