Index: psi/src/contactlistaccountmenu.cpp
===================================================================
--- psi.orig/src/contactlistaccountmenu.cpp	2013-01-15 14:04:06.560826229 +0400
+++ psi/src/contactlistaccountmenu.cpp	2013-01-15 14:04:06.556826229 +0400
@@ -121,7 +121,7 @@
 		avatarMenu_ = menu->addMenu(tr("Avatar"));
 		avatarMenu_->addAction(setAvatarAction_);
 		avatarMenu_->addAction(unsetAvatarAction_);
-		bookmarksMenu_ = menu->addMenu(tr("Bookmarks"));
+		bookmarksMenu_ = menu->addMenu(IconsetFactory::icon("psi/bookmarks").icon(), tr("Bookmarks"));
 		bookmarksMenu_->addAction(bookmarksManageAction_);
 		menu->addSeparator();
 		menu->addAction(addContactAction_);
Index: psi/src/contactview.cpp
===================================================================
--- psi.orig/src/contactview.cpp	2013-01-15 14:04:06.560826229 +0400
+++ psi/src/contactview.cpp	2013-01-15 14:04:06.556826229 +0400
@@ -978,7 +978,7 @@
 			bookmarks->setEnabled(false);
 		}

-		pm.insertItem(tr("Bookmarks"), bookmarks);
+		pm.insertItem(IconsetFactory::icon("psi/bookmarks").icon(), tr("Bookmarks"), bookmarks);

 		pm.insertSeparator();
 		pm.insertItem(IconsetFactory::icon("psi/addContact").icon(), tr("&Add a Contact"), 7);
@@ -1266,7 +1266,7 @@
 		// Voice call
 		if(d->pa->avCallManager() && !isAgent) {
 		//if(d->pa->voiceCaller() && !isAgent) {
-			pm.insertItem(IconsetFactory::icon("psi/voice").icon(), tr("Voice Call"), 24);
+			pm.insertItem(IconsetFactory::icon("psi/avcall").icon(), tr("Voice Call"), 24);
 			if(!online) {
 				pm.setItemEnabled(24, false);
 			}
Index: psi/src/options/opt_avcall.cpp
===================================================================
--- psi.orig/src/options/opt_avcall.cpp	2013-01-15 14:04:06.560826229 +0400
+++ psi/src/options/opt_avcall.cpp	2013-01-15 14:04:06.556826229 +0400
@@ -166,7 +166,7 @@
 //----------------------------------------------------------------------------

 OptionsTabAvCall::OptionsTabAvCall(QObject *parent)
-: OptionsTab(parent, "avcall", "", tr("Voice Calling"), AvCallManager::isVideoSupported() ? tr("Audio and video device configuration") : tr("Audio device configuration"), "psi/voice")
+: OptionsTab(parent, "avcall", "", tr("Voice Calling"), AvCallManager::isVideoSupported() ? tr("Audio and video device configuration") : tr("Audio device configuration"), "psi/avcall")
 {
 	w = 0;
 }
Index: psi/src/options/opt_plugins.cpp
===================================================================
--- psi.orig/src/options/opt_plugins.cpp	2013-01-15 14:04:06.560826229 +0400
+++ psi/src/options/opt_plugins.cpp	2013-01-15 14:04:06.556826229 +0400
@@ -23,7 +23,7 @@
 //----------------------------------------------------------------------------

 OptionsTabPlugins::OptionsTabPlugins(QObject *parent)
-: OptionsTab(parent, "plugins", "", tr("Plugins"), tr("Options for Psi plugins"), "psi/advanced")
+: OptionsTab(parent, "plugins", "", tr("Plugins"), tr("Options for Psi plugins"), "psi/plugins")
 {
 	w = 0;
 }
Index: psi/src/options/opt_tree.cpp
===================================================================
--- psi.orig/src/options/opt_tree.cpp	2013-01-15 14:04:06.560826229 +0400
+++ psi/src/options/opt_tree.cpp	2013-01-15 14:04:06.556826229 +0400
@@ -5,7 +5,7 @@
 #include "psioptionseditor.h"

 OptionsTabTree::OptionsTabTree(QObject *parent)
-		: OptionsTab(parent, "tree", "", tr("Advanced"), tr("Options for advanced users"), "psi/advanced")
+		: OptionsTab(parent, "tree", "", tr("Advanced"), tr("Options for advanced users"), "psi/advanced-plus")
 {
 	w = 0;
 }
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:06.560826229 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:06.556826229 +0400
@@ -344,7 +344,7 @@
 	act_icon_->setMenu(account()->psi()->iconSelectPopup());
 	ui_.tb_emoticons->setMenu(account()->psi()->iconSelectPopup());

-	act_voice_ = new IconAction(tr("Voice Call"), "psi/voice", tr("Voice Call"), 0, this);
+	act_voice_ = new IconAction(tr("Voice Call"), "psi/avcall", tr("Voice Call"), 0, this);
 	connect(act_voice_, SIGNAL(triggered()), SLOT(doVoice()));
 	//act_voice_->setEnabled(false);
 	ui_.tb_voice->setDefaultAction(act_voice_);
Index: psi/src/psicontactmenu.cpp
===================================================================
--- psi.orig/src/psicontactmenu.cpp	2013-01-15 14:04:06.560826229 +0400
+++ psi/src/psicontactmenu.cpp	2013-01-15 14:04:06.556826229 +0400
@@ -331,7 +331,7 @@
 		openWhiteboardAction_ = new IconAction(tr("Open a &Whiteboard"), this, "psi/whiteboard");
 		connect(openWhiteboardAction_, SIGNAL(triggered()), SLOT(openWhiteboard()));

-		voiceCallAction_ = new IconAction(tr("Voice Call"), this, "psi/voice");
+		voiceCallAction_ = new IconAction(tr("Voice Call"), this, "psi/avcall");
 		connect(voiceCallAction_, SIGNAL(triggered()), SLOT(voiceCall()));

 		sendFileAction_ = new IconAction(tr("Send &File"), this, "psi/upload");
