Index: psi/iris/src/xmpp/xmpp-im/xmpp_xmlcommon.cpp
===================================================================
--- psi.orig/iris/src/xmpp/xmpp-im/xmpp_xmlcommon.cpp	2013-01-15 14:04:06.860826242 +0400
+++ psi/iris/src/xmpp/xmpp-im/xmpp_xmlcommon.cpp	2013-01-15 14:04:06.856826241 +0400
@@ -322,14 +322,18 @@
 
 	// find closest xmlns
 	QDomNode n = e;
-	while(!n.isNull() && !n.toElement().hasAttribute("xmlns"))
+	while(!n.isNull() && !n.toElement().hasAttribute("xmlns") && n.toElement().namespaceURI() == "" )
 		n = n.parentNode();
 	QString ns;
-	if(n.isNull() || !n.toElement().hasAttribute("xmlns"))
-		ns = "jabber:client";
-	else
+	if(n.isNull() || !n.toElement().hasAttribute("xmlns")){
+		if (n.toElement().namespaceURI () == ""){		
+			ns = "jabber:client";
+		} else {
+			ns = n.toElement().namespaceURI();
+		}
+	} else {
 		ns = n.toElement().attribute("xmlns");
-
+	}
 	// make a new node
 	QDomElement i = e.ownerDocument().createElementNS(ns, e.tagName());
 
