--- git.orig/src/discodlg.cpp
+++ git/src/discodlg.cpp
@@ -891,7 +891,7 @@ public: // data
 	IconAction *actBrowse, *actBack, *actForward, *actRefresh, *actStop;
 
 	// custom actions, that will be added to toolbar and context menu
-	IconAction *actRegister, *actSearch, *actJoin, *actAHCommand, *actVCard, *actAdd;
+	IconAction *actRegister, *actSearch, *actJoin, *actAHCommand, *actVCard, *actAdd, *actQueryVersion;
 
 	typedef QList<History*> HistoryList;
 	HistoryList backHistory, forwardHistory;
@@ -1009,6 +1009,9 @@ DiscoDlg::Private::Private(DiscoDlg *par
 	actAdd = new IconAction (tr("Add to roster"), "psi/addContact", tr("&Add to roster"), 0, dlg);
 	connect (actAdd, SIGNAL(triggered()), sm, SLOT(map()));
 	sm->setMapping(actAdd, Features::FID_Add);
+	actQueryVersion = new IconAction (tr("Query version"), "psi/info", tr("&Query version"), 0, dlg);
+	connect (actQueryVersion, SIGNAL(triggered()), sm, SLOT(map()));
+	sm->setMapping(actQueryVersion, Features::FID_QueryVersion);
 
 	// create toolbar
 	toolBar = new QToolBar(tr("Service Discovery toolbar"), dlg);
@@ -1034,6 +1037,7 @@ DiscoDlg::Private::Private(DiscoDlg *par
 	toolBar->addAction(actAdd);
 	toolBar->addAction(actVCard);
 	toolBar->addAction(actAHCommand);
+	toolBar->addAction(actQueryVersion);
 
 	// select protocol
 	if (PsiOptions::instance()->getOption("options.ui.show-deprecated.service-discovery.protocol-selector").toBool()) {
@@ -1203,6 +1207,7 @@ void DiscoDlg::Private::enableButtons(co
 	actAdd->setEnabled( itemSelected );
 	actVCard->setEnabled( f.haveVCard() );
 	actAHCommand->setEnabled( f.canCommand() );
+	actQueryVersion->setEnabled( f.hasVersion() );
 }
 
 void DiscoDlg::Private::itemSelected (QTreeWidgetItem *item)
@@ -1331,7 +1336,8 @@ bool DiscoDlg::Private::eventFilter (QOb
 			actAdd->addTo(&p);
 			actVCard->addTo(&p);
 			actAHCommand->addTo(&p);
-
+			actQueryVersion->addTo(&p);
+			
 			// popup with all available features
 			QMenu *fm = new QMenu(&p);
 			QHash<QAction*, int> actions;
--- git.orig/src/psiaccount.cpp
+++ git/src/psiaccount.cpp
@@ -3220,6 +3220,9 @@ void PsiAccount::featureActivated(QStrin
 		QStringList sl;
 		dj_add(jid, QString::null, sl, true);
 	}
+	else if ( f.hasVersion() ) {
+		actionQueryVersion(jid);
+	}
 }
 
 void PsiAccount::actionManageBookmarks()
@@ -3719,6 +3722,25 @@ void PsiAccount::actionSendFiles(const J
 	sendFiles(j, l);
 }
 
+void PsiAccount::actionQueryVersion(const Jid& j)
+{
+	JT_ClientVersion *task = new JT_ClientVersion(d->client->rootTask());
+	task->get(j);
+	connect(task, SIGNAL(finished()), SLOT(queryVersionFinished()));
+	task->go(true);
+}
+
+void PsiAccount::queryVersionFinished()
+{
+	JT_ClientVersion* j = static_cast<JT_ClientVersion*>(sender());
+	QString text;
+	text += tr("Name:\t") + j->name();
+	text += "\n" + tr("Version:\t") + j->version();
+	text += "\n" + tr("Os:\t") + j->os();
+	
+	QMessageBox::information(NULL, QString(tr("Version Query Information")), text);
+}
+
 void PsiAccount::actionExecuteCommand(const Jid& j, const QString& node)
 {
 	Jid j2 = j;
--- git.orig/src/psiaccount.h
+++ git/src/psiaccount.h
@@ -353,6 +353,7 @@ public slots:
 	void actionSetActivity();
 	void actionSetAvatar();
 	void actionUnsetAvatar();
+	void actionQueryVersion(const Jid& j);
 	void featureActivated(QString feature, Jid jid, QString node);
 
 	void actionAssignKey(const Jid &);
@@ -425,6 +426,7 @@ private slots:
 
 	void processReadNext(const UserListItem &);
 	void processReadNext(const Jid &);
+	void queryVersionFinished();
 
 protected:
 	bool validRosterExchangeItem(const RosterExchangeItem&);
