Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:07.316826260 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:07.312826260 +0400
@@ -1049,6 +1049,7 @@
 	features << "http://jabber.org/protocol/muc";
 #endif
 	features << "jabber:x:data";
+	features << "jabber:iq:version";
 	d->client->setFeatures(Features(features));

 #ifdef FILETRANSFER
