--- git.orig/src/psiaccount.cpp
+++ git/src/psiaccount.cpp
@@ -372,6 +372,7 @@ public:
 		, rcSetStatusServer(0)
 		, rcSetOptionsServer(0)
 		, rcForwardServer(0)
+		, rcLeaveMucServer(0)
 		, avatarFactory(0)
 		, voiceCaller(0)
 		, avCallManager(0)
@@ -438,6 +439,7 @@ public:
 	RCSetStatusServer* rcSetStatusServer;
 	RCSetOptionsServer* rcSetOptionsServer;
 	RCForwardServer* rcForwardServer;
+	RCLeaveMucServer* rcLeaveMucServer;
 
 	// Avatars
 	AvatarFactory* avatarFactory;
@@ -1129,6 +1131,7 @@ PsiAccount::PsiAccount(const UserAccount
 	d->rcSetStatusServer = 0;
 	d->rcSetOptionsServer = 0;
 	d->rcForwardServer = 0;
+	d->rcLeaveMucServer =0;
 	setRCEnabled(PsiOptions::instance()->getOption("options.external-control.adhoc-remote-control.enable").toBool());
 
 	// Plugins
@@ -5829,6 +5832,7 @@ void PsiAccount::setRCEnabled(bool b)
 	if (b && !d->rcSetStatusServer) {
 		d->rcSetStatusServer = new RCSetStatusServer(d->ahcManager);
 		d->rcForwardServer = new RCForwardServer(d->ahcManager);
+		d->rcLeaveMucServer = new RCLeaveMucServer(d->ahcManager);
 		d->rcSetOptionsServer = new RCSetOptionsServer(d->ahcManager, d->psi);
 	}
 	else if (!b && d->rcSetStatusServer) {
@@ -5836,6 +5840,8 @@ void PsiAccount::setRCEnabled(bool b)
 		d->rcSetStatusServer = 0;
 		delete d->rcForwardServer;
 		d->rcForwardServer = 0;
+		delete d->rcLeaveMucServer;
+		d->rcLeaveMucServer = 0;
 		delete d->rcSetOptionsServer;
 		d->rcSetOptionsServer = 0;
 	}
--- git.orig/src/rc.cpp
+++ git/src/rc.cpp
@@ -27,6 +27,7 @@
 #include "xmpp_xdata.h"
 #include "ahcservermanager.h"
 #include "ahcommand.h"
+#include "groupchatdlg.h"
 
 using namespace XMPP;
 
@@ -245,3 +246,16 @@ AHCommand RCSetOptionsServer::execute(co
 		return AHCommand::completedReply(c);
 	}
 }
+
+AHCommand RCLeaveMucServer::execute(const AHCommand& c, const Jid& /*j*/)
+{
+	foreach (QString gc, manager()->account()->groupchats()) {
+		Jid mj(gc);
+		GCMainDlg *gcDlg = manager()->account()->findDialog<GCMainDlg*>(mj.bare());
+		if (gcDlg) gcDlg->close();
+	}
+	XData form;
+	form.setTitle(QObject::tr("Leave All Conferences"));
+	form.setType(XData::Data_Form);
+	return AHCommand::completedReply(c);
+}
--- git.orig/src/rc.h
+++ git/src/rc.h
@@ -54,6 +54,15 @@ public:
 	virtual AHCommand execute(const AHCommand& c, const Jid&);
 };
 
+class RCLeaveMucServer : public RCCommandServer
+{
+public:
+	RCLeaveMucServer(AHCServerManager* m) : RCCommandServer(m) { }
+	virtual QString name() const { return "Leave All Conferences"; }
+	virtual QString rcNode() const { return "leave-muc"; }
+	virtual AHCommand execute(const AHCommand& c, const Jid&);
+};
+
 class RCSetOptionsServer : public RCCommandServer
 {
 public:
