Entering 'iris'
Entering 'src/libpsi'
Index: psi/options/default.xml
===================================================================
--- psi.orig/options/default.xml	2013-01-15 14:04:08.828826322 +0400
+++ psi/options/default.xml	2013-01-15 14:04:08.824826321 +0400
@@ -377,6 +377,9 @@
 			</service-discovery>
 			<tabs>
 				<show-tab-icons type="bool">true</show-tab-icons>
+				<disable-wheel-scroll type="bool">false</disable-wheel-scroll>
+				<can-close-inactive-tab type="bool">true</can-close-inactive-tab>
+				<show-tab-close-buttons type="bool">true</show-tab-close-buttons>
 				<put-tabs-at-bottom type="bool">false</put-tabs-at-bottom>
 				<use-tabs type="bool">true</use-tabs>
 				<size type="QString"></size> <!-- will be invalid when converted to QSize so we can detect first load -->
Index: psi/src/tabs/tabdlg.cpp
===================================================================
--- psi.orig/src/tabs/tabdlg.cpp	2013-01-15 14:04:08.828826322 +0400
+++ psi/src/tabs/tabdlg.cpp	2013-01-15 14:04:08.824826321 +0400
@@ -161,6 +161,7 @@
 	connect(tabWidget_, SIGNAL(tabContextMenu(int, QPoint, QContextMenuEvent*)), SLOT(showTabMenu(int, QPoint, QContextMenuEvent*)));
 	connect(tabWidget_, SIGNAL(closeButtonClicked()), SLOT(closeCurrentTab()));
 	connect(tabWidget_, SIGNAL(currentChanged(QWidget*)), SLOT(tabSelected(QWidget*)));
+	connect(tabWidget_, SIGNAL(tabCloseRequested(int)), SLOT(tabCloseRequested(int)));

 	if(delegate_)
 		delegate_->tabWidgetCreated(this, tabWidget_);
@@ -817,6 +818,23 @@
 	updateCaption();
 }

+/**
+  * the slot is invoked, when small close button is clicked on a tab
+  * dont close tabs, that are not active.
+  * \param tab number requested to close
+  */
+void TabDlg::tabCloseRequested(int i)
+{
+	if (tabWidget_->currentPageIndex() == i)
+		closeTab(static_cast<TabbableWidget*>(tabWidget_->page(i)));
+	else {
+		if (PsiOptions::instance()->getOption("options.ui.tabs.can-close-inactive-tab").toBool())
+			closeTab(static_cast<TabbableWidget*>(tabWidget_->page(i)));
+		else
+			selectTab(static_cast<TabbableWidget*>(tabWidget_->page(i)));
+	}
+}
+
 TabbableWidget* TabDlg::getCurrentTab() const
 {
 	return dynamic_cast<TabbableWidget*>(tabWidget_->currentPage());
Index: psi/src/tabs/tabdlg.h
===================================================================
--- psi.orig/src/tabs/tabdlg.h	2013-01-15 14:04:08.828826322 +0400
+++ psi/src/tabs/tabdlg.h	2013-01-15 14:04:08.824826321 +0400
@@ -122,6 +122,7 @@
 	void optionsUpdate();
 	void detachTab(TabbableWidget*);
 	void sendTabTo(TabbableWidget*, TabDlg *);
+	void tabCloseRequested(int i);

 signals:
 	void resized(QSize size);
Index: psi/src/widgets/psitabbar.cpp
===================================================================
--- psi.orig/src/widgets/psitabbar.cpp	2013-01-15 14:04:08.828826322 +0400
+++ psi/src/widgets/psitabbar.cpp	2013-01-15 14:04:08.824826321 +0400
@@ -24,6 +24,9 @@
 #include <QApplication>
 #include <QDrag>
 #include <QMimeData>
+#include <QPainter>
+
+#include "psioptions.h"

 /**
  * Constructor
@@ -32,6 +35,13 @@
 		: QTabBar(parent)
 		, dragsEnabled_(true) {
 	//setAcceptDrops(true);
+
+#if QT_VERSION >= 0x040500
+	setMovable(true);
+	setTabsClosable(true);
+	setSelectionBehaviorOnRemove ( QTabBar::SelectPreviousTab );
+#endif
+	currTab=-1;
 }

 /**
@@ -71,20 +81,42 @@
 }

 void PsiTabBar::mousePressEvent(QMouseEvent *event) {
-	int tabno = findTabUnder(event->pos());
-	if (event->button() == Qt::LeftButton) {
-		dragStartPosition_ = event->pos();
-		dragTab_ = tabno;
-		if (tabno != -1) {
-			setCurrentIndex(tabno);
+	QTabBar::mousePressEvent(event);
+	event->accept();
+}
+
+void PsiTabBar::mouseReleaseEvent ( QMouseEvent * event )
+{
+	if (event->button() == Qt::MidButton && findTabUnder(event->pos())!=-1) {
+		emit mouseMiddleClickTab(findTabUnder(event->pos()));
+		event->accept();
+	}
+	QTabBar::mouseReleaseEvent(event);
+
+#if QT_VERSION >= 0x040500
+	if ((dragTab_ != -1) && (event->button() != Qt::MidButton)) {
+		this->setCurrentIndex(currentIndex());
+	}
+#else
+	if ((dragTab_ != currTab)&&(currTab!=-1)&&(dragTab_!=-1)) {
+		if (dragTab_ < currTab) {
+			this->insertTab(isOnTheLeft ? currTab : currTab+1, this->tabIcon(dragTab_), this->tabText(dragTab_));
+			this->removeTab(dragTab_);
+			emit tabMoved(dragTab_, isOnTheLeft ? currTab-1 : currTab);
 		}
-	} else if (event->button() == Qt::MidButton) {
-		if (tabno != -1) {
-			emit mouseMiddleClickTab(tabno);
+		else {
+			this->insertTab(isOnTheLeft ? currTab : currTab+1, this->tabIcon(dragTab_), this->tabText(dragTab_));
+			this->removeTab(dragTab_+1);
+			emit tabMoved(dragTab_, isOnTheLeft ? currTab : currTab+1);
 		}
 	}
+
+	dragTab_ = -1;
+	currTab = -1;
+	this->update();
 	event->accept();
-}
+#endif
+};

 /*
  * Used for starting drags of tabs
@@ -94,6 +126,7 @@
 		return;
 	}
 	if (!(event->buttons() & Qt::LeftButton)) {
+		currTab=-1;
 		return;
 	}
 	if ((event->pos() - dragStartPosition_).manhattanLength()
@@ -102,22 +135,21 @@
 	}

 	if (dragTab_ != -1) {
-		QDrag *drag = new QDrag(this);
-		QMimeData *mimeData = new QMimeData;
-		QByteArray data;
-		QPixmap icon;
-
-		data.setNum(dragTab_);
-
-		mimeData->setData(PSITABDRAGMIMETYPE, data);
-		drag->setMimeData(mimeData);
-		drag->setPixmap(icon);
-
-		Qt::DropAction dropAction = drag->start(Qt::MoveAction);
-		Q_UNUSED(dropAction);
-	}
-
+#if QT_VERSION < 0x040500
+	currTab = findTabUnder(event->pos());
+	this->update(this->tabRect(currTab));
+
+	if (tabRect(currTab).center().x() - event->pos().x() > 0)
+		isOnTheLeft = 1;
+	else
+		isOnTheLeft = 0;
+#endif
+	}
+#if QT_VERSION >= 0x040500
+	QTabBar::mouseMoveEvent(event);
+#else
 	event->accept();
+#endif
 }

 void PsiTabBar::contextMenuEvent(QContextMenuEvent *event) {
@@ -126,6 +158,9 @@
 }

 void PsiTabBar::wheelEvent(QWheelEvent *event) {
+	if (PsiOptions::instance()->getOption("options.ui.tabs.disable-wheel-scroll").toBool())
+		return;
+
 	int numDegrees = event->delta() / 8;
 	int numSteps = numDegrees / 15;

@@ -147,3 +182,28 @@
 void PsiTabBar::setDragsEnabled(bool enabled) {
 	dragsEnabled_ = enabled;
 }
+
+void PsiTabBar::paintEvent(QPaintEvent *event)
+{
+	QTabBar::paintEvent(event);
+#if QT_VERSION < 0x040500
+	QPainter painter(this);
+	QPen pen(Qt::cyan);
+	pen.setWidth(4);
+	pen.setStyle(Qt::SolidLine);
+	painter.setPen(pen);
+	if ( currTab != -1) {
+		if (isOnTheLeft)
+			painter.drawLine(tabRect(currTab).topLeft().x()+3, tabRect(currTab).topLeft().y(),
+			tabRect(currTab).bottomLeft().x()+3, tabRect(currTab ).bottomLeft().y());
+		else
+			painter.drawLine(tabRect(currTab).topRight().x(), tabRect(currTab).topRight().y(),
+			tabRect(currTab).bottomRight().x(), tabRect(currTab ).bottomRight().y());
+	}
+#endif
+};
+
+void PsiTabBar::resizeEvent(QResizeEvent * event)
+{
+	QTabBar::resizeEvent(event);
+};
Index: psi/src/widgets/psitabbar.h
===================================================================
--- psi.orig/src/widgets/psitabbar.h	2013-01-15 14:04:08.828826322 +0400
+++ psi/src/widgets/psitabbar.h	2013-01-15 14:04:08.824826321 +0400
@@ -41,6 +41,9 @@
 	void mouseDoubleClickTab(int tab);
 	void mouseMiddleClickTab(int tab);
 	void tabDropped(int tab, PsiTabBar *source);
+#if QT_VERSION <= 0x040500
+	void tabMoved(int from, int to);
+#endif
 	// context menu on the blank space will have tab==-1
 	void contextMenu(QContextMenuEvent *event, int tab);

@@ -52,11 +55,16 @@
 	void mousePressEvent(QMouseEvent *event);
 	void contextMenuEvent(QContextMenuEvent *event);
 	void wheelEvent(QWheelEvent *event);
+	void paintEvent(QPaintEvent *event);
+	void mouseReleaseEvent ( QMouseEvent * event );
+	void resizeEvent(QResizeEvent * event);

 private:
 	int findTabUnder(const QPoint &pos);
 	QPoint dragStartPosition_;
 	int dragTab_;
+	int currTab;
+	bool isOnTheLeft;
 	bool dragsEnabled_;
 };

Index: psi/src/widgets/psitabwidget.cpp
===================================================================
--- psi.orig/src/widgets/psitabwidget.cpp	2013-01-15 14:04:08.828826322 +0400
+++ psi/src/widgets/psitabwidget.cpp	2013-01-15 14:04:08.824826321 +0400
@@ -79,11 +79,18 @@
 		closeButton_->hide();
 		downButton_->hide();
 	}
+#if QT_VERSION >= 0x040500
+	if (!PsiOptions::instance()->getOption("options.ui.tabs.show-tab-close-buttons").toBool()){
+		tabBar_->setTabsClosable(false);
+	}
+#endif
 	connect( tabBar_, SIGNAL(mouseDoubleClickTab(int)), SLOT(mouseDoubleClickTab(int)));
 	connect( tabBar_, SIGNAL(mouseMiddleClickTab(int)), SLOT(mouseMiddleClickTab(int)));
 	connect( tabBar_, SIGNAL( currentChanged(int)), SLOT(tab_currentChanged(int)));
 	connect( tabBar_, SIGNAL( contextMenu(QContextMenuEvent*,int)), SLOT( tab_contextMenu(QContextMenuEvent*,int)));
 	connect( closeButton_, SIGNAL(clicked()), SIGNAL(closeButtonClicked()));
+	connect(tabBar_, SIGNAL(tabMoved(int,int)),SLOT(widgetMoved(int,int)));
+	connect(tabBar_, SIGNAL(tabCloseRequested(int)),SIGNAL(tabCloseRequested(int)));
 }

 void PsiTabWidget::setCloseIcon(const QIcon& icon) {
@@ -348,3 +355,25 @@
 void PsiTabWidget::setDragsEnabled(bool enabled) {
 	((PsiTabBar *)tabBar_)->setDragsEnabled(enabled);
 }
+
+void PsiTabWidget::widgetMoved(int from, int to)
+{
+	if (from > to) {
+		stacked_->removeWidget(widgets_[from]);
+		widgets_.insert(to, 1, widgets_[from]);
+		widgets_.remove(from+1);
+		stacked_->insertWidget(to,widgets_[to]);
+	}
+	else {
+		stacked_->removeWidget(widgets_[from]);
+		widgets_.insert(to+1, 1, widgets_[from]);
+		widgets_.remove(from,1);
+		stacked_->insertWidget(to,widgets_[to]);
+	}
+
+#if QT_VERSION <= 0x040500
+	setCurrentPage(to);
+#endif
+	emit currentChanged(currentPage());
+
+};
Index: psi/src/widgets/psitabwidget.h
===================================================================
--- psi.orig/src/widgets/psitabwidget.h	2013-01-15 14:04:08.828826322 +0400
+++ psi/src/widgets/psitabwidget.h	2013-01-15 14:04:08.824826321 +0400
@@ -74,6 +74,7 @@
 	void currentChanged(QWidget *selected);
 	void closeButtonClicked();
 	void aboutToShowMenu(QMenu *);
+	void tabCloseRequested(int i);

 	// context menu on the blank space will have tab==-1
 	void tabContextMenu(int tab, QPoint pos, QContextMenuEvent *event);
@@ -85,6 +86,7 @@
 	void tab_contextMenu(QContextMenuEvent *event, int tab);
 	void menu_aboutToShow();
 	void menu_triggered(QAction *act);
+	void widgetMoved(int from, int to);

 private:
 	QVector<QWidget*> widgets_;
