Entering 'iris'
Entering 'src/libpsi'
diff --git a/src/accountadddlg.cpp b/src/accountadddlg.cpp
index ff1d569..68181ba 100644
--- a/src/accountadddlg.cpp
+++ b/src/accountadddlg.cpp
@@ -28,8 +28,11 @@
 AccountAddDlg::AccountAddDlg(PsiCon *_psi, QWidget *parent)
 :QDialog(parent)
 {
-  	setupUi(this);
+	setupUi(this);
 	setModal(false);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	psi = _psi;
 	psi->dialogRegister(this);
 
diff --git a/src/accountmanagedlg.cpp b/src/accountmanagedlg.cpp
index aead779..406eb90 100644
--- a/src/accountmanagedlg.cpp
+++ b/src/accountmanagedlg.cpp
@@ -95,6 +95,9 @@ AccountRemoveDlg::AccountRemoveDlg(const UserAccount &acc, QWidget *parent)
 {
 	setupUi(this);
 	setModal(false);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	d = new Private;
 	d->acc = acc;
 
@@ -288,6 +291,7 @@ AccountManageDlg::AccountManageDlg(PsiCon *_psi)
 {
 	setupUi(this);
 	setModal(false);
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
 	psi = _psi;
 	psi->dialogRegister(this);
 
diff --git a/src/accountmodifydlg.cpp b/src/accountmodifydlg.cpp
index 40a2fa4..3c07b56 100644
--- a/src/accountmodifydlg.cpp
+++ b/src/accountmodifydlg.cpp
@@ -67,6 +67,9 @@ void AccountModifyDlg::init()
 #ifndef Q_OS_MAC
 	setWindowIcon(IconsetFactory::icon("psi/account").icon());
 #endif
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 
 	le_pass->setEnabled(true);
 	le_host->setEnabled(false);
diff --git a/src/adduserdlg.cpp b/src/adduserdlg.cpp
index 868eb52..5af0df6 100644
--- a/src/adduserdlg.cpp
+++ b/src/adduserdlg.cpp
@@ -109,6 +109,9 @@ void AddUserDlg::init(const QStringList &groups, PsiAccount *pa)
 
 	setWindowTitle(CAP(windowTitle()));
 	setWindowIcon(IconsetFactory::icon("psi/addContact").icon());
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 
 	d->busy = busy;
 
diff --git a/src/bookmarkmanagedlg.cpp b/src/bookmarkmanagedlg.cpp
index 863192b..c09a0f1 100644
--- a/src/bookmarkmanagedlg.cpp
+++ b/src/bookmarkmanagedlg.cpp
@@ -33,6 +33,9 @@ BookmarkManageDlg::BookmarkManageDlg(PsiAccount* account)
 	, model_(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose, true);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 	account_->dialogRegister(this);
 
diff --git a/src/changepwdlg.cpp b/src/changepwdlg.cpp
index 06fbb4e..3a8d0fe 100644
--- a/src/changepwdlg.cpp
+++ b/src/changepwdlg.cpp
@@ -40,6 +40,9 @@ ChangePasswordDlg::ChangePasswordDlg(PsiAccount *_pa, QWidget *parent)
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose, true);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 
 	pa = _pa;
diff --git a/src/discodlg.cpp b/src/discodlg.cpp
index 1e4de07..db205e3 100644
--- a/src/discodlg.cpp
+++ b/src/discodlg.cpp
@@ -1391,6 +1391,9 @@ DiscoDlg::DiscoDlg(PsiAccount *pa, const Jid &jid, const QString &node)
 	: QDialog(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setupUi(this);
 	// restore options
 	ck_autoItems->setChecked(PsiOptions::instance()->getOption("options.ui.service-discovery.automatically-get-items").toBool());
diff --git a/src/groupchatdlg.cpp b/src/groupchatdlg.cpp
index 97e8e59..6f51878 100644
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -1978,6 +1978,9 @@ GCFindDlg::GCFindDlg(const QString& str, QWidget* parent)
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+        setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setWindowTitle(tr("Find"));
 	QVBoxLayout *vb = new QVBoxLayout(this);
 	vb->setMargin(4);
diff --git a/src/infodlg.cpp b/src/infodlg.cpp
index 318408e..6b27ebf 100644
--- a/src/infodlg.cpp
+++ b/src/infodlg.cpp
@@ -127,6 +127,9 @@ InfoDlg::InfoDlg(int type, const Jid &j, const VCard &vcard, PsiAccount *pa, QWi
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 	d = new Private;
 	setModal(false);
diff --git a/src/mucjoindlg.cpp b/src/mucjoindlg.cpp
index 3c8c0a9..51b63b3 100644
--- a/src/mucjoindlg.cpp
+++ b/src/mucjoindlg.cpp
@@ -34,6 +34,9 @@ MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiAccount* pa)
 	: QDialog(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setModal(false);
 	ui_.setupUi(this);
 	controller_ = psi;
diff --git a/src/options/optionsdlg.cpp b/src/options/optionsdlg.cpp
index 77d4c3e..44ae93c 100644
--- a/src/options/optionsdlg.cpp
+++ b/src/options/optionsdlg.cpp
@@ -491,6 +491,9 @@ OptionsDlg::OptionsDlg(PsiCon *psi, QWidget *parent)
 	pb_apply = buttonBox->button(QDialogButtonBox::Apply);
 
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint | Qt::WindowContextHelpButtonHint);
+#endif
 	d = new Private(this, psi);
 	setModal(false);
 	d->psi->dialogRegister(this);
diff --git a/src/privacy/privacydlg.cpp b/src/privacy/privacydlg.cpp
index b09e224..92502e4 100644
--- a/src/privacy/privacydlg.cpp
+++ b/src/privacy/privacydlg.cpp
@@ -34,6 +34,9 @@ PrivacyDlg::PrivacyDlg(const QString& account_name, PrivacyManager* manager, QWi
 	ui_.setupUi(this);
 	setAttribute(Qt::WA_DeleteOnClose);
 	setWindowTitle(tr("%1: Privacy Lists").arg(account_name));
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 
 	connect(manager_,SIGNAL(listsReceived(const QString&, const QString&, const QStringList&)),SLOT(updateLists(const QString&, const QString&, const QStringList&)));
 	connect(manager_,SIGNAL(listReceived(const PrivacyList&)),SLOT(refreshList(const PrivacyList&)));
diff --git a/src/searchdlg.cpp b/src/searchdlg.cpp
index 3c98c95..bb42ae0 100644
--- a/src/searchdlg.cpp
+++ b/src/searchdlg.cpp
@@ -177,6 +177,9 @@ SearchDlg::SearchDlg(const Jid &jid, PsiAccount *pa)
 
 	setWindowTitle(windowTitle().arg(d->jid.full()));
 
+	#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+	#endif
 	d->busy = busy;
 
 	d->gr_form = new QWidget(gb_search);
diff --git a/src/vcardphotodlg.cpp b/src/vcardphotodlg.cpp
index 02a5b0e..4fef2ac 100644
--- a/src/vcardphotodlg.cpp
+++ b/src/vcardphotodlg.cpp
@@ -13,6 +13,9 @@ ShowPhotoDlg::ShowPhotoDlg(QWidget *parent, QPixmap &pixmap)
 	: QDialog(parent), initSize(true)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	photoPixmap = pixmap;
 	setWindowTitle(QString(tr("Photo Preview: %1")).arg(parent->windowTitle()));
 
