Entering 'iris'
Entering 'src/libpsi'
Index: psi/src/accountadddlg.cpp
===================================================================
--- psi.orig/src/accountadddlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/accountadddlg.cpp	2013-01-15 14:04:09.068826331 +0400
@@ -28,8 +28,11 @@
 AccountAddDlg::AccountAddDlg(PsiCon *_psi, QWidget *parent)
 :QDialog(parent)
 {
-  	setupUi(this);
+	setupUi(this);
 	setModal(false);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	psi = _psi;
 	psi->dialogRegister(this);

Index: psi/src/accountmanagedlg.cpp
===================================================================
--- psi.orig/src/accountmanagedlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/accountmanagedlg.cpp	2013-01-15 14:04:09.068826331 +0400
@@ -95,6 +95,9 @@
 {
 	setupUi(this);
 	setModal(false);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	d = new Private;
 	d->acc = acc;

@@ -288,6 +291,7 @@
 {
 	setupUi(this);
 	setModal(false);
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
 	psi = _psi;
 	psi->dialogRegister(this);

Index: psi/src/accountmodifydlg.cpp
===================================================================
--- psi.orig/src/accountmodifydlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/accountmodifydlg.cpp	2013-01-15 14:04:09.068826331 +0400
@@ -67,6 +67,9 @@
 #ifndef Q_OS_MAC
 	setWindowIcon(IconsetFactory::icon("psi/account").icon());
 #endif
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif

 	le_pass->setEnabled(true);
 	le_host->setEnabled(false);
Index: psi/src/adduserdlg.cpp
===================================================================
--- psi.orig/src/adduserdlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/adduserdlg.cpp	2013-01-15 14:04:09.068826331 +0400
@@ -109,6 +109,9 @@

 	setWindowTitle(CAP(windowTitle()));
 	setWindowIcon(IconsetFactory::icon("psi/addContact").icon());
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif

 	d->busy = busy;

Index: psi/src/bookmarkmanagedlg.cpp
===================================================================
--- psi.orig/src/bookmarkmanagedlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/bookmarkmanagedlg.cpp	2013-01-15 14:04:09.068826331 +0400
@@ -33,6 +33,9 @@
 	, model_(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose, true);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 	account_->dialogRegister(this);

Index: psi/src/changepwdlg.cpp
===================================================================
--- psi.orig/src/changepwdlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/changepwdlg.cpp	2013-01-15 14:04:09.068826331 +0400
@@ -40,6 +40,9 @@
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose, true);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);

 	pa = _pa;
Index: psi/src/discodlg.cpp
===================================================================
--- psi.orig/src/discodlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/discodlg.cpp	2013-01-15 14:04:09.068826331 +0400
@@ -1391,6 +1391,9 @@
 	: QDialog(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setupUi(this);
 	// restore options
 	ck_autoItems->setChecked(PsiOptions::instance()->getOption("options.ui.service-discovery.automatically-get-items").toBool());
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:09.072826331 +0400
@@ -1978,6 +1978,9 @@
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+        setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setWindowTitle(tr("Find"));
 	QVBoxLayout *vb = new QVBoxLayout(this);
 	vb->setMargin(4);
Index: psi/src/infodlg.cpp
===================================================================
--- psi.orig/src/infodlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/infodlg.cpp	2013-01-15 14:04:09.072826331 +0400
@@ -127,6 +127,9 @@
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 	d = new Private;
 	setModal(false);
Index: psi/src/mucjoindlg.cpp
===================================================================
--- psi.orig/src/mucjoindlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/mucjoindlg.cpp	2013-01-15 14:04:09.072826331 +0400
@@ -34,6 +34,9 @@
 	: QDialog(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setModal(false);
 	ui_.setupUi(this);
 	controller_ = psi;
Index: psi/src/options/optionsdlg.cpp
===================================================================
--- psi.orig/src/options/optionsdlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/options/optionsdlg.cpp	2013-01-15 14:04:09.072826331 +0400
@@ -491,6 +491,9 @@
 	pb_apply = buttonBox->button(QDialogButtonBox::Apply);

 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint | Qt::WindowContextHelpButtonHint);
+#endif
 	d = new Private(this, psi);
 	setModal(false);
 	d->psi->dialogRegister(this);
Index: psi/src/privacy/privacydlg.cpp
===================================================================
--- psi.orig/src/privacy/privacydlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/privacy/privacydlg.cpp	2013-01-15 14:04:09.072826331 +0400
@@ -34,6 +34,9 @@
 	ui_.setupUi(this);
 	setAttribute(Qt::WA_DeleteOnClose);
 	setWindowTitle(tr("%1: Privacy Lists").arg(account_name));
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif

 	connect(manager_,SIGNAL(listsReceived(const QString&, const QString&, const QStringList&)),SLOT(updateLists(const QString&, const QString&, const QStringList&)));
 	connect(manager_,SIGNAL(listReceived(const PrivacyList&)),SLOT(refreshList(const PrivacyList&)));
Index: psi/src/searchdlg.cpp
===================================================================
--- psi.orig/src/searchdlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/searchdlg.cpp	2013-01-15 14:04:09.072826331 +0400
@@ -177,6 +177,9 @@

 	setWindowTitle(windowTitle().arg(d->jid.full()));

+	#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+	#endif
 	d->busy = busy;

 	d->gr_form = new QWidget(gb_search);
Index: psi/src/vcardphotodlg.cpp
===================================================================
--- psi.orig/src/vcardphotodlg.cpp	2013-01-15 14:04:09.076826332 +0400
+++ psi/src/vcardphotodlg.cpp	2013-01-15 14:04:09.072826331 +0400
@@ -13,6 +13,9 @@
 	: QDialog(parent), initSize(true)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	photoPixmap = pixmap;
 	setWindowTitle(QString(tr("Photo Preview: %1")).arg(parent->windowTitle()));

