--- psi.orig/src/accountadddlg.cpp
+++ psi/src/accountadddlg.cpp
@@ -28,8 +28,11 @@
 AccountAddDlg::AccountAddDlg(PsiCon *_psi, QWidget *parent)
 :QDialog(parent)
 {
-  	setupUi(this);
+	setupUi(this);
 	setModal(false);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	psi = _psi;
 	psi->dialogRegister(this);
 
--- psi.orig/src/accountmanagedlg.cpp
+++ psi/src/accountmanagedlg.cpp
@@ -96,6 +96,9 @@ AccountRemoveDlg::AccountRemoveDlg(Proxy
 {
 	setupUi(this);
 	setModal(false);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	d = new Private;
 	d->acc = acc;
 	d->proxyman = proxyman;
@@ -291,8 +294,11 @@ private slots:
 AccountManageDlg::AccountManageDlg(PsiCon *_psi)
 :QDialog(0)
 {
-  	setupUi(this);
+	setupUi(this);
 	setModal(false);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	psi = _psi;
 	psi->dialogRegister(this);
 
--- psi.orig/src/accountmodifydlg.cpp
+++ psi/src/accountmodifydlg.cpp
@@ -67,6 +67,9 @@ void AccountModifyDlg::init() 
 #ifndef Q_WS_MAC
 	setWindowIcon(IconsetFactory::icon("psi/account").icon());
 #endif
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 
 	le_pass->setEnabled(true);
 	le_host->setEnabled(false);
--- psi.orig/src/adduserdlg.cpp
+++ psi/src/adduserdlg.cpp
@@ -109,6 +109,9 @@ void AddUserDlg::init(const QStringList 
 
 	setWindowTitle(CAP(windowTitle()));
 	setWindowIcon(IconsetFactory::icon("psi/addContact").icon());
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 
 	d->busy = busy;
 
--- psi.orig/src/bookmarkmanagedlg.cpp
+++ psi/src/bookmarkmanagedlg.cpp
@@ -33,6 +33,9 @@ BookmarkManageDlg::BookmarkManageDlg(Psi
 	, model_(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose, true);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 	account_->dialogRegister(this);
 
--- psi.orig/src/changepwdlg.cpp
+++ psi/src/changepwdlg.cpp
@@ -40,6 +40,9 @@ ChangePasswordDlg::ChangePasswordDlg(Psi
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose, true);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 
 	pa = _pa;
--- psi.orig/src/discodlg.cpp
+++ psi/src/discodlg.cpp
@@ -1393,6 +1393,9 @@ DiscoDlg::DiscoDlg(PsiAccount *pa, const
 	: QDialog(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setupUi(this);
 	// restore options
 	ck_autoItems->setChecked(PsiOptions::instance()->getOption("options.ui.service-discovery.automatically-get-items").toBool());
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -1977,6 +1977,9 @@ GCFindDlg::GCFindDlg(const QString& str,
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+        setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setWindowTitle(tr("Find"));
 	QVBoxLayout *vb = new QVBoxLayout(this);
 	vb->setMargin(4);
--- psi.orig/src/infodlg.cpp
+++ psi/src/infodlg.cpp
@@ -125,6 +125,9 @@ InfoDlg::InfoDlg(int type, const Jid &j,
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	ui_.setupUi(this);
 	d = new Private;
 	setModal(false);
--- psi.orig/src/mucjoindlg.cpp
+++ psi/src/mucjoindlg.cpp
@@ -34,6 +34,9 @@ MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiA
 	: QDialog(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	setModal(false);
 	ui_.setupUi(this);
 	controller_ = psi;
--- psi.orig/src/options/optionsdlg.cpp
+++ psi/src/options/optionsdlg.cpp
@@ -489,6 +489,9 @@ OptionsDlg::OptionsDlg(PsiCon *psi, QWid
 	pb_apply = buttonBox->button(QDialogButtonBox::Apply);
 
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint | Qt::WindowContextHelpButtonHint);
+#endif
 	d = new Private(this, psi);
 	setModal(false);
 	d->psi->dialogRegister(this);
--- psi.orig/src/privacy/privacydlg.cpp
+++ psi/src/privacy/privacydlg.cpp
@@ -34,6 +34,9 @@ PrivacyDlg::PrivacyDlg(const QString& ac
 	ui_.setupUi(this);
 	setAttribute(Qt::WA_DeleteOnClose);
 	setWindowTitle(tr("%1: Privacy Lists").arg(account_name));
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 
 	connect(manager_,SIGNAL(listsReceived(const QString&, const QString&, const QStringList&)),SLOT(updateLists(const QString&, const QString&, const QStringList&)));
 	connect(manager_,SIGNAL(listReceived(const PrivacyList&)),SLOT(refreshList(const PrivacyList&)));
--- psi.orig/src/searchdlg.cpp
+++ psi/src/searchdlg.cpp
@@ -177,6 +177,9 @@ SearchDlg::SearchDlg(const Jid &jid, Psi
 
 	setWindowTitle(windowTitle().arg(d->jid.full()));
 
+	#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+	#endif
 	d->busy = busy;
 
 	d->gr_form = new QWidget(gb_search);
--- psi.orig/src/vcardphotodlg.cpp
+++ psi/src/vcardphotodlg.cpp
@@ -13,6 +13,9 @@ ShowPhotoDlg::ShowPhotoDlg(QWidget *pare
 	: QDialog(parent), initSize(true)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+#if QT_VERSION >= 0x040500
+	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
+#endif
 	photoPixmap = pixmap;
 	setWindowTitle(QString(tr("Photo Preview: %1")).arg(parent->windowTitle()));
 
