Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:09.628826354 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:09.624826354 +0400
@@ -2003,11 +2003,11 @@
 	hb->addStretch(1);
 	QPushButton *pb_close = new QPushButton(tr("&Close"), this);
 	connect(pb_close, SIGNAL(clicked()), SLOT(close()));
-	hb->addWidget(pb_close);
 	QPushButton *pb_find = new QPushButton(tr("&Find"), this);
 	pb_find->setDefault(true);
 	connect(pb_find, SIGNAL(clicked()), SLOT(doFind()));
 	hb->addWidget(pb_find);
+	hb->addWidget(pb_close);
 	pb_find->setAutoDefault(true);

 	resize(200, minimumSizeHint().height());
