Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:10.324826382 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:10.320826382 +0400
@@ -2385,6 +2385,8 @@
 			//ur.setSecurityEnabled(true);
 			if(local)
 				ur.setClient(ApplicationInfo::name(),ApplicationInfo::version(),SystemInfo::instance()->os());
+			else
+				ur.setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
 
 			u->userResourceList().append(ur);
 			rp = &u->userResourceList().last();
@@ -2454,8 +2456,8 @@
 		foreach(UserListItem* u, findRelevant(j)) {
 			UserResourceList::Iterator rit = u->userResourceList().find(j.resource());
 			if (rit != u->userResourceList().end()) {
-				//(*rit).setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
-				(*rit).setClient(QString(),QString(),"");
+				(*rit).setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
+				//(*rit).setClient(QString(),QString(),"");
 				cpUpdate(*u,(*rit).name());
 			}
 		}
@@ -2965,8 +2967,8 @@
 		bool found = (rit == u->userResourceList().end()) ? false: true;
 		if(!found)
 			continue;
-		//(*rit).setClient(name,version,"");
-		(*rit).setClient(QString(),QString(),"");
+		(*rit).setClient(name,version,"");
+		//(*rit).setClient(QString(),QString(),"");
 		cpUpdate(*u);
 	}
 }
Index: psi/src/gcuserview.cpp
===================================================================
--- psi.orig/src/gcuserview.cpp	2013-01-15 14:04:10.324826382 +0400
+++ psi/src/gcuserview.cpp	2013-01-15 14:04:10.320826382 +0400
@@ -366,8 +366,8 @@
 	UserResource ur;
 	ur.setName(nick);
 	ur.setStatus(s);
-	//ur.setClient(client_name,client_version,"");
-	ur.setClient(QString(),QString(),"");
+	ur.setClient(client_name,client_version,"");
+	//ur.setClient(QString(),QString(),"");
 	u.userResourceList().append(ur);
 
 	PsiToolTip::showText(mapToGlobal(pos), u.makeTip(), this);
