--- psi.orig/src/psiaccount.cpp
+++ psi/src/psiaccount.cpp
@@ -2422,8 +2422,8 @@ void PsiAccount::client_resourceAvailabl
 		foreach(UserListItem* u, findRelevant(j)) {
 			UserResourceList::Iterator rit = u->userResourceList().find(j.resource());
 			if (rit != u->userResourceList().end()) {
-				//(*rit).setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
-				(*rit).setClient(QString(),QString(),"");
+				(*rit).setClient(capsManager()->clientName(j),capsManager()->clientVersion(j),"");
+				//(*rit).setClient(QString(),QString(),"");
 				cpUpdate(*u,(*rit).name());
 			}
 		}
@@ -2925,8 +2925,8 @@ void PsiAccount::capsChanged(const Jid& 
 		bool found = (rit == u->userResourceList().end()) ? false: true;
 		if(!found)
 			continue;
-		//(*rit).setClient(name,version,"");
-		(*rit).setClient(QString(),QString(),"");
+		(*rit).setClient(name,version,"");
+		//(*rit).setClient(QString(),QString(),"");
 		cpUpdate(*u);
 	}
 }
--- psi.orig/src/gcuserview.cpp
+++ psi/src/gcuserview.cpp
@@ -363,8 +363,8 @@ bool GCUserView::maybeTip(const QPoint &
 	UserResource ur;
 	ur.setName(nick);
 	ur.setStatus(s);
-	//ur.setClient(client_name,client_version,"");
-	ur.setClient(QString(),QString(),"");
+	ur.setClient(client_name,client_version,"");
+	//ur.setClient(QString(),QString(),"");
 	u.userResourceList().append(ur);
 
 	PsiToolTip::showText(mapToGlobal(pos), u.makeTip(), this);
