--- git.orig/src/mainwin.cpp
+++ git/src/mainwin.cpp
@@ -119,7 +119,7 @@ public:
 	bool isLeftRoster;
 
 	PopupAction* optionsButton, *statusButton;
-	IconActionGroup* statusGroup;
+	IconActionGroup* statusGroup, *viewGroups;
 	EventNotifierAction* eventNotifier;
 	PsiCon* psi;
 	MainWin* mainWin;
@@ -154,6 +154,7 @@ MainWin::Private::Private(PsiCon* _psi, 
 {
 
 	statusGroup   = (IconActionGroup *)getAction("status_all");
+	viewGroups    = (IconActionGroup *)getAction("view_groups");
 	eventNotifier = (EventNotifierAction *)getAction("event_notifier");
 
 	optionsButton = (PopupAction *)getAction("button_options");
@@ -844,6 +845,7 @@ void MainWin::buildOptionsMenu()
 {
 	buildGeneralMenu( d->optionsMenu );
 	d->optionsMenu->addSeparator();
+	d->optionsMenu->addAction(d->viewGroups);
 
 	// help menu
 	QMenu* helpMenu = new QMenu(tr("&Help"), d->optionsMenu);
