Index: psi/src/mainwin.cpp
===================================================================
--- psi.orig/src/mainwin.cpp	2013-01-15 14:04:11.580826433 +0400
+++ psi/src/mainwin.cpp	2013-01-15 14:04:11.576826433 +0400
@@ -118,7 +118,7 @@
 	bool isLeftRoster;
 
 	PopupAction* optionsButton, *statusButton;
-	IconActionGroup* statusGroup;
+	IconActionGroup* statusGroup, *viewGroups;
 	EventNotifierAction* eventNotifier;
 	PsiCon* psi;
 	MainWin* mainWin;
@@ -153,6 +153,7 @@
 {
 
 	statusGroup   = (IconActionGroup *)getAction("status_all");
+	viewGroups    = (IconActionGroup *)getAction("view_groups");
 	eventNotifier = (EventNotifierAction *)getAction("event_notifier");
 
 	optionsButton = (PopupAction *)getAction("button_options");
@@ -841,6 +842,7 @@
 {
 	buildGeneralMenu( d->optionsMenu );
 	d->optionsMenu->addSeparator();
+	d->optionsMenu->addAction(d->viewGroups);
 
 	// help menu
 	QMenu* helpMenu = new QMenu(tr("&Help"), d->optionsMenu);
