--- git.temp.orig/src/options/opt_iconset.cpp
+++ git.temp/src/options/opt_iconset.cpp
@@ -1146,12 +1146,22 @@ QWidget *OptionsTabIconsetRoster::widget
 
 	// Initialize transport iconsets
 	addService("aim", "AIM");
+	addService("muc", "Chat (MUC)");
 	addService("gadugadu", "Gadu-Gadu");
 	addService("icq", "ICQ");
+	addService("irc", "IRC");
+	addService("disk", "Jabber Disk");
+	addService("mrim", "Mail.Ru-IM");
 	addService("msn", "MSN");
+	addService("rss", "RSS");
 	addService("sms", "SMS");
-	addService("yahoo", "Yahoo!");
+	addService("smtp", "SMTP");
 	addService("transport", tr("Transport"));
+	addService("vkontakte", tr("vk.com"));
+	addService("weather", tr("Weather"));
+	addService("xmpp", "XMPP");
+	addService("yahoo", "Yahoo!");
+
 	d->tw_isServices->resizeColumnToContents(0);
 
 	connect(d->pb_defRosterDetails, SIGNAL(clicked()), SLOT(defaultDetails()));
--- git.temp.orig/src/psiiconset.cpp
+++ git.temp/src/psiiconset.cpp
@@ -132,11 +132,20 @@ public:
 		if ( jid.node().isEmpty() || PsiOptions::instance()->getOption("options.ui.contactlist.use-transport-icons").toBool() ) {
 			QMap<QString, QRegExp> services;
 			services["aim"]		= QRegExp("^aim");
+			services["disk"]		= QRegExp("^disk");
 			services["gadugadu"]	= QRegExp("^gg");
-			services["icq"]		= QRegExp("^icq");
+			services["icq"]		= QRegExp("^icq|^jit");
+			services["irc"]		= QRegExp("^irc");
+			services["xmpp"]		= QRegExp("^j2j|^xmpp\\.[a-z1-9]+\\..*");
+			services["mrim"]		= QRegExp("^mrim");
 			services["msn"]		= QRegExp("^msn");
-			services["yahoo"]	= QRegExp("^yahoo");
+			services["muc"]		= QRegExp("^conference|^rooms");
+			services["rss"]		= QRegExp("^rss");
 			services["sms"]		= QRegExp("^sms");
+			services["smtp"]		= QRegExp("^smtp");
+			services["vkontakte"]		= QRegExp("^vk.com|^vkontakte|^vk-t");
+			services["weather"]		= QRegExp("^weather|^gism");
+			services["yahoo"]	= QRegExp("^yahoo");
 
 			bool found = false;
 
