From 414e49698a7fb64741d4b64ce01139a4bb31db2a Mon Sep 17 00:00:00 2001
From: Julius Schwartzenberg <julius.schwartzenberg@gmail.com>
Date: Wed, 24 Feb 2010 00:12:51 +0100
Subject: Allow multiple payload types

---
 src/avcall/avcall.cpp |   63 ++++++++++++++++++------------------------------
 1 files changed, 24 insertions(+), 39 deletions(-)

Index: psi/src/avcall/avcall.cpp
===================================================================
--- psi.orig/src/avcall/avcall.cpp	2013-01-15 14:04:13.136826497 +0400
+++ psi/src/avcall/avcall.cpp	2013-01-15 14:04:13.132826497 +0400
@@ -625,35 +625,38 @@
 
 		printf("rtp_started\n");
 
-		PsiMedia::PayloadInfo audio, *pAudio;
-		PsiMedia::PayloadInfo video, *pVideo;
-
-		pAudio = 0;
-		pVideo = 0;
-		if(transmitAudio)
+		if(!incoming)
+		{
+			sess = manager->rtpManager->createOutgoing();
+			setup_sess();
+		}
+		
+		if(transmitAudio && !rtp.localAudioPayloadInfo().isEmpty())
 		{
-			// confirm transmitting of audio is actually possible,
-			//   in the case that a file is used as input
-			if(rtp.canTransmitAudio())
+			QList<JingleRtpPayloadType> pis;
+			foreach(PsiMedia::PayloadInfo pi, rtp.localAudioPayloadInfo())
 			{
-				audio = rtp.localAudioPayloadInfo().first();
-				pAudio = &audio;
+				JingleRtpPayloadType pt = payloadInfoToPayloadType(pi);
+				pis << pt;
 			}
-			else
-				transmitAudio = false;
+			sess->setLocalAudioPayloadTypes(pis);
 		}
-		if(transmitVideo)
+		else
+			transmitAudio = false;
+
+		if(transmitVideo && !rtp.localVideoPayloadInfo().isEmpty())
 		{
-			// same for video
-			if(rtp.canTransmitVideo())
+			QList<JingleRtpPayloadType> pis;
+			foreach(PsiMedia::PayloadInfo pi, rtp.localVideoPayloadInfo())
 			{
-				video = rtp.localVideoPayloadInfo().first();
-				pVideo = &video;
+				JingleRtpPayloadType pt = payloadInfoToPayloadType(pi);
+				pis << pt;
 			}
-			else
-				transmitVideo = false;
+			sess->setLocalVideoPayloadTypes(pis);
 		}
-
+		else
+			transmitVideo = false;
+		
 		if(transmitAudio && transmitVideo)
 			mode = AvCall::Both;
 		else if(transmitAudio && !transmitVideo)
@@ -667,24 +670,6 @@
 		}
 
 		if(!incoming)
-		{
-			sess = manager->rtpManager->createOutgoing();
-			setup_sess();
-		}
-
-		if(pAudio)
-		{
-			JingleRtpPayloadType pt = payloadInfoToPayloadType(*pAudio);
-			sess->setLocalAudioPayloadTypes(QList<JingleRtpPayloadType>() << pt);
-		}
-
-		if(pVideo)
-		{
-			JingleRtpPayloadType pt = payloadInfoToPayloadType(*pVideo);
-			sess->setLocalVideoPayloadTypes(QList<JingleRtpPayloadType>() << pt);
-		}
-
-		if(!incoming)
 			sess->connectToJid(peer);
 		else
 			sess->localMediaUpdate();
