Index: psi/src/mainwin.cpp
===================================================================
--- psi.orig/src/mainwin.cpp	2013-01-15 14:04:13.588826515 +0400
+++ psi/src/mainwin.cpp	2013-01-15 14:04:13.580826515 +0400
@@ -332,6 +332,8 @@
 	} else
 		setCentralWidget(rosterBar);
 
+	connect(PsiOptions::instance(), SIGNAL(optionChanged(const QString&)), SLOT(optionChanged(const QString&)));
+
 	d->vb_roster = new QVBoxLayout(rosterBar);
 #ifndef NEWCONTACTLIST
 	cvlist = new ContactView(rosterBar);
@@ -538,6 +540,14 @@
 	d->tabsSize = d->isLeftRoster ? list.last() : list.first();
 }
 
+void MainWin::optionChanged(const QString& option)
+{
+	if (option == toolbarsStateOptionPath) { //for compatibility with skins plugin
+		loadToolbarsState();
+		//buildToolbars(); //FIXME!!! not a very good solution
+	}
+}
+
 void MainWin::registerAction( IconAction* action )
 {
 	const char *activated  = SIGNAL( triggered() );
Index: psi/src/mainwin.h
===================================================================
--- psi.orig/src/mainwin.h	2013-01-15 14:04:13.588826515 +0400
+++ psi/src/mainwin.h	2013-01-15 14:04:13.580826515 +0400
@@ -147,6 +147,8 @@
 
 	void doTrayToolTip(QObject *, QPoint);
 
+	void optionChanged(const QString&);
+
 public slots:
 	void setWindowIcon(const QPixmap&);
 	void showNoFocus();
