Заходим в «iris»
Заходим в «src/libpsi»
--- a/src/discodlg.cpp
+++ b/src/discodlg.cpp
@@ -1009,7 +1009,8 @@ DiscoDlg::Private::Private(DiscoDlg *parent, PsiAccount *pa)
 
 	// create toolbar
 	toolBar = new QToolBar(tr("Service Discovery toolbar"), dlg);
-	toolBar->setIconSize(QSize(16,16));
+	int s = PsiIconset::instance()->system().iconSize();
+	toolBar->setIconSize(QSize(s,s));
 	toolBar->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
 	replaceWidget(dlg->toolBarPlaceholder, toolBar);
 
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -716,7 +716,8 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, const Jid &j, TabManager *tabManager)
 
 	d->act_templates = new IconAction(tr("Templates"), "psi/action_templates", tr("Templates"), 0, this);
 	d->act_templates->setMenu(getTemplateMenu());
-	ui_.toolbar->setIconSize(QSize(16,16));
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.toolbar->setIconSize(QSize(s,s));
 	ui_.toolbar->addAction(d->act_clear);
 	ui_.toolbar->addAction(d->act_find);
 	ui_.toolbar->addAction(d->act_configure);
--- a/src/options/optionsdlg.cpp
+++ b/src/options/optionsdlg.cpp
@@ -25,6 +25,7 @@
 #include "fancylabel.h"
 #include "iconset.h"
 #include "iconwidget.h"
+#include "psiiconset.h"
 
 #include <QLayout>
 #include <QLabel>
@@ -88,7 +89,8 @@ public:
 			iconSize = icon.availableSizes().first();
 		else
 #endif
-			iconSize = QSize(16, 16);
+			int s = PsiIconset::instance()->system().iconSize();
+			iconSize = QSize(s,s);
 		QRect iconRect = opt.rect;
 		QRect textRect = opt.rect;
 		iconRect.setWidth(iconSize.width());
@@ -154,7 +156,8 @@ public:
 			iconSize = icon.availableSizes().first();
 		else
 #endif
-			iconSize = QSize(16, 16);
+			int s = PsiIconset::instance()->system().iconSize();
+			iconSize = QSize(s,s);
 
 		int width = iconSize.width();
 		width += 8;
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -274,6 +274,10 @@ void PsiChatDlg::setLooks()
 	ui_.splitter->optionsChanged();
 	ui_.mle->optionsChanged();
 
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.lb_status->setFixedSize(s,s);
+	ui_.lb_client->setFixedSize(s,s);
+
 	ui_.tb_pgp->hide();
 	if (smallChat_) {
 		ui_.lb_status->hide();
@@ -398,7 +402,8 @@ void PsiChatDlg::initToolButtons()
 void PsiChatDlg::initToolBar()
 {
 	ui_.toolbar->setWindowTitle(tr("Chat Toolbar"));
-	ui_.toolbar->setIconSize(QSize(16, 16));
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.toolbar->setIconSize(QSize(s, s));
 	ui_.toolbar->addAction(act_clear_);
 // typeahead find bar
 	ui_.toolbar->addAction(act_find);
--- a/src/psicontactlistviewdelegate.cpp
+++ b/src/psicontactlistviewdelegate.cpp
@@ -49,6 +49,7 @@ static const QString showStatusIconsPath = "options.ui.contactlist.show-status-i
 static const QString statusIconsOverAvatarsPath = "options.ui.contactlist.status-icon-over-avatar";
 static const QString allClientsOptionPath = "options.ui.contactlist.show-all-client-icons";
 static const QString enableGroupsOptionPath = "options.ui.contactlist.enable-groups";
+static const QString statusIconsetOptionPath = "options.iconsets.status";
 
 PsiContactListViewDelegate::PsiContactListViewDelegate(ContactListView* parent)
 	: ContactListViewDelegate(parent)
@@ -61,6 +62,8 @@ PsiContactListViewDelegate::PsiContactListViewDelegate(ContactListView* parent)
 	connect(alertTimer_, SIGNAL(timeout()), SLOT(updateAlerts()));
 
 	connect(PsiOptions::instance(), SIGNAL(optionChanged(const QString&)), SLOT(optionChanged(const QString&)));
+	connect(PsiIconset::instance(), SIGNAL(rosterIconsSizeChanged(int)), SLOT(rosterIconsSizeChanged(int)));
+	statusIconSize_ = PsiIconset::instance()->roster.value(PsiOptions::instance()->getOption(statusIconsetOptionPath).toString())->iconSize();
 	optionChanged(slimGroupsOptionPath);
 	optionChanged(outlinedGroupsOptionPath);
 	optionChanged(contactListFontOptionPath);
@@ -564,7 +567,7 @@ void PsiContactListViewDelegate::optionChanged(const QString& option)
 		font_ = new QFont();
 		font_->fromString(PsiOptions::instance()->getOption(contactListFontOptionPath).toString());
 		fontMetrics_ = new QFontMetrics(*font_);
-		rowHeight_ = qMax(fontMetrics_->height()+2, 18); // 18 - default row height
+		rowHeight_ = qMax(fontMetrics_->height()+2, statusIconSize_+2);
 		contactList()->viewport()->update();
 	}
 	else if (option == contactListBackgroundOptionPath) {
@@ -647,6 +650,13 @@ void PsiContactListViewDelegate::optionChanged(const QString& option)
 	}
 }
 
+void PsiContactListViewDelegate::rosterIconsSizeChanged(int size)
+{
+	statusIconSize_ = size;
+	rowHeight_ = qMax(fontMetrics_->height()+2, statusIconSize_+2);
+	contactList()->viewport()->update();
+}
+
 void PsiContactListViewDelegate::drawText(QPainter* painter, const QStyleOptionViewItem& o, const QRect& rect, const QString& text, const QModelIndex& index) const
 {
 	QRect r(rect);
--- a/src/psicontactlistviewdelegate.h
+++ b/src/psicontactlistviewdelegate.h
@@ -58,6 +58,7 @@ protected:
 private slots:
 	void optionChanged(const QString& option);
 	void updateAlerts();
+	void rosterIconsSizeChanged(int size);
 
 private:
 	QTimer* alertTimer_;
@@ -70,6 +71,7 @@ private:
 	int avatarSize_, avatarRadius_;
 	bool enableGroups_, allClients_;
 	mutable QHash<QModelIndex, bool> alertingIndexes_;
+	int statusIconSize_;
 };
 
 #endif
--- a/src/psiiconset.cpp
+++ b/src/psiiconset.cpp
@@ -335,6 +335,11 @@ bool PsiIconset::loadSystem()
 	QString cur_system = PsiOptions::instance()->getOption("options.iconsets.system").toString();
 	if (d->cur_system != cur_system) {
 		Iconset sys = d->systemIconset(&ok);
+
+		if(sys.iconSize() != d->system.iconSize()) {
+			emit systemIconsSizeChanged(sys.iconSize());
+		}
+
 		d->loadIconset(&d->system, &sys);
 
 		//d->system = d->systemIconset();
@@ -509,6 +514,10 @@ void PsiIconset::reloadRoster()
 	if (d->cur_status != cur_status) {
 		Iconset *newDef = d->defaultRosterIconset(&ok);
 		Iconset *oldDef = roster[d->cur_status];
+
+		if(oldDef->iconSize() != newDef->iconSize())
+			emit rosterIconsSizeChanged(newDef->iconSize());
+
 		d->loadIconset(oldDef, newDef);
 
 		roster.remove(d->cur_status);
--- a/src/psiiconset.h
+++ b/src/psiiconset.h
@@ -80,6 +80,8 @@ public:
 
 signals:
 	void emoticonsChanged();
+	void systemIconsSizeChanged(int);
+	void rosterIconsSizeChanged(int);
 
 public slots:
 	static void reset();
--- a/src/psitoolbar.cpp
+++ b/src/psitoolbar.cpp
@@ -29,6 +29,7 @@
 #include <QToolButton>
 
 #include "iconset.h"
+#include "psiiconset.h"
 #include "iconaction.h"
 #include "psioptions.h"
 #include "options/opt_toolbars.h"
@@ -63,7 +64,8 @@ PsiToolBar::PsiToolBar(const QString& base, QWidget* mainWindow, MetaActionList*
 	Q_ASSERT(mainWindow);
 	Q_ASSERT(actionList_);
 
-	setIconSize(QSize(16, 16));
+	int s = PsiIconset::instance()->system().iconSize();
+	setIconSize(QSize(s, s));
 
 	customizeAction_ = new QAction(tr("&Configure Toolbar..."), this);
 	connect(customizeAction_, SIGNAL(triggered()), this, SIGNAL(customize()));
--- a/src/tools/iconset/iconset.cpp
+++ b/src/tools/iconset/iconset.cpp
@@ -1028,6 +1028,7 @@ private:
 		//authors << "I. M. Anonymous";
 		//creation = "1900-01-01";
 		homeUrl = QString::null;
+		iconSize_ = 16;
 	}
 
 public:
@@ -1036,6 +1037,7 @@ public:
 	QHash<QString, PsiIcon *> dict; // unsorted hash for fast search
 	QList<PsiIcon *> list;          // sorted list
 	QHash<QString, QString> info;
+	int iconSize_;
 
 public:
 	Private()
@@ -1129,6 +1131,8 @@ public:
 	{
 		Q_UNUSED(dir);
 
+		iconSize_ = 16;
+
 		for (QDomNode node = i.firstChild(); !node.isNull(); node = node.nextSibling()) {
 			QDomElement e = node.toElement();
 			if( e.isNull() ) {
@@ -1139,6 +1143,10 @@ public:
 			if ( tag == "name" ) {
 				name = e.text();
 			}
+			else if( tag == "size") {
+				QString str = e.text();
+				iconSize_ = str.toInt();
+			}
 			else if ( tag == "version" ) {
 				version = e.text();
 			}
@@ -1410,6 +1418,7 @@ public:
 		filename = from.filename;
 		authors = from.authors;
 		info = from.info;
+		iconSize_ = from.iconSize_;
 	}
 };
 //! \endif
@@ -1599,6 +1608,14 @@ const QString &Iconset::name() const
 }
 
 /**
+ * Returns the icons size from Iconset.
+ */
+const int &Iconset::iconSize() const
+{
+	return d->iconSize_;
+}
+
+/**
  * Returns the Iconset version.
  */
 const QString &Iconset::version() const
--- a/src/tools/iconset/iconset.h
+++ b/src/tools/iconset/iconset.h
@@ -207,6 +207,7 @@ public:
 	const QStringList &authors() const;
 	const QString &creation() const;
 	const QString &homeUrl() const;
+	const int &iconSize() const;
 
 	const QString &fileName() const;
 	void setFileName(const QString &);
