--- psi.orig/src/mainwin.cpp
+++ psi/src/mainwin.cpp
@@ -328,6 +328,8 @@ MainWin::MainWin(bool _onTop, bool _asTo
 		d->mainTabs->resize(d->tabsGeometry);
 	} else setCentralWidget(rosterBar);
 
+	connect(PsiOptions::instance(), SIGNAL(optionChanged(const QString&)), SLOT(optionChanged(const QString&)));
+
 	d->vb_roster = new QVBoxLayout(rosterBar);
 #ifndef NEWCONTACTLIST
 	cvlist = new ContactView(rosterBar);
@@ -542,6 +544,14 @@ void MainWin::splitterMoved(int /*pos*/,
 	}
 }
 
+void MainWin::optionChanged(const QString& option)
+{
+	if (option == toolbarsStateOptionPath) { //for compatibility with skins plugin
+		loadToolbarsState();
+		//buildToolbars(); //FIXME!!! not a very good solution
+	}
+}
+
 void MainWin::registerAction( IconAction* action )
 {
 	const char *activated  = SIGNAL( triggered() );
--- psi.orig/src/mainwin.h
+++ psi/src/mainwin.h
@@ -146,6 +146,8 @@ private slots:
 
 	void doTrayToolTip(QObject *, QPoint);
 
+	void optionChanged(const QString&);
+
 public slots:
 	void setWindowIcon(const QPixmap&);
 	void showNoFocus();
