Entering 'iris'
Entering 'src/libpsi'
Index: psi/options/default.xml
===================================================================
--- psi.orig/options/default.xml	2013-01-15 14:04:13.792826523 +0400
+++ psi/options/default.xml	2013-01-15 14:04:13.784826523 +0400
@@ -113,6 +113,7 @@
 				<use-message-icons type="bool">true</use-message-icons>
 				<scaled-message-icons type="bool">false</scaled-message-icons>
 				<warn-before-clear type="bool">true</warn-before-clear>
+				<css type="QString" />
 				<only-paste-template type="bool">false</only-paste-template>
 				<disable-paste-send type="bool">false</disable-paste-send>
 			</chat>
@@ -135,6 +136,7 @@
 					<show comment="Show status messages" type="bool">false</show>
 				</status-messages>
 				<tooltip comment="Display options for the contact list tooltips">
+					<css type="QString"></css>
 					<avatar comment="Display avatar" type="bool">true</avatar>
 					<client-version comment="Display client version information" type="bool">true</client-version>
 					<geolocation comment="Display geolocation information" type="bool">true</geolocation>
@@ -143,6 +145,58 @@
 				</tooltip>
 				<temp-no-roster-animation comment="Disable animated icons. This option will be removed as soon as the contactlist code is optimized enough" type="bool">false</temp-no-roster-animation>
 				<account-sort-style type="QString">alpha</account-sort-style>
+				<css type="QString">/* frame of avatar - general settings */
+/* фрейм с аватаром - общие настройки */
+RosterAvatarFrame * {
+	margin: 1px;
+}
+
+/* frame of avatar - avatar */
+/* фрейм с аватаром - аватар */
+QLabel#lb_avatar {
+	margin-left: 0px;
+	margin-top: 2px;
+	margin-bottom: 2px;
+	border: 1px solid #949494;
+	border-radius: 3px;
+}
+
+/* frame of avatar - nick */
+/* фрейм с аватаром - ник */
+QLabel#lb_nick {
+	margin: 2px;
+	margin-right: 1px;
+	margin-bottom: 3px;
+	border: 1px solid #949494;
+	border-radius: 3px;
+}
+
+/* frame of avatar - buttons of PEP and status */
+/* фрейм с аватаром - кнопки PEP и статуса */
+QToolButton#tb_status, #tb_mood, #tb_activity {
+	margin-left: 1px;
+	margin-right: 1px;
+	margin-bottom: 2px;
+	width: 100%;
+}
+
+QToolButton#tb_status {
+	margin-top: 0px;
+	margin-bottom: 1px;
+}
+
+/* frame of avatar - button of status, arrow */
+/* фрейм с аватаром - кнопка статуса, стрелка */
+QToolButton#tb_status::menu-indicator {
+	image: url();
+}
+
+/* frame of avatar - field of status message */
+/* фрейм с аватаром - строка статусного сообщения */
+QLineEdit#le_status_text {
+	margin-right: 0px;
+	margin-bottom: 2px;
+}</css>
 				<always-on-top type="bool">false</always-on-top>
 				<automatically-resize-roster type="bool">false</automatically-resize-roster>
 				<grow-roster-upwards type="bool">true</grow-roster-upwards>
@@ -323,6 +377,7 @@
 					<message type="QString">Sans Serif,9,-1,5,50,0,0,0,0,0</message>
 					<passive-popup type="QString">Sans Serif,7,-1,5,50,0,0,0,0,0</passive-popup>
 				</font>
+				<css type="QString" />
 			</look>
 			<notifications>
 				<send-receipts type="bool" >true</send-receipts>
@@ -339,6 +394,7 @@
 						<status type="int">2000</status>
 					</delays>
 					<enabled type="bool">false</enabled>
+					<css type="QString"></css>
 					<top-to-bottom type="bool">false</top-to-bottom>
 					<at-left-corner type="bool">false</at-left-corner>
 					<incoming-chat type="bool">true</incoming-chat>
Index: psi/src/contactlistitemmenu.cpp
===================================================================
--- psi.orig/src/contactlistitemmenu.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/contactlistitemmenu.cpp	2013-01-15 14:04:13.784826523 +0400
@@ -21,12 +21,17 @@
 #include "contactlistitemmenu.h"
 
 #include "shortcutmanager.h"
+#include "psioptions.h"
 
 ContactListItemMenu::ContactListItemMenu(ContactListItem* item, ContactListModel* model)
 	: QMenu(0)
 	, item_(item)
 	, model_(model)
 {
+	const QString css = PsiOptions::instance()->getOption("options.ui.contactlist.css").toString();
+	if (!css.isEmpty()) {
+		setStyleSheet(css);
+	}
 }
 
 ContactListItemMenu::~ContactListItemMenu()
Index: psi/src/gcuserview.cpp
===================================================================
--- psi.orig/src/gcuserview.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/gcuserview.cpp	2013-01-15 14:04:13.784826523 +0400
@@ -523,6 +523,10 @@
 	pm->addAction(act);
 	act->setData(3);
 
+	const QString css = PsiOptions::instance()->getOption("options.ui.chat.css").toString();
+	if (!css.isEmpty()) {
+		pm->setStyleSheet(css);
+	}
 	int x = -1;
 	bool enabled = false;
 	act = pm->exec(QCursor::pos());
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:13.784826523 +0400
@@ -1704,6 +1704,10 @@
 
 void GCMainDlg::setLooks()
 {
+	const QString css = PsiOptions::instance()->getOption("options.ui.chat.css").toString();
+	if (!css.isEmpty()) {
+		setStyleSheet(css);
+	}
 	ui_.vsplitter->optionsChanged();
 	ui_.mle->optionsChanged();
 
Index: psi/src/mainwin.cpp
===================================================================
--- psi.orig/src/mainwin.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/mainwin.cpp	2013-01-15 14:04:13.784826523 +0400
@@ -514,6 +514,7 @@
 
 		/*QShortcut *sp_ss = new QShortcut(QKeySequence(tr("Ctrl+Shift+N")), this);
 		connect(sp_ss, SIGNAL(triggered()), SLOT(avcallConfig()));*/
+	optionChanged("options.ui.contactlist.css");
 }
 
 MainWin::~MainWin()
@@ -546,6 +547,12 @@
 		loadToolbarsState();
 		//buildToolbars(); //FIXME!!! not a very good solution
 	}
+	else if (option == "options.ui.contactlist.css") {
+		const QString css = PsiOptions::instance()->getOption("options.ui.contactlist.css").toString();
+		if (!css.isEmpty()) {
+			setStyleSheet(css);
+		}
+	}
 }
 
 void MainWin::registerAction( IconAction* action )
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:13.784826523 +0400
@@ -264,6 +264,10 @@
 {
 	ChatDlg::setLooks();
 
+	const QString css = PsiOptions::instance()->getOption("options.ui.chat.css").toString();
+	if (!css.isEmpty()) {
+		setStyleSheet(css);
+	}
 	ui_.splitter->optionsChanged();
 	ui_.mle->optionsChanged();
 
Index: psi/src/psicon.cpp
===================================================================
--- psi.orig/src/psicon.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/psicon.cpp	2013-01-15 14:04:13.788826523 +0400
@@ -501,6 +501,10 @@
 	d->iconSelect = new IconSelectPopup(0);
 	connect(PsiIconset::instance(), SIGNAL(emoticonsChanged()), d, SLOT(updateIconSelect()));
 
+	const QString css = PsiOptions::instance()->getOption("options.ui.chat.css").toString();
+	if (!css.isEmpty())
+		d->iconSelect->setStyleSheet(css);
+
 	// first thing, try to load the iconset
 	bool result = true;;
 	if( !PsiIconset::instance()->loadAll() ) {
@@ -1331,6 +1335,13 @@
 	if (option == "options.p2p.bytestreams.listen-port") {
 		s5b_init();
 	}
+	if (option == "options.ui.chat.css") {
+		QString css = PsiOptions::instance()->getOption(option).toString();
+		if (!css.isEmpty())
+			d->iconSelect->setStyleSheet(css);
+		return;
+	}
+
 #ifdef USE_PEP
 	if (option == tuneUrlFilterOptionPath || option == tuneTitleFilterOptionPath) {
 		d->tuneManager->setTuneFilters(PsiOptions::instance()->getOption(tuneUrlFilterOptionPath).toString().split(QRegExp("\\W+")),
Index: psi/src/tabs/tabdlg.cpp
===================================================================
--- psi.orig/src/tabs/tabdlg.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/tabs/tabdlg.cpp	2013-01-15 14:04:13.788826523 +0400
@@ -368,6 +368,11 @@
 		tabWidget_->setTabPosition(QTabWidget::South);
 
 	setWindowOpacity(double(qMax(MINIMUM_OPACITY,PsiOptions::instance()->getOption("options.ui.chat.opacity").toInt()))/100);
+
+	const QString css = PsiOptions::instance()->getOption("options.ui.chat.css").toString();
+	if (!css.isEmpty()) {
+		setStyleSheet(css);
+	}
 }
 
 void TabDlg::tabSelected(QWidget* _selected)
Index: psi/src/widgets/fancypopup.cpp
===================================================================
--- psi.orig/src/widgets/fancypopup.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/widgets/fancypopup.cpp	2013-01-15 14:04:13.788826523 +0400
@@ -213,6 +213,11 @@
 	ui_.closeButton->setIcon( popup->style()->standardPixmap(QStyle::SP_TitleBarCloseButton) );
 	ui_.closeButton->setFixedSize(BUTTON_WIDTH, BUTTON_HEIGHT);
 	connect(ui_.closeButton, SIGNAL(clicked()), popup, SLOT(hide()));
+
+	const QString css = PsiOptions::instance()->getOption("options.ui.notifications.passive-popups.css").toString();
+	if (!css.isEmpty()) {
+		popup->setStyleSheet(css);
+	}
 }
 
 bool FancyPopup::Private::eventFilter(QObject *o, QEvent *e)
Index: psi/src/widgets/iconaction.cpp
===================================================================
--- psi.orig/src/widgets/iconaction.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/widgets/iconaction.cpp	2013-01-15 14:04:13.788826523 +0400
@@ -30,6 +30,8 @@
 class Iconset;
 #endif
 
+#include "psioptions.h"
+
 #include <QLayout>
 #include <QMenu>
 #include <QTimer>
@@ -428,6 +430,11 @@
 	d->updatePopup();
 
 	d->exclusive = exclusive;
+
+	const QString css = PsiOptions::instance()->getOption("options.ui.contactlist.css").toString();
+	if (!css.isEmpty()) {
+		d->popup->setStyleSheet(css);
+	}
 }
 
 IconActionGroup::~IconActionGroup()
Index: psi/src/widgets/psitabwidget.cpp
===================================================================
--- psi.orig/src/widgets/psitabwidget.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/widgets/psitabwidget.cpp	2013-01-15 14:04:13.788826523 +0400
@@ -74,6 +74,7 @@
 	stacked_ = new QStackedLayout(layout_);
 
 	setTabPosition(QTabWidget::North);
+	setLooks();
 
 	if (!PsiOptions::instance()->getOption("options.ui.tabs.show-tab-buttons").toBool()){
 		closeButton_->hide();
@@ -179,7 +180,16 @@
 		tabBar_->addTab(icon, name);
 	else
 		tabBar_->addTab(name);
+	setLooks();
 	showPage(currentPage());
+}
+
+void PsiTabWidget::setLooks()
+{
+	const QString css = PsiOptions::instance()->getOption("options.ui.chat.css").toString();
+	if (!css.isEmpty()) {
+		setStyleSheet(css);
+	}
 } 
 
 /**
Index: psi/src/widgets/psitabwidget.h
===================================================================
--- psi.orig/src/widgets/psitabwidget.h	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/widgets/psitabwidget.h	2013-01-15 14:04:13.788826523 +0400
@@ -67,6 +67,7 @@
 public slots:
 	void setCurrentPage(int);
 	void removeCurrentPage();
+	void setLooks();
 
 signals:
 	void mouseDoubleClickTab(QWidget *tab);
Index: psi/src/widgets/psitiplabel.cpp
===================================================================
--- psi.orig/src/widgets/psitiplabel.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/widgets/psitiplabel.cpp	2013-01-15 14:04:13.788826523 +0400
@@ -51,6 +51,10 @@
 		}
 	}
 	setPalette(palette);
+	const QString css = PsiOptions::instance()->getOption("options.ui.contactlist.tooltip.css").toString();
+	if (!css.isEmpty()) {
+		setStyleSheet(css);
+	}
 }
 
 void PsiTipLabel::setText(const QString& text)
Index: psi/src/widgets/urlobject.cpp
===================================================================
--- psi.orig/src/widgets/urlobject.cpp	2013-01-15 14:04:13.792826523 +0400
+++ psi/src/widgets/urlobject.cpp	2013-01-15 14:04:13.788826523 +0400
@@ -19,6 +19,7 @@
  */
 
 #include "urlobject.h"
+#include "psioptions.h"
 
 #include <QApplication>
 #include <QClipboard>
@@ -233,6 +234,7 @@
 	}
 
 	m->addAction(d->act_copy);
+	m->setStyleSheet(PsiOptions::instance()->getOption("options.ui.look.css").toString());
 	return m;
 }
 
