--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -203,6 +203,7 @@ QLineEdit#le_status_text {
 				<disable-scrollbar type="bool">true</disable-scrollbar>
 				<contact-sort-style type="QString">status</contact-sort-style>
 				<disable-service-discovery type="bool">false</disable-service-discovery>
+				<enable-groups type="bool">true</enable-groups>
 				<ensure-contact-visible-on-event type="bool">true</ensure-contact-visible-on-event>
 				<group-sort-style type="QString">alpha</group-sort-style>
 				<lockdown-roster type="bool">false</lockdown-roster>
@@ -466,6 +467,7 @@ QLineEdit#le_status_text {
 				<menu_quit type="QKeySequence" comment="Quit"></menu_quit>
 				<button_options type="QKeySequence" comment="Options"></button_options>
 				<button_status type="QKeySequence" comment="Status"></button_status>
+				<enable_groups type="QKeySequence" comment="Enable Groups"></enable_groups>
 				<event_notifier type="QKeySequence" comment="Event Notifier"></event_notifier>
 				<view_groups type="QKeySequence" comment="View Groups"></view_groups>
 				<show_offline type="QKeySequence" comment="Show Offline Contacts"></show_offline>
--- psi.orig/src/mainwin.cpp
+++ psi/src/mainwin.cpp
@@ -587,6 +587,7 @@ void MainWin::registerAction( IconAction
 		{ "show_agents",    toggled, contactList, SLOT( setShowAgents(bool) ) },
 		{ "show_self",      toggled, contactList, SLOT( setShowSelf(bool) ) },
 		{ "show_statusmsg", toggled, d->rosterWidget_, SLOT( setShowStatusMsg(bool) ) },
+		{ "enable_groups",  toggled,   this, SLOT( actEnableGroupsActivated(bool) ) },
 #endif
 
 		{ "button_options", activated, this, SIGNAL( doOptions() ) },
@@ -648,6 +649,9 @@ void MainWin::registerAction( IconAction
 			if ( aName == "menu_play_sounds" ) {
 				action->setChecked(PsiOptions::instance()->getOption("options.ui.notifications.sounds.enable").toBool());
 			}
+			else if ( aName == "enable_groups" ) {
+				action->setChecked(PsiOptions::instance()->getOption("options.ui.contactlist.enable-groups").toBool());
+			}
 			//else if ( aName == "foobar" )
 			//	;
 		}
@@ -1084,6 +1088,11 @@ void MainWin::actPublishTuneActivated (b
 	PsiOptions::instance()->setOption("options.extended-presence.tune.publish",state);
 }
 
+void MainWin::actEnableGroupsActivated (bool state)
+{
+	PsiOptions::instance()->setOption("options.ui.contactlist.enable-groups", state);
+}
+
 void MainWin::activatedAccOption(PsiAccount* pa, int x)
 {
 	if(x == 0) {
--- psi.orig/src/mainwin.h
+++ psi/src/mainwin.h
@@ -131,6 +131,7 @@ private slots:
 	void actAboutPsiMediaActivated ();
 	void actPlaySoundsActivated (bool);
 	void actPublishTuneActivated (bool);
+	void actEnableGroupsActivated (bool);
 	void actTipActivated();
 	void actDiagQCAPluginActivated();
 	void actDiagQCAKeyStoreActivated();
--- psi.orig/src/psiactionlist.cpp
+++ psi/src/psiactionlist.cpp
@@ -140,6 +140,9 @@ void PsiActionList::Private::createMainW
 		viewGroups->setUsesDropDown(true);
 		viewGroups->setExclusive(false);
 
+		IconAction *actEnableGroups = new IconAction(tr("Show Roster Groups"), "psi/enable-groups", tr("Show Roster Groups"), 0, viewGroups, 0, true);
+		actEnableGroups->setWhatsThis(tr("Enable/disable groups in roster"));
+
 		IconAction *showOffline = new IconAction(tr("Show Offline Contacts"), "psi/show_offline", tr("Show Offline Contacts"), 0, viewGroups, 0, true);
 		showOffline->setWhatsThis(tr("Toggles visibility of offline contacts in roster"));
 
@@ -161,6 +164,7 @@ void PsiActionList::Private::createMainW
 
 		ActionNames actions[] = {
 			{ "view_groups",  viewGroups  },
+			{ "enable_groups",   actEnableGroups  },
 			{ "show_offline", showOffline },
 			//{ "show_away",    showAway    },
 			{ "show_hidden",  showHidden  },
--- psi.orig/src/psirosterwidget.cpp
+++ psi/src/psirosterwidget.cpp
@@ -49,6 +49,7 @@ static const QString showSelfOptionPath
 static const QString showStatusMessagesOptionPath = "options.ui.contactlist.status-messages.show";
 static const QString allowAutoResizeOptionPath = "options.ui.contactlist.automatically-resize-roster";
 static const QString showScrollBarOptionPath = "options.ui.contactlist.disable-scrollbar";
+static const QString enableGroupsOptionPath = "options.ui.contactlist.enable-groups";
 
 //----------------------------------------------------------------------------
 // PsiRosterFilterProxyModel
@@ -181,7 +182,7 @@ void PsiRosterWidget::setContactList(Psi
 
 	contactListModel_ = new PsiContactListModel(contactList_);
 	contactListModel_->invalidateLayout();
-	contactListModel_->setGroupsEnabled(true);
+	contactListModel_->setGroupsEnabled(PsiOptions::instance()->getOption(enableGroupsOptionPath).toBool());
 	contactListModel_->setAccountsEnabled(true);
 	contactListModel_->storeGroupState("contacts");
 #ifdef MODELTEST
@@ -244,6 +245,9 @@ void PsiRosterWidget::optionChanged(cons
 			PsiOptions::instance()->getOption(showScrollBarOptionPath).toBool()?
 			Qt::ScrollBarAlwaysOff : Qt::ScrollBarAsNeeded);
 	}
+	else if (option == enableGroupsOptionPath) {
+		contactListModel_->setGroupsEnabled(PsiOptions::instance()->getOption(enableGroupsOptionPath).toBool());
+	}
 }
 
 void PsiRosterWidget::removeSelection(QMimeData* selection)
