--- a/options/default.xml
+++ b/options/default.xml
@@ -551,6 +551,7 @@ QLineEdit#le_status_text {
 				<show-messageFirst type="QKeySequence" comment="Show first message" >Alt+End</show-messageFirst>
 				<show-messageNext type="QKeySequence" comment="Show next message" >Ctrl+Down</show-messageNext>
 				<show-messagePrev type="QKeySequence" comment="Show prev message" >Ctrl+Up</show-messagePrev>
+				<minimize type="QKeySequence" comment="Minimize active chat window"/>
 			</chat>
 			<contactlist comment="Shortcuts in the contactlist">
 				<assign-custom-avatar type="QKeySequence" comment="Assign a custom avatar to the selected contact"/>
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -214,7 +214,7 @@ public:
 #endif
 	QAction *act_send, *act_scrollup, *act_scrolldown, *act_close;
 
-	QAction *act_mini_cmd, *act_nick, *act_hide;
+	QAction *act_mini_cmd, *act_nick, *act_hide, *act_minimize;
 
 	MCmdSimpleSite mCmdSite;
 	MCmdManager mCmdManager;
@@ -704,6 +704,10 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, const Jid &j, TabManager *tabManager)
 	connect(bm, SIGNAL(conferencesChanged(QList<ConferenceBookmark>)), SLOT(updateBookmarkIcon()));
 	connect(bm, SIGNAL(bookmarksSaved()), SLOT(updateBookmarkIcon()));
 
+	d->act_minimize = new QAction(this);
+	connect(d->act_minimize, SIGNAL(triggered()), SLOT(doMinimize()));
+	addAction(d->act_minimize);
+
 	int s = PsiIconset::instance()->system().iconSize();
 	ui_.toolbar->setIconSize(QSize(s,s));
 	ui_.toolbar->addAction(d->act_clear);
@@ -812,6 +816,11 @@ void GCMainDlg::horizSplitterMoved()
 	PsiOptions::instance()->setOption("options.ui.muc.roster-width", d->rosterSize);
 }
 
+void GCMainDlg::doMinimize()
+{
+	window()->showMinimized();
+}
+
 void GCMainDlg::ensureTabbedCorrectly()
 {
 	TabbableWidget::ensureTabbedCorrectly();
@@ -854,6 +863,7 @@ void GCMainDlg::setShortcuts()
 	d->act_scrollup->setShortcuts(ShortcutManager::instance()->shortcuts("common.scroll-up"));
 	d->act_scrolldown->setShortcuts(ShortcutManager::instance()->shortcuts("common.scroll-down"));
 	d->act_mini_cmd->setShortcuts(ShortcutManager::instance()->shortcuts("chat.quick-command"));
+	d->act_minimize->setShortcuts(ShortcutManager::instance()->shortcuts("chat.minimize"));
 }
 
 void GCMainDlg::scrollUp()
--- a/src/groupchatdlg.h
+++ b/src/groupchatdlg.h
@@ -140,6 +140,9 @@ private slots:
 #endif
 	void chatEditCreated();
 	void horizSplitterMoved();
+	void doMinimize();
+
+
 
 public:
 	class Private;
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -312,6 +312,11 @@ void PsiChatDlg::setShortcuts()
 
 	act_mini_cmd_->setShortcuts(ShortcutManager::instance()->shortcuts("chat.quick-command"));
 
+	act_minimize_ = new QAction(this);
+
+	connect(act_minimize_, SIGNAL(triggered()), SLOT(doMinimize()));
+	addAction(act_minimize_);
+	act_minimize_->setShortcuts(ShortcutManager::instance()->shortcuts("chat.minimize"));
 }
 
 void PsiChatDlg::updateIdentityVisibility()
@@ -679,6 +684,10 @@ void PsiChatDlg::doMiniCmd()
 	mCmdManager_.open(new MCmdSimpleState(MCMDCHAT, tr("Command>")), QStringList() );
 }
 
+void PsiChatDlg::doMinimize()
+{
+	window()->showMinimized();
+}
 
 bool PsiChatDlg::eventFilter( QObject *obj, QEvent *ev ) {
 	if ( obj == chatEdit() && ev->type() == QEvent::KeyPress ) {
--- a/src/psichatdlg.h
+++ b/src/psichatdlg.h
@@ -33,6 +33,7 @@ private slots:
 	void toggleSmallChat();
 	void doClearButton();
 	void doMiniCmd();
+	void doMinimize();
 	void buildMenu();
 	void updateCounter();
 	void updateIdentityVisibility();
@@ -77,6 +78,7 @@ private:
 	IconAction* act_find_;
 
 	QAction *act_mini_cmd_;
+	QAction *act_minimize_;
 	TypeAheadFindBar *typeahead_;
 
 	MCmdManager mCmdManager_;
