--- git.orig/options/default.xml
+++ git/options/default.xml
@@ -551,6 +551,7 @@ QLineEdit#le_status_text {
 				<show-messageFirst type="QKeySequence" comment="Show first message" >Alt+End</show-messageFirst>
 				<show-messageNext type="QKeySequence" comment="Show next message" >Ctrl+Down</show-messageNext>
 				<show-messagePrev type="QKeySequence" comment="Show prev message" >Ctrl+Up</show-messagePrev>
+				<minimize type="QKeySequence" comment="Minimize active chat window"/>
 			</chat>
 			<contactlist comment="Shortcuts in the contactlist">
 				<assign-custom-avatar type="QKeySequence" comment="Assign a custom avatar to the selected contact"/>
--- git.orig/src/groupchatdlg.cpp
+++ git/src/groupchatdlg.cpp
@@ -214,7 +214,7 @@ public:
 #endif
 	QAction *act_send, *act_scrollup, *act_scrolldown, *act_close;
 
-	QAction *act_mini_cmd, *act_nick, *act_hide;
+	QAction *act_mini_cmd, *act_nick, *act_hide, *act_minimize;
 
 	MCmdSimpleSite mCmdSite;
 	MCmdManager mCmdManager;
@@ -695,6 +695,10 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	connect(d->act_mini_cmd, SIGNAL(triggered()), d, SLOT(doMiniCmd()));
 	addAction(d->act_mini_cmd);
 
+	d->act_minimize = new QAction(this);
+	connect(d->act_minimize, SIGNAL(triggered()), SLOT(doMinimize()));
+	addAction(d->act_minimize);
+
 	d->act_ps = new IconAction(tr("Paste and Send"), "psi/action_paste_and_send", tr("Paste and Send"), 0, this);
 	connect(d->act_ps, SIGNAL(triggered()), SLOT(doPasteAndSend()));
 
@@ -833,6 +837,11 @@ void GCMainDlg::horizSplitterMoved()
 	PsiOptions::instance()->setOption("options.ui.muc.roster-width", d->rosterSize);
 }
 
+void GCMainDlg::doMinimize()
+{
+	window()->showMinimized();
+}
+
 void GCMainDlg::ensureTabbedCorrectly()
 {
 	TabbableWidget::ensureTabbedCorrectly();
@@ -875,6 +884,7 @@ void GCMainDlg::setShortcuts()
 	d->act_scrollup->setShortcuts(ShortcutManager::instance()->shortcuts("common.scroll-up"));
 	d->act_scrolldown->setShortcuts(ShortcutManager::instance()->shortcuts("common.scroll-down"));
 	d->act_mini_cmd->setShortcuts(ShortcutManager::instance()->shortcuts("chat.quick-command"));
+	d->act_minimize->setShortcuts(ShortcutManager::instance()->shortcuts("chat.minimize"));
 }
 
 void GCMainDlg::scrollUp()
--- git.orig/src/groupchatdlg.h
+++ git/src/groupchatdlg.h
@@ -145,6 +145,9 @@ private slots:
 	void sendTemp(const QString &);
 	void psButtonEnabled();
 	void horizSplitterMoved();
+	void doMinimize();
+
+
 
 public:
 	class Private;
--- git.orig/src/psichatdlg.cpp
+++ git/src/psichatdlg.cpp
@@ -330,6 +330,11 @@ void PsiChatDlg::setShortcuts()
 
 	act_mini_cmd_->setShortcuts(ShortcutManager::instance()->shortcuts("chat.quick-command"));
 
+	act_minimize_ = new QAction(this);
+
+	connect(act_minimize_, SIGNAL(triggered()), SLOT(doMinimize()));
+	addAction(act_minimize_);
+	act_minimize_->setShortcuts(ShortcutManager::instance()->shortcuts("chat.minimize"));
 }
 
 void PsiChatDlg::updateIdentityVisibility()
@@ -709,6 +714,10 @@ void PsiChatDlg::doMiniCmd()
 	mCmdManager_.open(new MCmdSimpleState(MCMDCHAT, tr("Command>")), QStringList() );
 }
 
+void PsiChatDlg::doMinimize()
+{
+	window()->showMinimized();
+}
 
 bool PsiChatDlg::eventFilter( QObject *obj, QEvent *ev ) {
 	if ( obj == chatEdit() && ev->type() == QEvent::KeyPress ) {
--- git.orig/src/psichatdlg.h
+++ git/src/psichatdlg.h
@@ -34,6 +34,7 @@ private slots:
 	void toggleSmallChat();
 	void doClearButton();
 	void doMiniCmd();
+	void doMinimize();
 	void buildMenu();
 	void updateCounter();
 	void updateIdentityVisibility();
@@ -84,7 +85,7 @@ private:
 	IconAction* act_find;
 	IconAction* act_ps_;
 	IconAction* act_templates_;
-	QAction *act_mini_cmd_;
+	QAction *act_mini_cmd_, *act_minimize_;
 
 	MCmdManager mCmdManager_;
 	MCmdSimpleSite mCmdSite_;
