From 76f57a003af752fb712b01f2ca62bc0f2c036b28 Mon Sep 17 00:00:00 2001
From: Julius Schwartzenberg <julius.schwartzenberg@gmail.com>
Date: Tue, 13 Apr 2010 13:29:26 +0200
Subject: Video patch by Arseniy Lartsev, unified version for the original see: http://lists.affinix.com/pipermail/psi-devel-affinix.com/2010-March/009046.html

---
 src/avcall/jinglertp.cpp |    6 ------
 1 files changed, 0 insertions(+), 6 deletions(-)

Index: psi/src/avcall/jinglertp.cpp
===================================================================
--- psi.orig/src/avcall/jinglertp.cpp	2013-01-15 14:04:14.952826571 +0400
+++ psi/src/avcall/jinglertp.cpp	2013-01-15 14:04:14.948826570 +0400
@@ -621,12 +621,10 @@
 				if((types & JingleRtp::Audio) && c.desc.media == "audio" && !audioContent)
 				{
 					audioContent = &c;
-					break;
 				}
 				else if((types & JingleRtp::Video) && c.desc.media == "video" && !videoContent)
 				{
 					videoContent = &c;
-					break;
 				}
 			}

@@ -670,12 +668,10 @@
 				if((types & JingleRtp::Audio) && c.desc.media == "audio" && c.name == audioName && !audioContent)
 				{
 					audioContent = &c;
-					break;
 				}
 				else if((types & JingleRtp::Video) && c.desc.media == "video" && c.name == videoName && !videoContent)
 				{
 					videoContent = &c;
-					break;
 				}
 			}

@@ -743,12 +739,10 @@
 				if((types & JingleRtp::Audio) && c.name == audioName && !audioContent)
 				{
 					audioContent = &c;
-					break;
 				}
 				else if((types & JingleRtp::Video) && c.name == videoName && !videoContent)
 				{
 					videoContent = &c;
-					break;
 				}
 			}

