Index: psi/src/contactlistgroupmenu.cpp
===================================================================
--- psi.orig/src/contactlistgroupmenu.cpp	2013-01-15 14:04:15.164826579 +0400
+++ psi/src/contactlistgroupmenu.cpp	2013-01-15 14:04:15.160826579 +0400
@@ -108,7 +108,7 @@
 		actionMucLeave_ = new IconAction(tr("Leave All"), this, "psi/action_muc_leave");
 		connect(actionMucLeave_, SIGNAL(triggered()), SLOT(mucLeave()));

-		updateActions();
+		authMenu_ = 0;

 		if (ContactListGroup::SpecialType_Conference != group->specialGroupType()) {
 #ifdef YAPSI
@@ -135,6 +135,7 @@
 			menu->addAction(actionCustomStatus_);
 		}
 #endif
+		updateActions();
 	}

 private slots:
@@ -144,6 +145,10 @@
 			return;

 		sendMessageAction_->setVisible(PsiOptions::instance()->getOption("options.ui.message.enabled").toBool());
+		sendMessageAction_->setEnabled(group->contacts().first()->account()->isAvailable());
+		actionCustomStatus_->setEnabled(group->contacts().first()->account()->isAvailable());
+		if(authMenu_)
+			authMenu_->setEnabled(group->contacts().first()->account()->isAvailable());
 		renameAction_->setEnabled(group->isEditable());
 		removeGroupAndContactsAction_->setEnabled(group->isRemovable());
 		removeGroupWithoutContactsAction_->setEnabled(group->isRemovable());
