--- git.orig/src/contactlistaccountmenu.cpp
+++ git/src/contactlistaccountmenu.cpp
@@ -208,6 +208,9 @@ private slots:
 		}
 
 		newMessageAction_->setVisible(PsiOptions::instance()->getOption("options.ui.message.enabled").toBool());
+		newMessageAction_->setEnabled(account->account()->isAvailable());
+		addContactAction_->setEnabled(account->account()->isAvailable());
+		serviceDiscoveryAction_->setEnabled(account->account()->isAvailable());
 		contactManagerAction_->setEnabled(account->account()->isAvailable());
 		privacyListsAction_->setEnabled(account->account()->isAvailable());
 		if (!PsiOptions::instance()->getOption("options.ui.menu.account.admin").toBool()) {
--- git.orig/src/contactlistgroupmenu.cpp
+++ git/src/contactlistgroupmenu.cpp
@@ -108,7 +108,7 @@ public:
 		actionMucLeave_ = new IconAction(tr("Leave All"), this, "psi/action_muc_leave");
 		connect(actionMucLeave_, SIGNAL(triggered()), SLOT(mucLeave()));
 
-		updateActions();
+		authMenu_ = 0;
 
 		if (ContactListGroup::SpecialType_Conference != group->specialGroupType()) {
 #ifdef YAPSI
@@ -135,6 +135,7 @@ public:
 			menu->addAction(actionCustomStatus_);
 		}
 #endif
+		updateActions();
 	}
 
 private slots:
@@ -144,6 +145,10 @@ private slots:
 			return;
 
 		sendMessageAction_->setVisible(PsiOptions::instance()->getOption("options.ui.message.enabled").toBool());
+		sendMessageAction_->setEnabled(group->contacts().first()->account()->isAvailable());                
+		actionCustomStatus_->setEnabled(group->contacts().first()->account()->isAvailable());
+		if(authMenu_)
+			authMenu_->setEnabled(group->contacts().first()->account()->isAvailable());
 		renameAction_->setEnabled(group->isEditable());
 		removeGroupAndContactsAction_->setEnabled(group->isRemovable());
 		removeGroupWithoutContactsAction_->setEnabled(group->isRemovable());
--- git.orig/src/psicontactmenu.cpp
+++ git/src/psicontactmenu.cpp
@@ -519,6 +519,7 @@ private slots:
 			sendMessageAction_->setVisible(false);
 			sendMessageToMenu_->menuAction()->setVisible(false);
 		}
+		sendMessageAction_->setEnabled(contact_->account()->isAvailable());
 		sendMessageToMenu_->setEnabled(!sendMessageToMenu_->isEmpty());
 		openChatToMenu_->setEnabled(!openChatToMenu_->isEmpty());
 		openWhiteboardToMenu_->setEnabled(!openWhiteboardToMenu_->isEmpty());
@@ -534,7 +535,7 @@ private slots:
 		voiceCallAction_->setEnabled(contact_->account()->isAvailable());
 		sendFileAction_->setVisible(!contact_->isAgent());
 		sendFileAction_->setEnabled(contact_->account()->isAvailable());
-		inviteToGroupchatMenu_->setEnabled(!inviteToGroupchatMenu_->isEmpty());
+		inviteToGroupchatMenu_->setEnabled(!inviteToGroupchatMenu_->isEmpty() && contact_->account()->isAvailable());
 		renameAction_->setVisible(!PsiOptions::instance()->getOption("options.ui.contactlist.lockdown-roster").toBool());
 		renameAction_->setEnabled(contact_->isEditable());
 		if (contact_->isAgent()) {
