Index: psi/src/gcuserview.cpp
===================================================================
--- psi.orig/src/gcuserview.cpp	2013-01-15 14:04:15.332826586 +0400
+++ psi/src/gcuserview.cpp	2013-01-15 14:04:15.332826586 +0400
@@ -358,7 +358,7 @@
 	u.setName(nick);
 
 	// Find out capabilities info
-	Jid caps_jid(s.mucItem().jid().isEmpty() ? dlg->jid().withResource(nick) : s.mucItem().jid());
+	Jid caps_jid(/*s.mucItem().jid().isEmpty() ? */dlg->jid().withResource(nick)/* : s.mucItem().jid()*/);
 	QString client_name = dlg->account()->capsManager()->clientName(caps_jid);
 	QString client_version = (client_name.isEmpty() ? QString() : dlg->account()->capsManager()->clientVersion(caps_jid));
 
