--- git.orig/src/gcuserview.cpp
+++ git/src/gcuserview.cpp
@@ -354,7 +354,7 @@ bool GCUserView::maybeTip(const QPoint &
 	u.setName(nick);
 
 	// Find out capabilities info
-	Jid caps_jid(s.mucItem().jid().isEmpty() ? dlg->jid().withResource(nick) : s.mucItem().jid());
+	Jid caps_jid(/*s.mucItem().jid().isEmpty() ? */dlg->jid().withResource(nick)/* : s.mucItem().jid()*/);
 	QString client_name = dlg->account()->capsManager()->clientName(caps_jid);
 	QString client_version = (client_name.isEmpty() ? QString() : dlg->account()->capsManager()->clientVersion(caps_jid));
 
