Index: psi/src/chatdlg.ui
===================================================================
--- psi.orig/src/chatdlg.ui	2013-01-15 14:04:15.520826594 +0400
+++ psi/src/chatdlg.ui	2013-01-15 14:04:15.516826593 +0400
@@ -78,6 +78,25 @@
           </widget>
          </item>
          <item>
+          <widget class="QLabel" name="lb_client">
+           <property name="minimumSize">
+            <size>
+             <width>16</width>
+             <height>16</height>
+            </size>
+           </property>
+           <property name="maximumSize">
+            <size>
+             <width>16</width>
+             <height>16</height>
+            </size>
+           </property>
+           <property name="text">
+            <string/>
+           </property>
+          </widget>
+         </item>
+         <item>
           <widget class="QLabel" name="lb_count" >
            <property name="sizePolicy" >
             <sizepolicy>
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:15.520826594 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:15.520826594 +0400
@@ -279,8 +279,10 @@
 		ui_.tb_emoticons->hide();
 		ui_.toolbar->hide();
 		ui_.tb_voice->hide();
+		ui_.lb_client->hide();
 	}
 	else {
+		ui_.lb_client->show();
 		ui_.lb_status->show();
 		ui_.le_jid->show();
 		if (PsiOptions::instance()->getOption("options.ui.chat.central-toolbar").toBool()) {
@@ -482,6 +484,24 @@
 		ui_.le_jid->setText(name);
 		ui_.le_jid->setCursorPosition(0);
 		ui_.le_jid->setToolTip(name);
+
+		UserResourceList srl = u->userResourceList();
+		if(!srl.isEmpty()) {
+			UserResource r;
+			if(!jid().resource().isEmpty()) {
+				QString res = jid().resource();
+				UserResourceList::ConstIterator it = srl.find(res);
+				if(it != srl.end())
+					r = *it;
+			}
+			if(r.clientName().isEmpty()) {
+				srl.sort();
+				r = srl.first();
+			}
+			const QPixmap &pix = IconsetFactory::iconPixmap("clients/" + u->findClient(r.clientName().toLower()) );
+			ui_.lb_client->setPixmap(pix);
+			ui_.lb_client->setToolTip(r.versionString());
+		}
 	}
 }

