--- git.temp.orig/src/chatdlg.ui
+++ git.temp/src/chatdlg.ui
@@ -78,6 +78,25 @@
           </widget>
          </item>
          <item>
+          <widget class="QLabel" name="lb_client">
+           <property name="minimumSize">
+            <size>
+             <width>16</width>
+             <height>16</height>
+            </size>
+           </property>
+           <property name="maximumSize">
+            <size>
+             <width>16</width>
+             <height>16</height>
+            </size>
+           </property>
+           <property name="text">
+            <string/>
+           </property>
+          </widget>
+         </item>
+         <item>
           <widget class="QLabel" name="lb_count" >
            <property name="sizePolicy" >
             <sizepolicy>
--- git.temp.orig/src/psichatdlg.cpp
+++ git.temp/src/psichatdlg.cpp
@@ -279,8 +279,10 @@ void PsiChatDlg::setLooks()
 		ui_.tb_emoticons->hide();
 		ui_.toolbar->hide();
 		ui_.tb_voice->hide();
+		ui_.lb_client->hide();
 	}
 	else {
+		ui_.lb_client->show();
 		ui_.lb_status->show();
 		ui_.le_jid->show();
 		if (PsiOptions::instance()->getOption("options.ui.chat.central-toolbar").toBool()) {
@@ -481,6 +483,24 @@ void PsiChatDlg::contactUpdated(UserList
 		ui_.le_jid->setText(name);
 		ui_.le_jid->setCursorPosition(0);
 		ui_.le_jid->setToolTip(name);
+
+		UserResourceList srl = u->userResourceList();
+		if(!srl.isEmpty()) {
+			UserResource r;
+			if(!jid().resource().isEmpty()) {
+				QString res = jid().resource();
+				UserResourceList::ConstIterator it = srl.find(res);
+				if(it != srl.end())
+					r = *it;
+			}
+			if(r.clientName().isEmpty()) {
+				srl.sort();
+				r = srl.first();
+			}
+			const QPixmap &pix = IconsetFactory::iconPixmap("clients/" + u->findClient(r.clientName().toLower()) );
+			ui_.lb_client->setPixmap(pix);
+			ui_.lb_client->setToolTip(r.versionString());
+		}
 	}
 }
 
