--- git.orig/src/contactlistaccountmenu.cpp
+++ git/src/contactlistaccountmenu.cpp
@@ -33,6 +33,7 @@
 #include "psiiconset.h"
 #include "common.h"
 #include "contactmanager/contactmanagerdlg.h"
+#include "mucjoindlg.h"
 
 class ContactListAccountMenu::Private : public QObject
 {
@@ -62,6 +63,7 @@ class ContactListAccountMenu::Private : 
 	QAction* adminSetMotdAction_;
 	QAction* adminUpdateMotdAction_;
 	QAction* adminDeleteMotdAction_;
+	QAction *doGroupChatAction_;
 
 public:
 	Private(ContactListAccountMenu* menu, ContactListAccountGroup* _account)
@@ -100,9 +102,12 @@ public:
 		contactManagerAction_ = new IconAction(tr("&Contacts Manager"), this, "psi/action_contacts_manager");
 		connect(contactManagerAction_, SIGNAL(triggered()), SLOT(contactManager()));
 
-		bookmarksManageAction_ = new QAction(tr("Manage..."), this);
+		bookmarksManageAction_ = new QAction(tr("Manage Bookmarks"), this);
 		connect(bookmarksManageAction_, SIGNAL(triggered()), SLOT(bookmarksManage()));
 
+		doGroupChatAction_ = new IconAction(tr("Join Groupchat"), this, "psi/groupChat");
+		connect(doGroupChatAction_, SIGNAL(triggered()), SLOT(doGroupChat()));
+
 		addContactAction_ = new IconAction(tr("&Add a Contact"), this, "psi/addContact");
 		connect(addContactAction_, SIGNAL(triggered()), SLOT(addContact()));
 
@@ -143,7 +148,9 @@ public:
 		avatarMenu_ = menu->addMenu(IconsetFactory::icon("psi/vCard").icon(), tr("Avatar"));
 		avatarMenu_->addAction(setAvatarAction_);
 		avatarMenu_->addAction(unsetAvatarAction_);
-		bookmarksMenu_ = menu->addMenu(IconsetFactory::icon("psi/bookmarks").icon(), tr("Bookmarks"));
+		bookmarksMenu_ = menu->addMenu(IconsetFactory::icon("psi/bookmarks").icon(), tr("Groupchat"));
+		bookmarksMenu_->addAction(doGroupChatAction_);
+		bookmarksMenu_->addSeparator();
 		bookmarksMenu_->addAction(bookmarksManageAction_);
 		menu->addSeparator();
 		menu->addAction(addContactAction_);
@@ -191,6 +198,8 @@ private slots:
 		bookmarksMenu_->clear();
 		qDeleteAll(bookmarksJoinActions_);
 		bookmarksJoinActions_.clear();
+		bookmarksMenu_->addAction(doGroupChatAction_);
+		bookmarksMenu_->addSeparator();
 		bookmarksMenu_->addAction(bookmarksManageAction_);
 		if (account->account()->bookmarkManager()->isAvailable()) {
 			bookmarksMenu_->setEnabled(true);
@@ -279,6 +288,15 @@ private slots:
 		account->account()->actionManageBookmarks();
 	}
 
+	void doGroupChat()
+	{
+		if (!account)
+			return;
+
+		MUCJoinDlg *w = new MUCJoinDlg(account->account()->psi(), account->account());
+		w->show();
+	}
+
 	void bookmarksJoin()
 	{
 		if (!account)
