--- git.orig/src/ahcexecutetask.cpp
+++ git/src/ahcexecutetask.cpp
@@ -47,14 +47,10 @@ bool AHCExecuteTask::take(const QDomElem
 		QDomElement i = e.firstChildElement("command");
 		if (!i.isNull()) {
 			AHCommand c(i);
-			if (c.status() == AHCommand::Executing) {
-				AHCFormDlg *w = new AHCFormDlg(c,receiver_,client());
-				w->show();
-			}
-			else if (c.status() == AHCommand::Completed && i.childNodes().count() > 0) {
-				AHCFormDlg *w = new AHCFormDlg(c,receiver_,client(), true);
-				w->show();
-			}
+			if (c.status() == AHCommand::Executing)
+				new AHCFormDlg(c,receiver_,client());
+			else if (c.status() == AHCommand::Completed && i.childNodes().count() > 0)
+				new AHCFormDlg(c,receiver_,client(), true);
 			setSuccess();
 			return true;
 		}
--- git.orig/src/ahcformdlg.cpp
+++ git/src/ahcformdlg.cpp
@@ -138,6 +138,8 @@ AHCFormDlg::AHCFormDlg(const AHCommand&
 	}
 
 	adjustSize();
+	show();
+	setParent(0);
 }
 
 void AHCFormDlg::doPrev()
