Index: psi/src/dbus.cpp
===================================================================
--- psi.orig/src/dbus.cpp	2013-01-15 14:04:16.584826637 +0400
+++ psi/src/dbus.cpp	2013-01-15 14:04:16.580826637 +0400
@@ -31,6 +31,14 @@
 	void raise();
 	void sleep();
 	void wake();
+
+	void setstatusonline(const QString &message);
+	void setstatusaway(const QString &message);
+	void setstatusffc(const QString &message);
+	void setstatusxa(const QString &message);
+	void setstatusdnd(const QString &message);
+	void setstatusoffline(const QString &message);
+
 /*Q_SIGNALS:
 	void psi_pong();
 */
@@ -74,6 +82,41 @@
 	psicon->doWakeup();
 }

+void PsiConAdapter::setstatusonline(const QString &message)
+{
+	Status s(Status::Online, message);
+	psicon->setGlobalStatus(s);
+}
+
+void PsiConAdapter::setstatusaway(const QString &message)
+{
+	Status s(Status::Away, message);
+	psicon->setGlobalStatus(s);
+}
+
+void PsiConAdapter::setstatusffc(const QString &message)
+{
+	Status s(Status::FFC, message);
+	psicon->setGlobalStatus(s);
+}
+
+void PsiConAdapter::setstatusxa(const QString &message)
+{
+	Status s(Status::XA, message);
+	psicon->setGlobalStatus(s);
+}
+
+void PsiConAdapter::setstatusdnd(const QString &message)
+{
+	Status s(Status::DND, message);
+	psicon->setGlobalStatus(s);
+}
+
+void PsiConAdapter::setstatusoffline(const QString &message)
+{
+	Status s(Status::Offline, message);
+	psicon->setGlobalStatus(s);
+}

 void addPsiConAdapter(PsiCon *psicon)
 {
