--- git.orig/options/default.xml
+++ git/options/default.xml
@@ -401,6 +401,8 @@ QLineEdit#le_status_text {
 						<status type="int">2000</status>
 					</delays>
 					<enabled type="bool">false</enabled>
+					<top-to-bottom type="bool">false</top-to-bottom>
+					<at-left-corner type="bool">false</at-left-corner>
 					<css type="QString"></css>
 					<incoming-chat type="bool">true</incoming-chat>
 					<incoming-file-transfer type="bool">true</incoming-file-transfer>
--- git.orig/src/widgets/fancypopup.cpp
+++ git/src/widgets/fancypopup.cpp
@@ -149,12 +149,17 @@ void FancyPopup::Private::popupDestroyed
 QPoint FancyPopup::Private::position()
 {
 	QRect geom = qApp->desktop()->availableGeometry(popup);
-	QPoint destination(geom.x() + geom.width(), geom.y() + geom.height()); // in which corner popup should appear
+	bool topToBottom = PsiOptions::instance()->getOption("options.ui.notifications.passive-popups.top-to-bottom").toBool();
+	bool atLeft = PsiOptions::instance()->getOption("options.ui.notifications.passive-popups.at-left-corner").toBool();
+	QPoint destination(geom.x() + (atLeft ? 0 : geom.width()),
+			   geom.y() + (topToBottom ? 0 : geom.height()) ); // in which corner popup should appear
 
-	if ( destination.y() > (qApp->desktop()->screenGeometry().height()/2) )
+	/*if ( destination.y() > (qApp->desktop()->screenGeometry().height()/2) )
 		popupLayout = Private::BottomToTop;
 	else
-		popupLayout = Private::TopToBottom;
+		popupLayout = Private::TopToBottom;*/
+
+	popupLayout = topToBottom ? Private::TopToBottom : Private::BottomToTop;
 
 	if ( (destination.x() + popup->width()) > (geom.x() + geom.width()) )
 		destination.setX( geom.x() + geom.width() - popup->width() );
