--- psi.orig/src/discodlg.cpp
+++ psi/src/discodlg.cpp
@@ -1015,7 +1015,8 @@ DiscoDlg::Private::Private(DiscoDlg *par
 
 	// create toolbar
 	toolBar = new QToolBar(tr("Service Discovery toolbar"), dlg);
-	toolBar->setIconSize(QSize(16,16));
+	int s = PsiIconset::instance()->system().iconSize();
+	toolBar->setIconSize(QSize(s,s));
 	toolBar->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
 	replaceWidget(dlg->toolBarPlaceholder, toolBar);
 
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -716,7 +716,8 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	connect(bm, SIGNAL(conferencesChanged(QList<ConferenceBookmark>)), SLOT(updateBookmarkIcon()));
 	connect(bm, SIGNAL(bookmarksSaved()), SLOT(updateBookmarkIcon()));
 
-	ui_.toolbar->setIconSize(QSize(16,16));
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.toolbar->setIconSize(QSize(s,s));
 	ui_.toolbar->addAction(d->act_clear);
 	ui_.toolbar->addAction(d->act_find);
 	ui_.toolbar->addAction(d->act_configure);
--- psi.orig/src/options/optionsdlg.cpp
+++ psi/src/options/optionsdlg.cpp
@@ -25,6 +25,7 @@
 #include "fancylabel.h"
 #include "iconset.h"
 #include "iconwidget.h"
+#include "psiiconset.h"
 
 #include <QLayout>
 #include <QLabel>
@@ -87,7 +88,8 @@ public:
 			iconSize = icon.availableSizes().first();
 		else
 #endif
-			iconSize = QSize(16, 16);
+			int s = PsiIconset::instance()->system().iconSize();
+			iconSize = QSize(s,s);
 		QRect iconRect = opt.rect;
 		iconRect.setLeft(4);
 		iconRect.setWidth(iconSize.width());
@@ -147,7 +149,8 @@ public:
 			iconSize = icon.availableSizes().first();
 		else
 #endif
-			iconSize = QSize(16, 16);
+			int s = PsiIconset::instance()->system().iconSize();
+			iconSize = QSize(s,s);
 
 		int width = iconSize.width();
 		width += 8;
--- psi.orig/src/psichatdlg.cpp
+++ psi/src/psichatdlg.cpp
@@ -270,6 +270,10 @@ void PsiChatDlg::setLooks()
 	ui_.splitter->optionsChanged();
 	ui_.mle->optionsChanged();
 
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.lb_status->setFixedSize(s,s);
+	ui_.lb_client->setFixedSize(s,s);
+
 	ui_.tb_pgp->hide();
 	if (smallChat_) {
 		ui_.lb_status->hide();
@@ -394,7 +398,8 @@ void PsiChatDlg::initToolButtons()
 void PsiChatDlg::initToolBar()
 {
 	ui_.toolbar->setWindowTitle(tr("Chat Toolbar"));
-	ui_.toolbar->setIconSize(QSize(16, 16));
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.toolbar->setIconSize(QSize(s, s));
 	ui_.toolbar->addAction(act_clear_);
 // typeahead find bar
 	ui_.toolbar->addAction(act_find);
--- psi.orig/src/psicontactlistviewdelegate.cpp
+++ psi/src/psicontactlistviewdelegate.cpp
@@ -45,6 +45,7 @@ static const QString avatarAtLeftOptionP
 static const QString showStatusIconsPath = "options.ui.contactlist.show-status-icons";
 static const QString statusIconsOverAvatarsPath = "options.ui.contactlist.status-icon-over-avatar";
 static const QString statusSingleOptionPath = "options.ui.contactlist.status-messages.single-line";
+static const QString statusIconsetOptionPath = "options.iconsets.status";
 
 PsiContactListViewDelegate::PsiContactListViewDelegate(ContactListView* parent)
 	: ContactListViewDelegate(parent)
@@ -57,6 +58,8 @@ PsiContactListViewDelegate::PsiContactLi
 	connect(alertTimer_, SIGNAL(timeout()), SLOT(updateAlerts()));
 
 	connect(PsiOptions::instance(), SIGNAL(optionChanged(const QString&)), SLOT(optionChanged(const QString&)));
+	connect(PsiIconset::instance(), SIGNAL(rosterIconsSizeChanged(int)), SLOT(rosterIconsSizeChanged(int)));
+	statusIconSize_ = PsiIconset::instance()->roster.value(PsiOptions::instance()->getOption(statusIconsetOptionPath).toString())->iconSize();
 	optionChanged(contactListFontOptionPath);
 	optionChanged(contactListBackgroundOptionPath);
 	optionChanged(showStatusMessagesOptionPath);
@@ -512,7 +515,7 @@ void PsiContactListViewDelegate::optionC
 		font_ = new QFont();
 		font_->fromString(PsiOptions::instance()->getOption(contactListFontOptionPath).toString());
 		fontMetrics_ = new QFontMetrics(*font_);
-		rowHeight = qMax(font_->pointSize()*2, 18);
+		rowHeight = qMax(font_->pointSize()*2, statusIconSize_+2);
 		contactList()->viewport()->update();
 	}
 	else if (option == contactListBackgroundOptionPath) {
@@ -579,6 +582,13 @@ void PsiContactListViewDelegate::optionC
 	}
 }
 
+void PsiContactListViewDelegate::rosterIconsSizeChanged(int size)
+{
+	statusIconSize_ = size;
+	rowHeight = qMax(font_->pointSize()*2, statusIconSize_+2);
+	contactList()->viewport()->update();
+}
+
 void PsiContactListViewDelegate::drawText(QPainter* painter, const QStyleOptionViewItem& o, const QRect& rect, const QString& text, const QModelIndex& index) const
 {
 	QRect r(rect);
--- psi.orig/src/psicontactlistviewdelegate.h
+++ psi/src/psicontactlistviewdelegate.h
@@ -58,6 +58,7 @@ protected:
 private slots:
 	void optionChanged(const QString& option);
 	void updateAlerts();
+	void rosterIconsSizeChanged(int size);
 
 private:
 	QTimer* alertTimer_;
@@ -68,6 +69,7 @@ private:
 	int avatarSize_, avatarRadius_, rowHeight;
 	bool statusSingle_;
 	mutable QHash<QModelIndex, bool> alertingIndexes_;
+	int statusIconSize_;
 };
 
 #endif
--- psi.orig/src/psiiconset.cpp
+++ psi/src/psiiconset.cpp
@@ -350,6 +350,11 @@ bool PsiIconset::loadSystem()
 	QString cur_system = PsiOptions::instance()->getOption("options.iconsets.system").toString();
 	if (d->cur_system != cur_system) {
 		Iconset sys = d->systemIconset(&ok);
+
+		if(sys.iconSize() != d->system.iconSize()) {
+			emit systemIconsSizeChanged(sys.iconSize());
+		}
+
 		d->loadIconset(&d->system, &sys);
 
 		//d->system = d->systemIconset();
@@ -543,6 +548,10 @@ void PsiIconset::reloadRoster()
 	if (d->cur_status != cur_status) {
 		Iconset *newDef = d->defaultRosterIconset(&ok);
 		Iconset *oldDef = roster[d->cur_status];
+
+		if(oldDef->iconSize() != newDef->iconSize())
+			emit rosterIconsSizeChanged(newDef->iconSize());
+
 		d->loadIconset(oldDef, newDef);
 
 		roster.remove(d->cur_status);
--- psi.orig/src/psiiconset.h
+++ psi/src/psiiconset.h
@@ -81,6 +81,8 @@ public:
 
 signals:
 	void emoticonsChanged();
+	void systemIconsSizeChanged(int);
+	void rosterIconsSizeChanged(int);
 
 public slots:
 	static void reset();
--- psi.orig/src/psitoolbar.cpp
+++ psi/src/psitoolbar.cpp
@@ -29,6 +29,7 @@
 #include <QToolButton>
 
 #include "iconset.h"
+#include "psiiconset.h"
 #include "iconaction.h"
 #include "psioptions.h"
 #include "options/opt_toolbars.h"
@@ -63,7 +64,8 @@ PsiToolBar::PsiToolBar(const QString& ba
 	Q_ASSERT(mainWindow);
 	Q_ASSERT(actionList_);
 
-	setIconSize(QSize(16, 16));
+	int s = PsiIconset::instance()->system().iconSize();
+	setIconSize(QSize(s, s));
 
 	customizeAction_ = new QAction(tr("Configure& Toolbar..."), this);
 	connect(customizeAction_, SIGNAL(triggered()), this, SIGNAL(customize()));
--- psi.orig/src/tools/iconset/iconset.cpp
+++ psi/src/tools/iconset/iconset.cpp
@@ -1028,6 +1028,7 @@ private:
 		//authors << "I. M. Anonymous";
 		//creation = "1900-01-01";
 		homeUrl = QString::null;
+		iconSize_ = 16;
 	}
 
 public:
@@ -1036,6 +1037,7 @@ public:
 	QHash<QString, PsiIcon *> dict; // unsorted hash for fast search
 	QList<PsiIcon *> list;          // sorted list
 	QHash<QString, QString> info;
+	int iconSize_;
 
 public:
 	Private()
@@ -1126,6 +1128,8 @@ public:
 	{
 		Q_UNUSED(dir);
 
+		iconSize_ = 16;
+
 		for (QDomNode node = i.firstChild(); !node.isNull(); node = node.nextSibling()) {
 			QDomElement e = node.toElement();
 			if( e.isNull() ) {
@@ -1136,6 +1140,10 @@ public:
 			if ( tag == "name" ) {
 				name = e.text();
 			}
+			else if( tag == "size") {
+				QString str = e.text();
+				iconSize_ = str.toInt();
+			}
 			else if ( tag == "version" ) {
 				version = e.text();
 			}
@@ -1407,6 +1415,7 @@ public:
 		filename = from.filename;
 		authors = from.authors;
 		info = from.info;
+		iconSize_ = from.iconSize_;
 	}
 };
 //! \endif
@@ -1587,6 +1596,14 @@ const QString &Iconset::name() const
 }
 
 /**
+ * Returns the icons size from Iconset.
+ */
+const int &Iconset::iconSize() const
+{
+	return d->iconSize_;
+}
+
+/**
  * Returns the Iconset version.
  */
 const QString &Iconset::version() const
--- psi.orig/src/tools/iconset/iconset.h
+++ psi/src/tools/iconset/iconset.h
@@ -205,6 +205,7 @@ public:
 	const QStringList &authors() const;
 	const QString &creation() const;
 	const QString &homeUrl() const;
+	const int &iconSize() const;
 
 	const QString &fileName() const;
 	void setFileName(const QString &);
