Entering 'iris'
Entering 'src/libpsi'
Index: psi/src/discodlg.cpp
===================================================================
--- psi.orig/src/discodlg.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/discodlg.cpp	2013-01-15 14:04:18.468826713 +0400
@@ -1009,7 +1009,8 @@

 	// create toolbar
 	toolBar = new QToolBar(tr("Service Discovery toolbar"), dlg);
-	toolBar->setIconSize(QSize(16,16));
+	int s = PsiIconset::instance()->system().iconSize();
+	toolBar->setIconSize(QSize(s,s));
 	toolBar->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Preferred);
 	replaceWidget(dlg->toolBarPlaceholder, toolBar);

Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:18.472826714 +0400
@@ -715,7 +715,8 @@
 	connect(bm, SIGNAL(conferencesChanged(QList<ConferenceBookmark>)), SLOT(updateBookmarkIcon()));
 	connect(bm, SIGNAL(bookmarksSaved()), SLOT(updateBookmarkIcon()));

-	ui_.toolbar->setIconSize(QSize(16,16));
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.toolbar->setIconSize(QSize(s,s));
 	ui_.toolbar->addAction(d->act_clear);
 	ui_.toolbar->addAction(d->act_find);
 	ui_.toolbar->addAction(d->act_configure);
Index: psi/src/options/optionsdlg.cpp
===================================================================
--- psi.orig/src/options/optionsdlg.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/options/optionsdlg.cpp	2013-01-15 14:04:18.472826714 +0400
@@ -25,6 +25,7 @@
 #include "fancylabel.h"
 #include "iconset.h"
 #include "iconwidget.h"
+#include "psiiconset.h"

 #include <QLayout>
 #include <QLabel>
@@ -88,7 +89,8 @@
 			iconSize = icon.availableSizes().first();
 		else
 #endif
-			iconSize = QSize(16, 16);
+			int s = PsiIconset::instance()->system().iconSize();
+			iconSize = QSize(s,s);
 		QRect iconRect = opt.rect;
 		iconRect.setLeft(4);
 		iconRect.setWidth(iconSize.width());
@@ -148,7 +150,8 @@
 			iconSize = icon.availableSizes().first();
 		else
 #endif
-			iconSize = QSize(16, 16);
+			int s = PsiIconset::instance()->system().iconSize();
+			iconSize = QSize(s,s);

 		int width = iconSize.width();
 		width += 8;
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:18.472826714 +0400
@@ -271,6 +271,10 @@
 	ui_.splitter->optionsChanged();
 	ui_.mle->optionsChanged();

+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.lb_status->setFixedSize(s,s);
+	ui_.lb_client->setFixedSize(s,s);
+
 	ui_.tb_pgp->hide();
 	if (smallChat_) {
 		ui_.lb_status->hide();
@@ -395,7 +399,8 @@
 void PsiChatDlg::initToolBar()
 {
 	ui_.toolbar->setWindowTitle(tr("Chat Toolbar"));
-	ui_.toolbar->setIconSize(QSize(16, 16));
+	int s = PsiIconset::instance()->system().iconSize();
+	ui_.toolbar->setIconSize(QSize(s, s));
 	ui_.toolbar->addAction(act_clear_);
 // typeahead find bar
 	ui_.toolbar->addAction(act_find);
Index: psi/src/psicontactlistviewdelegate.cpp
===================================================================
--- psi.orig/src/psicontactlistviewdelegate.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/psicontactlistviewdelegate.cpp	2013-01-15 14:04:18.472826714 +0400
@@ -47,6 +47,7 @@
 static const QString avatarAtLeftOptionPath = "options.ui.contactlist.avatars.avatars-at-left";
 static const QString showStatusIconsPath = "options.ui.contactlist.show-status-icons";
 static const QString statusIconsOverAvatarsPath = "options.ui.contactlist.status-icon-over-avatar";
+static const QString statusIconsetOptionPath = "options.iconsets.status";

 PsiContactListViewDelegate::PsiContactListViewDelegate(ContactListView* parent)
 	: ContactListViewDelegate(parent)
@@ -59,6 +60,8 @@
 	connect(alertTimer_, SIGNAL(timeout()), SLOT(updateAlerts()));

 	connect(PsiOptions::instance(), SIGNAL(optionChanged(const QString&)), SLOT(optionChanged(const QString&)));
+	connect(PsiIconset::instance(), SIGNAL(rosterIconsSizeChanged(int)), SLOT(rosterIconsSizeChanged(int)));
+	statusIconSize_ = PsiIconset::instance()->roster.value(PsiOptions::instance()->getOption(statusIconsetOptionPath).toString())->iconSize();
 	optionChanged(slimGroupsOptionPath);
 	optionChanged(outlinedGroupsOptionPath);
 	optionChanged(contactListFontOptionPath);
@@ -538,7 +541,7 @@
 		font_ = new QFont();
 		font_->fromString(PsiOptions::instance()->getOption(contactListFontOptionPath).toString());
 		fontMetrics_ = new QFontMetrics(*font_);
-		rowHeight_ = qMax(fontMetrics_->height()+2, 18); // 18 - default row height
+		rowHeight_ = qMax(fontMetrics_->height()+2, statusIconSize_+2);
 		contactList()->viewport()->update();
 	}
 	else if (option == contactListBackgroundOptionPath) {
@@ -613,6 +616,13 @@
 	}
 }

+void PsiContactListViewDelegate::rosterIconsSizeChanged(int size)
+{
+	statusIconSize_ = size;
+	rowHeight_ = qMax(fontMetrics_->height()+2, statusIconSize_+2);
+	contactList()->viewport()->update();
+}
+
 void PsiContactListViewDelegate::drawText(QPainter* painter, const QStyleOptionViewItem& o, const QRect& rect, const QString& text, const QModelIndex& index) const
 {
 	QRect r(rect);
Index: psi/src/psicontactlistviewdelegate.h
===================================================================
--- psi.orig/src/psicontactlistviewdelegate.h	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/psicontactlistviewdelegate.h	2013-01-15 14:04:18.472826714 +0400
@@ -58,6 +58,7 @@
 private slots:
 	void optionChanged(const QString& option);
 	void updateAlerts();
+	void rosterIconsSizeChanged(int size);

 private:
 	QTimer* alertTimer_;
@@ -69,6 +70,7 @@
 	bool showAvatars_, useDefaultAvatar_, avatarAtLeft_, showStatusIcons_, statusIconsOverAvatars_;
 	int avatarSize_, avatarRadius_;
 	mutable QHash<QModelIndex, bool> alertingIndexes_;
+	int statusIconSize_;
 };

 #endif
Index: psi/src/psiiconset.cpp
===================================================================
--- psi.orig/src/psiiconset.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/psiiconset.cpp	2013-01-15 14:04:18.472826714 +0400
@@ -333,6 +333,11 @@
 	QString cur_system = PsiOptions::instance()->getOption("options.iconsets.system").toString();
 	if (d->cur_system != cur_system) {
 		Iconset sys = d->systemIconset(&ok);
+
+		if(sys.iconSize() != d->system.iconSize()) {
+			emit systemIconsSizeChanged(sys.iconSize());
+		}
+
 		d->loadIconset(&d->system, &sys);

 		//d->system = d->systemIconset();
@@ -507,6 +512,10 @@
 	if (d->cur_status != cur_status) {
 		Iconset *newDef = d->defaultRosterIconset(&ok);
 		Iconset *oldDef = roster[d->cur_status];
+
+		if(oldDef->iconSize() != newDef->iconSize())
+			emit rosterIconsSizeChanged(newDef->iconSize());
+
 		d->loadIconset(oldDef, newDef);

 		roster.remove(d->cur_status);
Index: psi/src/psiiconset.h
===================================================================
--- psi.orig/src/psiiconset.h	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/psiiconset.h	2013-01-15 14:04:18.472826714 +0400
@@ -80,6 +80,8 @@

 signals:
 	void emoticonsChanged();
+	void systemIconsSizeChanged(int);
+	void rosterIconsSizeChanged(int);

 public slots:
 	static void reset();
Index: psi/src/psitoolbar.cpp
===================================================================
--- psi.orig/src/psitoolbar.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/psitoolbar.cpp	2013-01-15 14:04:18.472826714 +0400
@@ -29,6 +29,7 @@
 #include <QToolButton>

 #include "iconset.h"
+#include "psiiconset.h"
 #include "iconaction.h"
 #include "psioptions.h"
 #include "options/opt_toolbars.h"
@@ -63,7 +64,8 @@
 	Q_ASSERT(mainWindow);
 	Q_ASSERT(actionList_);

-	setIconSize(QSize(16, 16));
+	int s = PsiIconset::instance()->system().iconSize();
+	setIconSize(QSize(s, s));

 	customizeAction_ = new QAction(tr("&Configure Toolbar..."), this);
 	connect(customizeAction_, SIGNAL(triggered()), this, SIGNAL(customize()));
Index: psi/src/tools/iconset/iconset.cpp
===================================================================
--- psi.orig/src/tools/iconset/iconset.cpp	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/tools/iconset/iconset.cpp	2013-01-15 14:04:18.472826714 +0400
@@ -1028,6 +1028,7 @@
 		//authors << "I. M. Anonymous";
 		//creation = "1900-01-01";
 		homeUrl = QString::null;
+		iconSize_ = 16;
 	}

 public:
@@ -1036,6 +1037,7 @@
 	QHash<QString, PsiIcon *> dict; // unsorted hash for fast search
 	QList<PsiIcon *> list;          // sorted list
 	QHash<QString, QString> info;
+	int iconSize_;

 public:
 	Private()
@@ -1129,6 +1131,8 @@
 	{
 		Q_UNUSED(dir);

+		iconSize_ = 16;
+
 		for (QDomNode node = i.firstChild(); !node.isNull(); node = node.nextSibling()) {
 			QDomElement e = node.toElement();
 			if( e.isNull() ) {
@@ -1139,6 +1143,10 @@
 			if ( tag == "name" ) {
 				name = e.text();
 			}
+			else if( tag == "size") {
+				QString str = e.text();
+				iconSize_ = str.toInt();
+			}
 			else if ( tag == "version" ) {
 				version = e.text();
 			}
@@ -1410,6 +1418,7 @@
 		filename = from.filename;
 		authors = from.authors;
 		info = from.info;
+		iconSize_ = from.iconSize_;
 	}
 };
 //! \endif
@@ -1599,6 +1608,14 @@
 }

 /**
+ * Returns the icons size from Iconset.
+ */
+const int &Iconset::iconSize() const
+{
+	return d->iconSize_;
+}
+
+/**
  * Returns the Iconset version.
  */
 const QString &Iconset::version() const
Index: psi/src/tools/iconset/iconset.h
===================================================================
--- psi.orig/src/tools/iconset/iconset.h	2013-01-15 14:04:18.476826714 +0400
+++ psi/src/tools/iconset/iconset.h	2013-01-15 14:04:18.472826714 +0400
@@ -207,6 +207,7 @@
 	const QStringList &authors() const;
 	const QString &creation() const;
 	const QString &homeUrl() const;
+	const int &iconSize() const;

 	const QString &fileName() const;
 	void setFileName(const QString &);
