--- a/options/default.xml
+++ b/options/default.xml
@@ -120,6 +120,8 @@
 				<css type="QString" />
 				<hide-when-closing type="bool">false</hide-when-closing>
 				<status-with-priority comment="Show priority with status change" type="bool">false</status-with-priority>
+				<log-height type="int">325</log-height>
+				<chatedit-height type="int">10</chatedit-height>
 			</chat>
 			<save>
 				<toolbars-state type="QByteArray"/>
--- a/src/chatsplitter.cpp
+++ b/src/chatsplitter.cpp
@@ -75,6 +75,15 @@ void ChatSplitter::setSizes(const QList<int>& list)
 		splitter_->setSizes(list);
 }
 
+QList<int> ChatSplitter::sizes()
+{
+	QList<int> list;
+	if(splitter_)
+		list = splitter_->sizes();
+
+	return list;
+}
+
 /**
  * Moves \a child either to the real QSplitter, or adds it to the
  * private layout.
@@ -130,6 +139,8 @@ void ChatSplitter::updateLayout()
 
 	foreach(QWidget* child, children_)
 		updateChildLayout(child);
+
+	connect(splitter_, SIGNAL(splitterMoved(int,int)), SIGNAL(splitterMoved(int,int)));
 }
 
 /**
--- a/src/chatsplitter.h
+++ b/src/chatsplitter.h
@@ -35,6 +35,7 @@ public:
 	void setOrientation(Qt::Orientation orientation);
 	void addWidget(QWidget* widget);
 	void setSizes(const QList<int>& list);
+	QList<int> sizes();
 
 protected:
 	/**
@@ -49,6 +50,9 @@ public slots:
 private slots:
 	void childDestroyed(QObject* obj);
 
+signals:
+	void splitterMoved(int,int);
+
 private:
 	void updateChildLayout(QWidget* child);
 	void updateLayout();
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -240,6 +240,9 @@ public:
 	int logSize;
 	int rosterSize;
 
+	int logHeight;
+	int chateditHeight;
+
 public:
 	bool trackBar;
 
@@ -769,6 +772,7 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, const Jid &j, TabManager *tabManager)
 	ui_.tb_actions->setStyleSheet(" QToolButton::menu-indicator { image:none } ");
 
 	connect(ui_.hsplitter, SIGNAL(splitterMoved(int,int)), this, SLOT(horizSplitterMoved()));
+	connect(ui_.vsplitter, SIGNAL(splitterMoved(int,int)), this, SLOT(verticalSplitterMoved(int,int)));
 
 	// resize the horizontal splitter
 	d->logSize = PsiOptions::instance()->getOption("options.ui.muc.log-width").toInt();
@@ -786,10 +790,9 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, const Jid &j, TabManager *tabManager)
 		ui_.hsplitter->insertWidget(0, ui_.lv_users);  // Swap widgets
 
 	// resize the vertical splitter
-	list.clear();
-	list << 324;
-	list << 10;
-	ui_.vsplitter->setSizes(list);
+	d->logHeight = PsiOptions::instance()->getOption("options.ui.chat.log-height").toInt();
+	d->chateditHeight = PsiOptions::instance()->getOption("options.ui.chat.chatedit-height").toInt();
+	setVSplitterPosition(d->logHeight, d->chateditHeight);
 
 	X11WM_CLASS("groupchat");
 
@@ -835,6 +838,26 @@ void GCMainDlg::horizSplitterMoved()
 	PsiOptions::instance()->setOption("options.ui.muc.roster-width", d->rosterSize);
 }
 
+void GCMainDlg::verticalSplitterMoved(int, int)
+{
+	QList<int> list = ui_.vsplitter->sizes();
+	d->logHeight = list.first();
+	d->chateditHeight = list.last();
+	PsiOptions::instance()->setOption("options.ui.chat.log-height", d->logHeight);
+	PsiOptions::instance()->setOption("options.ui.chat.chatedit-height", d->chateditHeight);
+
+	emit vSplitterMoved(d->logHeight, d->chateditHeight);
+}
+
+void GCMainDlg::setVSplitterPosition(int log,int chat)
+{
+	QList<int> list;
+	d->logHeight = log;
+	d->chateditHeight = chat;
+	list << log << chat;
+	ui_.vsplitter->setSizes(list);
+}
+
 void GCMainDlg::doMinimize()
 {
 	window()->showMinimized();
--- a/src/groupchatdlg.h
+++ b/src/groupchatdlg.h
@@ -85,6 +85,7 @@ public:
 	virtual TabbableWidget::State state() const;
 	virtual int unreadMessageCount() const;
 	virtual QString desiredCaption() const;
+	virtual void setVSplitterPosition(int log,int chat);
 
 protected:
 	void setShortcuts();
@@ -140,6 +141,7 @@ private slots:
 #endif
 	void chatEditCreated();
 	void horizSplitterMoved();
+	void verticalSplitterMoved(int, int);
 	void doMinimize();
 
 
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -219,10 +219,11 @@ void PsiChatDlg::initUi()
 
 	connect(account()->capsManager(), SIGNAL(capsChanged(const Jid&)), SLOT(capsChanged(const Jid&)));
 
-	QList<int> list;
-	list << 324;
-	list << 96;
-	ui_.splitter->setSizes(list);
+	logHeight = PsiOptions::instance()->getOption("options.ui.chat.log-height").toInt();
+	chateditHeight = PsiOptions::instance()->getOption("options.ui.chat.chatedit-height").toInt();
+	setVSplitterPosition(logHeight, chateditHeight);
+
+	connect(ui_.splitter, SIGNAL(splitterMoved(int,int)), this, SLOT(verticalSplitterMoved(int,int)));
 
 	smallChat_ = PsiOptions::instance()->getOption("options.ui.chat.use-small-chats").toBool();
  	ui_.pb_send->setIcon(IconsetFactory::icon("psi/action_button_send").icon());
@@ -248,6 +249,26 @@ void PsiChatDlg::initUi()
 	unacked_messages = 0;
 }
 
+void PsiChatDlg::verticalSplitterMoved(int, int)
+{
+	QList<int> list = ui_.splitter->sizes();
+	logHeight = list.first();
+	chateditHeight = list.last();
+	PsiOptions::instance()->setOption("options.ui.chat.log-height", logHeight);
+	PsiOptions::instance()->setOption("options.ui.chat.chatedit-height", chateditHeight);
+
+	emit vSplitterMoved(logHeight, chateditHeight);
+}
+
+void PsiChatDlg::setVSplitterPosition(int log,int chat)
+{
+	QList<int> list;
+	logHeight = log;
+	chateditHeight = chat;
+	list << log << chat;
+	ui_.splitter->setSizes(list);
+}
+
 void PsiChatDlg::updateCountVisibility()
 {
 	if (PsiOptions::instance()->getOption("options.ui.message.show-character-count").toBool() && !smallChat_) {
--- a/src/psichatdlg.h
+++ b/src/psichatdlg.h
@@ -21,6 +21,8 @@ class PsiChatDlg : public ChatDlg
 public:
 	PsiChatDlg(const Jid& jid, PsiAccount* account, TabManager* tabManager);
 
+	virtual void setVSplitterPosition(int log,int chat);
+
 protected:
 	// reimplemented
 	void contextMenuEvent(QContextMenuEvent *);
@@ -43,6 +45,7 @@ private slots:
 	void updateCountVisibility();
 	void updateContactAdding(PsiContact* c = 0);
 	void updateContactAdding(const Jid &j);
+	void verticalSplitterMoved(int, int);
 
 	// reimplemented
 	void chatEditCreated();
@@ -104,6 +107,9 @@ private:
 	static PsiIcon *throbber_icon;
 	bool tabmode;
 	QPointer <PsiWindowHeader> winHeader_;
+
+	int logHeight;
+	int chateditHeight;
 };
 
 #endif
--- a/src/tabs/tabbablewidget.h
+++ b/src/tabs/tabbablewidget.h
@@ -67,11 +67,13 @@ public:
 	virtual State state() const = 0;
 	virtual int unreadMessageCount() const = 0;
 	virtual QString desiredCaption() const = 0;
+	virtual void setVSplitterPosition(int,int){}; // default implementation do nothing
 
 signals:
 	void invalidateTabInfo();
 	void updateFlashState();
 	void eventsRead(const Jid &);
+	void vSplitterMoved(int,int);
 
 public slots:
 	virtual void deactivated();
--- a/src/tabs/tabdlg.cpp
+++ b/src/tabs/tabdlg.cpp
@@ -426,6 +426,7 @@ void TabDlg::addTab(TabbableWidget* tab)
 
 	connect(tab, SIGNAL(invalidateTabInfo()), SLOT(updateTab()));
 	connect(tab, SIGNAL(updateFlashState()), SLOT(updateFlashState()));
+	connect(tab, SIGNAL(vSplitterMoved(int,int)), SLOT(updateVSplitters(int,int)));
 
 	updateTab(tab);
 	setUpdatesEnabled(true);
@@ -496,6 +497,7 @@ void TabDlg::removeTabWithNoChecks(TabbableWidget *tab)
 {
 	disconnect(tab, SIGNAL(invalidateTabInfo()), this, SLOT(updateTab()));
 	disconnect(tab, SIGNAL(updateFlashState()), this, SLOT(updateFlashState()));
+	disconnect(tab, SIGNAL(vSplitterMoved(int,int)), this, SLOT(updateVSplitters(int,int)));
 
 	tabs_.removeAll(tab);
 	tabWidget_->removePage(tab);
@@ -901,3 +903,10 @@ void TabDlg::setTabIcon(QWidget *widget,const QIcon &icon)
 {
 	tabWidget_->setTabIcon(widget, icon);
 }
+
+void TabDlg::updateVSplitters(int log, int chat)
+{
+	foreach(TabbableWidget *w, tabs_) {
+		w->setVSplitterPosition(log, chat);
+	}
+}
--- a/src/tabs/tabdlg.h
+++ b/src/tabs/tabdlg.h
@@ -145,6 +145,7 @@ private slots:
 	void menu_sendTabTo(QAction *act);
 	void queuedSendTabTo(TabbableWidget* chat, TabDlg *dest);
 	void showTabMenu(int tab, QPoint pos, QContextMenuEvent * event);
+	void updateVSplitters(int log, int chat);
 
 private:
 	TabDlgDelegate *delegate_;
