--- psi.orig/src/psicontactlistviewdelegate.cpp
+++ psi/src/psicontactlistviewdelegate.cpp
@@ -422,13 +422,23 @@ void PsiContactListViewDelegate::drawGro
 	o.font = *font_;
 	o.fontMetrics = *fontMetrics_;
 	QPalette palette = o.palette;
-	palette.setColor(QPalette::Base, ColorOpt::instance()->color("options.ui.look.colors.contactlist.grouping.header-background"));
-	palette.setColor(QPalette::Text, ColorOpt::instance()->color("options.ui.look.colors.contactlist.grouping.header-foreground"));
+	bool b = PsiOptions::instance()->getOption("options.ui.look.contactlist.use-slim-group-headings").toBool();
+	QColor background = ColorOpt::instance()->color("options.ui.look.colors.contactlist.grouping.header-background");
+	QColor foreground = ColorOpt::instance()->color("options.ui.look.colors.contactlist.grouping.header-foreground");
+	if (!b)
+		palette.setColor(QPalette::Base, background);
+	palette.setColor(QPalette::Text, foreground);
 	o.palette = palette;
 
 	drawBackground(painter, o, index);
 
 	QRect r = option.rect;
+	if (!b && PsiOptions::instance()->getOption("options.ui.look.contactlist.use-outlined-group-headings").toBool()) {
+		painter->setPen(QPen(foreground));
+		QRect gr(r);
+		gr.setLeft(contactList()->x());
+		painter->drawRect(gr);
+	}
 
 	const QPixmap& pixmap = index.data(ContactListModel::ExpandedRole).toBool() ?
 	                        IconsetFactory::iconPtr("psi/groupOpen")->pixmap() :
@@ -443,6 +453,13 @@ void PsiContactListViewDelegate::drawGro
 
 	QString text = index.data(Qt::ToolTipRole).toString();
 	drawText(painter, o, r, text, index);
+
+	if(b && !(option.state & QStyle::State_Selected)) {
+		int h = r.y() + (r.height() / 2);
+		int x = r.left() + fontMetrics_->width(text) + 1;
+		painter->setPen(QPen(background,2));
+		painter->drawLine(x, h, r.right(), h);
+	}
 }
 
 void PsiContactListViewDelegate::drawAccount(QPainter* painter, const QStyleOptionViewItem& option, const QModelIndex& index) const
@@ -452,12 +469,17 @@ void PsiContactListViewDelegate::drawAcc
 	o.fontMetrics = *fontMetrics_;
 	QPalette palette = o.palette;
 	palette.setColor(QPalette::Base, ColorOpt::instance()->color("options.ui.look.colors.contactlist.profile.header-background"));
-	palette.setColor(QPalette::Text, ColorOpt::instance()->color("options.ui.look.colors.contactlist.profile.header-foreground"));
+	QColor foreground = ColorOpt::instance()->color("options.ui.look.colors.contactlist.profile.header-foreground");
+	palette.setColor(QPalette::Text, foreground);
 	o.palette = palette;
 
 	drawBackground(painter, o, index);
 
 	QRect r = option.rect;
+	if (PsiOptions::instance()->getOption("options.ui.look.contactlist.use-outlined-group-headings").toBool()) {
+		painter->setPen(QPen(foreground));
+		painter->drawRect(r);
+	}
 
 	QRect avatarRect(r);
 	const QPixmap statusPixmap = this->statusPixmap(index);
