--- a/options/default.xml
+++ b/options/default.xml
@@ -117,6 +117,7 @@
 				<use-small-chats type="bool">false</use-small-chats>
 				<use-message-icons type="bool">true</use-message-icons>
 				<scaled-message-icons type="bool">false</scaled-message-icons>
+				<show-status-changes type="bool">true</show-status-changes>
 				<warn-before-clear type="bool">true</warn-before-clear>
 				<css type="QString" />
 				<hide-when-closing type="bool">false</hide-when-closing>
--- a/src/chatdlg.cpp
+++ b/src/chatdlg.cpp
@@ -481,7 +481,9 @@ void ChatDlg::updateContact(const Jid &j, bool fromPresence)
 			key_ = userStatus.publicKeyID;
 			updatePGP();
 
-			if (fromPresence && statusChanged) {
+			if (PsiOptions::instance()->getOption("options.ui.chat.show-status-changes").toBool()
+				&& fromPresence && statusChanged)
+			{
 				chatView()->dispatchMessage(MessageView::statusMessage(
 												dispNick_, status_,
 												statusString_, priority_));
