--- git.orig/src/mainwin.cpp
+++ git/src/mainwin.cpp
@@ -508,9 +508,9 @@ MainWin::MainWin(bool _onTop, bool _asTo
 
 	mainMenuBar()->addMenu(viewMenu);
 	d->getAction("show_offline")->addTo(viewMenu);
-	if (PsiOptions::instance()->getOption("options.ui.menu.view.show-away").toBool()) {
+	/*if (PsiOptions::instance()->getOption("options.ui.menu.view.show-away").toBool()) {
 		d->getAction("show_away")->addTo(viewMenu);
-	}
+	}*/
 	d->getAction("show_hidden")->addTo(viewMenu);
 	d->getAction("show_agents")->addTo(viewMenu);
 	d->getAction("show_self")->addTo(viewMenu);
--- git.orig/src/psiactionlist.cpp
+++ git/src/psiactionlist.cpp
@@ -146,9 +146,9 @@ void PsiActionList::Private::createMainW
 		IconAction *showOffline = new IconAction(tr("Show Offline Contacts"), "psi/show_offline", tr("Show Offline Contacts"), 0, viewGroups, 0, true);
 		showOffline->setWhatsThis(tr("Toggles visibility of offline contacts in roster"));
 
-		IconAction *showAway = new IconAction(tr("Show Away/XA/DnD Contacts"), "psi/show_away", tr("Show Away/XA/DnD Contacts"), 0,
+		/*IconAction *showAway = new IconAction(tr("Show Away/XA/DnD Contacts"), "psi/show_away", tr("Show Away/XA/DnD Contacts"), 0,
 		                                      PsiOptions::instance()->getOption("options.ui.menu.view.show-away").toBool() ? (QObject*)viewGroups : (QObject*)this, 0, true);
-		showAway->setWhatsThis(tr("Toggles visibility of away/xa/dnd contacts in roster"));
+		showAway->setWhatsThis(tr("Toggles visibility of away/xa/dnd contacts in roster"));*/
 
 		IconAction *showHidden = new IconAction(tr("Show Hidden Contacts"), "psi/show_hidden", tr("Show Hidden Contacts"), 0, viewGroups, 0, true);
 		showHidden->setWhatsThis(tr("Toggles visibility of hidden contacts in roster"));
@@ -166,7 +166,7 @@ void PsiActionList::Private::createMainW
 			{ "view_groups",  viewGroups  },
 			{ "enable_groups",   actEnableGroups  },
 			{ "show_offline", showOffline },
-			{ "show_away",    showAway    },
+			//{ "show_away",    showAway    },
 			{ "show_hidden",  showHidden  },
 			{ "show_agents",  showAgents  },
 			{ "show_self",    showSelf    },
