Index: psi/src/registrationdlg.cpp
===================================================================
--- psi.orig/src/registrationdlg.cpp	2013-01-15 14:04:19.892826771 +0400
+++ psi/src/registrationdlg.cpp	2013-01-15 14:04:19.888826771 +0400
@@ -159,6 +159,7 @@
 	QVBoxLayout *vb1 = new QVBoxLayout(this);
 	vb1->setMargin(4);
 	d->lb_top = new QLabel(this);
+	d->lb_top->setWordWrap(true);
 	d->lb_top->setFrameStyle( QFrame::Panel | QFrame::Sunken );
 	d->lb_top->hide();
 	vb1->addWidget(d->lb_top);
Index: psi/src/searchdlg.cpp
===================================================================
--- psi.orig/src/searchdlg.cpp	2013-01-15 14:04:19.892826771 +0400
+++ psi/src/searchdlg.cpp	2013-01-15 14:04:19.888826771 +0400
@@ -160,6 +160,8 @@
 	QList<QLineEdit*> le_field;
 	XDataWidget *xdata;
 	XData xdata_form;
+	QScrollArea *scrollArea;
+
 };
 
 SearchDlg::SearchDlg(const Jid &jid, PsiAccount *pa)
@@ -174,6 +176,8 @@
 	d->pa->dialogRegister(this, d->jid);
 	d->jt = 0;
 	d->xdata = 0;
+	d->scrollArea = scrollArea;
+
 
 	setWindowTitle(windowTitle().arg(d->jid.full()));
 
@@ -184,8 +188,8 @@
 
 	d->gr_form = new QWidget(gb_search);
 	d->gr_form_layout = new QGridLayout(d->gr_form);
-	d->gr_form_layout->setSpacing(4);
-	replaceWidget(lb_form, d->gr_form);
+	d->gr_form_layout->setSpacing(0);
+	d->scrollArea->setWidget(d->gr_form);
 	d->gr_form->hide();
 
 	pb_add->setEnabled(false);
Index: psi/src/search.ui
===================================================================
--- psi.orig/src/search.ui	2013-01-15 14:04:19.892826771 +0400
+++ psi/src/search.ui	2013-01-15 14:04:19.888826771 +0400
@@ -1,308 +1,315 @@
-<ui version="4.0" stdsetdef="1" >
-  <author></author>
-  <comment></comment>
-  <exportmacro></exportmacro>
-  <class>Search</class>
-  <widget class="QWidget" name="Search" >
-    <property name="geometry" >
-      <rect>
-        <x>0</x>
-        <y>0</y>
-        <width>568</width>
-        <height>422</height>
-      </rect>
-    </property>
-    <property name="windowTitle" >
-      <string>Search: %1</string>
-    </property>
-    <layout class="QVBoxLayout" >
-      <property name="margin" >
-        <number>11</number>
-      </property>
-      <property name="spacing" >
-        <number>6</number>
-      </property>
-      <item>
-        <layout class="QHBoxLayout" >
-          <property name="margin" >
-            <number>0</number>
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
+ <class>Search</class>
+ <widget class="QWidget" name="Search">
+  <property name="geometry">
+   <rect>
+    <x>0</x>
+    <y>0</y>
+    <width>780</width>
+    <height>480</height>
+   </rect>
+  </property>
+  <property name="minimumSize">
+   <size>
+    <width>640</width>
+    <height>480</height>
+   </size>
+  </property>
+  <property name="windowTitle">
+   <string>Search: %1</string>
+  </property>
+  <layout class="QVBoxLayout">
+   <property name="margin">
+    <number>5</number>
+   </property>
+   <item>
+    <layout class="QHBoxLayout">
+     <property name="margin">
+      <number>0</number>
+     </property>
+     <item>
+      <widget class="QGroupBox" name="gb_search">
+       <property name="minimumSize">
+        <size>
+         <width>280</width>
+         <height>0</height>
+        </size>
+       </property>
+       <property name="maximumSize">
+        <size>
+         <width>256</width>
+         <height>32767</height>
+        </size>
+       </property>
+       <property name="title">
+        <string>Search Fields</string>
+       </property>
+       <layout class="QVBoxLayout">
+        <property name="margin">
+         <number>1</number>
+        </property>
+        <item>
+         <layout class="QVBoxLayout">
+          <property name="margin">
+           <number>0</number>
           </property>
           <item>
-            <widget class="QGroupBox" name="gb_search" >
-              <property name="maximumSize" >
-                <size>
-                  <width>256</width>
-                  <height>32767</height>
-                </size>
-              </property>
-              <property name="title" >
-                <string>Search Fields</string>
-              </property>
-              <layout class="QVBoxLayout" >
-                <property name="margin" >
-                  <number>11</number>
-                </property>
-                <property name="spacing" >
-                  <number>6</number>
-                </property>
-                <item>
-                  <layout class="QVBoxLayout" >
-                    <property name="margin" >
-                      <number>0</number>
-                    </property>
-                    <item>
-                      <widget class="QLabel" name="lb_instructions" >
-                        <property name="frameShape" >
-                          <enum>QFrame::Panel</enum>
-                        </property>
-                        <property name="frameShadow" >
-                          <enum>QFrame::Sunken</enum>
-                        </property>
-                        <property name="text" >
-                          <string>Instructions go here.</string>
-                        </property>
-                        <property name="wordWrap" >
-                          <bool>true</bool>
-                        </property>
-                      </widget>
-                    </item>
-                    <item>
-                      <widget class="QLabel" name="lb_form" >
-                        <property name="text" >
-                          <string>form</string>
-                        </property>
-                      </widget>
-                    </item>
-                    <item>
-                      <layout class="QHBoxLayout" >
-                        <property name="margin" >
-                          <number>0</number>
-                        </property>
-                        <item>
-                          <widget class="IconButton" name="pb_search" >
-                            <property name="text" >
-                              <string>&amp;Search</string>
-                            </property>
-                            <property name="psiIconName" stdset="0" >
-                              <string>psi/search</string>
-                            </property>
-                            <property name="default" >
-                              <bool>true</bool>
-                            </property>
-                          </widget>
-                        </item>
-                        <item>
-                          <spacer name="Spacer1" >
-                            <property name="sizeHint" >
-                              <size>
-                                <width>30</width>
-                                <height>16</height>
-                              </size>
-                            </property>
-                            <property name="sizeType" >
-                              <enum>Expanding</enum>
-                            </property>
-                            <property name="orientation" >
-                              <enum>Horizontal</enum>
-                            </property>
-                          </spacer>
-                        </item>
-                        <item>
-                          <widget class="IconButton" name="pb_stop" >
-                            <property name="text" >
-                              <string>&amp;Stop</string>
-                            </property>
-                            <property name="psiIconName" stdset="0" >
-                              <string>psi/stop</string>
-                            </property>
-                          </widget>
-                        </item>
-                      </layout>
-                    </item>
-                  </layout>
-                </item>
-              </layout>
-            </widget>
+           <widget class="QLabel" name="lb_instructions">
+            <property name="frameShape">
+             <enum>QFrame::StyledPanel</enum>
+            </property>
+            <property name="frameShadow">
+             <enum>QFrame::Sunken</enum>
+            </property>
+            <property name="text">
+             <string>Instructions go here.</string>
+            </property>
+            <property name="wordWrap">
+             <bool>true</bool>
+            </property>
+           </widget>
           </item>
           <item>
-            <layout class="QVBoxLayout" >
-              <property name="margin" >
-                <number>0</number>
-              </property>
-              <item>
-                <widget class="QTreeWidget" name="lv_results" >
-                  <property name="alternatingRowColors" >
-                    <bool>true</bool>
-                  </property>
-                  <property name="selectionMode" >
-                    <enum>QAbstractItemView::ExtendedSelection</enum>
-                  </property>
-                  <property name="selectionBehavior" >
-                    <enum>QAbstractItemView::SelectRows</enum>
-                  </property>
-                  <property name="indentation" >
-                    <number>0</number>
-                  </property>
-                  <property name="uniformRowHeights" >
-                    <bool>true</bool>
-                  </property>
-                  <property name="itemsExpandable" >
-                    <bool>false</bool>
-                  </property>
-                  <property name="allColumnsShowFocus" >
-                    <bool>true</bool>
-                  </property>
-                  <property name="sortingEnabled" >
-                    <bool>true</bool>
-                  </property>
-                  <column>
-                    <property name="text" >
-                      <string>Nickname</string>
-                    </property>
-                    <property name="clickable" >
-                      <bool>true</bool>
-                    </property>
-                    <property name="resizable" >
-                      <bool>true</bool>
-                    </property>
-                  </column>
-                  <column>
-                    <property name="text" >
-                      <string>First Name</string>
-                    </property>
-                    <property name="clickable" >
-                      <bool>true</bool>
-                    </property>
-                    <property name="resizable" >
-                      <bool>true</bool>
-                    </property>
-                  </column>
-                  <column>
-                    <property name="text" >
-                      <string>Last Name</string>
-                    </property>
-                    <property name="clickable" >
-                      <bool>true</bool>
-                    </property>
-                    <property name="resizable" >
-                      <bool>true</bool>
-                    </property>
-                  </column>
-                  <column>
-                    <property name="text" >
-                      <string>E-Mail Address</string>
-                    </property>
-                    <property name="clickable" >
-                      <bool>true</bool>
-                    </property>
-                    <property name="resizable" >
-                      <bool>true</bool>
-                    </property>
-                  </column>
-                  <column>
-                    <property name="text" >
-                      <string>XMPP Address</string>
-                    </property>
-                    <property name="clickable" >
-                      <bool>true</bool>
-                    </property>
-                    <property name="resizable" >
-                      <bool>true</bool>
-                    </property>
-                  </column>
-                </widget>
-              </item>
-              <item>
-                <layout class="QHBoxLayout" >
-                  <property name="margin" >
-                    <number>0</number>
-                  </property>
-                  <item>
-                    <spacer name="Spacer3" >
-                      <property name="sizeHint" >
-                        <size>
-                          <width>222</width>
-                          <height>16</height>
-                        </size>
-                      </property>
-                      <property name="sizeType" >
-                        <enum>Expanding</enum>
-                      </property>
-                      <property name="orientation" >
-                        <enum>Horizontal</enum>
-                      </property>
-                    </spacer>
-                  </item>
-                  <item>
-                    <widget class="IconButton" name="pb_add" >
-                      <property name="text" >
-                        <string>&amp;Add Contact</string>
-                      </property>
-                      <property name="psiIconName" stdset="0" >
-                        <string>psi/addContact</string>
-                      </property>
-                    </widget>
-                  </item>
-                  <item>
-                    <widget class="IconButton" name="pb_info" >
-                      <property name="text" >
-                        <string>User &amp;Info</string>
-                      </property>
-                      <property name="psiIconName" stdset="0" >
-                        <string>psi/vCard</string>
-                      </property>
-                    </widget>
-                  </item>
-                </layout>
-              </item>
-            </layout>
-          </item>
-        </layout>
-      </item>
-      <item>
-        <widget class="Line" name="line" >
-          <property name="frameShape" >
-            <enum>QFrame::HLine</enum>
-          </property>
-          <property name="frameShadow" >
-            <enum>QFrame::Sunken</enum>
-          </property>
-        </widget>
-      </item>
-      <item>
-        <layout class="QHBoxLayout" >
-          <property name="margin" >
-            <number>0</number>
-          </property>
-          <item>
-            <widget class="BusyWidget" name="busy" />
+           <widget class="QScrollArea" name="scrollArea">
+            <property name="sizePolicy">
+             <sizepolicy hsizetype="Preferred" vsizetype="Expanding">
+              <horstretch>0</horstretch>
+              <verstretch>0</verstretch>
+             </sizepolicy>
+            </property>
+            <property name="horizontalScrollBarPolicy">
+             <enum>Qt::ScrollBarAlwaysOff</enum>
+            </property>
+            <property name="widgetResizable">
+             <bool>true</bool>
+            </property>
+            <widget class="QWidget" name="scrollAreaWidgetContents_2">
+             <property name="geometry">
+              <rect>
+               <x>0</x>
+               <y>0</y>
+               <width>270</width>
+               <height>363</height>
+              </rect>
+             </property>
+            </widget>
+           </widget>
           </item>
           <item>
-            <spacer name="Spacer2" >
-              <property name="sizeHint" >
-                <size>
-                  <width>485</width>
-                  <height>16</height>
-                </size>
+           <layout class="QHBoxLayout">
+            <property name="margin">
+             <number>0</number>
+            </property>
+            <item>
+             <widget class="IconButton" name="pb_search" native="true">
+              <property name="text" stdset="0">
+               <string>&amp;Search</string>
               </property>
-              <property name="sizeType" >
-                <enum>Expanding</enum>
+              <property name="psiIconName" stdset="0">
+               <string>psi/search</string>
               </property>
-              <property name="orientation" >
-                <enum>Horizontal</enum>
+              <property name="default" stdset="0">
+               <bool>true</bool>
               </property>
-            </spacer>
-          </item>
-          <item>
-            <widget class="IconButton" name="pb_close" >
-              <property name="text" >
-                <string>&amp;Close</string>
+             </widget>
+            </item>
+            <item>
+             <spacer name="Spacer1">
+              <property name="orientation">
+               <enum>Qt::Horizontal</enum>
               </property>
-            </widget>
+              <property name="sizeType">
+               <enum>QSizePolicy::Expanding</enum>
+              </property>
+              <property name="sizeHint" stdset="0">
+               <size>
+                <width>30</width>
+                <height>16</height>
+               </size>
+              </property>
+             </spacer>
+            </item>
+            <item>
+             <widget class="IconButton" name="pb_stop" native="true">
+              <property name="text" stdset="0">
+               <string>&amp;Stop</string>
+              </property>
+              <property name="psiIconName" stdset="0">
+               <string>psi/stop</string>
+              </property>
+             </widget>
+            </item>
+           </layout>
           </item>
+         </layout>
+        </item>
+       </layout>
+      </widget>
+     </item>
+     <item>
+      <layout class="QVBoxLayout">
+       <property name="margin">
+        <number>0</number>
+       </property>
+       <item>
+        <widget class="QTreeWidget" name="lv_results">
+         <property name="alternatingRowColors">
+          <bool>true</bool>
+         </property>
+         <property name="selectionMode">
+          <enum>QAbstractItemView::ExtendedSelection</enum>
+         </property>
+         <property name="selectionBehavior">
+          <enum>QAbstractItemView::SelectRows</enum>
+         </property>
+         <property name="indentation">
+          <number>0</number>
+         </property>
+         <property name="uniformRowHeights">
+          <bool>true</bool>
+         </property>
+         <property name="itemsExpandable">
+          <bool>false</bool>
+         </property>
+         <property name="sortingEnabled">
+          <bool>true</bool>
+         </property>
+         <property name="allColumnsShowFocus">
+          <bool>true</bool>
+         </property>
+         <column>
+          <property name="text">
+           <string>Nickname</string>
+          </property>
+         </column>
+         <column>
+          <property name="text">
+           <string>First Name</string>
+          </property>
+         </column>
+         <column>
+          <property name="text">
+           <string>Last Name</string>
+          </property>
+         </column>
+         <column>
+          <property name="text">
+           <string>E-Mail Address</string>
+          </property>
+         </column>
+         <column>
+          <property name="text">
+           <string>XMPP Address</string>
+          </property>
+         </column>
+        </widget>
+       </item>
+       <item>
+        <layout class="QHBoxLayout">
+         <property name="margin">
+          <number>0</number>
+         </property>
+         <item>
+          <spacer name="Spacer3">
+           <property name="orientation">
+            <enum>Qt::Horizontal</enum>
+           </property>
+           <property name="sizeType">
+            <enum>QSizePolicy::Expanding</enum>
+           </property>
+           <property name="sizeHint" stdset="0">
+            <size>
+             <width>222</width>
+             <height>16</height>
+            </size>
+           </property>
+          </spacer>
+         </item>
+         <item>
+          <widget class="IconButton" name="pb_add" native="true">
+           <property name="text" stdset="0">
+            <string>&amp;Add Contact</string>
+           </property>
+           <property name="psiIconName" stdset="0">
+            <string>psi/addContact</string>
+           </property>
+          </widget>
+         </item>
+         <item>
+          <widget class="IconButton" name="pb_info" native="true">
+           <property name="text" stdset="0">
+            <string>User &amp;Info</string>
+           </property>
+           <property name="psiIconName" stdset="0">
+            <string>psi/vCard</string>
+           </property>
+          </widget>
+         </item>
         </layout>
-      </item>
+       </item>
+      </layout>
+     </item>
+    </layout>
+   </item>
+   <item>
+    <widget class="Line" name="line">
+     <property name="frameShape">
+      <enum>QFrame::HLine</enum>
+     </property>
+     <property name="frameShadow">
+      <enum>QFrame::Sunken</enum>
+     </property>
+    </widget>
+   </item>
+   <item>
+    <layout class="QHBoxLayout">
+     <property name="margin">
+      <number>0</number>
+     </property>
+     <item>
+      <widget class="BusyWidget" name="busy" native="true"/>
+     </item>
+     <item>
+      <spacer name="Spacer2">
+       <property name="orientation">
+        <enum>Qt::Horizontal</enum>
+       </property>
+       <property name="sizeType">
+        <enum>QSizePolicy::Expanding</enum>
+       </property>
+       <property name="sizeHint" stdset="0">
+        <size>
+         <width>485</width>
+         <height>16</height>
+        </size>
+       </property>
+      </spacer>
+     </item>
+     <item>
+      <widget class="IconButton" name="pb_close" native="true">
+       <property name="text" stdset="0">
+        <string>&amp;Close</string>
+       </property>
+      </widget>
+     </item>
     </layout>
-  </widget>
-  <layoutdefault spacing="6" margin="11" />
+   </item>
+  </layout>
+ </widget>
+ <layoutdefault spacing="6" margin="11"/>
+ <customwidgets>
+  <customwidget>
+   <class>IconButton</class>
+   <extends>QWidget</extends>
+   <header>iconbutton.h</header>
+  </customwidget>
+  <customwidget>
+   <class>BusyWidget</class>
+   <extends>QWidget</extends>
+   <header>busywidget.h</header>
+  </customwidget>
+ </customwidgets>
+ <resources/>
+ <connections/>
 </ui>
Index: psi/src/xdata_widget.cpp
===================================================================
--- psi.orig/src/xdata_widget.cpp	2013-01-15 14:04:19.892826771 +0400
+++ psi/src/xdata_widget.cpp	2013-01-15 14:04:19.888826771 +0400
@@ -33,6 +33,7 @@
 #include <QNetworkAccessManager>
 #include <QNetworkRequest>
 #include <QNetworkReply>
+#include <QSpacerItem>
 
 #include "desktoputil.h"
 #include "xmpp_xdata.h"
@@ -148,12 +149,12 @@
 		return _field;
 	}
 
-	QString labelText() const
+	QString labelText(QString str=": ") const
 	{
 		QString text = _field.label();
 		if ( text.isEmpty() )
 			text = _field.var();
-		return text + ": ";
+		return text + str;
 	}
 
 	QString reqText() const
@@ -207,16 +208,22 @@
 			if ( s == "1" || s == "true" || s == "yes" )
 				checked = true;
 		}
-
-		QLabel *label = new QLabel(labelText(), parent);
-		grid->addWidget(label, row, 0);
+		
+		QHBoxLayout *layout = new QHBoxLayout;
+		QSpacerItem *spacerItem = new QSpacerItem(40, 20, QSizePolicy::Expanding, QSizePolicy::Minimum);
 
 		check = new QCheckBox(parent);
 		check->setChecked(checked);
-		grid->addWidget(check, row, 1);
+		layout->addWidget(check);
+
+		QLabel *label = new QLabel(labelText(""), parent);
+		layout->addWidget(label);
+		layout->addSpacerItem(spacerItem);
+		
+		grid->addLayout(layout, row, 0);
 
 		QLabel *req = new QLabel(reqText(), parent);
-		grid->addWidget(req, row, 2);
+		grid->addWidget(req, row, 1);
 
 		if ( !f.desc().isEmpty() ) {
 			label->setToolTip(f.desc());
@@ -256,8 +263,9 @@
 			text += *it;
 		}
 
-		QLabel *fixed = new QLabel("<qt>" + text + "</qt>", parent);
-		grid->addWidget(fixed, row, 0, 2, 0);
+		QLabel *fixed = new QLabel("<qt>" + text + "<br/></qt>", parent);
+		fixed->setWordWrap(true);
+		grid->addWidget(fixed, row, 0);
 
 		if ( !f.desc().isEmpty() ) {
 			fixed->setToolTip(f.desc());
@@ -294,16 +302,20 @@
 		QString text;
 		if ( f.value().count() )
 			text = f.value().first();
-
+		QVBoxLayout *layout = new QVBoxLayout;
+		
 		QLabel *label = new QLabel(labelText(), parent);
-		grid->addWidget(label, row, 0);
+		label->setWordWrap(true);
+		layout->addWidget(label);
 
 		edit = new QLineEdit(parent);
 		edit->setText(text);
-		grid->addWidget(edit, row, 1);
-
+		layout->addWidget(edit);
+		
+		grid->addLayout(layout, row, 0);
+		
 		QLabel *req = new QLabel(reqText(), parent);
-		grid->addWidget(req, row, 2);
+		grid->addWidget(req, row, 1);
 
 		if ( !f.desc().isEmpty() ) {
 			label->setToolTip(f.desc());
@@ -357,13 +369,18 @@
 	XDataField_ListSingle(XData::Field f, QGridLayout *grid, QWidget *parent)
 	: XDataField(f, ((XDataWidget *)parent)->client())
 	{
+		QHBoxLayout *layout = new QHBoxLayout;
+		
 		int row = grid->rowCount();
 		QLabel *label = new QLabel(labelText(), parent);
-		grid->addWidget(label, row, 0);
+		label->setWordWrap(true);
+		layout->addWidget(label);
 
 		combo = new QComboBox(parent);
-		grid->addWidget(combo, row, 1);
+		layout->addWidget(combo);
 		combo->setInsertPolicy(QComboBox::NoInsert);
+		
+		grid->addLayout(layout, row, 0);
 
 		QString sel;
 		if ( !f.value().isEmpty() )
@@ -382,7 +399,7 @@
 		}
 
 		QLabel *req = new QLabel(reqText(), parent);
-		grid->addWidget(req, row, 2);
+		grid->addWidget(req, row, 1);
 
 		if ( !f.desc().isEmpty() ) {
 			label->setToolTip(f.desc());
@@ -425,6 +442,7 @@
 	{
 		int row = grid->rowCount();
 		QLabel *label = new QLabel(labelText(), parent);
+		label->setWordWrap(true);
 		grid->addWidget(label, row, 0);
 
 		list = new QListWidget(parent);
@@ -494,11 +512,14 @@
 	: XDataField(f, ((XDataWidget *)parent)->client())
 	{
 		int row = grid->rowCount();
+		QHBoxLayout *layout = new QHBoxLayout;
+		
 		QLabel *label = new QLabel(labelText(), parent);
-		grid->addWidget(label, row, 0);
+		label->setWordWrap(true);
+		layout->addWidget(label);
 
 		edit = new QTextEdit(parent);
-		grid->addWidget(edit, row, 1);
+		layout->addWidget(edit);
 
 		QString text;
 		QStringList val = f.value();
@@ -510,8 +531,10 @@
 		}
 		edit->setText(text);
 
+		grid->addLayout(layout, row, 0);
+
 		QLabel *req = new QLabel(reqText(), parent);
-		grid->addWidget(req, row, 2);
+		grid->addWidget(req, row, 1);
 
 		if ( !f.desc().isEmpty() ) {
 			label->setToolTip(f.desc());
@@ -554,6 +577,8 @@
 {
 	owner_ = owner;
 	layout_ = new QVBoxLayout(this);
+	layout_->setContentsMargins(0,0,0,0);
+
 }
 
 XDataWidget::~XDataWidget()
@@ -684,6 +709,7 @@
 	if ( f.count() ) {
 		// FIXME
 		QGridLayout *grid = new QGridLayout(fields);
+		grid->setSpacing(3);
 
 		XData::FieldList::ConstIterator it = f.begin();
 		for ( ; it != f.end(); ++it) {
