diff --git a/src/chatdlg.cpp b/src/chatdlg.cpp
--- a/src/chatdlg.cpp
+++ b/src/chatdlg.cpp
@@ -112,6 +112,7 @@ ChatDlg::ChatDlg(const Jid& jid, PsiAccount* pa, TabManager* tabManager)
 	sendComposingEvents_ = false;
 	isComposing_ = false;
 	composingTimer_ = 0;
+	updateRealJid();
 }
 
 void ChatDlg::init()
@@ -365,6 +366,7 @@ void ChatDlg::setJid(const Jid &j)
 	if (!j.compare(jid())) {
 		account()->dialogUnregister(this);
 		TabbableWidget::setJid(j);
+		updateRealJid();
 		account()->dialogRegister(this, jid());
 		updateContact(jid(), false);
 	}
@@ -633,6 +635,16 @@ void ChatDlg::invalidateTab()
 	TabbableWidget::invalidateTab();
 }
 
+void ChatDlg::updateRealJid()
+{
+	realJid_ = account()->realJid(jid());
+}
+
+Jid ChatDlg::realJid() const
+{
+	return realJid_;
+}
+
 bool ChatDlg::isEncryptionEnabled() const
 {
 	return false;
diff --git a/src/chatdlg.h b/src/chatdlg.h
--- a/src/chatdlg.h
+++ b/src/chatdlg.h
@@ -75,6 +75,7 @@ public:
 public:
 	PsiAccount* account() const;
 	void setText(const QString& );
+	Jid realJid() const;
 
 signals:
 	void aInfo(const Jid &);
@@ -141,6 +142,7 @@ protected:
 	// reimplemented
 	virtual void invalidateTab();
 
+	void updateRealJid();
 	void resetComposing();
 	void doneSend();
 	virtual void setLooks();
@@ -184,6 +186,7 @@ private:
 	int transid_;
 	Message m_;
 	bool lastWasEncrypted_;
+	Jid realJid_;
 
 	// Message Events & Chat States
 	QTimer* composingTimer_;
diff --git a/src/psiaccount.cpp b/src/psiaccount.cpp
--- a/src/psiaccount.cpp
+++ b/src/psiaccount.cpp
@@ -3592,6 +3592,20 @@ void PsiAccount::itemPublished(const Jid& j, const QString& n, const PubSubItem&
 	}*/
 }
 
+Jid PsiAccount::realJid(const Jid &j) const
+{
+	GCContact *c = findGCContact(j);
+	if (c) {
+		if (c->status.hasMUCItem()) {
+			return c->status.mucItem().jid();
+		} else {
+			return Jid();
+		}
+	} else {
+		return j;
+	}
+}
+
 QList<UserListItem*> PsiAccount::findRelevant(const Jid &j) const
 {
 	QList<UserListItem*> list;
diff --git a/src/psiaccount.h b/src/psiaccount.h
--- a/src/psiaccount.h
+++ b/src/psiaccount.h
@@ -204,6 +204,7 @@ public:
 	void groupChatChangeNick(const QString &host, const QString &room, const QString& nick, const Status &);
 	void groupChatLeave(const QString &host, const QString &room);
 
+	Jid realJid(const Jid &j) const;
 	PsiContact* selfContact() const;
 	const QList<PsiContact*>& contactList() const;
 	int onlineContactsCount() const;
diff --git a/src/psichatdlg.cpp b/src/psichatdlg.cpp
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -42,6 +42,7 @@
 #include "capsmanager.h"
 #include "psicontactlist.h"
 #include "userlist.h"
+#include "psicontact.h"
 #include "jidutil.h"
 #include "xmpp_tasks.h"
 #include "lastactivitytask.h"
@@ -172,6 +173,9 @@ PsiChatDlg::PsiChatDlg(const Jid& jid, PsiAccount* pa, TabManager* tabManager)
 		: ChatDlg(jid, pa, tabManager), mCmdManager_(&mCmdSite_), tabCompletion(&mCmdManager_)
 {
 	connect(account()->psi(), SIGNAL(accountCountChanged()), this, SLOT(updateIdentityVisibility()));
+	connect(account(), SIGNAL(addedContact(PsiContact*)), SLOT(updateContactAdding(PsiContact*)));
+	connect(account(), SIGNAL(removedContact(PsiContact*)), SLOT(updateContactAdding(PsiContact*)));
+	connect(account(), SIGNAL(updateContact(const Jid &)), SLOT(updateContactAdding(const Jid &)));
 	mCmdManager_.registerProvider(new ChatDlgMCmdProvider(this));
 	SendButtonTemplatesMenu* menu = getTemplateMenu();
 	if (menu) {
@@ -312,6 +316,7 @@ void PsiChatDlg::setLooks()
 
 	updateIdentityVisibility();
 	updateCountVisibility();
+	updateContactAdding();
 
 	// toolbuttons
 	QIcon i;
@@ -359,6 +364,26 @@ void PsiChatDlg::updateIdentityVisibility()
 	}
 }
 
+void PsiChatDlg::updateContactAdding(PsiContact* c)
+{
+	if (!c || realJid().compare(c->jid(), false)) {
+		Jid rj = realJid();
+		UserListItem *uli;
+		if (rj.isNull() || ((uli = account()->findFirstRelevant(rj)) && (uli->inList() || uli->isSelf()))) {
+			act_add_contact->setVisible(false);
+		} else {
+			act_add_contact->setVisible(true);
+		}
+	}
+}
+
+void PsiChatDlg::updateContactAdding(const Jid &j)
+{
+	if (realJid().compare(j, false)) {
+		updateContactAdding();
+	}
+}
+
 void PsiChatDlg::initToolButtons()
 {
 // typeahead find
@@ -379,6 +404,9 @@ void PsiChatDlg::initToolButtons()
 	act_html_text = new IconAction(tr("Set Text Format"), "psi/text", tr("Set Text Format"), 0, this);
 	connect(act_html_text, SIGNAL(triggered()), chatEdit(), SLOT(doHTMLTextMenu()));
 
+	act_add_contact = new IconAction(tr("Add Contact To Roster"), "psi/addContact", tr("Add Contact"), 0, this);
+	connect(act_add_contact, SIGNAL(triggered()), SLOT(addContact()));
+
 	connect(account()->psi()->iconSelectPopup(), SIGNAL(textSelected(QString)), this, SLOT(addEmoticon(QString)));
 	act_icon_ = new IconAction(tr("Select Icon"), "psi/smile", tr("Select Icon"), 0, this);
 	act_icon_->setMenu(account()->psi()->iconSelectPopup());
@@ -436,6 +464,7 @@ void PsiChatDlg::initToolBar()
 	if (AvCallManager::isSupported()) {
 		ui_.toolbar->addAction(act_voice_);
 	}
+	ui_.toolbar->addAction(act_add_contact);
 }
 
 void PsiChatDlg::contextMenuEvent(QContextMenuEvent *)
@@ -720,6 +749,14 @@ void PsiChatDlg::doMiniCmd()
 	mCmdManager_.open(new MCmdSimpleState(MCMDCHAT, tr("Command>")), QStringList() );
 }
 
+void PsiChatDlg::addContact()
+{
+	Jid j(realJid());
+	UserListItem *uli = account()->findFirstRelevant(jid());
+	QString name = uli && !uli->name().isEmpty()? uli->name() : j.node();
+	account()->openAddUserDlg(j.withResource(""), name.isEmpty()?j.node():name, "");
+}
+
 void PsiChatDlg::doMinimize()
 {
 	window()->showMinimized();
diff --git a/src/psichatdlg.h b/src/psichatdlg.h
--- a/src/psichatdlg.h
+++ b/src/psichatdlg.h
@@ -12,6 +12,7 @@
 #include "typeaheadfind.h"
 
 class IconAction;
+class PsiContact;
 
 class PsiChatDlg : public ChatDlg
 {
@@ -34,11 +35,14 @@ private slots:
 	void toggleSmallChat();
 	void doClearButton();
 	void doMiniCmd();
+	void addContact();
 	void doMinimize();
 	void buildMenu();
 	void updateCounter();
 	void updateIdentityVisibility();
 	void updateCountVisibility();
+	void updateContactAdding(PsiContact* c = 0);
+	void updateContactAdding(const Jid &j);
 
 	// reimplemented
 	void chatEditCreated();
@@ -86,6 +90,7 @@ private:
 	IconAction* act_ps_;
 	IconAction* act_templates_;
 	IconAction* act_html_text;
+	IconAction* act_add_contact;
 	QAction *act_mini_cmd_, *act_minimize_;
 
 	MCmdManager mCmdManager_;
