Index: psi/src/chatdlg.cpp
===================================================================
--- psi.orig/src/chatdlg.cpp	2013-01-15 14:04:21.004826817 +0400
+++ psi/src/chatdlg.cpp	2013-01-15 14:04:20.996826816 +0400
@@ -114,6 +114,7 @@
 	isComposing_ = false;
 	composingTimer_ = 0;
 	connect(URLObject::getInstance(), SIGNAL(vcardsig(QString)), SLOT(vcard(QString)));
+	updateRealJid();
 }

 void ChatDlg::vcard(const QString& string)
@@ -378,6 +379,7 @@
 	if (!j.compare(jid())) {
 		account()->dialogUnregister(this);
 		TabbableWidget::setJid(j);
+		updateRealJid();
 		account()->dialogRegister(this, jid());
 		updateContact(jid(), false);
 	}
@@ -645,6 +647,16 @@
 	TabbableWidget::invalidateTab();
 }

+void ChatDlg::updateRealJid()
+{
+	realJid_ = account()->realJid(jid());
+}
+
+Jid ChatDlg::realJid() const
+{
+	return realJid_;
+}
+
 bool ChatDlg::isEncryptionEnabled() const
 {
 	return false;
Index: psi/src/chatdlg.h
===================================================================
--- psi.orig/src/chatdlg.h	2013-01-15 14:04:21.004826817 +0400
+++ psi/src/chatdlg.h	2013-01-15 14:04:20.996826816 +0400
@@ -73,6 +73,7 @@
 public:
 	PsiAccount* account() const;
 	void setText(const QString& );
+	Jid realJid() const;

 signals:
 	void aInfo(const Jid &);
@@ -142,6 +143,7 @@
 	// reimplemented
 	virtual void invalidateTab();

+	void updateRealJid();
 	void resetComposing();
 	void doneSend();
 	virtual void setLooks();
@@ -185,6 +187,7 @@
 	int transid_;
 	Message m_;
 	bool lastWasEncrypted_;
+	Jid realJid_;

 	// Message Events & Chat States
 	QTimer* composingTimer_;
Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:21.004826817 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:20.996826816 +0400
@@ -3571,6 +3571,20 @@
 	}*/
 }

+Jid PsiAccount::realJid(const Jid &j) const
+{
+	GCContact *c = findGCContact(j);
+	if (c) {
+		if (c->status.hasMUCItem()) {
+			return c->status.mucItem().jid();
+		} else {
+			return Jid();
+		}
+	} else {
+		return j;
+	}
+}
+
 QList<UserListItem*> PsiAccount::findRelevant(const Jid &j) const
 {
 	QList<UserListItem*> list;
Index: psi/src/psiaccount.h
===================================================================
--- psi.orig/src/psiaccount.h	2013-01-15 14:04:21.004826817 +0400
+++ psi/src/psiaccount.h	2013-01-15 14:04:21.000826816 +0400
@@ -204,6 +204,7 @@
 	void groupChatChangeNick(const QString &host, const QString &room, const QString& nick, const Status &);
 	void groupChatLeave(const QString &host, const QString &room);

+	Jid realJid(const Jid &j) const;
 	PsiContact* selfContact() const;
 	const QList<PsiContact*>& contactList() const;
 	int onlineContactsCount() const;
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:21.004826817 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:21.000826816 +0400
@@ -41,6 +41,7 @@
 #include "capsmanager.h"
 #include "psicontactlist.h"
 #include "userlist.h"
+#include "psicontact.h"
 #include "jidutil.h"
 #include "xmpp_tasks.h"
 #include "lastactivitytask.h"
@@ -171,6 +172,9 @@
 		: ChatDlg(jid, pa, tabManager), mCmdManager_(&mCmdSite_), tabCompletion(&mCmdManager_)
 {
 	connect(account()->psi(), SIGNAL(accountCountChanged()), this, SLOT(updateIdentityVisibility()));
+	connect(account(), SIGNAL(addedContact(PsiContact*)), SLOT(updateContactAdding(PsiContact*)));
+	connect(account(), SIGNAL(removedContact(PsiContact*)), SLOT(updateContactAdding(PsiContact*)));
+	connect(account(), SIGNAL(updateContact(const Jid &)), SLOT(updateContactAdding(const Jid &)));
 	mCmdManager_.registerProvider(new ChatDlgMCmdProvider(this));

 	SendButtonTemplatesMenu* menu = getTemplateMenu();
@@ -310,6 +314,7 @@

 	updateIdentityVisibility();
 	updateCountVisibility();
+	updateContactAdding();

 	// toolbuttons
 	QIcon i;
@@ -353,6 +358,26 @@
 	}
 }

+void PsiChatDlg::updateContactAdding(PsiContact* c)
+{
+	if (!c || realJid().compare(c->jid(), false)) {
+		Jid rj = realJid();
+		UserListItem *uli;
+		if (rj.isNull() || ((uli = account()->findFirstRelevant(rj)) && (uli->inList() || uli->isSelf()))) {
+			act_add_contact->setVisible(false);
+		} else {
+			act_add_contact->setVisible(true);
+		}
+	}
+}
+
+void PsiChatDlg::updateContactAdding(const Jid &j)
+{
+	if (realJid().compare(j, false)) {
+		updateContactAdding();
+	}
+}
+
 void PsiChatDlg::initToolButtons()
 {
 // typeahead find
@@ -373,6 +398,9 @@
 	act_html_text = new IconAction(tr("Set Text Format"), "psi/text", tr("Set Text Format"), 0, this);
 	connect(act_html_text, SIGNAL(triggered()), chatEdit(), SLOT(doHTMLTextMenu()));

+	act_add_contact = new IconAction(tr("Add Contact To Roster"), "psi/addContact", tr("Add Contact"), 0, this);
+	connect(act_add_contact, SIGNAL(triggered()), SLOT(addContact()));
+
 	connect(account()->psi()->iconSelectPopup(), SIGNAL(textSelected(QString)), this, SLOT(addEmoticon(QString)));
 	act_icon_ = new IconAction(tr("Select Icon"), "psi/smile", tr("Select Icon"), 0, this);
 	act_icon_->setMenu(account()->psi()->iconSelectPopup());
@@ -431,6 +459,7 @@
 	if (AvCallManager::isSupported()) {
 		ui_.toolbar->addAction(act_voice_);
 	}
+	ui_.toolbar->addAction(act_add_contact);
 }

 void PsiChatDlg::contextMenuEvent(QContextMenuEvent *)
@@ -688,6 +717,14 @@
 	mCmdManager_.open(new MCmdSimpleState(MCMDCHAT, tr("Command>")), QStringList() );
 }

+void PsiChatDlg::addContact()
+{
+	Jid j(realJid());
+	UserListItem *uli = account()->findFirstRelevant(jid());
+	QString name = uli && !uli->name().isEmpty()? uli->name() : j.node();
+	account()->openAddUserDlg(j.withResource(""), name.isEmpty()?j.node():name, "");
+}
+
 void PsiChatDlg::doMinimize()
 {
 	window()->showMinimized();
Index: psi/src/psichatdlg.h
===================================================================
--- psi.orig/src/psichatdlg.h	2013-01-15 14:04:21.004826817 +0400
+++ psi/src/psichatdlg.h	2013-01-15 14:04:21.000826816 +0400
@@ -11,6 +11,7 @@
 #include "typeaheadfind.h"

 class IconAction;
+class PsiContact;

 class PsiChatDlg : public ChatDlg
 {
@@ -33,11 +34,14 @@
 	void toggleSmallChat();
 	void doClearButton();
 	void doMiniCmd();
+	void addContact();
 	void doMinimize();
 	void buildMenu();
 	void updateCounter();
 	void updateIdentityVisibility();
 	void updateCountVisibility();
+	void updateContactAdding(PsiContact* c = 0);
+	void updateContactAdding(const Jid &j);

 	// reimplemented
 	void chatEditCreated();
@@ -85,6 +89,7 @@
 	IconAction* act_ps_;
 	IconAction* act_templates_;
 	IconAction* act_html_text;
+	IconAction* act_add_contact;
 	QAction *act_mini_cmd_, *act_minimize_;

 	MCmdManager mCmdManager_;
