--- git.temp.orig/options/default.xml
+++ git.temp/options/default.xml
@@ -466,6 +466,8 @@ QLineEdit#le_status_text {
 				<show-tab-close-buttons type="bool">true</show-tab-close-buttons>
 				<put-tabs-at-bottom type="bool">false</put-tabs-at-bottom>
 				<use-tabs type="bool">true</use-tabs>
+ 				<mouse-middle-button type="QString">close</mouse-middle-button>
+				<mouse-doubleclick-action type="QString">detach</mouse-doubleclick-action>
 				<size type="QString"></size> <!-- will be invalid when converted to QSize so we can detect first load -->
 				<grouping type="QString" comment="A ':' seperated list of groups of kinds of tabs to keep in the same tabset. 'C' for chat and 'M' for mucs. 'A' means using all in one window patch.">CM</grouping>
 				<group-state comment="Saved state data of the tabsets defined by options.ui.tabs.grouping"/>
--- git.temp.orig/src/tabs/tabdlg.cpp
+++ git.temp/src/tabs/tabdlg.cpp
@@ -449,12 +449,23 @@ void TabDlg::detachCurrentTab()
 
 void TabDlg::mouseDoubleClickTab(QWidget* widget)
 {
-	if(userManagement_)
+	const QString act = PsiOptions::instance()->getOption("options.ui.tabs.mouse-doubleclick-action").toString();
+	if(act == "hide")
+		hideTab(static_cast<TabbableWidget*>(widget));
+	else if(act == "close")
+		closeTab(static_cast<TabbableWidget*>(widget));
+	else if(act == "detach" && userManagement_)
 		detachTab(static_cast<TabbableWidget*>(widget));
 }
 
 void TabDlg::mouseMiddleClickTab(QWidget* widget) {
-	closeTab(static_cast<TabbableWidget*>(widget));
+	const QString act = PsiOptions::instance()->getOption("options.ui.tabs.mouse-middle-button").toString();
+	if(act == "hide")
+		hideTab(static_cast<TabbableWidget*>(widget));
+	else if(act == "close")
+		closeTab(static_cast<TabbableWidget*>(widget));
+	else if(act == "detach" && userManagement_)
+		detachTab(static_cast<TabbableWidget*>(widget));
 }
 
 void TabDlg::detachTab(TabbableWidget* tab)
