--- psi.orig/src/main.cpp
+++ psi/src/main.cpp
@@ -470,6 +470,12 @@ int main(int argc, char *argv[])
 
 	// it must be initialized first in order for ApplicationInfo::resourcesDir() to work
 	PsiApplication app(argc, argv);
+
+	//Fix encoding after QApplication initialization
+	for(QMap<QString, QString>::iterator i = cmdline.begin(); i != cmdline.end(); ++i) {
+		*i = (*i).toAscii().constData();
+	}
+
 #if defined(Q_WS_WIN)
 	QFileInfo fi(argv[0]);
 	if (fi.baseName().toLower().indexOf("portable") != -1) {
