--- psi.orig/iris/src/libidn/libidn.pri
+++ psi/iris/src/libidn/libidn.pri
@@ -4,7 +4,7 @@ DEPENDPATH  += $$PWD/..
 unix:{
   QMAKE_CFLAGS_WARN_ON -= -W
 }
-win32:{
+win32-msvc*:{
   QMAKE_CFLAGS += -Zm400
 }
 
--- psi.orig/iris/src/xmpp/xmpp-core/xmpp_stanza.h
+++ psi/iris/src/xmpp/xmpp-core/xmpp_stanza.h
@@ -71,7 +71,7 @@ namespace XMPP 
 				UnexpectedRequest
 			};
 
-			Error(int type=Cancel, int condition=UndefinedCondition, const QString &text="", const QDomElement &appSpec=QDomElement());
+			Error(int type=Cancel, int condition=UndefinedCondition, const QString &text=QString(), const QDomElement &appSpec=QDomElement());
 
 			int type;
 			int condition;
--- psi.orig/iris/src/xmpp/xmpp-im/xmpp_httpauthrequest.h
+++ psi/iris/src/xmpp/xmpp-im/xmpp_httpauthrequest.h
@@ -31,7 +31,7 @@ namespace XMPP
 	{
 	public:
 		HttpAuthRequest(const QString &m, const QString &u, const QString &i);
-		HttpAuthRequest(const QString &m = "", const QString &u = "");
+		HttpAuthRequest(const QString &m = QString(), const QString &u = QString());
 		HttpAuthRequest(const QDomElement &);
 
 		bool isEmpty() const;
--- psi.orig/iris/src/xmpp/xmpp-im/xmpp_status.h
+++ psi/iris/src/xmpp/xmpp-im/xmpp_status.h
@@ -35,8 +35,8 @@ namespace XMPP 
 	public:
 		enum Type { Offline, Online, Away, XA, DND, Invisible, FFC };
 
-		Status(const QString &show="", const QString &status="", int priority=0, bool available=true);
-		Status(Type type, const QString& status="", int priority=0);
+		Status(const QString &show=QString(), const QString &status=QString(), int priority=0, bool available=true);
+		Status(Type type, const QString& status=QString(), int priority=0);
 		~Status();
 
 		int priority() const;
