Entering 'iris'
Entering 'src/libpsi'
diff --git a/src/ahcexecutetask.cpp b/src/ahcexecutetask.cpp
--- a/src/ahcexecutetask.cpp
+++ b/src/ahcexecutetask.cpp
@@ -47,14 +47,10 @@ bool AHCExecuteTask::take(const QDomElement& e)
 		QDomElement i = e.firstChildElement("command");
 		if (!i.isNull()) {
 			AHCommand c(i);
-			if (c.status() == AHCommand::Executing) {
-				AHCFormDlg *w = new AHCFormDlg(c,receiver_,client());
-				w->show();
-			}
-			else if (c.status() == AHCommand::Completed && i.childNodes().count() > 0) {
-				AHCFormDlg *w = new AHCFormDlg(c,receiver_,client(), true);
-				w->show();
-			}
+			if (c.status() == AHCommand::Executing)
+				new AHCFormDlg(c,receiver_,client());
+			else if (c.status() == AHCommand::Completed && i.childNodes().count() > 0)
+				new AHCFormDlg(c,receiver_,client(), true);
 			setSuccess();
 			return true;
 		}
diff --git a/src/ahcformdlg.cpp b/src/ahcformdlg.cpp
--- a/src/ahcformdlg.cpp
+++ b/src/ahcformdlg.cpp
@@ -133,6 +133,8 @@ AHCFormDlg::AHCFormDlg(const AHCommand& r, const Jid& receiver, XMPP::Client* cl
 	}
 
 	adjustSize();
+	show();
+	setParent(0);
 }
 
 void AHCFormDlg::doPrev()
