Index: psi/src/ahcexecutetask.cpp
===================================================================
--- psi.orig/src/ahcexecutetask.cpp	2013-01-15 14:04:21.852826851 +0400
+++ psi/src/ahcexecutetask.cpp	2013-01-15 14:04:21.848826851 +0400
@@ -48,14 +48,10 @@
 		QDomElement i = findSubTag(e, "command", &found);
 		if (found) {
 			AHCommand c(i);
-			if (c.status() == AHCommand::Executing) {
-				AHCFormDlg *w = new AHCFormDlg(c,receiver_,client());
-				w->show();
-			}
-			else if (c.status() == AHCommand::Completed && i.childNodes().count() > 0) {
-				AHCFormDlg *w = new AHCFormDlg(c,receiver_,client(), true);
-				w->show();
-			}
+			if (c.status() == AHCommand::Executing)
+				new AHCFormDlg(c,receiver_,client());
+			else if (c.status() == AHCommand::Completed && i.childNodes().count() > 0)
+				new AHCFormDlg(c,receiver_,client(), true);
 			setSuccess();
 			return true;
 		}
Index: psi/src/ahcformdlg.cpp
===================================================================
--- psi.orig/src/ahcformdlg.cpp	2013-01-15 14:04:21.852826851 +0400
+++ psi/src/ahcformdlg.cpp	2013-01-15 14:04:21.848826851 +0400
@@ -133,6 +133,8 @@
 	}
 
 	adjustSize();
+	show();
+	setParent(0);
 }
 
 void AHCFormDlg::doPrev()
