--- psi.orig/src/avcall/avcall.cpp
+++ psi/src/avcall/avcall.cpp
@@ -29,6 +29,7 @@
 #include "../psimedia/psimedia.h"
 #include "applicationinfo.h"
 #include "psiaccount.h"
+#include "psioptions.h"
 
 #define USE_THREAD
 
@@ -907,6 +908,9 @@ bool AvCallManager::isVideoSupported()
 	if(!isSupported())
 		return false;
 
+	if(PsiOptions::instance()->getOption("options.media.video-support").toBool())
+		return true;
+
 	if(!QString::fromLatin1(qgetenv("PSI_ENABLE_VIDEO")).isEmpty())
 		return true;
 	else
--- psi.orig/src/options/opt_avcall.cpp
+++ psi/src/options/opt_avcall.cpp
@@ -201,6 +201,7 @@ void OptionsTabAvCall::applyOptions()
 	PsiOptions::instance()->setOption("options.media.devices.audio-output", d->cb_audioOutDevice->itemData(d->cb_audioOutDevice->currentIndex()).toString());
 	PsiOptions::instance()->setOption("options.media.devices.audio-input", d->cb_audioInDevice->itemData(d->cb_audioInDevice->currentIndex()).toString());
 	PsiOptions::instance()->setOption("options.media.devices.video-input", d->cb_videoInDevice->itemData(d->cb_videoInDevice->currentIndex()).toString());
+	PsiOptions::instance()->setOption("options.media.video-support", d->cb_videoSupport->isChecked());
 }
 
 void OptionsTabAvCall::restoreOptions()
@@ -246,4 +247,5 @@ void OptionsTabAvCall::restoreOptions()
 	d->cb_audioOutDevice->setCurrentIndex(d->cb_audioOutDevice->findData(config.audioOutDeviceId));
 	d->cb_audioInDevice->setCurrentIndex(d->cb_audioInDevice->findData(config.audioInDeviceId));
 	d->cb_videoInDevice->setCurrentIndex(d->cb_videoInDevice->findData(config.videoInDeviceId));
+	d->cb_videoSupport->setChecked(PsiOptions::instance()->getOption("options.media.video-support").toBool());
 }
--- psi.orig/src/options/opt_avcall.ui
+++ psi/src/options/opt_avcall.ui
@@ -1,74 +1,90 @@
-<ui version="4.0" >
+<?xml version="1.0" encoding="UTF-8"?>
+<ui version="4.0">
  <class>OptAvCall</class>
- <widget class="QWidget" name="OptAvCall" >
-  <property name="geometry" >
+ <widget class="QWidget" name="OptAvCall">
+  <property name="geometry">
    <rect>
     <x>0</x>
     <y>0</y>
     <width>400</width>
-    <height>108</height>
+    <height>114</height>
    </rect>
   </property>
-  <property name="windowTitle" >
+  <property name="windowTitle">
    <string>Form</string>
   </property>
-  <layout class="QGridLayout" >
-   <item row="0" column="0" >
-    <widget class="QLabel" name="label" >
-     <property name="text" >
+  <layout class="QGridLayout">
+   <item row="0" column="0">
+    <widget class="QLabel" name="label">
+     <property name="text">
       <string>Audio output:</string>
      </property>
     </widget>
    </item>
-   <item row="0" column="1" >
-    <widget class="QComboBox" name="cb_audioOutDevice" >
-     <property name="sizePolicy" >
-      <sizepolicy vsizetype="Fixed" hsizetype="Expanding" >
+   <item row="0" column="2">
+    <widget class="QComboBox" name="cb_audioOutDevice">
+     <property name="sizePolicy">
+      <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
        <horstretch>0</horstretch>
        <verstretch>0</verstretch>
       </sizepolicy>
      </property>
     </widget>
    </item>
-   <item row="1" column="0" >
-    <widget class="QLabel" name="lb_audioInDevice" >
-     <property name="text" >
+   <item row="1" column="0">
+    <widget class="QLabel" name="lb_audioInDevice">
+     <property name="text">
       <string>Audio input:</string>
      </property>
     </widget>
    </item>
-   <item row="1" column="1" >
-    <widget class="QComboBox" name="cb_audioInDevice" >
-     <property name="sizePolicy" >
-      <sizepolicy vsizetype="Fixed" hsizetype="Expanding" >
+   <item row="1" column="2">
+    <widget class="QComboBox" name="cb_audioInDevice">
+     <property name="sizePolicy">
+      <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
        <horstretch>0</horstretch>
        <verstretch>0</verstretch>
       </sizepolicy>
      </property>
     </widget>
    </item>
-   <item row="2" column="0" >
-    <widget class="QLabel" name="lb_videoInDevice" >
-     <property name="text" >
+   <item row="3" column="0">
+    <widget class="QLabel" name="lb_videoInDevice">
+     <property name="text">
       <string>Video input:</string>
      </property>
     </widget>
    </item>
-   <item row="2" column="1" >
-    <widget class="QComboBox" name="cb_videoInDevice" >
-     <property name="sizePolicy" >
-      <sizepolicy vsizetype="Fixed" hsizetype="Expanding" >
+   <item row="3" column="2">
+    <widget class="QComboBox" name="cb_videoInDevice">
+     <property name="sizePolicy">
+      <sizepolicy hsizetype="Expanding" vsizetype="Fixed">
        <horstretch>0</horstretch>
        <verstretch>0</verstretch>
       </sizepolicy>
      </property>
     </widget>
    </item>
+   <item row="2" column="0">
+    <widget class="QLabel" name="label_2">
+     <property name="text">
+      <string>Video support:</string>
+     </property>
+    </widget>
+   </item>
+   <item row="2" column="2">
+    <widget class="QCheckBox" name="cb_videoSupport">
+     <property name="text">
+      <string/>
+     </property>
+    </widget>
+   </item>
   </layout>
  </widget>
  <tabstops>
   <tabstop>cb_audioOutDevice</tabstop>
   <tabstop>cb_audioInDevice</tabstop>
+  <tabstop>cb_videoSupport</tabstop>
   <tabstop>cb_videoInDevice</tabstop>
  </tabstops>
  <resources/>
