--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -52,6 +52,7 @@
 #include <QCheckBox>
 #include <QDialogButtonBox>
 #include <QFormLayout>
+#include <QClipboard>
 
 #include "psicon.h"
 #include "psiaccount.h"
@@ -210,6 +211,7 @@ public:
 	QString topic;
 	bool nonAnonymous;		 // got status code 100 ?
 	IconAction *act_find, *act_clear, *act_icon, *act_configure, *act_ps, *act_templates, *act_bookmark, *act_html_text;
+	QAction *act_copy_muc_jid;
 	TypeAheadFindBar *typeahead;
 //#ifdef WHITEBOARDING
 //	IconAction *act_whiteboard;
@@ -728,6 +730,11 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, const Jid &j, TabManager *tabManager)
 	d->act_bookmark = new IconAction(this);
 	connect(d->act_bookmark, SIGNAL(triggered()), SLOT(doBookmark()));
 	ui_.le_topic->addAction(d->act_bookmark);
+
+	d->act_copy_muc_jid = new QAction(tr("Copy Groupchat JID"), this);
+	connect(d->act_copy_muc_jid, SIGNAL(triggered()), SLOT(copyMucJid()));
+	ui_.le_topic->addAction(d->act_copy_muc_jid);
+
 	BookmarkManager *bm = account()->bookmarkManager();
 	d->act_bookmark->setVisible(bm->isAvailable());
 	if (bm->isAvailable()) {
@@ -2117,6 +2124,11 @@ void GCMainDlg::doBookmark()
 	delete dlg;
 }
 
+void GCMainDlg::copyMucJid()
+{
+	QApplication::clipboard()->setText(jid().bare());
+}
+
 void GCMainDlg::doRemoveBookmark()
 {
 	BookmarkManager *bm = account()->bookmarkManager();
--- a/src/groupchatdlg.h
+++ b/src/groupchatdlg.h
@@ -142,6 +142,7 @@ private slots:
 	void editTemplates();
 	void doPasteAndSend();
 	void doBookmark();
+	void copyMucJid();
 	void doRemoveBookmark();
 	void sendTemp(const QString &);
 	void psButtonEnabled();
--- a/src/infodlg.cpp
+++ b/src/infodlg.cpp
@@ -161,7 +161,8 @@ InfoDlg::InfoDlg(int type, const Jid &j, const VCard &vcard, PsiAccount *pa, QWi
 	d->namesDlg->setLayout(fl);
 	// end names editing dialog
 
-	QAction *editnames = new QAction(IconsetFactory::icon(d->type == Self?"psi/options":"psi/info").icon(), "", this);
+	QAction *editnames = new QAction(IconsetFactory::icon(d->type == Self ? "psi/options" : "psi/info").icon(),
+									 d->type == Self ? tr("Edit names") : tr("View names") , this);
 	ui_.le_fullname->addAction(editnames);
 	ui_.le_fullname->widgetForAction(editnames)->setPopup(d->namesDlg);
 	d->homepageAction = new QAction(IconsetFactory::icon("psi/arrowRight").icon(), tr("Open web browser"), this);
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -19,6 +19,7 @@
 #include <QMessageBox>
 #include <QDebug>
 #include <QTimer>
+#include <QClipboard>
 
 #include "psicon.h"
 #include "psiaccount.h"
@@ -219,6 +220,10 @@ void PsiChatDlg::initUi()
 		updateAutojidIcon();
 		connect(act_autojid, SIGNAL(triggered()), SLOT(doSwitchJidMode()));
 		le_autojid->addAction(act_autojid);
+
+		QAction *act_copy_user_jid = new QAction(tr("Copy user JID"), this);
+		le_autojid->addAction(act_copy_user_jid);
+		connect(act_copy_user_jid, SIGNAL(triggered()), SLOT(copyUserJid()));
 	}
 
 	ui_.lb_ident->setAccount(account());
@@ -426,6 +431,11 @@ void PsiChatDlg::updateContactAdding(PsiContact* c)
 	}
 }
 
+void PsiChatDlg::copyUserJid()
+{
+	QApplication::clipboard()->setText(jid().bare());
+}
+
 void PsiChatDlg::updateContactAdding(const Jid &j)
 {
 	if (realJid().compare(j, false)) {
--- a/src/psichatdlg.h
+++ b/src/psichatdlg.h
@@ -48,6 +48,7 @@ private slots:
 	void updateCountVisibility();
 	void updateContactAdding(PsiContact* c = 0);
 	void updateContactAdding(const Jid &j);
+	void copyUserJid();
 
 	// reimplemented
 	void chatEditCreated();
