--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -52,6 +52,7 @@
 #include <QCheckBox>
 #include <QDialogButtonBox>
 #include <QFormLayout>
+#include <QClipboard>
 
 #include "psicon.h"
 #include "psiaccount.h"
@@ -218,6 +219,7 @@ public:
 	QAction *act_send, *act_scrollup, *act_scrolldown, *act_close;
 
 	QAction *act_mini_cmd, *act_nick, *act_hide, *act_minimize;
+	QAction *act_copy_muc_jid;
 
 	MCmdSimpleSite mCmdSite;
 	MCmdManager mCmdManager;
@@ -716,6 +718,11 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, const Jid &j, TabManager *tabManager)
 	d->act_bookmark = new IconAction(this);
 	connect(d->act_bookmark, SIGNAL(triggered()), SLOT(doBookmark()));
 	ui_.le_topic->addAction(d->act_bookmark);
+
+	d->act_copy_muc_jid = new QAction(tr("Copy Groupchat JID"), this);
+	connect(d->act_copy_muc_jid, SIGNAL(triggered()), SLOT(copyMucJid()));
+	ui_.le_topic->addAction(d->act_copy_muc_jid);
+
 	BookmarkManager *bm = account()->bookmarkManager();
 	d->act_bookmark->setVisible(bm->isAvailable());
 	if (bm->isAvailable()) {
@@ -1241,6 +1248,11 @@ void GCMainDlg::doBookmark()
 	delete dlg;
 }
 
+void GCMainDlg::copyMucJid()
+{
+	QApplication::clipboard()->setText(jid().bare());
+}
+
 void GCMainDlg::doRemoveBookmark()
 {
 	BookmarkManager *bm = account()->bookmarkManager();
--- a/src/groupchatdlg.h
+++ b/src/groupchatdlg.h
@@ -128,6 +128,7 @@ private slots:
 	void doClear();
 	void doClearButton();
 	void doBookmark();
+	void copyMucJid();
 	void doRemoveBookmark();
 	void buildMenu();
 	void logSelectionChanged();
--- a/src/infodlg.cpp
+++ b/src/infodlg.cpp
@@ -159,7 +159,8 @@ InfoDlg::InfoDlg(int type, const Jid &j, const VCard &vcard, PsiAccount *pa, QWi
 	d->namesDlg->setLayout(fl);
 	// end names editing dialog
 
-	QAction *editnames = new QAction(IconsetFactory::icon(d->type == Self?"psi/options":"psi/info").icon(), "", this);
+	QAction *editnames = new QAction(IconsetFactory::icon(d->type == Self ? "psi/options" : "psi/info").icon(),
+									 d->type == Self ? tr("Edit names") : tr("View names") , this);
 	ui_.le_fullname->addAction(editnames);
 	ui_.le_fullname->widgetForAction(editnames)->setPopup(d->namesDlg);
 	d->homepageAction = new QAction(IconsetFactory::icon("psi/arrowRight").icon(), tr("Open web browser"), this);
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -18,6 +18,7 @@
 #include <QDragEnterEvent>
 #include <QMessageBox>
 #include <QDebug>
+#include <QClipboard>
 
 #include "psicon.h"
 #include "psiaccount.h"
@@ -202,6 +203,10 @@ void PsiChatDlg::initUi()
 		updateAutojidIcon();
 		connect(act_autojid, SIGNAL(triggered()), SLOT(doSwitchJidMode()));
 		le_autojid->addAction(act_autojid);
+
+		QAction *act_copy_user_jid = new QAction(tr("Copy user JID"), this);
+		le_autojid->addAction(act_copy_user_jid);
+		connect(act_copy_user_jid, SIGNAL(triggered()), SLOT(copyUserJid()));
 	}
 
 	ui_.lb_ident->setAccount(account());
@@ -408,6 +413,11 @@ void PsiChatDlg::updateContactAdding(PsiContact* c)
 	}
 }
 
+void PsiChatDlg::copyUserJid()
+{
+	QApplication::clipboard()->setText(jid().bare());
+}
+
 void PsiChatDlg::updateContactAdding(const Jid &j)
 {
 	if (realJid().compare(j, false)) {
--- a/src/psichatdlg.h
+++ b/src/psichatdlg.h
@@ -51,6 +51,7 @@ private slots:
 	void contactChanged();
 	QString makeContactName(const QString &name, const Jid &jid) const;
 	void doSwitchJidMode();
+	void copyUserJid();
 
 	// reimplemented
 	void chatEditCreated();
