Index: psi/src/adduserdlg.cpp
===================================================================
--- psi.orig/src/adduserdlg.cpp	2013-01-15 14:04:23.172826905 +0400
+++ psi/src/adduserdlg.cpp	2013-01-15 14:04:23.168826905 +0400
@@ -329,10 +329,30 @@
 	JT_VCard *jt = (JT_VCard *)sender();

 	if(jt->success()) {
-		if ( !jt->vcard().nickName().isEmpty() )
-			le_nick->setText( jt->vcard().nickName() );
-		else if( !jt->vcard().fullName().isEmpty() )
-			le_nick->setText( jt->vcard().fullName() );
+		QString nickname;
+		const XMPP::VCard vcard = jt->vcard();
+		if ( !vcard.nickName().isEmpty() ) {
+			nickname = vcard.nickName();
+		} else if ( !vcard.fullName().isEmpty() ) {
+			nickname = vcard.fullName();
+		} else {
+			nickname = vcard.givenName();
+			if ( nickname.isEmpty() ) {
+				nickname = vcard.middleName();
+			} else if ( !vcard.middleName().isEmpty() ) {
+				nickname += " " + vcard.middleName();
+			}
+			if ( nickname.isEmpty() ) {
+				nickname = vcard.familyName();
+			} else if ( !vcard.familyName().isEmpty() ) {
+				nickname += " " + vcard.familyName();
+			}
+		}
+
+		if ( nickname.isEmpty() ) {
+			nickname = jt->jid().bare();
+		}
+		le_nick->setText(nickname);
 	}
 }

