--- git.orig/src/xmlconsole.cpp
+++ git/src/xmlconsole.cpp
@@ -63,7 +63,6 @@ XmlConsole::XmlConsole(PsiAccount *_pa)
 	connect(ui_.pb_input, SIGNAL(clicked()), SLOT(insertXml()));
 	connect(ui_.pb_close, SIGNAL(clicked()), SLOT(close()));
 	connect(ui_.pb_dumpRingbuf, SIGNAL(clicked()), SLOT(dumpRingbuf()));
-	connect(ui_.ck_enable, SIGNAL(clicked(bool)), ui_.gb_filter, SLOT(setEnabled(bool)));
 
 	resize(560,400);
 }
--- git.orig/src/xmlconsole.ui
+++ git/src/xmlconsole.ui
@@ -24,9 +24,6 @@
    </item>
    <item>
     <widget class="QGroupBox" name="gb_filter" >
-     <property name="enabled" >
-      <bool>false</bool>
-     </property>
      <property name="title" >
       <string>Filter</string>
      </property>
