Index: psi/src/xmlconsole.cpp
===================================================================
--- psi.orig/src/xmlconsole.cpp	2013-01-15 14:04:23.364826913 +0400
+++ psi/src/xmlconsole.cpp	2013-01-15 14:04:23.360826912 +0400
@@ -63,7 +63,6 @@
 	connect(ui_.pb_input, SIGNAL(clicked()), SLOT(insertXml()));
 	connect(ui_.pb_close, SIGNAL(clicked()), SLOT(close()));
 	connect(ui_.pb_dumpRingbuf, SIGNAL(clicked()), SLOT(dumpRingbuf()));
-	connect(ui_.ck_enable, SIGNAL(clicked(bool)), ui_.gb_filter, SLOT(setEnabled(bool)));
 
 	resize(560,400);
 }
Index: psi/src/xmlconsole.ui
===================================================================
--- psi.orig/src/xmlconsole.ui	2013-01-15 14:04:23.364826913 +0400
+++ psi/src/xmlconsole.ui	2013-01-15 14:04:23.360826912 +0400
@@ -24,9 +24,6 @@
    </item>
    <item>
     <widget class="QGroupBox" name="gb_filter" >
-     <property name="enabled" >
-      <bool>false</bool>
-     </property>
      <property name="title" >
       <string>Filter</string>
      </property>
