Index: psi/src/chatdlg.cpp
===================================================================
--- psi.orig/src/chatdlg.cpp	2013-01-15 14:04:23.792826930 +0400
+++ psi/src/chatdlg.cpp	2013-01-15 14:04:23.788826930 +0400
@@ -104,6 +104,7 @@
 	transid_ = -1;
 	key_ = "";
 	lastWasEncrypted_ = false;
+	trackBar_ = false;

 	status_ = -1;

@@ -173,6 +174,12 @@
 	chatEditCreated();
 }

+void ChatDlg::doTrackBar()
+{
+	trackBar_ = false;
+	chatView()->doTrackBar();
+}
+
 void ChatDlg::initActions()
 {
 	act_send_ = new QAction(this);
@@ -328,6 +335,8 @@
 void ChatDlg::deactivated()
 {
 	TabbableWidget::deactivated();
+
+	trackBar_ = true;
 }

 void ChatDlg::activated()
@@ -342,6 +351,8 @@
 	doFlash(false);

 	chatEdit()->setFocus();
+
+	trackBar_ = false;
 }

 void ChatDlg::dropEvent(QDropEvent* event)
@@ -841,6 +852,9 @@

 void ChatDlg::appendMessage(const Message &m, bool local)
 {
+	if(trackBar_)
+		doTrackBar();
+
 	// figure out the encryption state
 	bool encChanged = false;
 	bool encEnabled = false;
Index: psi/src/chatdlg.h
===================================================================
--- psi.orig/src/chatdlg.h	2013-01-15 14:04:23.792826930 +0400
+++ psi/src/chatdlg.h	2013-01-15 14:04:23.788826930 +0400
@@ -181,6 +181,9 @@
 	bool keepOpen_;
 	bool warnSend_;

+	bool trackBar_;
+	void doTrackBar();
+
 	QTimer* selfDestruct_;

 	QString key_;
