--- git.orig/src/chatdlg.cpp
+++ git/src/chatdlg.cpp
@@ -103,6 +103,7 @@ ChatDlg::ChatDlg(const Jid& jid, PsiAcco
 	transid_ = -1;
 	key_ = "";
 	lastWasEncrypted_ = false;
+	trackBar_ = false;
 
 	status_ = -1;
 
@@ -164,6 +165,12 @@ void ChatDlg::initComposing()
 	chatEditCreated();
 }
 
+void ChatDlg::doTrackBar()
+{
+	trackBar_ = false;
+	chatView()->doTrackBar();
+}
+
 void ChatDlg::initActions()
 {
 	act_send_ = new QAction(this);
@@ -325,6 +332,8 @@ void ChatDlg::logSelectionChanged()
 void ChatDlg::deactivated()
 {
 	TabbableWidget::deactivated();
+
+	trackBar_ = true;
 }
 
 void ChatDlg::activated()
@@ -339,6 +348,8 @@ void ChatDlg::activated()
 	doFlash(false);
 
 	chatEdit()->setFocus();
+
+	trackBar_ = false;
 }
 
 void ChatDlg::dropEvent(QDropEvent* event)
@@ -839,6 +850,9 @@ QString ChatDlg::whoNick(bool local) con
 
 void ChatDlg::appendMessage(const Message &m, bool local)
 {
+	if(trackBar_)
+		doTrackBar();
+
 	// figure out the encryption state
 	bool encChanged = false;
 	bool encEnabled = false;
--- git.orig/src/chatdlg.h
+++ git/src/chatdlg.h
@@ -180,6 +180,9 @@ private:
 	bool keepOpen_;
 	bool warnSend_;
 
+	bool trackBar_;
+	void doTrackBar();
+
 	QTimer* selfDestruct_;
 
 	QString key_;
