Entering 'iris'
Entering 'src/libpsi'
Index: psi/options/default.xml
===================================================================
--- psi.orig/options/default.xml	2013-01-15 14:04:24.580826962 +0400
+++ psi/options/default.xml	2013-01-15 14:04:24.576826962 +0400
@@ -291,6 +291,7 @@
 				</view>
 			</menu>
 			<muc comment="Multi-User Chat options">
+				<hide-on-autojoin type="bool">false</hide-on-autojoin>
 				<status-with-priority comment="Show priority with status change" type="bool">false</status-with-priority>
 				<show-initial-joins comment="Show joins on your MUC join (with status and priority, depends on options)" type="bool">false</show-initial-joins>
 				<log-width type="int">500</log-width>
Index: psi/src/mucjoindlg.cpp
===================================================================
--- psi.orig/src/mucjoindlg.cpp	2013-01-15 14:04:24.580826962 +0400
+++ psi/src/mucjoindlg.cpp	2013-01-15 14:04:24.576826962 +0400
@@ -48,6 +48,8 @@
 	joinButton_ = ui_.buttonBox->addButton(tr("&Join"), QDialogButtonBox::AcceptRole);
 	joinButton_->setDefault(true);
 
+	reason_ = MucCustomJoin;
+
 	updateIdentity(pa);
 
 	ui_.cb_ident->setController(controller_);
@@ -176,11 +178,13 @@
 	}
 }
 
-void MUCJoinDlg::doJoin()
+void MUCJoinDlg::doJoin(MucJoinReason r)
 {
 	if (!account_ || !account_->checkConnected(this))
 		return;
 
+	reason_ = r;
+
 	QString host = ui_.le_host->text();
 	QString room = ui_.le_room->text();
 	QString nick = ui_.le_nick->text();
@@ -244,6 +248,9 @@
 
 void MUCJoinDlg::error(int, const QString &str)
 {
+	if(!isVisible())
+		show();
+
 	ui_.busy->stop();
 	setWidgetsEnabled(true);
 
Index: psi/src/mucjoindlg.h
===================================================================
--- psi.orig/src/mucjoindlg.h	2013-01-15 14:04:24.580826962 +0400
+++ psi/src/mucjoindlg.h	2013-01-15 14:04:24.576826962 +0400
@@ -36,6 +36,12 @@
 	Q_OBJECT
 
 public:
+
+	enum MucJoinReason {
+		MucAutoJoin,
+		MucCustomJoin
+	};
+
 	MUCJoinDlg(PsiCon *, PsiAccount *);
 	~MUCJoinDlg();
 
@@ -48,11 +54,15 @@
 
 public slots:
 	void done(int);
-	void doJoin();
+	void doJoin(MucJoinReason reason = MucCustomJoin);
 
 	// reimplemented
 	void accept();
 
+public:
+	MucJoinReason getReason() const { return reason_; };
+
+
 private slots:
 	void updateIdentity(PsiAccount *);
 	void updateIdentityVisibility();
@@ -67,6 +77,7 @@
 	PsiAccount* account_;
 	QPushButton* joinButton_;
 	XMPP::Jid jid_;
+	MucJoinReason reason_;
 
 	void disableWidgets();
 	void enableWidgets();
Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:24.580826962 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:24.576826962 +0400
@@ -130,7 +130,6 @@
 #include "systeminfo.h"
 #include "avatars.h"
 #include "ahcommanddlg.h"
-#include "mucjoindlg.h"
 #include "ahcservermanager.h"
 #include "rc.h"
 #include "tabdlg.h"
@@ -2321,7 +2320,7 @@
 	foreach(ConferenceBookmark c, d->bookmarkManager->conferences()) {
 		if (!findDialog<GCMainDlg*>(Jid(c.jid().bare())) && c.autoJoin()) {
 			if  (!joinSkip.contains(c.jid().full().toUpper()))
-				actionJoin(c, true);
+				actionJoin(c, true, MUCJoinDlg::MucAutoJoin);
 		}
 	}
 #endif
@@ -3358,7 +3357,7 @@
 			   false);
 }
 
-void PsiAccount::actionJoin(const ConferenceBookmark& bookmark, bool connectImmediately)
+void PsiAccount::actionJoin(const ConferenceBookmark& bookmark, bool connectImmediately, MUCJoinDlg::MucJoinReason reason)
 {
 #ifdef GROUPCHAT
 	MUCJoinDlg* w = new MUCJoinDlg(psi(), this);
@@ -3367,9 +3366,11 @@
 	w->setNick(bookmark.nick().isEmpty() ? JIDUtil::nickOrJid(this->nick(), d->jid.node()) : bookmark.nick());
 	w->setPassword(bookmark.password());
 
-	w->show();
+	if(reason != MUCJoinDlg::MucAutoJoin
+	   || !PsiOptions::instance()->getOption("options.ui.muc.hide-on-autojoin").toBool())
+		w->show();
 	if (connectImmediately) {
-		w->doJoin();
+		w->doJoin(reason);
 	}
 #else
 	Q_UNUSED(bookmark);
@@ -5501,16 +5502,18 @@
 	delete h;
 }
 
-void PsiAccount::openGroupChat(const Jid &j, ActivationType activationType)
+void PsiAccount::openGroupChat(const Jid &j, ActivationType activationType, MUCJoinDlg::MucJoinReason reason)
 {
 #ifdef GROUPCHAT
 	GCMainDlg *w = new GCMainDlg(this, j, d->tabManager);
 	w->setPassword(d->client->groupChatPassword(j.domain(), j.node()));
 	connect(w, SIGNAL(aSend(const Message &)), SLOT(dj_sendMessage(const Message &)));
 	connect(d->psi, SIGNAL(emitOptionsUpdate()), w, SLOT(optionsUpdate()));
-	w->ensureTabbedCorrectly();
-	if (activationType == UserAction)
-		w->bringToFront();
+	if(reason != MUCJoinDlg::MucAutoJoin || !PsiOptions::instance()->getOption("options.ui.muc.hide-on-autojoin").toBool()) {
+		w->ensureTabbedCorrectly();
+		if (activationType == UserAction)
+			w->bringToFront();
+	}
 #endif
 }
 
@@ -5585,10 +5588,17 @@
 	MUCJoinDlg *w = findDialog<MUCJoinDlg*>(j);
 	if(!w)
 		return;
+	MUCJoinDlg::MucJoinReason r = w->getReason();
 	w->joined();
 
-	// TODO: Correctly handle auto-join groupchats
-	openGroupChat(j, UserAction);
+	openGroupChat(j, UserAction, r);
+//	if(r == MUCJoinDlg::MucAutoJoin && PsiOptions::instance()->getOption("options.ui.muc.hide-on-autojoin").toBool()) {
+//		m = findDialog<GCMainDlg*>(Jid(j.bare()));
+//		if(m) {
+//			QTimer::singleShot(0, m, SLOT(hideTab()));
+//		}
+//	}
+
 #endif
 }
 
Index: psi/src/psiaccount.h
===================================================================
--- psi.orig/src/psiaccount.h	2013-01-15 14:04:24.580826962 +0400
+++ psi/src/psiaccount.h	2013-01-15 14:04:24.576826962 +0400
@@ -35,6 +35,7 @@
 #include "mood.h"
 #include "activity.h"
 #include "geolocation.h"
+#include "mucjoindlg.h"
 
 namespace XMPP
 {
@@ -198,7 +199,7 @@
 	void showXmlConsole();
 	void openAddUserDlg();
 	void openAddUserDlg(const XMPP::Jid &jid, const QString &nick, const QString &group);
-	void openGroupChat(const Jid &, ActivationType activationType);
+	void openGroupChat(const Jid &, ActivationType activationType, MUCJoinDlg::MucJoinReason reason = MUCJoinDlg::MucCustomJoin);
 	bool groupChatJoin(const QString &host, const QString &room, const QString &nick, const QString& pass, bool nohistory = false);
 	void groupChatSetStatus(const QString &host, const QString &room, const Status &);
 	void groupChatChangeNick(const QString &host, const QString &room, const QString& nick, const Status &);
@@ -361,7 +362,7 @@
 	void actionSearch(const Jid &);
 	void actionManageBookmarks();
 	void actionJoin(const Jid& mucJid, const QString& password = QString());
-	void actionJoin(const ConferenceBookmark& bookmark, bool connectImmediately);
+	void actionJoin(const ConferenceBookmark& bookmark, bool connectImmediately, MUCJoinDlg::MucJoinReason reason = MUCJoinDlg::MucCustomJoin);
 	void actionDisco(const Jid &, const QString &);
 	void actionInvite(const Jid &, const QString &);
 	void actionVoice(const Jid&);
Index: psi/src/tabs/tabbablewidget.cpp
===================================================================
--- psi.orig/src/tabs/tabbablewidget.cpp	2013-01-15 14:04:24.580826962 +0400
+++ psi/src/tabs/tabbablewidget.cpp	2013-01-15 14:04:24.576826962 +0400
@@ -47,7 +47,7 @@
 		TabbableWidget::templateMenu->setStyleSheet(PsiOptions::instance()->getOption("options.ui.chat.css").toString());
 	}
 	++TabbableWidget::chatsCount;
-	QTimer::singleShot(0, this, SLOT(ensureTabbedCorrectly()));
+	//QTimer::singleShot(0, this, SLOT(ensureTabbedCorrectly()));
 }
 
 void TabbableWidget::ensureTabbedCorrectly()
