--- git.temp.orig/src/privacy/privacydlg.cpp
+++ git.temp/src/privacy/privacydlg.cpp
@@ -59,6 +59,7 @@ PrivacyDlg::PrivacyDlg(const QString& ac
 
 	connect(ui_.pb_newList,SIGNAL(clicked()),SLOT(newList()));
 	connect(ui_.pb_deleteList,SIGNAL(clicked()),SLOT(removeList()));
+	connect(ui_.pb_renameList,SIGNAL(clicked()),SLOT(renameList()));
 
 	connect(ui_.pb_add,SIGNAL(clicked()),SLOT(addRule()));
 	connect(ui_.pb_edit,SIGNAL(clicked()),SLOT(editCurrentRule()));
@@ -357,3 +358,23 @@ void PrivacyDlg::removeList()
 	manager_->changeList(model_.list());
 	manager_->requestListNames();
 }
+
+void PrivacyDlg::renameList()
+{
+	QString newName = QInputDialog::getText(this, tr("Rename List"), tr("Input new name"));
+	if(newName.isEmpty()) {
+		return;
+	}
+	PrivacyList tmp = model_.list();
+	model_.list().setName(newName);
+	tmp.clear();
+	manager_->changeList(model_.list());
+	if(ui_.cb_default->currentText() == tmp.name()) {
+		manager_->changeDefaultList(newName);
+	}
+	if(ui_.cb_active->currentText() == tmp.name()) {
+		manager_->changeActiveList(newName);
+	}
+	manager_->changeList(tmp);
+	manager_->requestListNames();
+}
--- git.temp.orig/src/privacy/privacydlg.h
+++ git.temp/src/privacy/privacydlg.h
@@ -70,6 +70,7 @@ protected slots:
 	void applyList();
 	void newList();
 	void removeList();
+	void renameList();
 
 private:
 	Ui::Privacy ui_;
--- git.temp.orig/src/privacy/privacy.ui
+++ git.temp/src/privacy/privacy.ui
@@ -117,6 +117,13 @@
          </widget>
         </item>
         <item>
+         <widget class="QPushButton" name="pb_renameList">
+          <property name="text">
+           <string>Rename List</string>
+          </property>
+         </widget>
+        </item>
+        <item>
          <widget class="QPushButton" name="pb_deleteList">
           <property name="text">
            <string>Delete List</string>
