--- git.temp.orig/src/gcuserview.cpp
+++ git.temp/src/gcuserview.cpp
@@ -613,6 +613,11 @@ void GCUserView::doContextMenu(QTreeWidg
 		pm->addAction(act);
 		act->setData(5);
 	}
+
+	act = new QAction(tr("E&xecute Command"), pm);
+	pm->addAction(act);
+	act->setData(50);
+
 	pm->addSeparator();
 
 	// Kick and Ban submenus
--- git.temp.orig/src/groupchatdlg.cpp
+++ git.temp/src/groupchatdlg.cpp
@@ -2034,6 +2034,9 @@ void GCMainDlg::lv_action(const QString
 		}
 
 	}
+	else if(x == 50) {
+		account()->actionExecuteCommandSpecific(jid().withResource(nick));
+	}
 }
 
 void GCMainDlg::contextMenuEvent(QContextMenuEvent *)
