--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -115,6 +115,7 @@
 				<use-small-chats type="bool">false</use-small-chats>
 				<use-message-icons type="bool">true</use-message-icons>
 				<scaled-message-icons type="bool">false</scaled-message-icons>
+				<show-status-changes type="bool">true</show-status-changes>
 				<warn-before-clear type="bool">true</warn-before-clear>
 				<css type="QString" />
 				<only-paste-template type="bool">false</only-paste-template>
--- psi.orig/src/chatdlg.cpp
+++ psi/src/chatdlg.cpp
@@ -519,7 +519,8 @@ void ChatDlg::updateContact(const Jid &j
 			key_ = userStatus.publicKeyID;
 			updatePGP();
 
-			if (fromPresence && statusChanged) {
+			if (PsiOptions::instance()->getOption("options.ui.chat.show-status-changes").toBool()
+				&& fromPresence && statusChanged) {
 				QString msg = tr("%1 is %2").arg(Qt::escape(dispNick_)).arg(status2txt(status_));
 				if (!statusString_.isEmpty()) {
 					QString ss = TextUtil::linkify(TextUtil::plain2rich(statusString_));
