--- a/options/default.xml
+++ b/options/default.xml
@@ -534,6 +534,7 @@ QLineEdit#le_status_text {
 				<show_self type="QKeySequence" comment="Show Self"></show_self>
 				<show_statusmsg type="QKeySequence" comment="Show Status Message"></show_statusmsg>
 				<publish_tune type="QKeySequence" comment="Publish Tune"></publish_tune>
+				<active_contacts type="QKeySequence" comment="Active contacts"/>
 				<status_all type="QKeySequence" comment="Status All"></status_all>
 				<status_group type="QKeySequence" comment="Status Group"></status_group>
 				<status_chat type="QKeySequence" comment="Status Chat"></status_chat>
new file mode 100644
--- /dev/null
+++ b/src/activecontactsmenu.cpp
@@ -0,0 +1,70 @@
+/*
+ * activecontactsmenu.cpp
+ * Copyright (C) 2011  Khryukin Evgeny
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+#include "activecontactsmenu.h"
+#include "psiaccount.h"
+#include "psicontact.h"
+#include "psiiconset.h"
+#include "psicon.h"
+
+
+class ActiveContactAction : public QAction
+{
+	Q_OBJECT
+public:
+	ActiveContactAction(const QString& jid, const QString& nick, const QIcon& icon, PsiAccount* pa, QMenu* parent)
+		: QAction(icon, nick, parent)
+		, pa_(pa)
+		, jid_(jid)
+	{
+		parent->addAction(this);
+		connect(this, SIGNAL(triggered()), SLOT(actionActivated()));
+	}
+
+private slots:
+	void actionActivated()
+	{
+		pa_->actionDefault(Jid(jid_));
+	}
+
+private:
+	PsiAccount* pa_;
+	QString jid_;
+};
+
+
+
+ActiveContactsMenu::ActiveContactsMenu(PsiCon *psi, QWidget *parent)
+	: QMenu(parent)
+	, psi_(psi)
+{
+	foreach(PsiAccount *pa, psi_->contactList()->accounts()) {
+		if(!pa->enabled())
+			continue;
+
+		QList<PsiContact*> list = pa->activeContacts();
+		foreach(PsiContact* pc, list) {
+			new ActiveContactAction(pc->jid().full(), pc->name(), PsiIconset::instance()->statusPtr(pc->jid(), pc->status())->icon(), pa, this);
+		}
+	}
+}
+
+
+#include "activecontactsmenu.moc"
new file mode 100644
index 0000000..d9edf4d
--- /dev/null
+++ b/src/activecontactsmenu.h
@@ -0,0 +1,38 @@
+/*
+ * activecontactsmenu.h
+ * Copyright (C) 2011  Khryukin Evgeny
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+#ifndef ACTIVECONTACTSMENU_H
+#define ACTIVECONTACTSMENU_H
+
+#include <QMenu>
+
+class PsiCon;
+
+class ActiveContactsMenu : public QMenu
+{
+	Q_OBJECT
+public:
+	ActiveContactsMenu(PsiCon* psi, QWidget* parent);
+
+private:
+	PsiCon* psi_;
+};
+
+#endif
--- a/src/mainwin.cpp
+++ b/src/mainwin.cpp
@@ -74,6 +74,7 @@
 #include "mooddlg.h"
 #include "activitydlg.h"
 #include "geolocationdlg.h"
+#include "activecontactsmenu.h"
 
 #include "mainwin_p.h"
 
@@ -660,6 +661,8 @@ void MainWin::registerAction( IconAction* action )
 	} actionlist[] = {
 		{ "choose_status", activated, this, SLOT( actChooseStatusActivated() ) },
 		{ "reconnect_all", activated, this, SLOT( actReconnectActivated() ) },
+
+		{ "active_contacts",activated, this, SLOT( actActiveContacts() ) },
 #ifndef NEWCONTACTLIST
 		{ "show_offline", toggled, cvlist, SLOT( setShowOffline(bool) ) },
 		{ "show_away",    toggled, cvlist, SLOT( setShowAway(bool) ) },
@@ -1247,6 +1250,14 @@ void MainWin::actSetGeolocActivated()
 	}
 }
 
+void MainWin::actActiveContacts()
+{
+	ActiveContactsMenu* acm = new ActiveContactsMenu(d->psi, this);
+	if(!acm->actions().isEmpty())
+		acm->exec(QCursor::pos());
+	delete acm;
+}
+
 void MainWin::activatedAccOption(PsiAccount* pa, int x)
 {
 	if(x == 0) {
--- a/src/mainwin.h
+++ b/src/mainwin.h
@@ -148,6 +148,7 @@ private slots:
 	void actSetMoodActivated();
 	void actSetActivityActivated();
 	void actSetGeolocActivated();
+	void actActiveContacts();
 
 	bool showDockMenu(const QPoint &);
 	void dockActivated();
--- a/src/psiaccount.cpp
+++ b/src/psiaccount.cpp
@@ -3326,6 +3326,17 @@ QList<ChatDlg*> PsiAccount::findChatDialogs(const Jid& jid, bool compareResource
 	return findDialogs<ChatDlg*>(jid, compareResource);
 }
 
+QList<PsiContact*> PsiAccount::activeContacts() const
+{
+	QList<PsiContact*> ret;
+	foreach(PsiContact *pc, contactList()) {
+		if(pc->isActiveContact()) {
+			ret.append(pc);
+		}
+	}
+	return ret;
+}
+
 QWidget* PsiAccount::findDialog(const QMetaObject& mo, const Jid& jid, bool compareResource) const
 {
 	return d->findDialog(mo, jid, compareResource);
--- a/src/psiaccount.h
+++ b/src/psiaccount.h
@@ -166,6 +166,8 @@ public:
 	ChatDlg* findChatDialogEx(const Jid& jid) const;
 	QList<ChatDlg*> findChatDialogs(const Jid& jid, bool compareResource = true) const;
 
+	QList<PsiContact*> activeContacts() const;
+
 	template<typename T>
 	inline T findDialog(const Jid& jid = Jid(), bool compareResource = true) const {
 		return static_cast<T>(findDialog(((T)0)->staticMetaObject, jid, compareResource));
--- a/src/psiactionlist.cpp
+++ b/src/psiactionlist.cpp
@@ -190,9 +190,13 @@ void PsiActionList::Private::createMainWin()
 		IconAction *eventNotifier = new EventNotifierAction(this, "EventNotifierAction");
 		eventNotifier->setWhatsThis (tr("Special item that displays number of pending events"));
 
+		IconAction *actActiveContacts = new IconAction (tr("Active contacts"), "psi/jabber", tr("Active contacts"), 0, this);
+		actActiveContacts->setWhatsThis (tr("Simple way to find contacts with opened chats"));
+
 		ActionNames actions[] = {
 			{ "button_options", optionsButton },
 			{ "button_status",  statusButton  },
+			{ "active_contacts",actActiveContacts},
 			{ "event_notifier", eventNotifier },
 			{ "", 0 }
 		};
--- a/src/psichatdlg.cpp
+++ b/src/psichatdlg.cpp
@@ -32,6 +32,7 @@
 #include "messageview.h"
 #include "iconselect.h"
 #include "avatars.h"
+#include "activecontactsmenu.h"
 #include "psitooltip.h"
 #include "psioptions.h"
 #include "coloropt.h"
@@ -472,6 +473,9 @@ void PsiChatDlg::initToolButtons()
 
 	act_compact_ = new IconAction(tr("Toggle Compact/Full Size"), "psi/compact", tr("Toggle Compact/Full Size"), 0, this);
 	connect(act_compact_, SIGNAL(triggered()), SLOT(toggleSmallChat()));
+
+	act_active_contacts = new IconAction (tr("Active contacts"), "psi/jabber", tr("Active contacts"), 0, this);
+	connect(act_active_contacts, SIGNAL(triggered()), SLOT(actActiveContacts()));
 }
 
 void PsiChatDlg::initToolBar()
@@ -496,6 +500,7 @@ void PsiChatDlg::initToolBar()
 		ui_.toolbar->addAction(act_voice_);
 	}
 	ui_.toolbar->addAction(act_add_contact);
+	ui_.toolbar->addAction(act_active_contacts);
 }
 
 void PsiChatDlg::contextMenuEvent(QContextMenuEvent *)
@@ -653,6 +658,14 @@ void PsiChatDlg::updateJidWidget(const QList<UserListItem*> &ul, int status, boo
 	internal_change = false;
 }
 
+void PsiChatDlg::actActiveContacts()
+{
+	ActiveContactsMenu* acm = new ActiveContactsMenu(account()->psi(), this);
+	if(!acm->actions().isEmpty())
+		acm->exec(QCursor::pos());
+	delete acm;
+}
+
 void PsiChatDlg::contactUpdated(UserListItem* u, int status, const QString& statusString)
 {
 	Q_UNUSED(statusString);
--- a/src/psichatdlg.h
+++ b/src/psichatdlg.h
@@ -52,6 +52,7 @@ private slots:
 	QString makeContactName(const QString &name, const Jid &jid) const;
 	void doSwitchJidMode();
 	void copyUserJid();
+	void actActiveContacts();
 
 	// reimplemented
 	void chatEditCreated();
@@ -103,6 +104,7 @@ private:
 
 	ActionLineEdit *le_autojid;
 	IconAction *act_autojid;
+	IconAction *act_active_contacts;
 
 	MCmdManager mCmdManager_;
 	MCmdSimpleSite mCmdSite_;
--- a/src/psicontact.cpp
+++ b/src/psicontact.cpp
@@ -540,6 +540,17 @@ bool PsiContact::isAnimated() const
 }
 
 /**
+ * Returns true if there is opened chat with this contact
+ */
+bool PsiContact::isActiveContact() const
+{
+	if(isConference()) {
+		return false;
+	}
+	return account()->findChatDialog(jid(), isPrivate() ? true : false);
+}
+
+/**
  * Contact should always be visible if it's alerting.
  */
 bool PsiContact::shouldBeVisible() const
--- a/src/psicontact.h
+++ b/src/psicontact.h
@@ -66,6 +66,8 @@ public:
 
 	virtual bool isAnimated() const;
 
+	bool isActiveContact() const;
+
 	void activate();
 
 	virtual void setEditing(bool editing);
--- a/src/src.pri
+++ b/src/src.pri
@@ -133,6 +133,7 @@ HEADERS += \
 	$$PWD/statusdlg.h \
 	$$PWD/statuscombobox.h \
 	$$PWD/eventdlg.h \
+	$$PWD/activecontactsmenu.h \
 	$$PWD/chatdlg.h \
 	$$PWD/psichatdlg.h \
 	$$PWD/chatsplitter.h \
@@ -281,6 +282,7 @@ SOURCES += \
 	$$PWD/statusdlg.cpp \
 	$$PWD/statuscombobox.cpp \
 	$$PWD/eventdlg.cpp \
+	$$PWD/activecontactsmenu.cpp \
 	$$PWD/chatdlg.cpp \
 	$$PWD/psichatdlg.cpp \
 	$$PWD/chatsplitter.cpp \
