Entering 'iris'
Entering 'src/libpsi'
Index: psi/src/psioptions.cpp
===================================================================
--- psi.orig/src/psioptions.cpp	2013-01-15 14:04:26.116827024 +0400
+++ psi/src/psioptions.cpp	2013-01-15 14:04:26.112827024 +0400
@@ -334,6 +334,42 @@
 	}
 }

+/**
+ * Reset node \a name to default value
+ * \return Nothing
+ */
+void PsiOptions::resetOption(const QString &name)
+{
+	defaults(); // to init defaults_
+	QStringList nodes = getChildOptionNames(name);
+
+	if (!isInternalNode(name)) {
+		const QVariant &dev = defaults_->getOption(name);
+		if (!dev.isValid()) { // don't touch non exist nodes
+			return;
+		}
+		const QVariant &prev = getOption(name);
+		if (prev == dev) { // do nothing if values is identical
+			return;
+		}
+		setOption(name, dev);
+	}
+	else { // internal node
+		foreach(QString node, nodes) {
+			const QVariant &dev = defaults_->getOption(node);
+			if (!dev.isValid()) {
+				continue;
+			}
+			const QVariant &prev = getOption(node);
+			if (prev == dev) {
+				continue;
+			}
+			setOption(node, dev);
+		}
+	}
+}
+
+
 PsiOptions* PsiOptions::instance_ = NULL;
 PsiOptions* PsiOptions::defaults_ = NULL;

Index: psi/src/psioptions.h
===================================================================
--- psi.orig/src/psioptions.h	2013-01-15 14:04:26.116827024 +0400
+++ psi/src/psioptions.h	2013-01-15 14:04:26.112827024 +0400
@@ -47,6 +47,7 @@
 	bool newProfile();
 	bool save(QString file);
 	void autoSave(bool autoSave, QString autoFile = "");
+	void resetOption(const QString &name);

 // don't call this normally
 	PsiOptions();
Index: psi/src/psioptionseditor.cpp
===================================================================
--- psi.orig/src/psioptionseditor.cpp	2013-01-15 14:04:26.116827024 +0400
+++ psi/src/psioptionseditor.cpp	2013-01-15 14:04:26.112827024 +0400
@@ -160,11 +160,13 @@

 	buttonLine->addStretch(1);

-	if (1) { // FIXME
-		pb_delete = new QPushButton(tr("Delete"), this);
-		buttonLine->addWidget(pb_delete);
-		connect(pb_delete, SIGNAL(clicked()), SLOT(deleteit()));
-	}
+	pb_delete = new QPushButton(tr("Delete..."), this);
+	buttonLine->addWidget(pb_delete);
+	connect(pb_delete, SIGNAL(clicked()), SLOT(deleteit()));
+
+	pb_reset = new QPushButton(tr("Reset..."), this);
+	buttonLine->addWidget(pb_reset);
+	connect(pb_reset, SIGNAL(clicked()), SLOT(resetit()));

 	pb_edit = new QPushButton(tr("Edit..."), this);
 	buttonLine->addWidget(pb_edit);
@@ -274,12 +276,28 @@
 		sub = true;
 		confirm = tr("Really delete all options starting with %1.?");
 	}
-	if (QMessageBox::Yes == QMessageBox::warning(this, tr("Psi: Option Editor"),
+	if (QMessageBox::Yes == QMessageBox::warning(this, tr("Psi+: Option Editor"),
 				   confirm.arg(option), QMessageBox::Yes | QMessageBox::Cancel, QMessageBox::Cancel)) {
 		PsiOptions::instance()->removeOption( option, sub);
 	}
 }

+void PsiOptionsEditor::resetit()
+{
+	QModelIndex idx = tv_->currentIndex();
+	QString option = tm_->indexToOptionName(idx);
+	bool sub = false;
+	QString confirm = tr("Really reset options %1 to default value?");
+	if (o_->isInternalNode(option)) {
+		sub = true;
+		confirm = tr("Really reset all options starting with %1. to default value?");
+	}
+	if (QMessageBox::Yes == QMessageBox::warning(this, tr("Psi+: Option Editor"),
+                   confirm.arg(option), QMessageBox::Yes | QMessageBox::Cancel, QMessageBox::Cancel)) {
+		PsiOptions::instance()->resetOption(option);
+	}
+}
+
 void PsiOptionsEditor::detach()
 {
 	new PsiOptionsEditor();
Index: psi/src/psioptionseditor.h
===================================================================
--- psi.orig/src/psioptionseditor.h	2013-01-15 14:04:26.116827024 +0400
+++ psi/src/psioptionseditor.h	2013-01-15 14:04:26.116827024 +0400
@@ -27,6 +27,7 @@
 	void add();
 	void edit();
 	void deleteit();
+	void resetit();
 	void detach();

 private:
@@ -39,6 +40,7 @@
 	QLabel* lb_path;
 	QLabel* lb_comment;
 	QPushButton *pb_delete;
+	QPushButton *pb_reset;
 	QPushButton *pb_edit;
 	QPushButton *pb_new;
 	QToolButton *pb_detach;
