--- git.orig/src/psiaccount.cpp
+++ git/src/psiaccount.cpp
@@ -1181,7 +1181,7 @@ PsiAccount::PsiAccount(const UserAccount
 	connect(capsManager(),SIGNAL(capsChanged(const Jid&)),SLOT(capsChanged(const Jid&)));
 
 	//printf("PsiAccount: [%s] loaded\n", name().latin1());
-	d->xmlConsole = new XmlConsole(this);
+
 	if(PsiOptions::instance()->getOption("options.xml-console.enable-at-login").toBool() && enabled()) {
 		this->showXmlConsole();
 		d->xmlConsole->enable();
@@ -1334,6 +1334,8 @@ void PsiAccount::setEnabled(bool e)
 				return;
 			}
 		}
+		delete d->xmlConsole;
+		d->xmlConsole = 0;
 	}
 
 	d->setEnabled( e );
@@ -3217,6 +3219,9 @@ void PsiAccount::changePW()
 
 void PsiAccount::showXmlConsole()
 {
+	if(!d->xmlConsole) {
+		d->xmlConsole = new XmlConsole(this);
+	}
 	bringToFront(d->xmlConsole);
 }
 
