Entering 'iris'
Entering 'src/libpsi'
diff --git a/src/activity.cpp b/src/activity.cpp
index d767628..936ce68 100644
--- a/src/activity.cpp
+++ b/src/activity.cpp
@@ -84,8 +84,8 @@ bool Activity::isNull() const
 
 QDomElement Activity::toXml(QDomDocument& doc)
 {
-	QDomElement activity = doc.createElement("activity");
-	activity.setAttribute("xmlns", "http://jabber.org/protocol/activity");
+	QDomElement activity = doc.createElement(PEP_ACTIVITY_TN);
+	activity.setAttribute("xmlns", PEP_ACTIVITY_NS);
 
 	if (!type() == Unknown) {
 		ActivityCatalog* ac = ActivityCatalog::instance();
@@ -98,9 +98,6 @@ QDomElement Activity::toXml(QDomDocument& doc)
 
 		activity.appendChild(el);
 	}
-	else {
-		qWarning("activity.cpp: ERROR: Converting unknown activity");
-	}
 
 	if (!text().isEmpty()) {
 		QDomElement el = doc.createElement("text");
@@ -114,14 +111,17 @@ QDomElement Activity::toXml(QDomDocument& doc)
 
 void Activity::fromXml(const QDomElement& element)
 {
-	if (element.tagName() != "activity")
-		return;
-
 	type_ = Activity::Unknown;
 	specificType_ = Activity::UnknownSpecific;
 
+	if (element.tagName() != PEP_ACTIVITY_TN)
+		return;
+
 	for (QDomNode node = element.firstChild(); !node.isNull(); node = node.nextSibling()) {
 		QDomElement child = node.toElement();
+		if(child.isNull()) {
+			continue;
+		}
 		if (child.tagName() == "text") {
 			text_ = child.text();
 		}
diff --git a/src/activity.h b/src/activity.h
index fde0086..877c4c6 100644
--- a/src/activity.h
+++ b/src/activity.h
@@ -23,6 +23,9 @@
 
 #include <QString>
 
+#define PEP_ACTIVITY_TN "activity"
+#define PEP_ACTIVITY_NS "http://jabber.org/protocol/activity"
+
 class QDomElement;
 class QDomDocument;
 
diff --git a/src/activitycatalog.cpp b/src/activitycatalog.cpp
index 68cda63..1e1f503 100644
--- a/src/activitycatalog.cpp
+++ b/src/activitycatalog.cpp
@@ -27,6 +27,8 @@
 		
 ActivityCatalog::Entry::Entry()
 {
+	type_ = Activity::Unknown;
+	specificType_ = Activity::UnknownSpecific;
 }
 
 ActivityCatalog::Entry::Entry(Activity::Type type, const QString& value, const QString& text)
diff --git a/src/activitydlg.cpp b/src/activitydlg.cpp
index 9a9cea2..3a9d83e 100644
--- a/src/activitydlg.cpp
+++ b/src/activitydlg.cpp
@@ -94,7 +94,7 @@ void ActivityDlg::setActivity()
 
 	foreach(PsiAccount *pa, pa_) {
 		if (generalActivityStr == tr("<unset>")) {
-			pa->pepManager()->retract("http://jabber.org/protocol/activity", "current");
+			pa->pepManager()->disable(PEP_ACTIVITY_TN, PEP_ACTIVITY_NS, "current");
 		}
 		else {
 			ActivityCatalog* ac = ActivityCatalog::instance();
@@ -104,7 +104,7 @@ void ActivityDlg::setActivity()
 			if (specificActivityStr != tr("<unset>")) {
 				specificType = ac->findEntryByText(specificActivityStr).specificType();
 			}
-			pa->pepManager()->publish("http://jabber.org/protocol/activity", PubSubItem("current",Activity(generalType,specificType,ui_.le_description->text()).toXml(*pa->client()->rootTask()->doc())), PEPManager::PresenceAccess);
+			pa->pepManager()->publish(PEP_ACTIVITY_NS, PubSubItem("current",Activity(generalType,specificType,ui_.le_description->text()).toXml(*pa->client()->rootTask()->doc())));
 		}
 	}
 	close();
diff --git a/src/avatars.cpp b/src/avatars.cpp
index 0846257..22a54bb 100644
--- a/src/avatars.cpp
+++ b/src/avatars.cpp
@@ -241,7 +241,7 @@ signals:
 
 protected:
 	void requestAvatar() {
-		factory()->account()->pepManager()->get(jid_,"urn:xmpp:avatar:data",hash());
+		factory()->account()->pepManager()->get(jid_,PEP_AVATAR_DATA_NS,hash());
 	}
 
 	void avatarUpdated()
@@ -557,20 +557,16 @@ void AvatarFactory::setSelfAvatar(const QString& fileName)
 			// Publish data
 			QDomDocument* doc = account()->client()->doc();
 			QString hash = Hash("sha1").hashToString(avatar_data);
-			QDomElement el = doc->createElement("data");
-			el.setAttribute("xmlns","urn:xmpp:avatar:data");
+			QDomElement el = doc->createElement(PEP_AVATAR_DATA_TN);
+			el.setAttribute("xmlns",PEP_AVATAR_DATA_NS);
 			el.appendChild(doc->createTextNode(Base64().arrayToString(avatar_data)));
 			selfAvatarData_ = avatar_data;
 			selfAvatarHash_ = hash;
-			account()->pepManager()->publish("urn:xmpp:avatar:data",PubSubItem(hash,el));
+			account()->pepManager()->publish(PEP_AVATAR_DATA_NS,PubSubItem(hash,el));
 		}
 	}
 	else {
-		QDomDocument* doc = account()->client()->doc();
-		QDomElement meta_el =  doc->createElement("metadata");
-		meta_el.setAttribute("xmlns","urn:xmpp:avatar:metadata");
-		meta_el.appendChild(doc->createElement("stop"));
-		account()->pepManager()->publish("urn:xmpp:avatar:metadata",PubSubItem("current",meta_el));
+		account()->pepManager()->disable(PEP_AVATAR_METADATA_TN, PEP_AVATAR_METADATA_NS, "current");
 	}
 }
 
@@ -640,8 +636,8 @@ int AvatarFactory::maxAvatarSize()
 
 void AvatarFactory::itemPublished(const Jid& jid, const QString& n, const PubSubItem& item)
 {
-	if (n == "urn:xmpp:avatar:data") {
-		if (item.payload().tagName() == "data") {
+	if (n == PEP_AVATAR_DATA_NS) {
+		if (item.payload().tagName() == PEP_AVATAR_DATA_TN) {
 			if (pep_avatars_.contains(jid.bare())) {
 				pep_avatars_[jid.bare()]->setData(item.id(),item.payload().text());
 			}
@@ -650,15 +646,17 @@ void AvatarFactory::itemPublished(const Jid& jid, const QString& n, const PubSub
 			qWarning("avatars.cpp: Unexpected item payload");
 		}
 	}
-	else if (n == "urn:xmpp:avatar:metadata") {
+	else if (n == PEP_AVATAR_METADATA_NS) {
 		if (!pep_avatars_.contains(jid.bare())) {
 			pep_avatars_[jid.bare()] = new PEPAvatar(this, jid.bare());
 			connect(pep_avatars_[jid.bare()],SIGNAL(avatarChanged(const Jid&)),this, SLOT(updateAvatar(const Jid&)));
 		}
+		// check for the deprecated <stop/> child element
 		QDomElement e;
 		bool found;
 		e = findSubTag(item.payload(), "stop", &found);
-		if (found) {
+		// if there's a <stop/> element or the <metadata/> element is empty, remove user's avatar
+		if (found || !item.payload().hasChildNodes()) {
 			pep_avatars_[jid.bare()]->updateHash("");
 		}
 		else {
@@ -669,12 +667,12 @@ void AvatarFactory::itemPublished(const Jid& jid, const QString& n, const PubSub
 
 void AvatarFactory::publish_success(const QString& n, const PubSubItem& item)
 {
-	if (n == "urn:xmpp:avatar:data" && item.id() == selfAvatarHash_) {
+	if (n == PEP_AVATAR_DATA_NS && item.id() == selfAvatarHash_) {
 		// Publish metadata
 		QDomDocument* doc = account()->client()->doc();
 		QImage avatar_image = QImage::fromData(selfAvatarData_);
-		QDomElement meta_el = doc->createElement("metadata");
-		meta_el.setAttribute("xmlns","urn:xmpp:avatar:metadata");
+		QDomElement meta_el = doc->createElement(PEP_AVATAR_METADATA_TN);
+		meta_el.setAttribute("xmlns",PEP_AVATAR_METADATA_NS);
 		QDomElement info_el = doc->createElement("info");
 		info_el.setAttribute("id",selfAvatarHash_);
 		info_el.setAttribute("bytes",avatar_image.byteCount());
@@ -682,7 +680,7 @@ void AvatarFactory::publish_success(const QString& n, const PubSubItem& item)
 		info_el.setAttribute("width",avatar_image.width());
 		info_el.setAttribute("type",image2type(selfAvatarData_));
 		meta_el.appendChild(info_el);
-		account()->pepManager()->publish("urn:xmpp:avatar:metadata",PubSubItem(selfAvatarHash_,meta_el));
+		account()->pepManager()->publish(PEP_AVATAR_METADATA_NS,PubSubItem(selfAvatarHash_,meta_el));
 	}
 }
 
diff --git a/src/avatars.h b/src/avatars.h
index 9f17783..2125390 100644
--- a/src/avatars.h
+++ b/src/avatars.h
@@ -28,6 +28,12 @@
 
 #include "iconset.h"
 
+#define PEP_AVATAR_DATA_TN     "data"
+#define PEP_AVATAR_DATA_NS     "urn:xmpp:avatar:data"
+
+#define PEP_AVATAR_METADATA_TN "metadata"
+#define PEP_AVATAR_METADATA_NS "urn:xmpp:avatar:metadata"
+
 class PsiAccount;
 class Avatar;
 class VCardAvatar;
diff --git a/src/geolocation.cpp b/src/geolocation.cpp
index 995b6d2..c8975d8 100644
--- a/src/geolocation.cpp
+++ b/src/geolocation.cpp
@@ -35,8 +35,8 @@ GeoLocation::GeoLocation(const QDomElement& el)
 
 QDomElement GeoLocation::toXml(QDomDocument& doc)
 {
-	QDomElement geoloc = doc.createElement("geoloc");
-	geoloc.setAttribute("xmlns", "http://jabber.org/protocol/geoloc");
+	QDomElement geoloc = doc.createElement(PEP_GEOLOC_TN);
+	geoloc.setAttribute("xmlns", PEP_GEOLOC_NS);
 	
 	if (alt_.hasValue()) {
 		QDomElement e = doc.createElement("alt");
@@ -129,11 +129,14 @@ QDomElement GeoLocation::toXml(QDomDocument& doc)
 
 void GeoLocation::fromXml(const QDomElement& e)
 {
-	if (e.tagName() != "geoloc")
+	if (e.tagName() != PEP_GEOLOC_TN)
 		return;
 
 	for(QDomNode n = e.firstChild(); !n.isNull(); n = n.nextSibling()) {
 		QDomElement m = n.toElement();
+		if (m.isNull()) {
+			continue;
+		}
 		if (m.tagName() == "alt") 
 			alt_ = Maybe<float>(m.text().toFloat());
 		else if (m.tagName() == "bearing")
diff --git a/src/geolocation.h b/src/geolocation.h
index b51c90d..a7ae83f 100644
--- a/src/geolocation.h
+++ b/src/geolocation.h
@@ -25,6 +25,9 @@
 
 #include "maybe.h"
 
+#define PEP_GEOLOC_TN "geoloc"
+#define PEP_GEOLOC_NS "http://jabber.org/protocol/geoloc"
+
 class QDomElement;
 class QDomDocument;
 
diff --git a/src/geolocationdlg.cpp b/src/geolocationdlg.cpp
index dfb5edd..6adaae5 100644
--- a/src/geolocationdlg.cpp
+++ b/src/geolocationdlg.cpp
@@ -160,10 +160,10 @@ void GeoLocationDlg::setGeoLocation()
 
 	foreach(PsiAccount *pa, pa_) {
 		if (geoloc.isNull()) {
-			pa->pepManager()->retract("http://jabber.org/protocol/geoloc", "current");
+			pa->pepManager()->disable(PEP_GEOLOC_TN, PEP_GEOLOC_NS, "current");
 		}
 		else {
-			pa->pepManager()->publish("http://jabber.org/protocol/geoloc", PubSubItem("current",geoloc.toXml(*pa->client()->rootTask()->doc())), PEPManager::PresenceAccess);
+			pa->pepManager()->publish(PEP_GEOLOC_NS, PubSubItem("current",geoloc.toXml(*pa->client()->rootTask()->doc())));
 		}
 	}
 	close();
diff --git a/src/mood.cpp b/src/mood.cpp
index eb969db..f749fcb 100644
--- a/src/mood.cpp
+++ b/src/mood.cpp
@@ -65,16 +65,13 @@ bool Mood::isNull() const
 
 QDomElement Mood::toXml(QDomDocument& doc)
 {
-	QDomElement mood = doc.createElement("mood");
-	mood.setAttribute("xmlns", "http://jabber.org/protocol/mood");
+	QDomElement mood = doc.createElement(PEP_MOOD_TN);
+	mood.setAttribute("xmlns", PEP_MOOD_NS);
 
 	if (!type() == Unknown) {
 		QDomElement el = doc.createElement(MoodCatalog::instance()->findEntryByType(type()).value());
 		mood.appendChild(el);
 	}
-	else {
-		qWarning("mood.cpp: ERROR: Converting unknown mood");
-	}
 
 	if (!text().isEmpty()) {
 		QDomElement el = doc.createElement("text");
@@ -88,11 +85,16 @@ QDomElement Mood::toXml(QDomDocument& doc)
 
 void Mood::fromXml(const QDomElement& e)
 {
-	if (e.tagName() != "mood")
+	type_ = Mood::Unknown;
+
+	if (e.tagName() != PEP_MOOD_TN)
 		return;
 
 	for(QDomNode n = e.firstChild(); !n.isNull(); n = n.nextSibling()) {
 		QDomElement m = n.toElement();
+		if(m.isNull()) {
+			continue;
+		}
 		if (m.tagName() == "text") {
 			text_ = m.text();
 		}
diff --git a/src/mood.h b/src/mood.h
index f319ad5..03ae60f 100644
--- a/src/mood.h
+++ b/src/mood.h
@@ -23,6 +23,9 @@
 
 #include <QString>
 
+#define PEP_MOOD_TN "mood"
+#define PEP_MOOD_NS "http://jabber.org/protocol/mood"
+
 class QDomElement;
 class QDomDocument;
 
diff --git a/src/moodcatalog.cpp b/src/moodcatalog.cpp
index b2f21d1..6f62b87 100644
--- a/src/moodcatalog.cpp
+++ b/src/moodcatalog.cpp
@@ -27,6 +27,7 @@
 		
 MoodCatalog::Entry::Entry()
 {
+	type_ = Mood::Unknown;
 }
 
 MoodCatalog::Entry::Entry(Mood::Type type, const QString& value, const QString& text) : type_(type), value_(value), text_(text)
diff --git a/src/mooddlg.cpp b/src/mooddlg.cpp
index 84109e9..1e9101d 100644
--- a/src/mooddlg.cpp
+++ b/src/mooddlg.cpp
@@ -58,11 +58,11 @@ void MoodDlg::setMood()
 	QString moodstr = ui_.cb_type->currentText();
 	foreach(PsiAccount *pa, pa_) {
 		if (moodstr == tr("<unset>")) {
-			pa->pepManager()->retract("http://jabber.org/protocol/mood", "current");
+			pa->pepManager()->disable(PEP_MOOD_TN, PEP_MOOD_NS, "current");
 		}
 		else {
 			Mood::Type type = MoodCatalog::instance()->findEntryByText(moodstr).type();
-			pa->pepManager()->publish("http://jabber.org/protocol/mood", PubSubItem("current",Mood(type,ui_.le_text->text()).toXml(*pa->client()->rootTask()->doc())), PEPManager::PresenceAccess);
+			pa->pepManager()->publish(PEP_MOOD_NS, PubSubItem("current",Mood(type,ui_.le_text->text()).toXml(*pa->client()->rootTask()->doc())));
 		}
 	}
 	close();
diff --git a/src/pepmanager.cpp b/src/pepmanager.cpp
index b2d6d5c..532b045 100644
--- a/src/pepmanager.cpp
+++ b/src/pepmanager.cpp
@@ -576,6 +576,16 @@ void PEPManager::retract(const QString& node, const QString& id)
 }
 
 
+void PEPManager::disable(const QString& tagName, const QString& node, const QString& id)
+{
+	// disable by publishing an empty element
+	QDomElement element = client_->rootTask()->doc()->createElement(tagName);
+	element.setAttribute("xmlns", node);
+
+	publish(node, PubSubItem(id,element));
+}
+
+
 void PEPManager::publishFinished()
 {
 	PEPPublishTask* task = (PEPPublishTask*) sender();
diff --git a/src/pepmanager.h b/src/pepmanager.h
index 0f5c0cd..51a69c5 100644
--- a/src/pepmanager.h
+++ b/src/pepmanager.h
@@ -59,6 +59,7 @@ public:
 
 	void publish(const QString& node, const PubSubItem&, Access = DefaultAccess);
 	void retract(const QString& node, const QString& id);
+	void disable(const QString& tagName, const QString& node, const QString& id);
 	void get(const Jid& jid, const QString& node, const QString& id);
 
 	//void getSubscriptions(const Jid& jid);
