--- git.orig/src/contactlistview.h
+++ git/src/contactlistview.h
@@ -48,13 +48,13 @@ public:
 
 	void activate(const QModelIndex& index);
 	void toggleExpandedState(const QModelIndex& index);
-	void ensureVisible(const QModelIndex& index);
 
 	// reimplemented
 	void setModel(QAbstractItemModel* model);
 
 public slots:
 	virtual void rename();
+	void ensureVisible(const QModelIndex& index);
 
 signals:
 	void realExpanded(const QModelIndex&);
--- git.orig/src/psicontactlistview.cpp
+++ git/src/psicontactlistview.cpp
@@ -62,7 +62,9 @@ void PsiContactListView::contactAlert(co
 
 	bool alerting = index.data(ContactListModel::IsAlertingRole).toBool();
 	if (alerting && PsiOptions::instance()->getOption("options.ui.contactlist.ensure-contact-visible-on-event").toBool()) {
-		ensureVisible(index);
+		QMetaObject::invokeMethod(this, "ensureVisible",
+					  Qt::QueuedConnection,
+					  Q_ARG(const QModelIndex&, index));
 	}
 }
 
