--- git.orig/src/gcuserview.cpp
+++ git/src/gcuserview.cpp
@@ -552,6 +552,11 @@ void GCUserView::mousePressEvent(QMouseE
 	QTreeWidget::mousePressEvent(event);
 	QTreeWidgetItem *item = itemAt(event->pos());
 
+	if(!item && event->button() == Qt::LeftButton) {
+		setCurrentIndex(QModelIndex()); // Hack to reset current selection
+		return;
+	}
+
 	if (!item || !item->parent() || !gcDlg_)
 		return;
 	if (event->button() == Qt::MidButton ||
