--- a/src/contactlistmodel.h
+++ b/src/contactlistmodel.h
@@ -97,6 +97,7 @@ public:
 		TotalContactsRole = Qt::UserRole + 16,
 		InternalGroupNameRole = Qt::UserRole + 17,
 		SpecialGroupTypeRole = Qt::UserRole + 18,
+		DisplayGroupRole = Qt::UserRole + 33,
 
 		// accounts
 		UsingSSLRole = Qt::UserRole + 19,
--- a/src/psicontactlistmodel.cpp
+++ b/src/psicontactlistmodel.cpp
@@ -19,6 +19,7 @@
  */
 
 #include <QTimer>
+#include <QTextDocument> // for Qt::escape()
 
 #include "psicontactlistmodel.h"
 
@@ -57,8 +58,14 @@ QVariant PsiContactListModel::contactData(const PsiContact* contact, int role) c
 
 QVariant PsiContactListModel::contactGroupData(const ContactListGroup* group, int role) const
 {
-	if (role == Qt::ToolTipRole) {
+	if (role == DisplayGroupRole || role == Qt::ToolTipRole) {
 		QString text = itemData(group, Qt::DisplayRole).toString();
+		if (role == Qt::ToolTipRole)
+#ifdef HAVE_QT5
+			text = text.toHtmlEscaped();
+#else
+			text = Qt::escape(text);
+#endif
 		text += QString(" (%1/%2)")
 		        .arg(itemData(group, ContactListModel::OnlineContactsRole).toInt())
 		        .arg(itemData(group, ContactListModel::TotalContactsRole).toInt());
--- a/src/psicontactlistviewdelegate.cpp
+++ b/src/psicontactlistviewdelegate.cpp
@@ -452,7 +452,7 @@ void PsiContactListViewDelegate::drawGroup(QPainter* painter, const QStyleOption
 	r = relativeRect(option, QSize(), pixmapRect, 3);
 	painter->drawPixmap(pixmapRect.topLeft(), pixmap);
 
-	QString text = index.data(Qt::ToolTipRole).toString();
+	QString text = index.data(ContactListModel::DisplayGroupRole).toString();
 	drawText(painter, o, r, text, index);
 
 	if(slimGroup_ && !(option.state & QStyle::State_Selected)) {
