Entering 'iris'
Entering 'src/libpsi'
diff --git a/src/chatdlg.cpp b/src/chatdlg.cpp
index 458c23e..a145c67 100644
--- a/src/chatdlg.cpp
+++ b/src/chatdlg.cpp
@@ -1133,9 +1133,15 @@ void ChatDlg::chatEditCreated()
 
 TabbableWidget::State ChatDlg::state() const
 {
-	return contactChatState_ == XMPP::StateComposing ?
-		   TabbableWidget::StateComposing :
-		   TabbableWidget::StateNone;
+	TabbableWidget::State state = TabbableWidget::StateNone;
+	if(contactChatState_ == XMPP::StateComposing) {
+		state = TabbableWidget::StateComposing;
+	}
+	else if(unreadMessageCount()) {
+		state = TabbableWidget::StateHighlighted;
+	}
+
+	return state;
 }
 
 int ChatDlg::unreadMessageCount() const
diff --git a/src/groupchatdlg.cpp b/src/groupchatdlg.cpp
index abc3551..6332c9d 100644
--- a/src/groupchatdlg.cpp
+++ b/src/groupchatdlg.cpp
@@ -1994,7 +1994,7 @@ void GCMainDlg::chatEditCreated()
 
 TabbableWidget::State GCMainDlg::state() const
 {
-	return TabbableWidget::StateNone;
+	return d->hPending ? TabbableWidget::StateHighlighted : TabbableWidget::StateNone;
 }
 
 int GCMainDlg::unreadMessageCount() const
diff --git a/src/tabs/tabbablewidget.h b/src/tabs/tabbablewidget.h
index b39b9f6..af4116b 100644
--- a/src/tabs/tabbablewidget.h
+++ b/src/tabs/tabbablewidget.h
@@ -63,7 +63,8 @@ public:
 
 	enum State {
 		StateNone = 0,
-		StateComposing
+		StateComposing,
+		StateHighlighted
 	};
 	virtual State state() const = 0;
 	virtual int unreadMessageCount() const = 0;
diff --git a/src/tabs/tabdlg.cpp b/src/tabs/tabdlg.cpp
index 14d1009..3028e93 100644
--- a/src/tabs/tabdlg.cpp
+++ b/src/tabs/tabdlg.cpp
@@ -651,17 +651,25 @@ void TabDlg::updateTab(TabbableWidget* chat)
 	tabWidget_->setTabText(chat, captionForTab(chat));
 	//now set text colour based upon whether there are new messages/composing etc
 
-	if (chat->state() == TabbableWidget::StateComposing) {
+	TabbableWidget::State state = chat->state();
+	if (state == TabbableWidget::StateComposing) {
 		tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.composing-color").value<QColor>());
 		tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/typing")->icon());
 	}
-	else if (chat->unreadMessageCount()) {
-		tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.unread-message-color").value<QColor>());
-		tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/chat")->icon());
-	}
 	else {
-		tabWidget_->setTabTextColor(chat, palette().color(QPalette::Text));
-		tabWidget_->setTabIcon(chat, chat->icon());
+		if (state == TabbableWidget::StateHighlighted) {
+			tabWidget_->setTabTextColor(chat, PsiOptions::instance()->getOption("options.ui.look.colors.chat.unread-message-color").value<QColor>());
+		}
+		else {
+			tabWidget_->setTabTextColor(chat, palette().color(QPalette::Text));
+		}
+
+		if(chat->unreadMessageCount()) {
+			tabWidget_->setTabIcon(chat, IconsetFactory::iconPtr("psi/chat")->icon());
+		}
+		else {
+			tabWidget_->setTabIcon(chat, chat->icon());
+		}
 	}
 	updateCaption();
 }
