--- git.orig/src/contactlistgroup.cpp
+++ git/src/contactlistgroup.cpp
@@ -334,6 +334,11 @@ bool ContactListGroup::haveOnlineContact
 	return haveOnlineContacts_;
 }
 
+bool ContactListGroup::haveAlwaysVisibleContacts() const
+{
+	return haveAlwaysVisibleContacts_;
+}
+
 int ContactListGroup::onlineContactsCount() const
 {
 	return onlineContactsCount_;
@@ -356,6 +361,7 @@ void ContactListGroup::updateOnlineConta
 		return;
 
 	bool haveOnlineContacts = false;
+	bool haveAlwaysVisibleContacts = false;
 	int onlineContactsCount = 0;
 	int totalContactsCount = 0;
 	foreach(ContactListItemProxy* item, items_) {
@@ -368,6 +374,9 @@ void ContactListGroup::updateOnlineConta
 				++onlineContactsCount;
 				// break;
 			}
+			if (contact->isAlwaysVisible()) {
+				haveAlwaysVisibleContacts = true;
+			}
 		}
 		else if ((group = dynamic_cast<ContactListGroup*>(item->item()))) {
 			totalContactsCount += group->totalContactsCount();
@@ -387,6 +396,11 @@ void ContactListGroup::updateOnlineConta
 		}
 	}
 
+	if (haveAlwaysVisibleContacts != haveAlwaysVisibleContacts_) {
+		haveAlwaysVisibleContacts_ = haveAlwaysVisibleContacts;
+		model_->updatedGroupVisibility(this);
+	}
+
 	if (onlineContactsCount != onlineContactsCount_ ||
 	    totalContactsCount != totalContactsCount_)
 	{
--- git.orig/src/contactlistgroup.h
+++ git/src/contactlistgroup.h
@@ -52,6 +52,7 @@ public:
 
 	bool isFake() const;
 	bool haveOnlineContacts() const;
+	bool haveAlwaysVisibleContacts() const;
 	virtual int onlineContactsCount() const;
 	virtual int totalContactsCount() const;
 	int contactsCount() const;
@@ -122,6 +123,7 @@ private:
 	QVector<PsiContact*> contacts_;
 	QVector<ContactListItemProxy*> items_;
 	bool haveOnlineContacts_;
+	bool haveAlwaysVisibleContacts_;
 	int onlineContactsCount_;
 	int totalContactsCount_;
 
--- git.orig/src/contactlistproxymodel.cpp
+++ git/src/contactlistproxymodel.cpp
@@ -92,6 +92,9 @@ bool ContactListProxyModel::filterAccept
 		else if (psiContact->isAgent()) {
 			return showTransports();
 		}
+		else if (psiContact->isAlwaysVisible()) {
+			return true;
+		}
 
 		bool show = true;
 		if (psiContact->isHidden()) {
@@ -112,6 +115,10 @@ bool ContactListProxyModel::filterAccept
 				if (specialGroupType == ContactListGroup::SpecialType_Transports)
 					return showTransports();
 			}
+			ContactListGroup* group = static_cast<ContactListGroup*>(item);
+			if (group->haveAlwaysVisibleContacts()) {
+				return true;
+			}
 
 			bool show = true;
 			if (index.data(Qt::DisplayRole) == PsiContact::hiddenGroupName()) {
@@ -119,7 +126,6 @@ bool ContactListProxyModel::filterAccept
 			}
 
 			if (!showOffline()) {
-				ContactListGroup* group = static_cast<ContactListGroup*>(item);
 				return show && group->haveOnlineContacts();
 			}
 			else {
--- git.orig/src/profiles.h
+++ git/src/profiles.h
@@ -84,6 +84,8 @@ public:
 	XMPP::Jid dtProxy;
 	bool ibbOnly;
 
+	QStringList alwaysVisibleContacts;
+
 	QString optionsBase;
 
 	QString stunHost;
--- git.orig/src/psiaccount.cpp
+++ git/src/psiaccount.cpp
@@ -870,6 +870,37 @@ public:
 		avCallManager->setStunRelayTcpService(acc.stunHost, acc.stunPort, convert_proxy(acc, jid), acc.stunUser, acc.stunPass);
 	}
 
+	bool isAlwaysVisibleContact(const Jid& jid) const
+	{
+		return acc.alwaysVisibleContacts.contains(jid.bare());
+	}
+
+	void setAlwaysVisibleContact(const Jid& jid, bool visible)
+	{
+		if(!visible) {
+			acc.alwaysVisibleContacts.removeAll(jid.bare());
+		}
+		else if(!acc.alwaysVisibleContacts.contains(jid.bare())) {
+			acc.alwaysVisibleContacts.append(jid.bare());
+		}
+	}
+
+	void updateContacts()
+	{
+		QStringList jids;
+		foreach(PsiContact *pc, contacts) {
+			const Jid jid = pc->jid();
+			jids.append(jid.bare());
+			bool vis = isAlwaysVisibleContact(jid);
+			if(vis != pc->isAlwaysVisible())
+				pc->setAlwaysVisible(vis);
+		}
+		foreach(const QString& j,  acc.alwaysVisibleContacts) {
+			if(!jids.contains(j))
+				acc.alwaysVisibleContacts.removeAll(j);
+		}
+	}
+
 private:
 	Status lastManualStatus_;
 
@@ -1248,6 +1279,8 @@ PsiAccount::PsiAccount(const UserAccount
 	QTimer::singleShot(0, d, SLOT(loadQueue()));
 
 	d->contactList->link(this);
+
+	d->updateContacts(); //update always visible contacts state
 }
 
 PsiAccount::~PsiAccount()
@@ -1313,6 +1346,11 @@ void PsiAccount::cleanupStream()
 	d->localAddress = QHostAddress();
 }
 
+void PsiAccount::updateAlwaysVisibleContact(PsiContact *pc)
+{
+	d->setAlwaysVisibleContact(pc->jid(), pc->isAlwaysVisible());
+}
+
 bool PsiAccount::enabled() const
 {
 	return d->acc.opt_enabled;
--- git.orig/src/psiaccount.h
+++ git/src/psiaccount.h
@@ -180,6 +180,8 @@ public:
 
 	bool notifyOnline() const;
 
+	void updateAlwaysVisibleContact(PsiContact* pc);
+
 	void modify();
 	void reconfigureFTManager();
 	void changeVCard();
--- git.orig/src/psicontact.cpp
+++ git/src/psicontact.cpp
@@ -67,6 +67,7 @@ public:
 		, statusTimer_(0)
 		, isValid_(true)
 		, isAnimated_(false)
+		, isAlwaysVisible_(false)
 		, contact_(contact)
 #ifdef YAPSI
 		, gender_(XMPP::VCard::UnknownGender)
@@ -108,6 +109,7 @@ public:
 	Status oldStatus_;
 	bool isValid_;
 	bool isAnimated_;
+	bool isAlwaysVisible_;
 #ifdef YAPSI
 	bool showOnlineTemporarily_;
 	bool reconnecting_;
@@ -550,6 +552,18 @@ bool PsiContact::shouldBeVisible() const
 	// return ContactListItem::shouldBeVisible();
 }
 
+bool PsiContact::isAlwaysVisible() const
+{
+	return d->isAlwaysVisible_;
+}
+
+void PsiContact::setAlwaysVisible(bool visible)
+{
+	d->isAlwaysVisible_ = visible;
+	account()->updateAlwaysVisibleContact(this);
+	emit updated();
+}
+
 /**
  * Standard desired parent.
  */
--- git.orig/src/psicontact.h
+++ git/src/psicontact.h
@@ -70,6 +70,9 @@ public:
 
 	virtual void setEditing(bool editing);
 
+	bool isAlwaysVisible() const;
+	void setAlwaysVisible(bool visible);
+
 	bool addAvailable() const;
 	bool removeAvailable() const;
 	bool authAvailable() const;
--- git.orig/src/psicontactmenu.cpp
+++ git/src/psicontactmenu.cpp
@@ -257,6 +257,7 @@ public:
 	QAction* mucShowAction_;
 	QAction* mucLeaveAction_;
 	QAction* blockAction_;
+	QAction* visibleAction_;
 
 #endif
 
@@ -427,6 +428,9 @@ public:
 		blockAction_ = new IconAction(tr("Block"), "psi/stop", tr("Block"), 0, this, 0, true);
 		connect(blockAction_, SIGNAL(triggered(bool)), SLOT(block(bool)));
 
+		visibleAction_ = new IconAction(tr("Always Visible"), "psi/eye", tr("Always Visible"), 0, this, 0, true);
+		connect(visibleAction_, SIGNAL(triggered(bool)), SLOT(setAlwaysVisible(bool)));
+
 		if (!contact_->isConference()) {
 			menu_->addAction(addAuthAction_);
 			menu_->addSeparator();
@@ -452,6 +456,7 @@ public:
 			menu_->addAction(transportLogonAction_);
 			menu_->addAction(transportLogoffAction_);
 			menu_->addAction(blockAction_);
+			menu_->addAction(visibleAction_);
 			authMenu_ = menu_->addMenu(tr("Authorization"));
 			authMenu_->addAction(authResendAction_);
 			authMenu_->addAction(authRerequestAction_);
@@ -568,6 +573,7 @@ private slots:
 		}
 		authMenu_->setEnabled(contact_->account()->isAvailable());
 		updateBlockActionState();
+		visibleAction_->setChecked(contact_->isAlwaysVisible());
 		removeAction_->setVisible(!PsiOptions::instance()->getOption("options.ui.contactlist.lockdown-roster").toBool()  && !contact_->isSelf());
 		removeAction_->setEnabled(contact_->removeAvailable());
 		if (!PsiOptions::instance()->getOption("options.ui.menu.contact.custom-picture").toBool()) {
@@ -656,6 +662,14 @@ private slots:
 		contact_->toggleBlockedStateConfirmation();
 	}
 
+	void setAlwaysVisible(bool visible)
+	{
+		if (!contact_)
+			return;
+
+		contact_->setAlwaysVisible(visible);
+	}
+
 	void addAuth()
 	{
 		if (!contact_)
--- git.orig/src/psi_profiles.cpp
+++ git/src/psi_profiles.cpp
@@ -326,6 +326,8 @@ void UserAccount::fromOptions(OptionsTre
 		tlsOverrideCert = o->getOption(base + ".tls.override-certificate").toByteArray();
 		tlsOverrideDomain = o->getOption(base + ".tls.override-domain").toString();
 	}
+
+	alwaysVisibleContacts = o->getOption(base + ".always-visible-contacts").toStringList();
 }
 
 void UserAccount::toOptions(OptionsTree *o, QString base)
@@ -463,6 +465,8 @@ void UserAccount::toOptions(OptionsTree 
 	o->setOption(base + ".tls.override-certificate", tlsOverrideCert);
 	o->setOption(base + ".tls.override-domain", tlsOverrideDomain);
 	saveLastStatus(o, base);
+
+	o->setOption(base + ".always-visible-contacts", alwaysVisibleContacts);
 }
 
 void UserAccount::fromXml(const QDomElement &a)
