--- psi.orig/src/contactlistnestedgroup.cpp
+++ psi/src/contactlistnestedgroup.cpp
@@ -289,12 +289,12 @@ ContactListGroup* ContactListNestedGroup
 	}
 
 	ContactListGroup::SpecialType type = ContactListGroup::SpecialType_None;
-	if (!contact->inList()) {
-		type = ContactListGroup::SpecialType_NotInList;
-	}
-	else if (contact->isPrivate()) {
+	if (contact->isPrivate()) {
 		type = ContactListGroup::SpecialType_MUCPrivateChats;
 	}
+	else if (!contact->inList()) {
+		type = ContactListGroup::SpecialType_NotInList;
+	}
 	else if (contact->isAgent()) {
 		type = ContactListGroup::SpecialType_Transports;
 	}
