Entering 'iris'
Entering 'src/libpsi'
Index: psi/src/aboutdlg.cpp
===================================================================
--- psi.orig/src/aboutdlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/aboutdlg.cpp	2013-01-15 14:04:29.724827171 +0400
@@ -24,12 +24,14 @@

 #include "applicationinfo.h"
 #include "aboutdlg.h"
+#include "iconset.h"

 AboutDlg::AboutDlg(QWidget* parent)
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/psiplus_logo").icon());

 	setModal(false);

Index: psi/src/accountmanagedlg.cpp
===================================================================
--- psi.orig/src/accountmanagedlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/accountmanagedlg.cpp	2013-01-15 14:04:29.724827171 +0400
@@ -41,6 +41,7 @@
 #include "psicontactlist.h"
 #include "iconaction.h"
 #include "shortcutmanager.h"
+#include "iconset.h"

 using namespace XMPP;

@@ -94,6 +95,7 @@
 :QDialog(parent)
 {
 	setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/account").icon());
 	setModal(false);
 #if QT_VERSION >= 0x040500
 	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
Index: psi/src/activitydlg.cpp
===================================================================
--- psi.orig/src/activitydlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/activitydlg.cpp	2013-01-15 14:04:29.724827171 +0400
@@ -33,7 +33,9 @@
 	if(pa_.isEmpty())
 		close();
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("activities/other").icon());
 	setModal(false);
+
 	connect(ui_.cb_general_type, SIGNAL(currentIndexChanged(const QString&)), SLOT(loadSpecificActivities(const QString&)));
 	connect(ui_.pb_cancel, SIGNAL(clicked()), SLOT(close()));
 	connect(ui_.pb_ok, SIGNAL(clicked()), SLOT(setActivity()));
Index: psi/src/avcall/calldlg.cpp
===================================================================
--- psi.orig/src/avcall/calldlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/avcall/calldlg.cpp	2013-01-15 14:04:29.724827171 +0400
@@ -28,6 +28,7 @@
 #include "common.h"
 #include "psiaccount.h"
 #include "psioptions.h"
+#include "iconset.h"
 
 // from opt_avcall.cpp
 extern void options_avcall_update();
@@ -68,6 +69,7 @@ public:
 	{
 		ui.setupUi(q);
 		q->setWindowTitle(tr("Voice Call"));
+		q->setWindowIcon(IconsetFactory::icon("psi/avcall").icon());
 
 		ui.lb_bandwidth->setEnabled(false);
 		ui.cb_bandwidth->setEnabled(false);

Index: psi/src/bookmarkmanagedlg.cpp
===================================================================
--- psi.orig/src/bookmarkmanagedlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/bookmarkmanagedlg.cpp	2013-01-15 14:04:29.724827171 +0400
@@ -26,6 +26,7 @@

 #include "bookmarkmanager.h"
 #include "psiaccount.h"
+#include "iconset.h"

 BookmarkManageDlg::BookmarkManageDlg(PsiAccount* account)
 	: QDialog()
@@ -37,6 +38,8 @@
 	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
 #endif
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/bookmarks").icon());
+
 	account_->dialogRegister(this);

 	QAction* removeBookmarkAction = new QAction(this);
Index: psi/src/contactmanager/contactmanagerdlg.cpp
===================================================================
--- psi.orig/src/contactmanager/contactmanagerdlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/contactmanager/contactmanagerdlg.cpp	2013-01-15 14:04:29.724827171 +0400
@@ -44,6 +44,7 @@
 	ui_.setupUi(this);
 	pa_->dialogRegister(this);
 	setWindowTitle(tr("Contacts Manager")+" - "+pa->jid().bare());
+	setWindowIcon(IconsetFactory::icon("psi/action_contacts_manager").icon());

 	um = new ContactManagerModel(this, pa_);
 	um->reloadUsers();
Index: psi/src/geolocationdlg.cpp
===================================================================
--- psi.orig/src/geolocationdlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/geolocationdlg.cpp	2013-01-15 14:04:29.724827171 +0400
@@ -27,6 +27,7 @@
 #include "pepmanager.h"
 #include "geolocation.h"
 #include <QLineEdit>
+#include "iconset.h"

 GeoLocationDlg::GeoLocationDlg(QList<PsiAccount*> list) : QDialog(0), pa_(list)
 {
@@ -34,7 +35,9 @@
 	if(pa_.isEmpty())
 		close();
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("system/geolocation").icon());
 	setModal(false);
+
 	connect(ui_.pb_cancel, SIGNAL(clicked()), SLOT(close()));
  	connect(ui_.pb_ok, SIGNAL(clicked()), SLOT(setGeoLocation()));
 	connect(ui_.pb_reset, SIGNAL(clicked()), SLOT(reset()));
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -634,7 +634,7 @@
 	setAcceptDrops(true);

 #ifndef Q_OS_MAC
-	setWindowIcon(IconsetFactory::icon("psi/groupChat").icon());
+	setWindowIcon(IconsetFactory::icon("psi/start-chat").icon());
 #endif

 	ui_.setupUi(this);
@@ -1858,7 +1858,7 @@

 	// update the widget icon
 #ifndef Q_OS_MAC
-	setWindowIcon(IconsetFactory::icon("psi/groupChat").icon());
+	setWindowIcon(IconsetFactory::icon("psi/start-chat").icon());
 #endif

 	ui_.lv_users->setLooks();
@@ -2097,6 +2097,7 @@
 	formLayout->addRow(tr("&Auto join:"), chkAJoin);
 	layout->addLayout(formLayout);
 	layout->addLayout(blayout);
+	dlg->setWindowIcon(IconsetFactory::icon("psi/bookmark_remove").icon());
 	dlg->setLayout(layout);
 	dlg->setMinimumWidth(300);
 	dlg->connect(saveBtn, SIGNAL(clicked()), dlg, SLOT(accept()));
Index: psi/src/mooddlg.cpp
===================================================================
--- psi.orig/src/mooddlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/mooddlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -34,7 +34,9 @@
 	if(pa_.isEmpty())
 		close();
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("mood/").icon());
 	setModal(false);
+
 	connect(ui_.pb_cancel, SIGNAL(clicked()), SLOT(close()));
 	connect(ui_.pb_ok, SIGNAL(clicked()), SLOT(setMood()));

Index: psi/src/mucjoindlg.cpp
===================================================================
--- psi.orig/src/mucjoindlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/mucjoindlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -30,6 +30,7 @@
 #include "psicontactlist.h"
 #include "groupchatdlg.h"
 #include "bookmarkmanager.h"
+#include "iconset.h"

 MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiAccount* pa)
 	: QDialog(0)
@@ -40,6 +41,7 @@
 #endif
 	setModal(false);
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/groupChat").icon());
 	controller_ = psi;
 	account_ = 0;
 	controller_->dialogRegister(this);
Index: psi/src/options/optionsdlg.cpp
===================================================================
--- psi.orig/src/options/optionsdlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/options/optionsdlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -499,6 +499,7 @@
 	: QDialog(parent)
 {
 	setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/options").icon());
 	pb_apply = buttonBox->button(QDialogButtonBox::Apply);

 	setAttribute(Qt::WA_DeleteOnClose);
Index: psi/src/privacy/privacydlg.cpp
===================================================================
--- psi.orig/src/privacy/privacydlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/privacy/privacydlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -26,6 +26,7 @@
 #include "privacylist.h"
 #include "privacymanager.h"
 #include "privacylistmodel.h"
+#include "iconset.h"

 // fixme: subscribe on the destroyed() signal of the manager

@@ -34,6 +35,7 @@
 	ui_.setupUi(this);
 	setAttribute(Qt::WA_DeleteOnClose);
 	setWindowTitle(tr("%1: Privacy Lists").arg(account_name));
+	setWindowIcon(IconsetFactory::icon("psi/eye").icon());
 #if QT_VERSION >= 0x040500
 	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
 #endif
Index: psi/src/profiledlg.cpp
===================================================================
--- psi.orig/src/profiledlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/profiledlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -81,6 +81,7 @@
 	setupUi(this);
 	setModal(true);
 	setWindowTitle(CAP(windowTitle()));
+	setWindowIcon(IconsetFactory::icon("psi/profile").icon());
 	pb_open->setDefault(true);

 	langs = _langs;
Index: psi/src/psioptionseditor.cpp
===================================================================
--- psi.orig/src/psioptionseditor.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/psioptionseditor.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -184,6 +184,9 @@
 		pb_detach->setToolTip(tr("Open a detached option editor window."));
 		buttonLine->addWidget(pb_detach);
 		connect(pb_detach, SIGNAL(clicked()), SLOT(detach()));
+	} else {
+		setWindowIcon(IconsetFactory::icon("psi/advanced-plus").icon());
+		setWindowTitle(CAP(tr("Advanced")));
 	}


Index: psi/src/statusdlg.cpp
===================================================================
--- psi.orig/src/statusdlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/statusdlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -71,6 +71,7 @@
 	te->setText(u.makeDesc());

 	setWindowTitle(tr("Status for %1").arg(JIDUtil::nickOrJid(u.name(), u.jid().full())));
+	setWindowIcon(IconsetFactory::icon("psi/action_direct_presence").icon());
 	resize(400,240);

 	pb->setFocus();
@@ -112,6 +113,7 @@
 	d->withPriority = withPriority;

 	setWindowTitle(CAP(tr("Set Status: All accounts")));
+	setWindowIcon(IconsetFactory::icon("psi/action_direct_presence").icon());
 	d->setStatusMode = setStatusForAccount;
 	init();
 }
@@ -128,6 +130,7 @@
 	d->withPriority = withPriority;

 	setWindowTitle(CAP(tr("Set Status: %1").arg(d->pa->name())));
+	setWindowIcon(IconsetFactory::icon("psi/action_direct_presence").icon());
 	d->setStatusMode = setStatusForAccount;
 	init();
 }
Index: psi/src/tipdlg.cpp
===================================================================
--- psi.orig/src/tipdlg.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/tipdlg.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -27,6 +27,7 @@

 #include "psioptions.h"
 #include "psicon.h"
+#include "iconset.h"

 void TipDlg::show(PsiCon* psi)
 {
@@ -48,6 +49,7 @@
 {
 	setAttribute(Qt::WA_DeleteOnClose);
 	setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/tip").icon());
 	setModal(false);

 	psi_->dialogRegister(this);
Index: psi/src/xmlconsole.cpp
===================================================================
--- psi.orig/src/xmlconsole.cpp	2013-01-15 14:04:29.732827171 +0400
+++ psi/src/xmlconsole.cpp	2013-01-15 14:04:29.728827171 +0400
@@ -32,6 +32,7 @@
 #include "psiaccount.h"
 #include "psicon.h"
 #include "psicontactlist.h"
+#include "iconset.h"

 //----------------------------------------------------------------------------
 // XmlConsole
@@ -40,7 +41,7 @@
 :QWidget()
 {
 	ui_.setupUi(this);
-
+	setWindowIcon(IconsetFactory::icon("psi/xml").icon());
 	pa = _pa;
 	pa->dialogRegister(this);
 	connect(pa, SIGNAL(updatedAccount()), SLOT(updateCaption()));
