--- git.orig/src/aboutdlg.cpp
+++ git/src/aboutdlg.cpp
@@ -24,12 +24,14 @@
 
 #include "applicationinfo.h"
 #include "aboutdlg.h"
+#include "iconset.h"
 
 AboutDlg::AboutDlg(QWidget* parent)
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/psiplus_logo").icon());
 
 	setModal(false);
 
--- git.orig/src/accountmanagedlg.cpp
+++ git/src/accountmanagedlg.cpp
@@ -41,6 +41,7 @@
 #include "psicontactlist.h"
 #include "iconaction.h"
 #include "shortcutmanager.h"
+#include "iconset.h"
 
 using namespace XMPP;
 
@@ -94,6 +95,7 @@ AccountRemoveDlg::AccountRemoveDlg(const
 :QDialog(parent)
 {
 	setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/account").icon());
 	setModal(false);
 #if QT_VERSION >= 0x040500
 	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
--- git.orig/src/activitydlg.cpp
+++ git/src/activitydlg.cpp
@@ -33,7 +33,9 @@ ActivityDlg::ActivityDlg(QList<PsiAccoun
 	if(pa_.isEmpty())
 		close();
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("activities/other").icon());
 	setModal(false);
+
 	connect(ui_.cb_general_type, SIGNAL(currentIndexChanged(const QString&)), SLOT(loadSpecificActivities(const QString&)));
 	connect(ui_.pb_cancel, SIGNAL(clicked()), SLOT(close()));
 	connect(ui_.pb_ok, SIGNAL(clicked()), SLOT(setActivity()));
--- git.orig/src/avcall/calldlg.cpp
+++ git/src/avcall/calldlg.cpp
@@ -28,6 +28,7 @@
 #include "common.h"
 #include "psiaccount.h"
 #include "psioptions.h"
+#include "iconset.h"
 
 // from opt_avcall.cpp
 extern void options_avcall_update();
@@ -68,6 +69,7 @@ public:
 	{
 		ui.setupUi(q);
 		q->setWindowTitle(tr("Voice Call"));
+		q->setWindowIcon(IconsetFactory::icon("psi/avcall").icon());
 
 		ui.lb_bandwidth->setEnabled(false);
 		ui.cb_bandwidth->setEnabled(false);
--- git.orig/src/bookmarkmanagedlg.cpp
+++ git/src/bookmarkmanagedlg.cpp
@@ -26,6 +26,7 @@
 
 #include "bookmarkmanager.h"
 #include "psiaccount.h"
+#include "iconset.h"
 
 BookmarkManageDlg::BookmarkManageDlg(PsiAccount* account)
 	: QDialog()
@@ -37,6 +38,8 @@ BookmarkManageDlg::BookmarkManageDlg(Psi
 	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
 #endif
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/bookmarks").icon());
+
 	account_->dialogRegister(this);
 
 	QAction* removeBookmarkAction = new QAction(this);
--- git.orig/src/contactmanager/contactmanagerdlg.cpp
+++ git/src/contactmanager/contactmanagerdlg.cpp
@@ -44,6 +44,7 @@ ContactManagerDlg::ContactManagerDlg(Psi
 	ui_.setupUi(this);
 	pa_->dialogRegister(this);
 	setWindowTitle(tr("Contacts Manager")+" - "+pa->jid().bare());
+	setWindowIcon(IconsetFactory::icon("psi/action_contacts_manager").icon());
 
 	um = new ContactManagerModel(this, pa_);
 	um->reloadUsers();
--- git.orig/src/geolocationdlg.cpp
+++ git/src/geolocationdlg.cpp
@@ -27,6 +27,7 @@
 #include "pepmanager.h"
 #include "geolocation.h"
 #include <QLineEdit>
+#include "iconset.h"
 
 GeoLocationDlg::GeoLocationDlg(QList<PsiAccount*> list) : QDialog(0), pa_(list)
 {
@@ -34,7 +35,9 @@ GeoLocationDlg::GeoLocationDlg(QList<Psi
 	if(pa_.isEmpty())
 		close();
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("system/geolocation").icon());
 	setModal(false);
+
 	connect(ui_.pb_cancel, SIGNAL(clicked()), SLOT(close()));
  	connect(ui_.pb_ok, SIGNAL(clicked()), SLOT(setGeoLocation()));
 	connect(ui_.pb_reset, SIGNAL(clicked()), SLOT(reset()));
--- git.orig/src/groupchatdlg.cpp
+++ git/src/groupchatdlg.cpp
@@ -631,7 +631,7 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	setAcceptDrops(true);
 
 #ifndef Q_OS_MAC
-	setWindowIcon(IconsetFactory::icon("psi/groupChat").icon());
+	setWindowIcon(IconsetFactory::icon("psi/start-chat").icon());
 #endif
 
 	ui_.setupUi(this);
@@ -1856,7 +1856,7 @@ void GCMainDlg::setLooks()
 
 	// update the widget icon
 #ifndef Q_OS_MAC
-	setWindowIcon(IconsetFactory::icon("psi/groupChat").icon());
+	setWindowIcon(IconsetFactory::icon("psi/start-chat").icon());
 #endif
 
 	ui_.lv_users->setLooks();
@@ -2094,6 +2094,7 @@ void GCMainDlg::doBookmark()
 	formLayout->addRow(tr("&Auto join:"), chkAJoin);
 	layout->addLayout(formLayout);
 	layout->addLayout(blayout);
+	dlg->setWindowIcon(IconsetFactory::icon("psi/bookmark_remove").icon());
 	dlg->setLayout(layout);
 	dlg->setMinimumWidth(300);
 	dlg->connect(saveBtn, SIGNAL(clicked()), dlg, SLOT(accept()));
--- git.orig/src/mooddlg.cpp
+++ git/src/mooddlg.cpp
@@ -34,7 +34,9 @@ MoodDlg::MoodDlg(QList<PsiAccount*> list
 	if(pa_.isEmpty())
 		close();
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("mood/").icon());
 	setModal(false);
+
 	connect(ui_.pb_cancel, SIGNAL(clicked()), SLOT(close()));
 	connect(ui_.pb_ok, SIGNAL(clicked()), SLOT(setMood()));
 
--- git.orig/src/mucjoindlg.cpp
+++ git/src/mucjoindlg.cpp
@@ -30,6 +30,7 @@
 #include "psicontactlist.h"
 #include "groupchatdlg.h"
 #include "bookmarkmanager.h"
+#include "iconset.h"
 
 MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiAccount* pa)
 	: QDialog(0)
@@ -40,6 +41,7 @@ MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiA
 #endif
 	setModal(false);
 	ui_.setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/groupChat").icon());
 	controller_ = psi;
 	account_ = 0;
 	controller_->dialogRegister(this);
--- git.orig/src/options/optionsdlg.cpp
+++ git/src/options/optionsdlg.cpp
@@ -499,6 +499,7 @@ OptionsDlg::OptionsDlg(PsiCon *psi, QWid
 	: QDialog(parent)
 {
 	setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/options").icon());
 	pb_apply = buttonBox->button(QDialogButtonBox::Apply);
 
 	setAttribute(Qt::WA_DeleteOnClose);
--- git.orig/src/privacy/privacydlg.cpp
+++ git/src/privacy/privacydlg.cpp
@@ -26,6 +26,7 @@
 #include "privacylist.h"
 #include "privacymanager.h"
 #include "privacylistmodel.h"
+#include "iconset.h"
 
 // fixme: subscribe on the destroyed() signal of the manager
 
@@ -34,6 +35,7 @@ PrivacyDlg::PrivacyDlg(const QString& ac
 	ui_.setupUi(this);
 	setAttribute(Qt::WA_DeleteOnClose);
 	setWindowTitle(tr("%1: Privacy Lists").arg(account_name));
+	setWindowIcon(IconsetFactory::icon("psi/eye").icon());
 #if QT_VERSION >= 0x040500
 	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
 #endif
--- git.orig/src/profiledlg.cpp
+++ git/src/profiledlg.cpp
@@ -81,6 +81,7 @@ ProfileOpenDlg::ProfileOpenDlg(const QSt
 	setupUi(this);
 	setModal(true);
 	setWindowTitle(CAP(windowTitle()));
+	setWindowIcon(IconsetFactory::icon("psi/profile").icon());
 	pb_open->setDefault(true);
 
 	langs = _langs;
--- git.orig/src/psioptionseditor.cpp
+++ git/src/psioptionseditor.cpp
@@ -184,6 +184,9 @@ PsiOptionsEditor::PsiOptionsEditor(QWidg
 		pb_detach->setToolTip(tr("Open a detached option editor window."));
 		buttonLine->addWidget(pb_detach);
 		connect(pb_detach, SIGNAL(clicked()), SLOT(detach()));
+	} else {
+		setWindowIcon(IconsetFactory::icon("psi/advanced-plus").icon());
+		setWindowTitle(CAP(tr("Advanced")));
 	}
 
 
--- git.orig/src/statusdlg.cpp
+++ git/src/statusdlg.cpp
@@ -72,6 +72,7 @@ StatusShowDlg::StatusShowDlg(const UserL
 	te->setText(u.makeDesc());
 
 	setWindowTitle(tr("Status for %1").arg(JIDUtil::nickOrJid(u.name(), u.jid().full())));
+	setWindowIcon(IconsetFactory::icon("psi/action_direct_presence").icon());
 	resize(400,240);
 
 	pb->setFocus();
@@ -113,6 +114,7 @@ StatusSetDlg::StatusSetDlg(PsiCon *psi,
 	d->withPriority = withPriority;
 
 	setWindowTitle(CAP(tr("Set Status: All accounts")));
+	setWindowIcon(IconsetFactory::icon("psi/action_direct_presence").icon());
 	d->setStatusMode = setStatusForAccount;
 	init();
 }
@@ -129,6 +131,7 @@ StatusSetDlg::StatusSetDlg(PsiAccount *p
 	d->withPriority = withPriority;
 
 	setWindowTitle(CAP(tr("Set Status: %1").arg(d->pa->name())));
+	setWindowIcon(IconsetFactory::icon("psi/action_direct_presence").icon());
 	d->setStatusMode = setStatusForAccount;
 	init();
 }
--- git.orig/src/tipdlg.cpp
+++ git/src/tipdlg.cpp
@@ -27,6 +27,7 @@
 
 #include "psioptions.h"
 #include "psicon.h"
+#include "iconset.h"
 
 void TipDlg::show(PsiCon* psi)
 {
@@ -48,6 +49,7 @@ TipDlg::TipDlg(PsiCon* psi)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
 	setupUi(this);
+	setWindowIcon(IconsetFactory::icon("psi/tip").icon());
 	setModal(false);
 
 	psi_->dialogRegister(this);
--- git.orig/src/xmlconsole.cpp
+++ git/src/xmlconsole.cpp
@@ -33,6 +33,7 @@
 #include "psicon.h"
 #include "psicontactlist.h"
 #include "textutil.h"
+#include "iconset.h"
 
 //----------------------------------------------------------------------------
 // XmlConsole
@@ -41,7 +42,7 @@ XmlConsole::XmlConsole(PsiAccount *_pa)
 :QWidget()
 {
 	ui_.setupUi(this);
-
+	setWindowIcon(IconsetFactory::icon("psi/xml").icon());
 	pa = _pa;
 	pa->dialogRegister(this);
 	connect(pa, SIGNAL(updatedAccount()), SLOT(updateCaption()));
