--- git.orig/options/default.xml
+++ git/options/default.xml
@@ -108,6 +108,7 @@
 				<auto-popup type="bool">false</auto-popup>
 				<auto-capitalize type="bool">false</auto-capitalize>
 				<auto-scroll-to-bottom type="bool">true</auto-scroll-to-bottom>
+				<caption type="QString"></caption>
 				<delete-contents-after type="QString">never</delete-contents-after>
 				<raise-chat-windows-on-new-messages type="bool">false</raise-chat-windows-on-new-messages>
 				<use-chat-says-style type="bool">false</use-chat-says-style>
--- git.orig/src/chatdlg.cpp
+++ git/src/chatdlg.cpp
@@ -646,7 +646,10 @@ void ChatDlg::setSelfDestruct(int minute
 
 QString ChatDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 
 	if (pending_ > 0) {
 		cap += "* ";
--- git.orig/src/groupchatdlg.cpp
+++ git/src/groupchatdlg.cpp
@@ -1804,7 +1804,10 @@ void GCMainDlg::doAlert()
 
 QString GCMainDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 
 	if (d->pending > 0) {
 		cap += "* ";
--- git.orig/src/tabs/tabdlg.cpp
+++ git/src/tabs/tabdlg.cpp
@@ -540,7 +540,10 @@ void TabDlg::activated()
 
 QString TabDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 	uint pending = 0;
 	foreach(TabbableWidget* tab, tabs_) {
 		pending += tab->unreadMessageCount();
