Index: psi/options/default.xml
===================================================================
--- psi.orig/options/default.xml	2013-01-15 14:04:30.384827198 +0400
+++ psi/options/default.xml	2013-01-15 14:04:30.380827198 +0400
@@ -108,6 +108,7 @@
 				<auto-popup type="bool">false</auto-popup>
 				<auto-capitalize type="bool">false</auto-capitalize>
 				<auto-scroll-to-bottom type="bool">true</auto-scroll-to-bottom>
+				<caption type="QString"></caption>
 				<delete-contents-after type="QString">never</delete-contents-after>
 				<raise-chat-windows-on-new-messages type="bool">false</raise-chat-windows-on-new-messages>
 				<use-chat-says-style type="bool">false</use-chat-says-style>
Index: psi/src/chatdlg.cpp
===================================================================
--- psi.orig/src/chatdlg.cpp	2013-01-15 14:04:30.384827198 +0400
+++ psi/src/chatdlg.cpp	2013-01-15 14:04:30.380827198 +0400
@@ -647,7 +647,10 @@
 
 QString ChatDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 
 	if (pending_ > 0) {
 		cap += "* ";
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:30.384827198 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:30.380827198 +0400
@@ -1804,7 +1804,10 @@
 
 QString GCMainDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 
 	if (d->pending > 0) {
 		cap += "* ";
Index: psi/src/tabs/tabdlg.cpp
===================================================================
--- psi.orig/src/tabs/tabdlg.cpp	2013-01-15 14:04:30.384827198 +0400
+++ psi/src/tabs/tabdlg.cpp	2013-01-15 14:04:30.380827198 +0400
@@ -555,7 +555,10 @@
 
 QString TabDlg::desiredCaption() const
 {
-	QString cap = "";
+	QString cap = PsiOptions::instance()->getOption("options.ui.chat.caption").toString();
+	if(!cap.isEmpty()) {
+		return cap;
+	}
 	uint pending = 0;
 	foreach(TabbableWidget* tab, tabs_) {
 		pending += tab->unreadMessageCount();
