Index: psi/src/mucjoindlg.cpp
===================================================================
--- psi.orig/src/mucjoindlg.cpp	2013-01-15 14:04:31.312827236 +0400
+++ psi/src/mucjoindlg.cpp	2013-01-15 14:04:31.308827236 +0400
@@ -32,8 +32,12 @@
 #include "bookmarkmanager.h"
 #include "iconset.h"
 
+static const int nickConflictCode = 409;
+static const QString additionalSymbol = "_";
+
 MUCJoinDlg::MUCJoinDlg(PsiCon* psi, PsiAccount* pa)
 	: QDialog(0)
+	, nickAlreadyCompleted_(false)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
 #if QT_VERSION >= 0x040500
@@ -243,22 +247,32 @@
 	controller_->recentGCAdd(jid_.full());
 	ui_.busy->stop();
 
+	nickAlreadyCompleted_ = false;
 	closeDialogs(this);
 	deleteLater();
 	account_->addMucItem(jid_.bare());
 }
 
-void MUCJoinDlg::error(int, const QString &str)
+void MUCJoinDlg::error(int error, const QString &str)
 {
-	if(!isVisible())
-		show();
-
 	ui_.busy->stop();
 	setWidgetsEnabled(true);
 
 	account_->dialogUnregister(this);
 	controller_->dialogRegister(this);
 
+	if(!nickAlreadyCompleted_ && reason_ == MucAutoJoin && error == nickConflictCode) {
+		nickAlreadyCompleted_ = true;
+		ui_.le_nick->setText(ui_.le_nick->text() + additionalSymbol);
+		doJoin(reason_);
+		return;
+	}
+
+	if(!isVisible())
+		show();
+
+	nickAlreadyCompleted_ = false;
+
 	QMessageBox* msg = new QMessageBox(QMessageBox::Information, tr("Error"), tr("Unable to join groupchat.\nReason: %1").arg(str), QMessageBox::Ok, this);
 	msg->setAttribute(Qt::WA_DeleteOnClose, true);
 	msg->setModal(false);
@@ -271,7 +285,7 @@
 	ui_.le_room->setText(mucJid.node());
 }
 
-void MUCJoinDlg::setNick(const QString nick)
+void MUCJoinDlg::setNick(const QString& nick)
 {
 	ui_.le_nick->setText(nick);
 }
Index: psi/src/mucjoindlg.h
===================================================================
--- psi.orig/src/mucjoindlg.h	2013-01-15 14:04:31.312827236 +0400
+++ psi/src/mucjoindlg.h	2013-01-15 14:04:31.308827236 +0400
@@ -46,7 +46,7 @@
 	~MUCJoinDlg();
 
 	void setJid(const XMPP::Jid& jid);
-	void setNick(const QString nick);
+	void setNick(const QString& nick);
 	void setPassword(const QString& password);
 
 	void joined();
@@ -78,6 +78,7 @@
 	QPushButton* joinButton_;
 	XMPP::Jid jid_;
 	MucJoinReason reason_;
+	bool nickAlreadyCompleted_;
 
 	void disableWidgets();
 	void enableWidgets();
