--- psi.orig/src/infodlg.cpp
+++ psi/src/infodlg.cpp
@@ -820,7 +820,7 @@ void InfoDlg::entityTimeFinished()
 {
 	EntityTimeTask *j = (EntityTimeTask *)sender();
 	if(j->success()) {
-		foreach(UserListItem* u, d->pa->findRelevant(j->jid())) {
+		foreach(UserListItem* u, d->findRelevant(j->jid())) {
 			UserResourceList::Iterator rit = u->userResourceList().find(j->jid().resource());
 			bool found = (rit == u->userResourceList().end()) ? false: true;
 			if(!found)
