--- psi.orig/src/contactlistmodel.cpp
+++ psi/src/contactlistmodel.cpp
@@ -482,6 +482,9 @@ QVariant ContactListModel::contactData(c
 	if (role == JidRole) {
 		return QVariant(contact->jid().full());
 	}
+	else if (role == Qt::ToolTipRole) {
+		return QVariant(contact->userListItem().makeTip(true, false));
+	}
 	else if (role == PictureRole) {
 		return QVariant(contact->picture());
 	}
--- psi.orig/src/psicontactlistmodel.cpp
+++ psi/src/psicontactlistmodel.cpp
@@ -24,7 +24,6 @@
 
 #include "psicontact.h"
 #include "psiaccount.h"
-#include "userlist.h"
 #include "contactlistgroup.h"
 #include "contactlistaccountgroup.h"
 
@@ -49,10 +48,7 @@ bool PsiContactListModel::setData(const 
 
 QVariant PsiContactListModel::contactData(const PsiContact* contact, int role) const
 {
-	if (role == Qt::ToolTipRole) {
-		return QVariant(contact->userListItem().makeTip(true, false));
-	}
-	else if (role == ContactListModel::PhaseRole) {
+	if (role == ContactListModel::PhaseRole) {
 		return QVariant(secondPhase_);
 	}
 
