Index: psi/src/discodlg.cpp
===================================================================
--- psi.orig/src/discodlg.cpp	2013-01-15 14:04:31.900827260 +0400
+++ psi/src/discodlg.cpp	2013-01-15 14:04:31.896827259 +0400
@@ -919,7 +919,7 @@
 	IconAction *actBrowse, *actBack, *actForward, *actRefresh, *actStop;
 
 	// custom actions, that will be added to toolbar and context menu
-	IconAction *actRegister, *actSearch, *actJoin, *actAHCommand, *actVCard, *actAdd, *actQueryVersion;
+	IconAction *actRegister, *actUnregister, *actSearch, *actJoin, *actAHCommand, *actVCard, *actAdd, *actQueryVersion;
 
 	typedef QList<History*> HistoryList;
 	HistoryList backHistory, forwardHistory;
@@ -1024,6 +1024,9 @@
 	actRegister = new IconAction (tr("Register"), "psi/register", tr("&Register"), 0, dlg);
 	connect (actRegister, SIGNAL(triggered()), sm, SLOT(map()));
 	sm->setMapping(actRegister, Features::FID_Register);
+	actUnregister = new IconAction (tr("Unregister"), "psi/cancel", tr("&Unregister"), 0, dlg);
+	connect (actUnregister, SIGNAL(triggered()), sm, SLOT(map()));
+	sm->setMapping(actUnregister, Features::FID_Gateway);
 	actSearch = new IconAction (tr("Search"), "psi/search", tr("&Search"), 0, dlg);
 	connect (actSearch, SIGNAL(triggered()), sm, SLOT(map()));
 	sm->setMapping(actSearch, Features::FID_Search);
@@ -1061,6 +1064,7 @@
 	// custom actions
 	toolBar->addSeparator();
 	toolBar->addAction(actRegister);
+	toolBar->addAction(actUnregister);
 	toolBar->addAction(actSearch);
 	toolBar->addAction(actJoin);
 
@@ -1237,6 +1241,7 @@
 	// custom actions
 	Features f = it.features();
 	actRegister->setEnabled( f.canRegister() );
+	actUnregister->setEnabled( f.canRegister() );
 	actSearch->setEnabled( f.canSearch() );
 	actJoin->setEnabled( f.canGroupchat() );
 	actAdd->setEnabled( itemSelected );
@@ -1364,6 +1369,7 @@
 			// custom actions
 			p.addSeparator();
 			actRegister->addTo(&p);
+			actUnregister->addTo(&p);
 			actSearch->addTo(&p);
 			actJoin->addTo(&p);
 
@@ -1372,7 +1378,7 @@
 			actVCard->addTo(&p);
 			actAHCommand->addTo(&p);
 			actQueryVersion->addTo(&p);
-			
+
 			// popup with all available features
 			QMenu *fm = new QMenu(&p);
 			QHash<QAction*, int> actions;
Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:31.900827260 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:31.896827259 +0400
@@ -3397,8 +3397,12 @@
 		actionExecuteCommand(jid, node);
 	else if ( f.canDisco() )
 		actionDisco(jid, node);
-	else if ( f.isGateway() )
-		; // TODO
+	else if ( f.isGateway() ) {
+		if(QMessageBox::question(0, tr("Unregister from %1").arg(jid.bare()),
+					 tr("Are you sure?"),
+					 QMessageBox::Yes | QMessageBox::No) == QMessageBox::Yes)
+			actionUnregister(jid);
+	}
 	else if ( f.haveVCard() )
 		actionInfo(jid);
 	else if ( f.id() == Features::FID_Add ) {
@@ -4433,6 +4437,13 @@
 	}
 }
 
+void PsiAccount::actionUnregister(const Jid &j)
+{
+	JT_UnRegister *ju = new JT_UnRegister(d->client->rootTask());
+	ju->unreg(j);
+	ju->go(true);
+}
+
 void PsiAccount::actionSearch(const Jid &j)
 {
 	if(!checkConnected())
@@ -4799,9 +4810,7 @@
 
 		// if it looks like a transport, unregister (but not if it is the server!!)
 		if(u->isTransport() && !Jid(d->client->host()).compare(u->jid())) {
-			JT_UnRegister *ju = new JT_UnRegister(d->client->rootTask());
-			ju->unreg(j);
-			ju->go(true);
+			actionUnregister(j);
 		}
 	}
 }
Index: psi/src/psiaccount.h
===================================================================
--- psi.orig/src/psiaccount.h	2013-01-15 14:04:31.900827260 +0400
+++ psi/src/psiaccount.h	2013-01-15 14:04:31.896827259 +0400
@@ -363,6 +363,7 @@
 	void actionGroupsSet(const Jid &, const QStringList &);
 	void actionHistoryBox(PsiEvent *);
 	void actionRegister(const Jid &);
+	void actionUnregister(const Jid &);
 	void actionSearch(const Jid &);
 	void actionManageBookmarks();
 	void actionJoin(const Jid& mucJid, const QString& password = QString());
