Index: psi/options/default.xml
===================================================================
--- psi.orig/options/default.xml	2013-01-15 14:04:32.148827270 +0400
+++ psi/options/default.xml	2013-01-15 14:04:32.140827269 +0400
@@ -679,6 +679,7 @@
 		</p2p>
 		<service-discovery>
 			<enable-entity-capabilities type="bool">true</enable-entity-capabilities>
+			<last-activity type="bool">true</last-activity>
 		</service-discovery>
 		<status>
 			<ask-for-message-on-offline type="bool">false</ask-for-message-on-offline>
Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:32.148827270 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:32.144827270 +0400
@@ -318,6 +318,48 @@
 	return false;
 }
 
+
+//----------------------------------------------------------------------------
+// IdleServer
+//----------------------------------------------------------------------------
+class IdleServer : public XMPP::Task
+{
+	Q_OBJECT
+public:
+	IdleServer(PsiAccount *pa, Task *parent)
+		: Task(parent)
+		, pa_(pa)
+	{
+	}
+
+	~IdleServer()
+	{
+	}
+
+	bool take(const QDomElement &e)
+	{
+		if (e.tagName() != "iq" || e.attribute("type") != "get")
+			return false;
+
+		const QString ns = queryNS(e, "query");
+		if (ns == "jabber:iq:last") {
+			QDomElement iq = createIQ(doc(), "result", e.attribute("from"), e.attribute("id"));
+			QDomElement query = doc()->createElement("query");
+			query.setAttribute("xmlns", ns);
+			iq.appendChild(query);
+			query.setAttribute("seconds", pa_->psi()->idle());
+
+			send(iq);
+			return true;
+		}
+		return false;
+	}
+
+private:
+	PsiAccount *pa_;
+};
+
+
 //----------------------------------------------------------------------------
 // PsiAccount
 //----------------------------------------------------------------------------
@@ -1200,6 +1242,12 @@
 	PluginManager::instance()->addAccount(this, d->client);
 #endif
 
+	//Idle server
+	if(PsiOptions::instance()->getOption("options.service-discovery.last-activity").toBool()) {
+		new IdleServer(this, d->client->rootTask());
+		d->client->addExtension("last-act", QStringList("jabber:iq:last"));
+	}
+
 	// HTML
 	if(PsiOptions::instance()->getOption("options.html.chat.render").toBool())
 		d->client->addExtension("html",Features("http://jabber.org/protocol/xhtml-im"));
Index: psi/src/psicon.cpp
===================================================================
--- psi.orig/src/psicon.cpp	2013-01-15 14:04:32.148827270 +0400
+++ psi/src/psicon.cpp	2013-01-15 14:04:32.144827270 +0400
@@ -309,7 +309,7 @@
 
 	struct IdleSettings
 	{
-		IdleSettings()
+		IdleSettings() : secondsIdle(0)
 		{}
 
 		void update()
@@ -322,10 +322,13 @@
 			notAvailableAfter = o->getOption("options.status.auto-away.not-availible-after").toInt();
 			awayAfter = o->getOption("options.status.auto-away.away-after").toInt();
 			menuXA = o->getOption("options.ui.menu.status.xa").toBool();
+			useIdleServer = o->getOption("options.service-discovery.last-activity").toBool();
 		}
 
 		bool useOffline, useNotAvailable, useAway, menuXA;
 		int offlineAfter, notAvailableAfter, awayAfter;
+		int secondsIdle;
+		bool useIdleServer;
 	};
 
 	IdleSettings idleSettings_;
@@ -1400,10 +1403,12 @@
 
 	//Idle server
 	d->idleSettings_.update();
-	if(d->idleSettings_.useAway || d->idleSettings_.useNotAvailable || d->idleSettings_.useOffline)
+	if(d->idleSettings_.useAway || d->idleSettings_.useNotAvailable || d->idleSettings_.useOffline || d->idleSettings_.useIdleServer)
 		d->idle.start();
-	else
+	else {
 		d->idle.stop();
+		d->idleSettings_.secondsIdle = 0;
+	}
 
 	if (option == "options.ui.notifications.alert-style") {
 		alertIconUpdateAlertStyle();
@@ -1921,6 +1926,7 @@
 
 void PsiCon::secondsIdle(int sec)
 {
+	d->idleSettings_.secondsIdle = sec;
 	int minutes = sec / 60;
 	PsiAccount::AutoAway aa;
 
@@ -1938,6 +1944,11 @@
 	}
 }
 
+int PsiCon::idle() const
+{
+	return d->idleSettings_.secondsIdle;
+}
+
 ContactUpdatesManager* PsiCon::contactUpdatesManager() const
 {
 	return contactUpdatesManager_;
Index: psi/src/psicon.h
===================================================================
--- psi.orig/src/psicon.h	2013-01-15 14:04:32.148827270 +0400
+++ psi/src/psicon.h	2013-01-15 14:04:32.144827270 +0400
@@ -90,6 +90,7 @@
 	QWidget *dialogFind(const char *className);
 	void dialogRegister(QWidget *w);
 	void dialogUnregister(QWidget *w);
+	int idle() const;
 
 	QMenuBar* defaultMenuBar() const;
 
