--- psi.orig/src/mucaffiliationsmodel.cpp
+++ psi/src/mucaffiliationsmodel.cpp
@@ -27,7 +27,7 @@
 using namespace XMPP;
 
 
-MUCAffiliationsModel::MUCAffiliationsModel() : QStandardItemModel(Unknown,1) 
+MUCAffiliationsModel::MUCAffiliationsModel() : QStandardItemModel(Unknown,2)
 { 
 	QFont font;
 	font.setBold(true);
@@ -37,8 +37,10 @@ MUCAffiliationsModel::MUCAffiliationsMod
 		setData(ind,QVariant(affiliationlistindexToString((AffiliationListIndex) i)));
 		setData(ind,font_variant,Qt::FontRole);
 		insertColumns(0,1,ind);
+		insertColumns(1,1,ind);
 		enabled_[(AffiliationListIndex) i] = false;
 	}
+	setHorizontalHeaderLabels(QStringList() << tr("JID") << tr("Reason"));
 }
 
 Qt::ItemFlags MUCAffiliationsModel::flags(const QModelIndex &index) const
@@ -63,9 +65,13 @@ Qt::DropActions MUCAffiliationsModel::su
 
 bool MUCAffiliationsModel::dropMimeData(const QMimeData *data, Qt::DropAction action, int row, int, const QModelIndex &parent)
 {
-	if (!data || action != Qt::MoveAction || !(data->hasFormat("application/vnd.text.list") || data->hasFormat("text/plain")))
+	if ( (parent.isValid() && parent.column() != 0)
+		|| !data || action != Qt::MoveAction
+		|| !( data->hasFormat("application/vnd.text.list") || data->hasFormat("text/plain") ) )
+	{
 		return false;
-	
+	}
+
 	// Decode the data
 	QStringList newItems;
 	int nb_rows = 0;
@@ -130,7 +136,7 @@ QMimeData* MUCAffiliationsModel::mimeDat
 	QByteArray encodedData;
 	QDataStream stream(&encodedData, QIODevice::WriteOnly);
 	foreach (QModelIndex index, indexes) {
-		if (index.isValid()) {
+		if (index.isValid() && index.column() == 0) {
 			QString text = data(index, Qt::DisplayRole).toString();
 			stream << text;
 		}
@@ -220,6 +226,7 @@ void MUCAffiliationsModel::addItems(cons
 			}
 			insertRows(row,1,list);
 			setData(index(row,0,list),QVariant(item.jid().full()));
+			setData(index(row,1,list),QVariant(item.reason()));
 			MUCItem i(MUCItem::UnknownRole,item.affiliation());
 			i.setJid(item.jid());
 			items_ += i;
--- psi.orig/src/mucaffiliationsview.cpp
+++ psi/src/mucaffiliationsview.cpp
@@ -26,11 +26,12 @@
 MUCAffiliationsView::MUCAffiliationsView(QWidget* parent) : QTreeView(parent)
 {
 	setRootIsDecorated(false);
-	header()->hide();
+	header()->setResizeMode(QHeaderView::ResizeToContents);
 	setItemsExpandable(false);
 	setDragEnabled(true);
 	setAcceptDrops(true);
 	setDropIndicatorShown(true);
+	setSelectionBehavior(QAbstractItemView::SelectRows);
 }
 
 void MUCAffiliationsView::removeCurrent()
