--- psi.orig/src/privacy/privacyrule.ui
+++ psi/src/privacy/privacyrule.ui
@@ -9,7 +9,7 @@
     <x>0</x>
     <y>0</y>
     <width>365</width>
-    <height>245</height>
+    <height>210</height>
    </rect>
   </property>
   <property name="windowTitle" >
@@ -123,6 +123,12 @@
      </item>
      <item row="0" column="2" >
       <widget class="QComboBox" name="cb_value" >
+       <property name="sizePolicy" >
+        <sizepolicy hsizetype="Preferred" vsizetype="Fixed" >
+         <horstretch>1</horstretch>
+         <verstretch>0</verstretch>
+        </sizepolicy>
+       </property>
        <property name="editable" >
         <bool>true</bool>
        </property>
@@ -138,6 +144,19 @@
     </layout>
    </item>
    <item>
+    <spacer name="verticalSpacer">
+     <property name="orientation">
+      <enum>Qt::Vertical</enum>
+      </property>
+     <property name="sizeHint" stdset="0">
+      <size>
+       <width>20</width>
+       <height>40</height>
+      </size>
+     </property>
+    </spacer>
+   </item>
+   <item>
     <layout class="QHBoxLayout" >
      <property name="margin" >
       <number>0</number>
