Index: psi/src/chatdlg.ui
===================================================================
--- psi.orig/src/chatdlg.ui	2013-01-15 14:04:32.636827290 +0400
+++ psi/src/chatdlg.ui	2013-01-15 14:04:32.632827289 +0400
@@ -372,6 +372,9 @@
            <property name="text" >
             <string>Avatar</string>
            </property>
+           <property name="alignment">
+            <set>Qt::AlignCenter</set>
+           </property>
           </widget>
          </item>
         </layout>
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:32.636827290 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:32.632827289 +0400
@@ -732,13 +732,19 @@
 			account()->avatarFactory()->getMucAvatar(jid().withResource(res)) :
 			account()->avatarFactory()->getAvatar(jid().withResource(res));
 	if (p.isNull()) {
-		ui_.avatar->hide();
-	}
-	else {
-		int size = PsiOptions::instance()->getOption("options.ui.chat.avatars.size").toInt();
-		ui_.avatar->setPixmap(p.scaled(QSize(size, size), Qt::KeepAspectRatio, Qt::SmoothTransformation));
-		ui_.avatar->show();
+		if (!PsiOptions::instance()->getOption("options.ui.contactlist.avatars.use-default-avatar").toBool()) {
+			ui_.avatar->hide();
+			return;
+		}
+		p = IconsetFactory::iconPixmap("psi/default_avatar");
 	}
+	int optSize = PsiOptions::instance()->getOption("options.ui.chat.avatars.size").toInt();
+	ui_.avatar->setFixedSize(optSize, optSize);
+	int avatarSize = p.width(); //qMax(p.width(), p.height());
+	if (avatarSize > optSize)
+		avatarSize = optSize;
+	ui_.avatar->setPixmap(p.scaled(QSize(avatarSize, avatarSize), Qt::KeepAspectRatio, Qt::SmoothTransformation));
+	ui_.avatar->show();
 }
 
 void PsiChatDlg::optionsUpdate()
