--- psi.orig/src/gcuserview.cpp
+++ psi/src/gcuserview.cpp
@@ -611,6 +611,11 @@ void GCUserView::doContextMenu(QTreeWidg
 		pm->addAction(act);
 		act->setData(5);
 	}
+
+	act = new QAction(tr("E&xecute Command"), pm);
+	pm->addAction(act);
+	act->setData(50);
+
 	pm->addSeparator();
 
 	// Kick and Ban submenus
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -2066,6 +2066,9 @@ void GCMainDlg::lv_action(const QString 
 		}
 
 	}
+	else if(x == 50) {
+		account()->actionExecuteCommandSpecific(jid().withResource(nick));
+	}
 }
 
 void GCMainDlg::contextMenuEvent(QContextMenuEvent *)
