Index: psi/src/gcuserview.cpp
===================================================================
--- psi.orig/src/gcuserview.cpp	2013-01-15 14:04:33.040827306 +0400
+++ psi/src/gcuserview.cpp	2013-01-15 14:04:33.036827306 +0400
@@ -612,6 +612,11 @@
 		pm->addAction(act);
 		act->setData(5);
 	}
+
+	act = new QAction(tr("E&xecute Command"), pm);
+	pm->addAction(act);
+	act->setData(50);
+
 	pm->addSeparator();

 	// Kick and Ban submenus
Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:33.040827306 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:33.036827306 +0400
@@ -2036,6 +2036,9 @@
 		}

 	}
+	else if(x == 50) {
+		account()->actionExecuteCommandSpecific(jid().withResource(nick));
+	}
 }

 void GCMainDlg::contextMenuEvent(QContextMenuEvent *)
