Entering 'iris'
Entering 'src/libpsi'
Index: psi/src/chatview_te.cpp
===================================================================
--- psi.orig/src/chatview_te.cpp	2013-01-15 14:04:33.636827330 +0400
+++ psi/src/chatview_te.cpp	2013-01-15 14:04:33.628827330 +0400
@@ -46,6 +46,7 @@
 ChatView::ChatView(QWidget *parent)
 	: PsiTextView(parent)
 	, isMuc_(false)
+	, isEncryptionEnabled_(false)
 	, oldTrackBarPosition(0)
 	, dialog_(0)
 {
@@ -74,12 +75,18 @@
 			logIconReceive = IconsetFactory::iconPixmap("psi/notification_chat_receive").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
 			logIconSend = IconsetFactory::iconPixmap("psi/notification_chat_send").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
 			logIconDelivered = IconsetFactory::iconPixmap("psi/notification_chat_delivery_ok").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
+			logIconReceivePgp = IconsetFactory::iconPixmap("psi/notification_chat_receive_pgp").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
+			logIconSendPgp = IconsetFactory::iconPixmap("psi/notification_chat_send_pgp").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
+			logIconDeliveredPgp = IconsetFactory::iconPixmap("psi/notification_chat_delivery_ok_pgp").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
 			logIconTime = IconsetFactory::iconPixmap("psi/notification_chat_time").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
 			logIconInfo = IconsetFactory::iconPixmap("psi/notification_chat_info").scaledToHeight(logIconsSize, Qt::SmoothTransformation);
 		} else {
 			logIconReceive = IconsetFactory::iconPixmap("psi/notification_chat_receive");
 			logIconSend = IconsetFactory::iconPixmap("psi/notification_chat_send");
 			logIconDelivered = IconsetFactory::iconPixmap("psi/notification_chat_delivery_ok");
+			logIconReceivePgp = IconsetFactory::iconPixmap("psi/notification_chat_receive_pgp");
+			logIconSendPgp = IconsetFactory::iconPixmap("psi/notification_chat_send_pgp");
+			logIconDeliveredPgp = IconsetFactory::iconPixmap("psi/notification_chat_delivery_ok_pgp");
 			logIconTime = IconsetFactory::iconPixmap("psi/notification_chat_time");
 			logIconInfo = IconsetFactory::iconPixmap("psi/notification_chat_info");
 		}
@@ -109,6 +116,11 @@
 	dialog_ = dialog;
 }

+void ChatView::setEncryptionEnabled(bool enabled)
+{
+	isEncryptionEnabled_ = enabled;
+}
+
 void ChatView::setSessionData(bool isMuc, const QString &jid, const QString name)
 {
 	isMuc_ = isMuc;
@@ -149,15 +161,18 @@
 {
 	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_receive"), logIconReceive);
 	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_send"), logIconSend);
+	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_receive_pgp"), logIconReceivePgp);
+	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_send_pgp"), logIconSendPgp);
 	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_time"), logIconTime);
 	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_info"), logIconInfo);
 	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_delivered"), logIconDelivered);
+	document()->addResource(QTextDocument::ImageResource, QUrl("icon:log_icon_delivered_pgp"), logIconDeliveredPgp);
 }

 void ChatView::markReceived(QString id)
 {
 	if (useMessageIcons_) {
-		document()->addResource(QTextDocument::ImageResource, QUrl(QString("icon:delivery") + id), logIconDelivered);
+		document()->addResource(QTextDocument::ImageResource, QUrl(QString("icon:delivery") + id), isEncryptionEnabled_? logIconDeliveredPgp : logIconDelivered);
 		setLineWrapColumnOrWidth(lineWrapColumnOrWidth());
 	}
 }
@@ -356,11 +371,14 @@
 	QString timestr = formatTimeStamp(mv.dateTime());
 	QString color = colorString(mv.isLocal(), mv.isSpooled());
 	if (useMessageIcons_ && mv.isAwaitingReceipt()) {
-		document()->addResource(QTextDocument::ImageResource, QUrl(QString("icon:delivery") + mv.messageId()), logIconSend);
+		document()->addResource(QTextDocument::ImageResource, QUrl(QString("icon:delivery") + mv.messageId()),
+					isEncryptionEnabled_ ? logIconSendPgp : logIconSend);
 	}
-	QString icon = useMessageIcons_?
+	QString icon = useMessageIcons_ ?
 		(QString("<img src=\"%1\" />").arg(mv.isLocal()?
-		(mv.isAwaitingReceipt()?QString("icon:delivery") + mv.messageId():"icon:log_icon_send"):"icon:log_icon_receive")):"";
+		(mv.isAwaitingReceipt() ? QString("icon:delivery") + mv.messageId()
+			: isEncryptionEnabled_ ? "icon:log_icon_send_pgp" : "icon:log_icon_send")
+		: isEncryptionEnabled_ ? "icon:log_icon_receive_pgp" : "icon:log_icon_receive")) : "";
 	if (mv.isEmote()) {
 		appendText(icon + QString("<span style=\"color: %1\">").arg(color) + QString("[%1]").arg(timestr) + QString(" *%1 ").arg(TextUtil::escape(mv.nick())) + mv.formattedText() + "</span>");
 	} else {
Index: psi/src/chatview_te.h
===================================================================
--- psi.orig/src/chatview_te.h	2013-01-15 14:04:33.636827330 +0400
+++ psi/src/chatview_te.h	2013-01-15 14:04:33.628827330 +0400
@@ -64,6 +64,8 @@
 	bool internalFind(QString str, bool startFromBeginning = false);
 	ChatView *textWidget();

+	void setEncryptionEnabled(bool enabled);
+
 protected:
 	// override the tab/esc behavior
 	bool focusNextPrevChild(bool next);
@@ -93,6 +95,7 @@

 private:
 	bool isMuc_;
+	bool isEncryptionEnabled_;
 	QString jid_;
 	QString name_;
 	int  oldTrackBarPosition;
@@ -102,6 +105,9 @@
 	QPixmap logIconSend;
 	QPixmap logIconReceive;
 	QPixmap logIconDelivered;
+	QPixmap logIconSendPgp;
+	QPixmap logIconReceivePgp;
+	QPixmap logIconDeliveredPgp;
 	QPixmap logIconTime;
 	QPixmap logIconInfo;
 	QAction *actQuote_;
Index: psi/src/chatview_webkit.cpp
===================================================================
--- psi.orig/src/chatview_webkit.cpp	2013-01-15 14:04:33.636827330 +0400
+++ psi/src/chatview_webkit.cpp	2013-01-15 14:04:33.628827330 +0400
@@ -137,6 +137,7 @@
 	, sessionReady_(false)
 	, dialog_(0)
 	, isMuc_(false)
+	, isEncryptionEnabled_(false)
 {
 	jsObject = new ChatViewJSObject(this);
 	webView = new WebView(this);
@@ -180,6 +181,11 @@
 	);
 }

+void ChatView::setEncryptionEnabled(bool enabled)
+{
+	isEncryptionEnabled_ = enabled;
+}
+
 void ChatView::embedJsObject()
 {
 	ChatViewTheme *theme = currentTheme();
@@ -196,6 +202,7 @@
 	QVariantMap m;
 	m["type"] = "receipt";
 	m["id"] = id;
+	m["encrypted"] = isEncryptionEnabled_;
 	sendJsObject(m);
 }

@@ -314,6 +321,7 @@
 	QVariantMap vm = mv.toVariantMap(isMuc_, true);
 	vm["mtype"] = vm["type"];
 	vm["type"] = "message";
+	vm["encrypted"] = isEncryptionEnabled_;
 	sendJsObject(vm);
 }

Index: psi/src/chatview_webkit.h
===================================================================
--- psi.orig/src/chatview_webkit.h	2013-01-15 14:04:33.636827330 +0400
+++ psi/src/chatview_webkit.h	2013-01-15 14:04:33.632827330 +0400
@@ -67,6 +67,8 @@
 	bool internalFind(QString str, bool startFromBeginning = false);
 	WebView * textWidget();

+	void setEncryptionEnabled(bool enabled);
+
 protected:
 	// override the tab/esc behavior
 	bool focusNextPrevChild(bool next);
@@ -98,6 +100,7 @@
 	bool sessionReady_;
 	QPointer<QWidget> dialog_;
 	bool isMuc_;
+	bool isEncryptionEnabled_;
 	QString jid_;
 	QString name_;
 	PsiAccount *account_;
Index: psi/src/psichatdlg.cpp
===================================================================
--- psi.orig/src/psichatdlg.cpp	2013-01-15 14:04:33.636827330 +0400
+++ psi/src/psichatdlg.cpp	2013-01-15 14:04:33.632827330 +0400
@@ -782,6 +782,7 @@
 	ui_.tb_pgp->setVisible(account()->hasPGP() &&
 						   !smallChat_ &&
 						   !PsiOptions::instance()->getOption("options.ui.chat.central-toolbar").toBool());
+	ui_.log->setEncryptionEnabled(isEncryptionEnabled());
 }

 void PsiChatDlg::doClearButton()
@@ -811,6 +812,7 @@
 void PsiChatDlg::setPGPEnabled(bool enabled)
 {
 	act_pgp_->setChecked(enabled);
+	ui_.log->setEncryptionEnabled(enabled);
 }

 void PsiChatDlg::toggleSmallChat()
Index: psi/themes/chatview/psi/classic/index.html
===================================================================
--- psi.orig/themes/chatview/psi/classic/index.html	2013-01-15 14:04:33.636827330 +0400
+++ psi/themes/chatview/psi/classic/index.html	2013-01-15 14:04:33.632827330 +0400
@@ -93,7 +93,7 @@
 				var img = document.getElementById("receipt"+shared.cdata.id);
 				if (img) {
 					var sizeUri = messageIconsSize?"?w="+messageIconsSize+"&h="+messageIconsSize:"";
-					img.src = "icon:psi/notification_chat_delivery_ok"+sizeUri;
+					img.src = (shared.cdata.encrypted?"icon:psi/notification_chat_delivery_ok_pgp":"icon:psi/notification_chat_delivery_ok")+sizeUri;
 				}
 			}
 			if (shared.cdata.mtype == "message") {
@@ -130,6 +130,9 @@
 							icon = shared.cdata.local?(shared.cdata.awaitingReceipt?
 								"psi/notification_chat_send":"psi/notification_chat_delivery_ok")
 								: "psi/notification_chat_receive";
+							if(shared.cdata.encrypted) {
+								icon += "_pgp"
+							}
 							break;
 						case "system":
 						case "subject":
