--- git.orig/src/mainwin_p.cpp
+++ git/src/mainwin_p.cpp
@@ -626,7 +626,8 @@ void EventNotifierAction::hide()
 			int found = 0;
 			foreach(QWidget* widget, toolBar->findChildren<QWidget*>()) {
 				if (!widget->objectName().startsWith("qt_") &&
-				    !QString(widget->metaObject()->className()).startsWith("QToolBar"))
+				    !QString(widget->metaObject()->className()).startsWith("QToolBar") &&
+				    !QString(widget->metaObject()->className()).startsWith("QMenu"))
 				{
 					found++;
 				}
