--- git.orig/src/mainwin_p.cpp
+++ git/src/mainwin_p.cpp
@@ -626,7 +626,9 @@ void EventNotifierAction::hide()
 			int found = 0;
 			foreach(QWidget* widget, toolBar->findChildren<QWidget*>()) {
 				if (!widget->objectName().startsWith("qt_") &&
-				    !QString(widget->metaObject()->className()).startsWith("QToolBar"))
+				    !QString(widget->metaObject()->className()).startsWith("QToolBar") &&
+				    !QString(widget->metaObject()->className()).startsWith("QMenu") &&
+				    QString(widget->metaObject()->className()) != "Oxygen::TransitionWidget") // dirty hack
 				{
 					found++;
 				}
