--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -62,6 +62,7 @@
 		</muc> 
 		<pgp comment="PGP encryption options">
 			<enable comment="Enable PGP globally" type="bool">true</enable>
+			<auto-start comment="Automatically enable PGP encryption" type="bool">false</auto-start>
 		</pgp>
 		<ui comment="User interface options">
 			<account comment="Account options">
--- psi.orig/src/psichatdlg.cpp
+++ psi/src/psichatdlg.cpp
@@ -770,6 +770,12 @@ void PsiChatDlg::updatePGP()
 {
 	if (account()->hasPGP()) {
 		act_pgp_->setEnabled(true);
+		if(PsiOptions::instance()->getOption("options.pgp.auto-start").toBool()) {
+			UserListItem *item = account()->findFirstRelevant(jid());
+			if(item && !item->publicKeyID().isEmpty()) {
+				setPGPEnabled(true);
+			}
+		}
 	}
 	else {
 		act_pgp_->setChecked(false);
