--- psi.orig/src/psicontactlistview.cpp
+++ psi/src/psicontactlistview.cpp
@@ -25,12 +25,17 @@
 #include <QApplication>
 #include <QDesktopWidget>
 #include <QTimer>
+#include <QFileInfo>
 
 #include "psicontactlistviewdelegate.h"
 #include "psitooltip.h"
 #include "psioptions.h"
 #include "contactlistmodel.h"
 #include "contactlistproxymodel.h"
+#include "contactlistgroup.h"
+#include "contactlistitemproxy.h"
+#include "psicontact.h"
+#include "psiaccount.h"
 
 static const int recalculateTimerTimeout = 2000;
 
@@ -172,6 +177,74 @@ void PsiContactListView::showToolTip(con
 	PsiToolTip::showText(globalPos, text, this);
 }
 
+bool PsiContactListView::acceptableDragOperation(QDropEvent *e)
+{
+	ContactListItemProxy* ip = itemProxy(indexAt(e->pos()));
+	if(ip) {
+		PsiContact *pc = dynamic_cast<PsiContact*>(ip->item());
+		if(pc) {
+			foreach(const QUrl& url, e->mimeData()->urls()) {
+				const QFileInfo fi(url.toLocalFile());
+				if (!fi.isDir() && fi.exists()) {
+					return true;
+				}
+			}
+		}
+	}
+
+	return false;
+}
+
+void PsiContactListView::dragEnterEvent(QDragEnterEvent *e)
+{
+	if(acceptableDragOperation(e)) {
+		setCurrentIndex(indexAt(e->pos()));
+		e->acceptProposedAction();
+		return;
+	}
+
+	ContactListDragView::dragEnterEvent(e);
+}
+
+void PsiContactListView::dragMoveEvent(QDragMoveEvent *e)
+{
+	if(acceptableDragOperation(e)) {
+		setCurrentIndex(indexAt(e->pos()));
+		e->acceptProposedAction();
+		return;
+	}
+
+	ContactListDragView::dragMoveEvent(e);
+}
+
+void PsiContactListView::dropEvent(QDropEvent *e)
+{
+	ContactListItemProxy* ip = itemProxy(indexAt(e->pos()));
+	if(ip) {
+		PsiContact *pc = dynamic_cast<PsiContact*>(ip->item());
+		if(pc) {
+			QStringList files;
+			foreach(const QUrl& url, e->mimeData()->urls()) {
+				const QFileInfo fi(url.toLocalFile());
+				if (!fi.isDir() && fi.exists()) {
+					const QString fileName = QFileInfo(fi.isSymLink() ?
+									    fi.symLinkTarget() : fi.absoluteFilePath()
+									    ).canonicalFilePath();
+					files.append(fileName);
+				}
+			}
+
+			if(!files.isEmpty()) {
+				e->acceptProposedAction();
+				pc->account()->sendFiles(pc->jid(), files, true);
+				return;
+			}
+		}
+	}
+
+	ContactListDragView::dropEvent(e);
+}
+
 void PsiContactListView::setModel(QAbstractItemModel* model)
 {
 	ContactListDragView::setModel(model);
--- psi.orig/src/psicontactlistview.h
+++ psi/src/psicontactlistview.h
@@ -49,9 +49,16 @@ protected:
 	virtual void showToolTip(const QModelIndex& index, const QPoint& globalPos) const;
 	virtual void doItemsLayoutStart();
 
+	virtual void dragEnterEvent(QDragEnterEvent *e);
+	virtual void dropEvent(QDropEvent *e);
+	virtual void dragMoveEvent(QDragMoveEvent *e);
+
 	PsiContactListViewDelegate* itemDelegate() const;
 
 private:
+	bool acceptableDragOperation(QDropEvent *e);
+
+private:
 	class Private;
 	Private* d;
 };
