--- git.orig/src/eventdlg.cpp
+++ git/src/eventdlg.cpp
@@ -591,7 +591,8 @@ EventDlg::EventDlg(const QString &to, Ps
 
 	setAccount(pa);
 
-	d->whois = new QTimer;
+	d->whois = new QTimer(d);
+	d->whois->setSingleShot(true);
 	connect(d->whois, SIGNAL(timeout()), SLOT(doWhois()));
 
 	init();
@@ -673,7 +674,6 @@ EventDlg::EventDlg(const Jid &j, PsiAcco
 EventDlg::~EventDlg()
 {
 	if(d->composing) {
-		delete d->whois;
 		d->psi->dialogUnregister(this);
 	}
 	else {
