--- git.orig/src/psicontactmenu.cpp
+++ git/src/psicontactmenu.cpp
@@ -506,9 +506,6 @@ private slots:
 			return;
 		}
 
-		renameAction_->setEnabled(contact_->isEditable());
-		removeAction_->setEnabled(contact_->removeAvailable());
-
 #ifdef YAPSI
 		YaProfile* profile = YaProfile::create(contact_->account(), contact_->jid());
 		openHistoryAction_->setEnabled(contact_->historyAvailable());
@@ -574,8 +571,8 @@ private slots:
 		authMenu_->setEnabled(contact_->account()->isAvailable());
 		updateBlockActionState();
 		visibleAction_->setChecked(contact_->isAlwaysVisible());
-		removeAction_->setVisible(!PsiOptions::instance()->getOption("options.ui.contactlist.lockdown-roster").toBool());
-		removeAction_->setEnabled(contact_->account()->isAvailable());// || !contact_->inList());
+		removeAction_->setVisible(!PsiOptions::instance()->getOption("options.ui.contactlist.lockdown-roster").toBool()  && !contact_->isSelf());
+		removeAction_->setEnabled(contact_->removeAvailable());
 		if (!PsiOptions::instance()->getOption("options.ui.menu.contact.custom-picture").toBool()) {
 			pictureMenu_->menuAction()->setVisible(false);
 		}
