Index: psi/src/groupchatdlg.cpp
===================================================================
--- psi.orig/src/groupchatdlg.cpp	2013-01-15 14:04:34.760827376 +0400
+++ psi/src/groupchatdlg.cpp	2013-01-15 14:04:34.756827376 +0400
@@ -1734,7 +1734,9 @@
 			if (!m.spooled() && !isActiveTab() && !m.from().resource().isEmpty()) {
 				XMPP::Jid jid = m.from()/*.withDomain("")*/;
 				MessageEvent *e = new MessageEvent(m, account());
-				account()->psi()->popupManager()->doPopup(account(), PopupManager::AlertGcHighlight, jid, m.from().resource(), 0, (PsiEvent *)e);
+				UserListItem i;
+				i.setPrivate(true);
+				account()->psi()->popupManager()->doPopup(account(), PopupManager::AlertGcHighlight, jid, m.from().resource(), &i, (PsiEvent *)e);
 				e->deleteLater();
 			}
 		}
Index: psi/src/psidbusnotifier.cpp
===================================================================
--- psi.orig/src/psidbusnotifier.cpp	2013-01-15 14:04:34.760827376 +0400
+++ psi/src/psidbusnotifier.cpp	2013-01-15 14:04:34.756827376 +0400
@@ -208,7 +208,11 @@
 //	}
 	QImage im;
 	if(account) {
-		im = account->avatarFactory()->getAvatar(jid.bare()).toImage();
+		if(uli && uli->isPrivate())
+			im = account->avatarFactory()->getMucAvatar(jid).toImage();
+		else
+			im = account->avatarFactory()->getAvatar(jid).toImage();
+
 		if(!im.isNull()) {
 			int size = PsiOptions::instance()->getOption("options.ui.notifications.passive-popups.avatar-size").toInt();
 			im = im.scaledToWidth(size, Qt::SmoothTransformation);
Index: psi/src/psigrowlnotifier.cpp
===================================================================
--- psi.orig/src/psigrowlnotifier.cpp	2013-01-15 14:04:34.760827376 +0400
+++ psi/src/psigrowlnotifier.cpp	2013-01-15 14:04:34.756827376 +0400
@@ -126,8 +126,12 @@
 		if (((int)statusMsg.length()) > len)
 			statusMsg = statusMsg.left(len) + "...";
 	QPixmap icon;
-	if(account)
-		icon = account->avatarFactory()->getAvatar(jid.bare());
+	if(account) {
+		if(uli && uli->isPrivate())
+			icon = account->avatarFactory()->getMucAvatar(jid);
+		else
+			icon = account->avatarFactory()->getAvatar(jid);
+	}

 	if (uli) {
 		contact = uli->name();
Index: psi/src/psipopup.cpp
===================================================================
--- psi.orig/src/psipopup.cpp	2013-01-15 14:04:34.760827376 +0400
+++ psi/src/psipopup.cpp	2013-01-15 14:04:34.756827376 +0400
@@ -340,7 +340,13 @@
 			}
 		}
 	}
-	QPixmap avatar = d->account->avatarFactory()->getAvatar(jid);
+	QPixmap avatar;
+	if(d->account) {
+		if(u && u->isPrivate())
+			avatar = d->account->avatarFactory()->getMucAvatar(j);
+		else
+			avatar = d->account->avatarFactory()->getAvatar(j);
+	}
 	// show popup
 	if ( d->popupType != PopupManager::AlertComposing && d->popupType != PopupManager::AlertHeadline &&
 	     (d->popupType != PopupManager::AlertFile || !PsiOptions::instance()->getOption("options.ui.file-transfer.auto-popup").toBool()) )
