--- git.orig/options/default.xml
+++ git/options/default.xml
@@ -463,6 +463,7 @@ QLineEdit#le_status_text {
 				</popup-dialogs>
 				<sounds>
 					<chat-message type="QString">sound/chat2.wav</chat-message>
+					<groupchat-message type="QString">sound/chat2.wav</groupchat-message>
 					<completed-file-transfer type="QString">sound/ft_complete.wav</completed-file-transfer>
 					<contact-offline type="QString">sound/offline.wav</contact-offline>
 					<contact-online type="QString">sound/online.wav</contact-online>
--- git.orig/src/groupchatdlg.cpp
+++ git/src/groupchatdlg.cpp
@@ -1729,7 +1729,7 @@ void GCMainDlg::message(const Message &_
 	}
 	else if(account()->status().type() != Status::DND) {
 		if(alert || (PsiOptions::instance()->getOption("options.ui.notifications.sounds.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) )
-			account()->playSound(PsiAccount::eChat2);
+			account()->playSound(PsiAccount::eGroupChat);
 		if(alert || (PsiOptions::instance()->getOption("options.ui.notifications.passive-popups.notify-every-muc-message").toBool() && !m.spooled() && !from.isEmpty()) )
 			if (!m.spooled() && !isActiveTab() && !m.from().resource().isEmpty()) {
 				XMPP::Jid jid = m.from().withDomain("");
--- git.orig/src/options/opt_sound.cpp
+++ git/src/options/opt_sound.cpp
@@ -51,7 +51,7 @@ QWidget *OptionsTabSound::widget()
 	w = new OptSoundUI();
 	OptSoundUI *d = (OptSoundUI *)w;
 
-	sounds_ << d->le_oeMessage << d->le_oeChat1 << d->le_oeChat2 << d->le_oeHeadline << d->le_oeSystem << d->le_oeOnline << d->le_oeOffline << d->le_oeSend << d->le_oeIncomingFT << d->le_oeFTComplete;
+	sounds_ << d->le_oeMessage << d->le_oeChat1 << d->le_oeChat2 << d->le_oeGroupChat << d->le_oeHeadline << d->le_oeSystem << d->le_oeOnline << d->le_oeOffline << d->le_oeSend << d->le_oeIncomingFT << d->le_oeFTComplete;
 
 	bg_se = new QButtonGroup;
 	bg_se->addButton(d->tb_seMessage);
@@ -60,6 +60,8 @@ QWidget *OptionsTabSound::widget()
 	modify_buttons_[d->tb_seChat1] = d->le_oeChat1;
 	bg_se->addButton(d->tb_seChat2);
 	modify_buttons_[d->tb_seChat2] = d->le_oeChat2;
+	bg_se->addButton(d->tb_seGroupChat);
+	modify_buttons_[d->tb_seGroupChat] = d->le_oeGroupChat;
 	bg_se->addButton(d->tb_seHeadline);
 	modify_buttons_[d->tb_seHeadline] = d->le_oeHeadline;
 	bg_se->addButton(d->tb_seSystem);
@@ -83,6 +85,8 @@ QWidget *OptionsTabSound::widget()
 	play_buttons_[d->tb_seChat1Play] = d->le_oeChat1;
 	bg_sePlay->addButton(d->tb_seChat2Play);
 	play_buttons_[d->tb_seChat2Play] = d->le_oeChat2;
+	bg_sePlay->addButton(d->tb_seGroupChatPlay);
+	play_buttons_[d->tb_seGroupChatPlay] = d->le_oeGroupChat;
 	bg_sePlay->addButton(d->tb_seHeadlinePlay);
 	play_buttons_[d->tb_seHeadlinePlay] = d->le_oeHeadline;
 	bg_sePlay->addButton(d->tb_seSystemPlay);
@@ -103,7 +107,7 @@ QWidget *OptionsTabSound::widget()
 
 	// set up proper tool button icons
 	int n;
-	for (n = 0; n < 10; n++) {
+	for (n = 0; n < 11; n++) {
 		IconToolButton *tb = (IconToolButton *)bg_se->buttons()[n];
 		tb->setPsiIcon( IconsetFactory::iconPtr("psi/browse") );
 		tb = (IconToolButton *)bg_sePlay->buttons()[n];
@@ -144,6 +148,7 @@ void OptionsTabSound::applyOptions()
 	PsiOptions::instance()->setOption("options.ui.notifications.sounds.incoming-message", d->le_oeMessage->text());
 	PsiOptions::instance()->setOption("options.ui.notifications.sounds.new-chat", d->le_oeChat1->text());
 	PsiOptions::instance()->setOption("options.ui.notifications.sounds.chat-message", d->le_oeChat2->text());
+	PsiOptions::instance()->setOption("options.ui.notifications.sounds.groupchat-message", d->le_oeGroupChat->text());
 	PsiOptions::instance()->setOption("options.ui.notifications.sounds.system-message", d->le_oeSystem->text());
 	PsiOptions::instance()->setOption("options.ui.notifications.sounds.incoming-headline", d->le_oeHeadline->text());
 	PsiOptions::instance()->setOption("options.ui.notifications.sounds.contact-online", d->le_oeOnline->text());
@@ -175,6 +180,7 @@ void OptionsTabSound::restoreOptions()
 	d->le_oeMessage->setText(PsiOptions::instance()->getOption("options.ui.notifications.sounds.incoming-message").toString());
 	d->le_oeChat1->setText(PsiOptions::instance()->getOption("options.ui.notifications.sounds.new-chat").toString());
 	d->le_oeChat2->setText(PsiOptions::instance()->getOption("options.ui.notifications.sounds.chat-message").toString());
+	d->le_oeGroupChat->setText(PsiOptions::instance()->getOption("options.ui.notifications.sounds.groupchat-message").toString());
 	d->le_oeSystem->setText(PsiOptions::instance()->getOption("options.ui.notifications.sounds.system-message").toString());
 	d->le_oeHeadline->setText(PsiOptions::instance()->getOption("options.ui.notifications.sounds.incoming-headline").toString());
 	d->le_oeOnline->setText(PsiOptions::instance()->getOption("options.ui.notifications.sounds.contact-online").toString());
@@ -212,6 +218,7 @@ void OptionsTabSound::soundReset()
 	d->le_oeMessage->setText("sound/chat2.wav");
 	d->le_oeChat1->setText("sound/chat1.wav");
 	d->le_oeChat2->setText("sound/chat2.wav");
+	d->le_oeGroupChat->setText("sound/chat2.wav");
 	d->le_oeSystem->setText("sound/chat2.wav");
 	d->le_oeHeadline->setText("sound/chat2.wav");
 	d->le_oeOnline->setText("sound/online.wav");
--- git.orig/src/options/opt_sound.ui
+++ git/src/options/opt_sound.ui
@@ -77,286 +77,314 @@
      <property name="spacing" >
       <number>6</number>
      </property>
-     <item row="9" column="3" >
+     <item row="10" column="3">
       <widget class="IconToolButton" name="tb_seFTCompletePlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="9" column="1" >
+     <item row="10" column="1">
       <widget class="QLineEdit" name="le_oeFTComplete" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="3" column="0" >
+     <item row="4" column="0">
       <widget class="QLabel" name="TextLabel_Headline" >
        <property name="text" >
         <string>Headline:</string>
        </property>
       </widget>
      </item>
-     <item row="1" column="1" >
+     <item row="1" column="1">
       <widget class="QLineEdit" name="le_oeChat1" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="5" column="0" >
+     <item row="6" column="0">
       <widget class="QLabel" name="TextLabel4_4_2" >
        <property name="text" >
         <string>Receive online status:</string>
        </property>
       </widget>
      </item>
-     <item row="5" column="1" >
+     <item row="6" column="1">
       <widget class="QLineEdit" name="le_oeOnline" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="6" column="2" >
+     <item row="7" column="2">
       <widget class="IconToolButton" name="tb_seOffline" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="0" column="2" >
+     <item row="0" column="2">
       <widget class="IconToolButton" name="tb_seMessage" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="8" column="1" >
+     <item row="9" column="1">
       <widget class="QLineEdit" name="le_oeIncomingFT" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="0" column="0" >
+     <item row="0" column="0">
       <widget class="QLabel" name="TextLabel4_6" >
        <property name="text" >
         <string>Receive message:</string>
        </property>
       </widget>
      </item>
-     <item row="7" column="2" >
+     <item row="8" column="2">
       <widget class="IconToolButton" name="tb_seSend" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="6" column="3" >
+     <item row="7" column="3">
       <widget class="IconToolButton" name="tb_seOfflinePlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="5" column="2" >
+     <item row="6" column="2">
       <widget class="IconToolButton" name="tb_seOnline" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="4" column="2" >
+     <item row="5" column="2">
       <widget class="IconToolButton" name="tb_seSystem" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="0" column="1" >
+     <item row="0" column="1">
       <widget class="QLineEdit" name="le_oeMessage" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="8" column="3" >
+     <item row="9" column="3">
       <widget class="IconToolButton" name="tb_seIncomingFTPlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="7" column="0" >
+     <item row="8" column="0">
       <widget class="QLabel" name="TextLabel4_5_2" >
        <property name="text" >
         <string>Send message:</string>
        </property>
       </widget>
      </item>
-     <item row="3" column="3" >
+     <item row="4" column="3">
       <widget class="IconToolButton" name="tb_seHeadlinePlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="3" column="1" >
+     <item row="4" column="1">
       <widget class="QLineEdit" name="le_oeHeadline" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="4" column="0" >
+     <item row="5" column="0">
       <widget class="QLabel" name="TextLabel4_3_2" >
        <property name="text" >
         <string>System message:</string>
        </property>
       </widget>
      </item>
-     <item row="8" column="2" >
+     <item row="9" column="2">
       <widget class="IconToolButton" name="tb_seIncomingFT" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="5" column="3" >
+     <item row="6" column="3">
       <widget class="IconToolButton" name="tb_seOnlinePlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="4" column="3" >
+     <item row="5" column="3">
       <widget class="IconToolButton" name="tb_seSystemPlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="2" column="3" >
+     <item row="2" column="3">
       <widget class="IconToolButton" name="tb_seChat2Play" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="3" column="2" >
+     <item row="4" column="2">
       <widget class="IconToolButton" name="tb_seHeadline" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="2" column="2" >
+     <item row="2" column="2">
       <widget class="IconToolButton" name="tb_seChat2" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="1" column="3" >
+     <item row="1" column="3">
       <widget class="IconToolButton" name="tb_seChat1Play" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="1" column="2" >
+     <item row="1" column="2">
       <widget class="IconToolButton" name="tb_seChat1" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="2" column="1" >
+     <item row="2" column="1">
       <widget class="QLineEdit" name="le_oeChat2" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="8" column="0" >
+     <item row="9" column="0">
       <widget class="QLabel" name="textLabel1" >
        <property name="text" >
         <string>Incoming file transfer:</string>
        </property>
       </widget>
      </item>
-     <item row="9" column="0" >
+     <item row="10" column="0">
       <widget class="QLabel" name="textLabel2" >
        <property name="text" >
         <string>File transfer complete:</string>
        </property>
       </widget>
      </item>
-     <item row="6" column="1" >
+     <item row="7" column="1">
       <widget class="QLineEdit" name="le_oeOffline" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="7" column="1" >
+     <item row="8" column="1">
       <widget class="QLineEdit" name="le_oeSend" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="7" column="3" >
+     <item row="8" column="3">
       <widget class="IconToolButton" name="tb_seSendPlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="4" column="1" >
+     <item row="5" column="1">
       <widget class="QLineEdit" name="le_oeSystem" >
        <property name="toolTip" >
         <string>Enter a filename or !beep for a system beep</string>
        </property>
       </widget>
      </item>
-     <item row="2" column="0" >
+     <item row="2" column="0">
       <widget class="QLabel" name="TextLabel4_2_3" >
        <property name="text" >
         <string>Receive next chat:</string>
        </property>
       </widget>
      </item>
-     <item row="0" column="3" >
+     <item row="0" column="3">
       <widget class="IconToolButton" name="tb_seMessagePlay" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="1" column="0" >
+     <item row="1" column="0">
       <widget class="QLabel" name="TextLabel4_2_2_2" >
        <property name="text" >
         <string>Receive first chat:</string>
        </property>
       </widget>
      </item>
-     <item row="9" column="2" >
+     <item row="10" column="2">
       <widget class="IconToolButton" name="tb_seFTComplete" >
        <property name="text" >
         <string/>
        </property>
       </widget>
      </item>
-     <item row="6" column="0" >
+     <item row="7" column="0">
       <widget class="QLabel" name="TextLabel4_4_2_2" >
        <property name="text" >
         <string>Receive offline status:</string>
        </property>
       </widget>
      </item>
+     <item row="3" column="0">
+      <widget class="QLabel" name="TextLabel4_4_1">
+       <property name="text">
+        <string>Receive MUC message</string>
+       </property>
+      </widget>
+     </item>
+     <item row="3" column="1">
+      <widget class="QLineEdit" name="le_oeGroupChat">
+       <property name="toolTip">
+        <string>Enter a filename or !beep for a system beep</string>
+       </property>
+      </widget>
+     </item>
+     <item row="3" column="2">
+      <widget class="IconToolButton" name="tb_seGroupChat">
+       <property name="text">
+        <string/>
+       </property>
+      </widget>
+     </item>
+     <item row="3" column="3">
+      <widget class="IconToolButton" name="tb_seGroupChatPlay" >
+       <property name="text" >
+        <string/>
+       </property>
+      </widget>
+     </item>
     </layout>
    </item>
    <item>
@@ -408,6 +436,9 @@
   <tabstop>le_oeChat2</tabstop>
   <tabstop>tb_seChat2</tabstop>
   <tabstop>tb_seChat2Play</tabstop>
+  <tabstop>le_oeGroupChat</tabstop>
+  <tabstop>tb_seGroupChat</tabstop>
+  <tabstop>tb_seGroupChatPlay</tabstop>
   <tabstop>le_oeHeadline</tabstop>
   <tabstop>tb_seHeadline</tabstop>
   <tabstop>tb_seHeadlinePlay</tabstop>
--- git.orig/src/psiaccount.cpp
+++ git/src/psiaccount.cpp
@@ -3162,6 +3162,9 @@ void PsiAccount::playSound(PsiAccount::S
 	case eChat2:
 		str = PsiOptions::instance()->getOption("options.ui.notifications.sounds.chat-message").toString();
 		break;
+	case eGroupChat:
+		str = PsiOptions::instance()->getOption("options.ui.notifications.sounds.groupchat-message").toString();
+		break;
 	case eHeadline:
 		str = PsiOptions::instance()->getOption("options.ui.notifications.sounds.incoming-headline").toString();
 		break;
--- git.orig/src/psiaccount.h
+++ git/src/psiaccount.h
@@ -221,6 +221,7 @@ public:
 		eMessage = 0,
 		eChat1,
 		eChat2,
+		eGroupChat,
 		eHeadline,
 		eSystem,
 		eOnline,
