Index: psi/src/libpsi/tools/atomicxmlfile/atomicxmlfile.cpp
===================================================================
--- psi.orig/src/libpsi/tools/atomicxmlfile/atomicxmlfile.cpp	2013-01-15 14:04:35.016827386 +0400
+++ psi/src/libpsi/tools/atomicxmlfile/atomicxmlfile.cpp	2013-01-15 14:04:35.012827386 +0400
@@ -61,16 +61,21 @@
 bool AtomicXmlFile::saveDocument(const QDomDocument& doc, QString fileName) const
 {
 	QFile file(fileName);
-	if (!file.open(QIODevice::WriteOnly)) {
+	if (!file.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
 		return false;
 	}

-	QTextStream text;
-	text.setDevice(&file);
-	text.setCodec("UTF-8");
-	text << doc.toString();
+	QTextStream *text = new QTextStream(&file);
+	text->setCodec("UTF-8");
+	*text << doc.toString();
+	delete text;

-	return file.error() == QFile::NoError;
+	bool res = (file.error() == QFile::NoError);
+	if (res)
+		res = file.flush();
+	file.close();
+
+	return res;
 }

 bool AtomicXmlFile::loadDocument(QDomDocument* doc, QString fileName) const
Index: psi/src/libpsi/tools/atomicxmlfile/atomicxmlfile.h
===================================================================
--- psi.orig/src/libpsi/tools/atomicxmlfile/atomicxmlfile.h	2013-01-15 14:04:35.016827386 +0400
+++ psi/src/libpsi/tools/atomicxmlfile/atomicxmlfile.h	2013-01-15 14:04:35.012827386 +0400
@@ -60,10 +60,9 @@
 			return false;
 		}

-		if (QFile::exists(backupFileName()))
-			QFile::remove(backupFileName());
-
 		if (QFile::exists(fileName_)) {
+			if (QFile::exists(backupFileName()))
+				QFile::remove(backupFileName());
 			if (!QFile::rename(fileName_, backupFileName())) {
 				qWarning("AtomicXmlFile::saveDocument(): Unable to rename '%s' to '%s'.",
 				         qPrintable(fileName_), qPrintable(backupFileName()));
@@ -91,6 +90,10 @@
 			if (loadDocument(reader, fileName)) {
 				return true;
 			}
+			if (QFile::exists(fileName)) {
+				// The file exists, but it is incorrect. Remove it. Otherwise, enter the backup.
+				QFile::remove(fileName);
+			}
 		}

 		return false;
