--- git.temp.orig/src/libpsi/tools/spellchecker/spellhighlighter.cpp
+++ git.temp/src/libpsi/tools/spellchecker/spellhighlighter.cpp
@@ -21,12 +21,14 @@ void SpellHighlighter::highlightBlock(co
 
 	// Match words (minimally)
 	QRegExp expression("\\b\\w+\\b");
+	QRegExp digit("\\d+");
 
 	// Iterate through all words
 	int index = text.indexOf(expression);
 	while (index >= 0) {
 		int length = expression.matchedLength();
-		if (!SpellChecker::instance()->isCorrect(expression.cap()))
+		QString word = expression.cap();
+		if (!digit.exactMatch(word) && !SpellChecker::instance()->isCorrect(word))
 			setFormat(index, length, tcf);
 		index = text.indexOf(expression, index + length);
 	}
--- git.temp.orig/src/msgmle.cpp
+++ git.temp/src/msgmle.cpp
@@ -348,7 +348,7 @@ void ChatEdit::contextMenuEvent(QContext
 		tc.movePosition(QTextCursor::StartOfWord, QTextCursor::MoveAnchor);
 		tc.movePosition(QTextCursor::EndOfWord, QTextCursor::KeepAnchor);
 		QString selected_word = tc.selectedText();
-		if (!selected_word.isEmpty() && !SpellChecker::instance()->isCorrect(selected_word)) {
+		if (!selected_word.isEmpty() && !QRegExp("\\d+").exactMatch(selected_word) && !SpellChecker::instance()->isCorrect(selected_word)) {
 			QList<QString> suggestions = SpellChecker::instance()->suggestions(selected_word);
 			if (!suggestions.isEmpty() || SpellChecker::instance()->writable()) {
 				QMenu spell_menu;
