--- psi.orig/src/libpsi/dialogs/grepshortcutkeydialog.cpp
+++ psi/src/libpsi/dialogs/grepshortcutkeydialog.cpp
@@ -25,6 +25,7 @@ GrepShortcutKeyDialog::GrepShortcutKeyDi
 	: QDialog()
 	, gotKey(false)
 {
+	setAttribute(Qt::WA_DeleteOnClose);
 	ui_.setupUi(this);
 	displayPressedKeys(QKeySequence());
 }
@@ -36,6 +37,7 @@ void GrepShortcutKeyDialog::show()
 {
 	QDialog::show();
 	grabKeyboard();
+	setFocus();
 }
 
 /**
