Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:36.240827436 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:36.240827436 +0400
@@ -448,6 +448,11 @@
 		reconnectTimeoutTimer_ = new QTimer(this);
 		reconnectTimeoutTimer_->setSingleShot(true);
 		connect(reconnectTimeoutTimer_, SIGNAL(timeout()), SLOT(reconnectTimerTimeout()));
+
+		updateOnlineContactsCountTimer_ = new QTimer(this);
+		updateOnlineContactsCountTimer_->setInterval(500);
+		updateOnlineContactsCountTimer_->setSingleShot(true);
+		connect(updateOnlineContactsCountTimer_, SIGNAL(timeout()), SLOT(updateOnlineContactsCountTimeout()));
 	}
 
 	PsiContactList* contactList;
@@ -476,6 +481,7 @@
 	bool pepAvailable;
 	QString currentConnectionError;
 	int currentConnectionErrorCondition;
+	QTimer *updateOnlineContactsCountTimer_;
 
 	// Tune
 	Tune lastTune;
@@ -642,7 +648,8 @@
 		return pathToProfile(activeProfile, ApplicationInfo::DataLocation) + "/events-" + JIDUtil::encode(acc.id).toLower() + ".xml";
 	}
 
-	void updateOnlineContactsCount()
+private slots:
+	void updateOnlineContactsCountTimeout()
 	{
 		int newOnlineContactsCount = 0;
 		foreach(const PsiContact* c, contacts) {
@@ -658,6 +665,12 @@
 		}
 	}
 
+public:
+	void updateOnlineContactsCount()
+	{
+		updateOnlineContactsCountTimer_->start();
+	}
+
 	// FIXME: Rename updateEntry -> updateContact
 	void updateEntry(const UserListItem& u)
 	{
