--- psi.orig/src/widgets/urlobject.cpp
+++ psi/src/widgets/urlobject.cpp
@@ -20,6 +20,7 @@
 
 #include "urlobject.h"
 #include "psioptions.h"
+#include "psiiconset.h"
 
 #include <QApplication>
 #include <QClipboard>
@@ -69,10 +70,12 @@ public:
 
 		tr = qApp->translate("URLLabel", "Send message to");
 		act_send_message = new IconAction(tr, "psi/message", tr, 0, this);
+		((PsiIcon*)act_send_message->psiIcon())->removeAnim(false);
 		connectXmppAction(act_send_message, "message");
 		
 		tr = qApp->translate("URLLabel", "Chat with");
 		act_chat = new IconAction(tr, "psi/chat", tr, 0, this);
+		((PsiIcon*)act_chat->psiIcon())->removeAnim(false);
 		connectXmppAction(act_chat, "message;type=chat");
 
 		tr = qApp->translate("URLLabel", "Join groupchat");
