--- psi.orig/src/psicon.cpp
+++ psi/src/psicon.cpp
@@ -1677,20 +1677,24 @@ bool PsiCon::filterEvent(const PsiAccoun
 
 void PsiCon::processEvent(PsiEvent *e, ActivationType activationType)
 {
+	if ( !e->account() ) {
+		delete e;
+		return;
+	}
+
 	if ( e->type() == PsiEvent::PGP ) {
 		e->account()->eventQueue()->dequeue(e);
 		e->account()->queueChanged();
+		delete e;
 		return;
 	}
 
-	if ( !e->account() )
-		return;
-
 	UserListItem *u = e->account()->find(e->jid());
 	if ( !u ) {
 		qWarning("SYSTEM MESSAGE: Bug #1. Contact the developers and tell them what you did to make this message appear. Thank you.");
 		e->account()->eventQueue()->dequeue(e);
 		e->account()->queueChanged();
+		delete e;
 		return;
 	}
 
@@ -1728,6 +1732,7 @@ void PsiCon::processEvent(PsiEvent *e, A
 				delete sess;
 			}
 		}
+		delete e;
 		return;
 	}
 
