Index: psi/src/psiaccount.cpp
===================================================================
--- psi.orig/src/psiaccount.cpp	2013-01-15 14:04:36.428827444 +0400
+++ psi/src/psiaccount.cpp	2013-01-15 14:04:36.424827444 +0400
@@ -4657,7 +4657,7 @@

 	if (!nm.body().isEmpty()) {
 		UserListItem *u = findFirstRelevant(m.to());
-		if (!u || (u->subscription().type() != Subscription::Both && u->subscription().type() != Subscription::From)) {
+		if (!u || (!u->isConference() && u->subscription().type() != Subscription::Both && u->subscription().type() != Subscription::From)) {
 			nm.setNick(nick());
 		}
 	}
