--- git.temp.orig/src/psiaccount.cpp
+++ git.temp/src/psiaccount.cpp
@@ -4656,7 +4656,7 @@ void PsiAccount::dj_sendMessage(const Me
 
 	if (!nm.body().isEmpty()) {
 		UserListItem *u = findFirstRelevant(m.to());
-		if (!u || (u->subscription().type() != Subscription::Both && u->subscription().type() != Subscription::From)) {
+		if (!u || (!u->isConference() && u->subscription().type() != Subscription::Both && u->subscription().type() != Subscription::From)) {
 			nm.setNick(nick());
 		}
 	}
