Index: psi/src/widgets/psirichtext.cpp
===================================================================
--- psi.orig/src/widgets/psirichtext.cpp	2013-01-15 14:04:36.656827453 +0400
+++ psi/src/widgets/psirichtext.cpp	2013-01-15 14:04:36.652827453 +0400
@@ -323,7 +323,7 @@
 			else {
 				// go here when  scheme in ["", "file"] and its not resource
 				QString path = QFileInfo(imgSrcUrl.scheme() == "file"?
-							   imgSrcUrl.toLocalFile() : imgSrc).canonicalFilePath();
+							   imgSrcUrl.toLocalFile() : imgSrc).absoluteFilePath();
 				bool baseDirFound = false;
 				foreach (const QString &baseDir, allowedImageDirs) {
 					if (path.startsWith(baseDir)) {
