--- psi.orig/src/userlist.cpp
+++ psi/src/userlist.cpp
@@ -23,6 +23,8 @@
 #include <QList>
 #include <QtCrypto>
 #include <QTextDocument> // for Qt::escape()
+#include <QBuffer>
+#include <QUrl>
 
 #include "userlist.h"
 #include "avatars.h"
@@ -848,33 +850,27 @@ QString UserListItem::makeBareTip(bool t
 			else
 				name = QObject::tr("[blank]");
 
-			int status = makeSTATUS(r.status());
-			QString istr = "status/offline";
-			if(status == STATUS_ONLINE)
-				istr = "status/online";
-			else if(status == STATUS_AWAY)
-				istr = "status/away";
-			else if(status == STATUS_XA)
-				istr = "status/xa";
-			else if(status == STATUS_DND)
-				istr = "status/dnd";
-			else if(status == STATUS_CHAT)
-				istr = "status/chat";
-			else if(status == STATUS_INVISIBLE)
-				istr = "status/invisible"; //this shouldn't happen
-
 			QString secstr;
 			if(isSecure(r.name()) && PsiOptions::instance()->getOption("options.ui.contactlist.tooltip.pgp").toBool())
 				secstr += QString(" <%1=\"psi/cryptoYes\">").arg(imgTag);
 			QString hr;
 			if (!mucItem)
 				hr = "<hr/>";
-			if(r.status().mucItem().jid().isEmpty())
-				str += hr + QString("<div style='white-space:pre'><%1=\"%2\"> ").arg(imgTag).arg(istr) + QString("<b>%1</b> ").arg(Qt::escape(name)) \
-				+ QString("(%1)").arg(r.priority()) + secstr + "</div>";
-			else
-				str += hr +  QString("<div style='white-space:pre'><%1=\"%2\"> ").arg(imgTag).arg(istr) + QString("<b>%1</b> ").arg(Qt::escape(name)) \
-				+ QString("(%1) &lt;%2&gt;").arg(r.priority()).arg(JIDUtil::toString(r.status().mucItem().jid(),true)) + secstr + "</div>";
+			str += hr + "<div style='white-space:pre'>";
+
+			PsiIcon *statusIcon = PsiIconset::instance()->statusPtr(jid(), makeSTATUS(r.status()));
+			if (statusIcon) {
+				QByteArray imageArray;
+				QBuffer buff(&imageArray);
+				statusIcon->image().save(&buff, "png");
+				QString imgBase64(QUrl::toPercentEncoding(imageArray.toBase64()));
+				str += QString("<img src=\"data:image/png;base64,%1\" alt=\"img\"/>").arg(imgBase64);
+			}
+
+			str += QString(" <b>%1</b> ").arg(Qt::escape(name)) + QString("(%1)").arg(r.priority());
+			if (!r.status().mucItem().jid().isEmpty())
+				str += QString(" &lt;%1&gt;").arg(JIDUtil::toString(r.status().mucItem().jid(),true));
+			str += secstr + "</div>";
 
 			if(!r.publicKeyID().isEmpty() && PsiOptions::instance()->getOption("options.ui.contactlist.tooltip.pgp").toBool()) {
 				int v = r.pgpVerifyStatus();
