--- git.temp.orig/iris/src/irisnet/corelib/objectsession.cpp
+++ git.temp/iris/src/irisnet/corelib/objectsession.cpp
@@ -148,6 +148,8 @@ public:
 		callTrigger->disconnect(this);
 		callTrigger->setParent(0);
 		callTrigger->deleteLater();
+		qDeleteAll(pendingCalls);
+		pendingCalls.clear();
 	}
 
 	void addPendingCall(MethodCall *call)
@@ -240,6 +242,7 @@ void ObjectSession::reset()
 	d->invalidateWatchers();
 	if(d->callTrigger->isActive())
 		d->callTrigger->stop();
+	qDeleteAll(d->pendingCalls);
 	d->pendingCalls.clear();
 }
 
--- git.temp.orig/iris/src/irisnet/noncore/cutestuff/bsocket.cpp
+++ git.temp/iris/src/irisnet/noncore/cutestuff/bsocket.cpp
@@ -137,6 +137,8 @@ BSocket::BSocket(QObject *parent)
 BSocket::~BSocket()
 {
 	resetConnection(true);
+	if (d->resolver)
+		delete d->resolver;
 	delete d;
 }
 
--- git.temp.orig/iris/src/xmpp/xmpp-core/securestream.cpp
+++ git.temp/iris/src/xmpp/xmpp-core/securestream.cpp
@@ -382,6 +382,12 @@ public:
 		}
 		return false;
 	}
+
+	void deleteLayers()
+	{
+		qDeleteAll(layers);
+		layers.clear();
+	}
 };
 
 SecureStream::SecureStream(ByteStream *s)
@@ -401,6 +407,7 @@ SecureStream::SecureStream(ByteStream *s
 
 SecureStream::~SecureStream()
 {
+	d->deleteLayers();
 	delete d;
 }
 
@@ -573,9 +580,7 @@ void SecureStream::layer_tlsClosed(const
 {
 	setOpenMode(QIODevice::NotOpen);
 	d->active = false;
-	while (!d->layers.isEmpty()) {
-		delete d->layers.takeFirst();
-	}
+	d->deleteLayers();
 	tlsClosed();
 }
 
@@ -628,9 +633,7 @@ void SecureStream::layer_error(int x)
 	d->errorCode = x;
 	setOpenMode(QIODevice::NotOpen);
 	d->active = false;
-	while (!d->layers.isEmpty()) {
-		delete d->layers.takeFirst();
-	}
+	d->deleteLayers();
 	if(type == SecureLayer::TLS)
 		setError(ErrTLS);
 	else if(type == SecureLayer::SASL)
--- git.temp.orig/src/contactlistnestedgroup.cpp
+++ git.temp/src/contactlistnestedgroup.cpp
@@ -47,6 +47,20 @@ CL_DEBUG("~ContactListNextedGroup(%x): %
 	qDeleteAll(groups_);
 	groups_.clear();
 
+	QHashIterator<ContactListGroup::SpecialType, QPointer<ContactListGroup> > it(specialGroups_);
+	while (it.hasNext())
+	{
+		it.next();
+		ContactListGroup* group = it.value().data();
+		if (group)
+		{
+CL_DEBUG("~ContactListNextedGroup(%x): %s", this, qPrintable(group->fullName()));
+			removeItem(ContactListGroup::findGroup(group));
+			delete group;
+		}
+	}
+	specialGroups_.clear();
+
 	ContactListGroup::clearGroup();
 }
 
--- git.temp.orig/src/psiaccount.cpp
+++ git.temp/src/psiaccount.cpp
@@ -1377,6 +1377,9 @@ PsiAccount::PsiAccount(const UserAccount
 PsiAccount::~PsiAccount()
 {
 	logout(true, loggedOutStatus());
+
+	setRCEnabled(false);
+
 	emit accountDestroyed();
 	// nuke all related dialogs
 	deleteAllDialogs();
@@ -1413,6 +1416,9 @@ PsiAccount::~PsiAccount()
 
 	delete d->blockTransportPopupList;
 
+	qDeleteAll(d->userList);
+	d->userList.clear();
+
 	d->contactList->unlink(this);
 	delete d;
 
--- git.temp.orig/src/psievent.cpp
+++ git.temp/src/psievent.cpp
@@ -1059,7 +1059,11 @@ void EventQueue::printContent() const
 void EventQueue::clear()
 {
 	while(!list_.isEmpty())
-		delete list_.takeFirst();
+	{
+		EventItem *i = list_.takeFirst();
+		delete i->event();
+		delete i;
+	}
 
 	emit queueChanged();
 }
@@ -1078,6 +1082,7 @@ void EventQueue::clear(const Jid &j, boo
 				GlobalEventQueue::instance()->dequeue(ei);
 			}
 			it = list_.erase(it);
+			delete e;
 			delete ei;
 			changed = true;
 		}
