diff --git a/src/systeminfo.cpp b/src/systeminfo.cpp
index 0f3fd31..f208ae0 100644
--- a/src/systeminfo.cpp
+++ b/src/systeminfo.cpp
@@ -88,6 +88,7 @@ static QString unixHeuristicDetect() {
 
 		LinuxASP, // Russian Linux distros
 		LinuxALT,
+		LinuxRFRemix,
 
 		LinuxPLD, // Polish Linux distros
 		LinuxAurox,
@@ -116,6 +117,7 @@ static QString unixHeuristicDetect() {
 		{ LinuxAurox,		OsUseName,	"/etc/aurox-release",		"Aurox Linux"		},
 		{ LinuxArch,		OsUseFile,	"/etc/arch-release",		"Arch Linux"		},
 		{ LinuxLFS,		OsAppendFile,	"/etc/lfs-release",		"LFS Linux"		},
+		{ LinuxRFRemix,		OsUseFile,	"/etc/rfremix-release",		"RFRemix Linux"		},
 
 		// untested
 		{ LinuxSuSE,		OsUseFile,	"/etc/SuSE-release",		"SuSE Linux"		},
