--- a/src/chatdlg.cpp
+++ b/src/chatdlg.cpp
@@ -935,7 +935,8 @@
 		if (PsiOptions::instance()->getOption("options.ui.chat.raise-chat-windows-on-new-messages").toBool()) {
 			if (isTabbed()) {
 				TabDlg* tabSet = getManagingTabDlg();
-				tabSet->selectTab(this);
+				if (PsiOptions::instance()->getOption("options.ui.chat.switch-tab-on-new-messages").toBool() || !tabSet->isActiveWindow())
+					tabSet->selectTab(this);
 				::bringToFront(tabSet, false);
 			}
 			else {
--- a/src/options/opt_chat.cpp
+++ b/src/options/opt_chat.cpp
@@ -74,6 +74,9 @@
 	d->ck_raiseChatWindow->setWhatsThis(
 		tr("Makes Psi bring an open chat window to the front of your screen when you receive a new message."
 		" It does not take the keyboard focus, so it will not interfere with your work."));
+	d->ck_switchTabOnMessage->setWhatsThis(
+		tr("Makes Psi switch tab on active tabbed window when you receive a new message."
+		" It does not take the keyboard focus, so it will not interfere with your work."));
 	d->ck_smallChats->setWhatsThis(
 		tr("Makes Psi open chat windows in compact mode."));
 	d->ck_tabChats->setWhatsThis(
@@ -102,6 +105,7 @@
 	PsiOptions::instance()->setOption("options.messages.default-outgoing-message-type", bg_defAct->buttons().indexOf(bg_defAct->checkedButton()) == 0 ? "message" : "chat");
 	PsiOptions::instance()->setOption("options.ui.chat.alert-for-already-open-chats", d->ck_alertOpenChats->isChecked());
 	PsiOptions::instance()->setOption("options.ui.chat.raise-chat-windows-on-new-messages", d->ck_raiseChatWindow->isChecked());
+	PsiOptions::instance()->setOption("options.ui.chat.switch-tab-on-new-messages", d->ck_switchTabOnMessage->isChecked());
 	PsiOptions::instance()->setOption("options.ui.chat.use-small-chats", d->ck_smallChats->isChecked());
 	PsiOptions::instance()->setOption("options.ui.tabs.show-tab-buttons", d->ck_showTabButtons->isChecked());
 	
@@ -179,6 +183,7 @@
 	bg_defAct->buttons()[PsiOptions::instance()->getOption("options.messages.default-outgoing-message-type").toString() == "message" ? 0 : 1]->setChecked(true);
 	d->ck_alertOpenChats->setChecked( PsiOptions::instance()->getOption("options.ui.chat.alert-for-already-open-chats").toBool() );
 	d->ck_raiseChatWindow->setChecked( PsiOptions::instance()->getOption("options.ui.chat.raise-chat-windows-on-new-messages").toBool() );
+	d->ck_switchTabOnMessage->setChecked( PsiOptions::instance()->getOption("options.ui.chat.switch-tab-on-new-messages").toBool() );
 	d->ck_smallChats->setChecked( PsiOptions::instance()->getOption("options.ui.chat.use-small-chats").toBool() );
 	d->ck_showTabButtons->setChecked( PsiOptions::instance()->getOption("options.ui.tabs.show-tab-buttons").toBool() );
 	d->ck_tabChats->setChecked( PsiOptions::instance()->getOption("options.ui.tabs.use-tabs").toBool() );
--- a/src/options/opt_chat.ui
+++ b/src/options/opt_chat.ui
@@ -41,6 +41,13 @@
     </widget>
    </item>
    <item>
+    <widget class="QCheckBox" name="ck_switchTabOnMessage" >
+     <property name="text" >
+      <string>Switch tab on receiving new messages</string>
+     </property>
+    </widget>
+   </item>
+   <item>
     <widget class="QCheckBox" name="ck_smallChats" >
      <property name="text" >
       <string>Open chats in compact mode</string>
--- a/options/default.xml
+++ b/options/default.xml
@@ -112,6 +112,7 @@
 				<caption type="QString"></caption>
 				<delete-contents-after type="QString">never</delete-contents-after>
 				<raise-chat-windows-on-new-messages type="bool">false</raise-chat-windows-on-new-messages>
+				<switch-tab-on-new-messages type="bool">false</switch-tab-on-new-messages>
 				<use-chat-says-style type="bool">false</use-chat-says-style>
 				<use-expanding-line-edit type="bool">false</use-expanding-line-edit>
 				<use-small-chats type="bool">false</use-small-chats>
