diff --git a/src/mainwin.cpp b/src/mainwin.cpp
index dafc1aa..8f94457 100644
--- a/src/mainwin.cpp
+++ b/src/mainwin.cpp
@@ -123,6 +123,11 @@ public:
 	bool onTop, asTool;
 	QMenu* mainMenu, *optionsMenu, *toolsMenu;
 	GlobalStatusMenu *statusMenu;
+#ifdef Q_OS_LINUX
+	// Status menu for MenuBar.
+	// Workaround a Unity bug.
+	GlobalStatusMenu *statusMenuMB;
+#endif
 	int sbState;
 	QString nickname;
 	PsiTrayIcon* tray;
@@ -461,6 +466,16 @@ MainWin::MainWin(bool _onTop, bool _asTool, PsiCon* psi)
 	connect(d->statusMenu, SIGNAL(statusSelected(XMPP::Status::Type, bool)), d->psi, SLOT(statusMenuChanged(XMPP::Status::Type, bool)));
 	connect(d->statusMenu, SIGNAL(statusPresetSelected(XMPP::Status,bool,bool)), d->psi, SLOT(setGlobalStatus(XMPP::Status,bool,bool)));
 	connect(d->statusMenu, SIGNAL(statusPresetDialogForced(const QString &)), d->psi, SLOT(showStatusDialog(const QString &)));
+
+#ifdef Q_OS_LINUX
+	d->statusMenuMB = new GlobalStatusMenu((QWidget*)this, d->psi);
+	d->statusMenuMB->setTitle(tr("Status"));
+	d->statusMenuMB->setObjectName("statusMenu");
+	connect(d->statusMenuMB, SIGNAL(statusSelected(XMPP::Status::Type, bool)), d->psi, SLOT(statusMenuChanged(XMPP::Status::Type, bool)));
+	connect(d->statusMenuMB, SIGNAL(statusPresetSelected(XMPP::Status,bool,bool)), d->psi, SLOT(setGlobalStatus(XMPP::Status,bool,bool)));
+	connect(d->statusMenuMB, SIGNAL(statusPresetDialogForced(const QString &)), d->psi, SLOT(showStatusDialog(const QString &)));
+#endif
+
 	d->optionsMenu = new QMenu(tr("General"), this);
 
 	d->optionsMenu->setObjectName("optionsMenu");
@@ -475,7 +490,10 @@ MainWin::MainWin(bool _onTop, bool _asTool, PsiCon* psi)
 #endif
 
 
-	buildStatusMenu();
+	buildStatusMenu(d->statusMenu);
+#ifdef Q_OS_LINUX
+	buildStatusMenu(d->statusMenuMB);
+#endif
 	buildOptionsMenu();
 	connect(d->optionsMenu, SIGNAL(aboutToShow()), SLOT(buildOptionsMenu()));
 
@@ -515,7 +533,11 @@ MainWin::MainWin(bool _onTop, bool _asTool, PsiCon* psi)
 	mainMenuBar()->addMenu(d->optionsMenu);
 #endif
 
+#ifdef Q_OS_LINUX
+	mainMenuBar()->addMenu(d->statusMenuMB);
+#else
 	mainMenuBar()->addMenu(d->statusMenu);
+#endif
 
 	mainMenuBar()->addMenu(viewMenu);
 	d->getAction("show_offline")->addTo(viewMenu);
@@ -870,11 +892,20 @@ void MainWin::setUseDock(bool use)
 
 void MainWin::buildStatusMenu()
 {
-	d->statusMenu->clear();
-	d->statusMenu->fill();
+	GlobalStatusMenu *statusMenu = qobject_cast<GlobalStatusMenu*>(sender());
+	Q_ASSERT(statusMenu);
+	if (statusMenu) {
+		buildStatusMenu(statusMenu);
+	}
+}
+
+void MainWin::buildStatusMenu(GlobalStatusMenu *statusMenu)
+{
+	statusMenu->clear();
+	statusMenu->fill();
 #ifdef USE_PEP
-	d->statusMenu->addSeparator();
-	d->getAction("publish_tune")->addTo(d->statusMenu);
+	statusMenu->addSeparator();
+	d->getAction("publish_tune")->addTo(statusMenu);
 #endif
 }
 
@@ -1440,6 +1471,9 @@ void MainWin::decorateButton(int status)
 			d->rosterAvatar->setStatusIcon(PsiIconset::instance()->statusPtr(STATUS_OFFLINE)->icon());
 		}
 
+#ifdef Q_OS_LINUX
+		d->statusMenuMB->statusChanged(makeStatus(STATUS_OFFLINE, ""));
+#endif
 		d->statusMenu->statusChanged(makeStatus(STATUS_OFFLINE, ""));
 
 		setWindowIcon(PsiIconset::instance()->status(STATUS_OFFLINE).impix());
@@ -1449,8 +1483,10 @@ void MainWin::decorateButton(int status)
 		d->statusButton->setIcon(PsiIconset::instance()->statusPtr(status));
 		d->statusSmallerAlt->setPsiIcon(PsiIconset::instance()->statusPtr(status));
 		d->rosterAvatar->setStatusIcon(PsiIconset::instance()->statusPtr(status)->icon());
+#ifdef Q_OS_LINUX
+		d->statusMenuMB->statusChanged(makeStatus(status, d->psi->currentStatusMessage()));
+#endif
 		d->statusMenu->statusChanged(makeStatus(status, d->psi->currentStatusMessage()));
-
 		setWindowIcon(PsiIconset::instance()->status(status).impix());
 	}
 
@@ -1635,8 +1671,10 @@ void MainWin::optionsUpdate()
 
 	setWindowOpacity(double(qMax(MINIMUM_OPACITY,PsiOptions::instance()->getOption("options.ui.contactlist.opacity").toInt()))/100);
 
-	buildStatusMenu();
-
+	buildStatusMenu(d->statusMenu);
+#ifdef Q_OS_LINUX
+	buildStatusMenu(d->statusMenuMB);
+#endif
 	updateTray();
 }
 
diff --git a/src/mainwin.h b/src/mainwin.h
index b4b0edf..20a21dc 100644
--- a/src/mainwin.h
+++ b/src/mainwin.h
@@ -45,6 +45,8 @@ class IconAction;
 class PsiIcon;
 class ContactView;
 class PsiTrayIcon;
+class GlobalStatusMenu;
+
 namespace XMPP {
 	class Status;
 }
@@ -196,6 +198,8 @@ private:
 	void saveToolbarsState();
 	void loadToolbarsState();
 
+	void buildStatusMenu(GlobalStatusMenu *statusMenu);
+
 private:
 	class Private;
 	Private *d;
