--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -547,6 +547,13 @@ public slots:
 public:
 
 	bool eventFilter( QObject *obj, QEvent *ev ) {
+		if (obj == te_log()->textWidget() || obj->parent() == te_log()->textWidget())
+		{
+			if (ev->type() == QEvent::MouseButtonPress)
+				mle()->setFocus();
+			return QObject::eventFilter(obj, ev);
+		}
+
 		if (te_log()->handleCopyEvent(obj, ev, mle()))
 			return true;
 
@@ -878,6 +885,7 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	X11WM_CLASS("groupchat");
 
 	ui_.mle->chatEdit()->setFocus();
+	logTextWidget()->installEventFilter(d);
 
 	// Connect signals from MUC manager
 	connect(d->mucManager,SIGNAL(action_error(MUCManager::Action, int, const QString&)), SLOT(action_error(MUCManager::Action, int, const QString&)));
@@ -1015,6 +1023,19 @@ void GCMainDlg::setShortcuts()
 	d->act_minimize->setShortcuts(ShortcutManager::instance()->shortcuts("chat.minimize"));
 }
 
+QWidget *GCMainDlg::logTextWidget()
+{
+#ifdef WEBKIT
+	return ui_.log->textWidget();
+#else
+	ChatView *logWid = ui_.log->textWidget();
+	QWidget *child = logWid->childAt(logWid->frameRect().center());
+	if (child)
+		return child;
+	return logWid;
+#endif
+}
+
 void GCMainDlg::scrollUp()
 {
 	ui_.log->scrollUp();
--- psi.orig/src/groupchatdlg.h
+++ psi/src/groupchatdlg.h
@@ -88,6 +88,7 @@ public:
 
 protected:
 	void setShortcuts();
+	QWidget *logTextWidget();
 
 	// reimplemented
 	void resizeEvent(QResizeEvent *e);
