diff --git a/src/common.cpp b/src/common.cpp
index 001676d..3eb3638 100644
--- a/src/common.cpp
+++ b/src/common.cpp
@@ -459,15 +459,18 @@ void reorderGridLayout(QGridLayout* layout, int maxCols)
 
 void x11wmClass(Display *dsp, WId wid, QString resName)
 {
-	const char *app_name = ApplicationInfo::sname().latin1();
-
 	//Display *dsp = x11Display();				 // get the display
 	//WId win = winId();						   // get the window
 	XClassHint classhint;						  // class hints
+	// Get old class hint. It is important to save old class name
+	XGetClassHint(dsp, wid, &classhint);
+	XFree(classhint.res_name);
+
 	const QByteArray latinResName = resName.toLatin1();
 	classhint.res_name = (char *)latinResName.data(); // res_name
-	classhint.res_class = const_cast<char*>(app_name);				// res_class
 	XSetClassHint(dsp, wid, &classhint);		   // set the class hints
+
+	XFree(classhint.res_class);
 }
 
 //>>>-- Nathaniel Gray -- Caltech Computer Science ------>
