--- psi.orig/src/contactlistdragview.cpp
+++ psi/src/contactlistdragview.cpp
@@ -352,7 +352,7 @@ static void acceptDropAction(QDropEvent*
 {
 	if (e->keyboardModifiers() == Qt::NoModifier) {
 		if (e->dropAction() != Qt::MoveAction) {
-			qWarning("acceptDropAction(): e->dropAction() != Qt::MoveAction");
+			//qWarning("acceptDropAction(): e->dropAction() != Qt::MoveAction");
 			return;
 		}
 		Q_ASSERT(e->dropAction() == Qt::MoveAction);
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -733,6 +733,7 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 	if (options->getOption("options.ui.contactlist.disable-scrollbar").toBool() ) {
 		ui_.lv_users->setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
 	}
+	ui_.lv_users->model()->setSupportedDragActions(Qt::CopyAction);
 	connect(ui_.lv_users, SIGNAL(action(const QString &, const Status &, int)), SLOT(lv_action(const QString &, const Status &, int)));
 	connect(ui_.lv_users, SIGNAL(insertNick(const QString&)), d, SLOT(insertNick(const QString&)));
 
