--- psi.orig/src/chatdlg.cpp
+++ psi/src/chatdlg.cpp
@@ -182,6 +182,7 @@ void ChatDlg::doTrackBar()
 void ChatDlg::initActions()
 {
 	act_send_ = new QAction(this);
+	act_send_->setShortcutContext(Qt::WidgetWithChildrenShortcut);
 	addAction(act_send_);
 	connect(act_send_, SIGNAL(triggered()), SLOT(doSend()));
 
--- psi.orig/src/groupchatdlg.cpp
+++ psi/src/groupchatdlg.cpp
@@ -834,6 +834,7 @@ GCMainDlg::GCMainDlg(PsiAccount *pa, con
 
 	// Common actions
 	d->act_send = new QAction(this);
+	d->act_send->setShortcutContext(Qt::WidgetWithChildrenShortcut);
 	addAction(d->act_send);
 	connect(d->act_send,SIGNAL(triggered()), SLOT(mle_returnPressed()));
 	ui_.pb_send->setIcon(IconsetFactory::icon("psi/action_button_send").icon());
