Entering 'iris'
Entering 'src/libpsi'
Index: psi/src/mainwin.cpp
===================================================================
--- psi.orig/src/mainwin.cpp	2013-01-15 14:04:38.876827543 +0400
+++ psi/src/mainwin.cpp	2013-01-15 14:04:38.872827543 +0400
@@ -1060,7 +1060,7 @@
 
 void MainWin::actOnlineForumActivated ()
 {
-	DesktopUtil::openUrl("http://forum.psi-im.org");
+	DesktopUtil::openUrl("http://forum.psi-plus.com");
 }
 
 void MainWin::actJoinPsiMUCActivated()
@@ -1070,12 +1070,12 @@
 		return;
 	}
 
-	account->actionJoin("psi@conference.psi-im.org");
+	account->actionJoin("psi-dev@conference.jabber.ru");
 }
 
 void MainWin::actBugReportActivated ()
 {
-	DesktopUtil::openUrl("http://forum.psi-im.org/forum/2");
+	DesktopUtil::openUrl("http://code.google.com/p/psi-dev/wiki/issues?tm=3");
 }
 
 void MainWin::actAboutActivated ()
Index: psi/src/psiactionlist.cpp
===================================================================
--- psi.orig/src/psiactionlist.cpp	2013-01-15 14:04:38.876827543 +0400
+++ psi/src/psiactionlist.cpp	2013-01-15 14:04:38.872827543 +0400
@@ -381,11 +381,11 @@
 		IconAction *actOnlineHome = new IconAction (tr("Home Page (Online)"), tr("&Home Page (Online)"), 0, this);
 		actOnlineHome->setWhatsThis (tr("Home Page (Online)"));
 
-		IconAction *actOnlineForum = new IconAction (tr("Psi Forum (Online)"), tr("Psi &Forum (Online)"), 0, this);
-		actOnlineForum->setWhatsThis (tr("Psi Forum (Online)"));
+		IconAction *actOnlineForum = new IconAction (tr("Psi+ Forum (Online)"), tr("Psi+ &Forum (Online)"), 0, this);
+		actOnlineForum->setWhatsThis (tr("Psi+ Forum (Online)"));
 
-		IconAction *actPsiMUC = new IconAction (tr("Join Psi Discussion Room (Online)"), tr("&Join Psi Discussion Room (Online)"), 0, this);
-		actOnlineHome->setWhatsThis (tr("Join Psi Discussion Room (Online)"));
+		IconAction *actPsiMUC = new IconAction (tr("Join Psi+ Discussion Room (Online)"), tr("&Join Psi+ Discussion Room (Online)"), 0, this);
+		actOnlineHome->setWhatsThis (tr("Join Psi+ Discussion Room (Online)"));
 
 		IconAction *actBugReport = new IconAction (tr("Report a Bug (Online)"), tr("Report a &Bug (Online)"), 0, this);
 		actBugReport->setWhatsThis (tr("Report a Bug (Online)"));
