--- a/src/mainwin.cpp
+++ b/src/mainwin.cpp
@@ -1019,7 +1019,7 @@ void MainWin::actOnlineHomeActivated ()
 
 void MainWin::actOnlineForumActivated ()
 {
-    DesktopUtil::openUrl("https://forum.psi-im.org");
+    DesktopUtil::openUrl("https://forum.psi-plus.com");
 }
 
 void MainWin::actJoinPsiMUCActivated()
@@ -1034,7 +1034,7 @@ void MainWin::actJoinPsiMUCActivated()
 
 void MainWin::actBugReportActivated ()
 {
-    DesktopUtil::openUrl("https://github.com/psi-im/psi/issues");
+    DesktopUtil::openUrl("https://github.com/psi-plus/main/issues");
 }
 
 void MainWin::actAboutActivated ()
--- a/src/psiactionlist.cpp
+++ b/src/psiactionlist.cpp
@@ -386,11 +386,11 @@ void PsiActionList::Private::createMainWin()
         IconAction *actOnlineHome = new IconAction (tr("Home Page (Online)"), tr("&Home Page (Online)"), 0, this);
         actOnlineHome->setWhatsThis (tr("Home Page (Online)"));
 
-        IconAction *actOnlineForum = new IconAction (tr("Psi Forum (Online)"), tr("Psi &Forum (Online)"), 0, this);
-        actOnlineForum->setWhatsThis (tr("Psi Forum (Online)"));
+        IconAction *actOnlineForum = new IconAction (tr("Psi+ Forum (Online)"), tr("Psi+ &Forum (Online)"), 0, this);
+        actOnlineForum->setWhatsThis (tr("Psi+ Forum (Online)"));
 
-        IconAction *actPsiMUC = new IconAction (tr("Join Psi Discussion Room (Online)"), tr("&Join Psi Discussion Room (Online)"), 0, this);
-        actOnlineHome->setWhatsThis (tr("Join Psi Discussion Room (Online)"));
+        IconAction *actPsiMUC = new IconAction (tr("Join Psi+ Discussion Room (Online)"), tr("&Join Psi+ Discussion Room (Online)"), 0, this);
+        actOnlineHome->setWhatsThis (tr("Join Psi+ Discussion Room (Online)"));
 
         IconAction *actBugReport = new IconAction (tr("Report a Bug (Online)"), tr("Report a &Bug (Online)"), 0, this);
         actBugReport->setWhatsThis (tr("Report a Bug (Online)"));
