--- git.orig/src/mainwin.cpp
+++ git/src/mainwin.cpp
@@ -968,9 +968,9 @@ void MainWin::actReadmeActivated()
     w->show();
 }
 
-void MainWin::actOnlineWikiActivated() { DesktopUtil::openUrl("https://github.com/psi-im/psi/wiki"); }
+void MainWin::actOnlineWikiActivated() { DesktopUtil::openUrl("https://psi-plus.com/wiki/main"); }
 
-void MainWin::actOnlineHomeActivated() { DesktopUtil::openUrl("https://psi-im.org"); }
+void MainWin::actOnlineHomeActivated() { DesktopUtil::openUrl("https://psi-plus.com"); }
 
 void MainWin::actOnlineForumActivated() { DesktopUtil::openUrl("https://groups.google.com/forum/#!forum/psi-users"); }
 
@@ -984,7 +984,7 @@ void MainWin::actJoinPsiMUCActivated()
     account->actionJoin("psi-dev@conference.jabber.ru");
 }
 
-void MainWin::actBugReportActivated() { DesktopUtil::openUrl("https://github.com/psi-im/psi/issues"); }
+void MainWin::actBugReportActivated() { DesktopUtil::openUrl("https://github.com/psi-plus/main/issues"); }
 
 void MainWin::actAboutActivated()
 {
--- git.orig/src/psiactionlist.cpp
+++ git/src/psiactionlist.cpp
@@ -377,11 +377,11 @@ void PsiActionList::Private::createMainW
         IconAction *actOnlineHome = new IconAction(tr("Home Page (Online)"), tr("&Home Page (Online)"), 0, this);
         actOnlineHome->setWhatsThis(tr("Home Page (Online)"));
 
-        IconAction *actOnlineForum = new IconAction(tr("Psi Forum (Online)"), tr("Psi &Forum (Online)"), 0, this);
-        actOnlineForum->setWhatsThis(tr("Psi Forum (Online)"));
+        IconAction *actOnlineForum = new IconAction(tr("Psi+ Forum (Online)"), tr("Psi+ &Forum (Online)"), 0, this);
+        actOnlineForum->setWhatsThis(tr("Psi+ Forum (Online)"));
 
-        IconAction *actPsiMUC = new IconAction(tr("Join Psi Discussion Room (Online)"),
-                                               tr("&Join Psi Discussion Room (Online)"), 0, this);
+        IconAction *actPsiMUC = new IconAction(tr("Join Psi+ Discussion Room (Online)"),
+                                               tr("&Join Psi+ Discussion Room (Online)"), 0, this);
         actOnlineHome->setWhatsThis(tr("Join Psi Discussion Room (Online)"));
 
         IconAction *actBugReport = new IconAction(tr("Report a Bug (Online)"), tr("Report a &Bug (Online)"), 0, this);
