Entering 'iris'
Entering 'src/libpsi'
--- psi.orig/src/mainwin.cpp
+++ psi/src/mainwin.cpp
@@ -1096,7 +1096,7 @@ void MainWin::actOnlineHomeActivated ()
 
 void MainWin::actOnlineForumActivated ()
 {
-	DesktopUtil::openUrl("http://forum.psi-im.org");
+	DesktopUtil::openUrl("http://forum.psi-plus.com");
 }
 
 void MainWin::actJoinPsiMUCActivated()
@@ -1106,12 +1106,12 @@ void MainWin::actJoinPsiMUCActivated()
 		return;
 	}
 
-	account->actionJoin("psi@conference.psi-im.org");
+	account->actionJoin("psi-dev@conference.jabber.ru");
 }
 
 void MainWin::actBugReportActivated ()
 {
-	DesktopUtil::openUrl("http://forum.psi-im.org/forum/2");
+	DesktopUtil::openUrl("http://code.google.com/p/psi-dev/wiki/issues?tm=3");
 }
 
 void MainWin::actAboutActivated ()
--- psi.orig/src/psiactionlist.cpp
+++ psi/src/psiactionlist.cpp
@@ -386,11 +386,11 @@ void PsiActionList::Private::createMainW
 		IconAction *actOnlineHome = new IconAction (tr("Home Page (Online)"), tr("&Home Page (Online)"), 0, this);
 		actOnlineHome->setWhatsThis (tr("Home Page (Online)"));
 
-		IconAction *actOnlineForum = new IconAction (tr("Psi Forum (Online)"), tr("Psi &Forum (Online)"), 0, this);
-		actOnlineForum->setWhatsThis (tr("Psi Forum (Online)"));
+		IconAction *actOnlineForum = new IconAction (tr("Psi+ Forum (Online)"), tr("Psi+ &Forum (Online)"), 0, this);
+		actOnlineForum->setWhatsThis (tr("Psi+ Forum (Online)"));
 
-		IconAction *actPsiMUC = new IconAction (tr("Join Psi Discussion Room (Online)"), tr("&Join Psi Discussion Room (Online)"), 0, this);
-		actOnlineHome->setWhatsThis (tr("Join Psi Discussion Room (Online)"));
+		IconAction *actPsiMUC = new IconAction (tr("Join Psi+ Discussion Room (Online)"), tr("&Join Psi+ Discussion Room (Online)"), 0, this);
+		actOnlineHome->setWhatsThis (tr("Join Psi+ Discussion Room (Online)"));
 
 		IconAction *actBugReport = new IconAction (tr("Report a Bug (Online)"), tr("Report a &Bug (Online)"), 0, this);
 		actBugReport->setWhatsThis (tr("Report a Bug (Online)"));
