Index: psi/src/applicationinfo.cpp
===================================================================
--- psi.orig/src/applicationinfo.cpp	2013-01-15 14:04:39.084827552 +0400
+++ psi/src/applicationinfo.cpp	2013-01-15 14:04:39.080827552 +0400
@@ -131,6 +131,17 @@
 	return  dirs;
 }

+QStringList ApplicationInfo::pluginDirs()
+{
+	QStringList l;
+	l += resourcesDir() + "/plugins";
+	l += homeDir(ApplicationInfo::DataLocation) + "/plugins";
+#if defined(Q_OS_UNIX)
+	l += libDir() + "/plugins";
+#endif
+	return l;
+}
+
 QString ApplicationInfo::getCertificateStoreSaveDir()
 {
 	QDir certsave(homeDir(DataLocation) + "/certs");
Index: psi/src/applicationinfo.h
===================================================================
--- psi.orig/src/applicationinfo.h	2013-01-15 14:04:39.084827552 +0400
+++ psi/src/applicationinfo.h	2013-01-15 14:04:39.080827552 +0400
@@ -40,6 +40,7 @@
 	static QStringList getCertificateStoreDirs();
 	static QString getCertificateStoreSaveDir();
 	static QStringList dataDirs();
+	static QStringList pluginDirs();

 	// Namespaces
 	static QString optionsNS();
Index: psi/src/avcall/avcall.cpp
===================================================================
--- psi.orig/src/avcall/avcall.cpp	2013-01-15 14:04:39.084827552 +0400
+++ psi/src/avcall/avcall.cpp	2013-01-15 14:04:39.080827552 +0400
@@ -117,7 +117,11 @@
 			pluginFile = findPlugin("../Plugins", "gstprovider");
 			resourcePath = QCoreApplication::applicationDirPath() + "/../Frameworks/gstreamer-0.10";
 #else
-			pluginFile = findPlugin(ApplicationInfo::libDir() + "/plugins", "gstprovider");
+			foreach(const QString& path, ApplicationInfo::pluginDirs()) {
+				pluginFile = findPlugin(path, "gstprovider");
+				if(!pluginFile.isEmpty())
+					break;
+			}
 #endif
 		}

Index: psi/src/pluginmanager.cpp
===================================================================
--- psi.orig/src/pluginmanager.cpp	2013-01-15 14:04:39.084827552 +0400
+++ psi/src/pluginmanager.cpp	2013-01-15 14:04:39.080827552 +0400
@@ -60,13 +60,7 @@
  */
 static QStringList pluginDirs()
 {
-	QStringList l;
-	l += ApplicationInfo::resourcesDir() + "/plugins";
-	l += ApplicationInfo::homeDir(ApplicationInfo::DataLocation) + "/plugins";
-#if defined(Q_OS_UNIX)
-	l += ApplicationInfo::libDir() + "/plugins";
-#endif
-	return l;
+	return ApplicationInfo::pluginDirs();
 }

 /**
