--- psi.orig/src/applicationinfo.cpp
+++ psi/src/applicationinfo.cpp
@@ -122,6 +122,17 @@ QStringList ApplicationInfo::dataDirs()
 	return  dirs;
 }
 
+QStringList ApplicationInfo::pluginDirs()
+{
+	QStringList l;
+	l += resourcesDir() + "/plugins";
+	l += homeDir(ApplicationInfo::DataLocation) + "/plugins";
+#if defined(Q_OS_UNIX)
+	l += libDir() + "/plugins";
+#endif
+	return l;
+}
+
 QString ApplicationInfo::getCertificateStoreSaveDir()
 {
 	QDir certsave(homeDir(DataLocation) + "/certs");
--- psi.orig/src/applicationinfo.h
+++ psi/src/applicationinfo.h
@@ -38,6 +38,7 @@ public:
 	static QStringList getCertificateStoreDirs();
 	static QString getCertificateStoreSaveDir();
 	static QStringList dataDirs();
+	static QStringList pluginDirs();
 
 	// Namespaces
 	static QString optionsNS();
--- psi.orig/src/avcall/avcall.cpp
+++ psi/src/avcall/avcall.cpp
@@ -117,7 +117,11 @@ static void ensureLoaded()
 			pluginFile = findPlugin("../Plugins", "gstprovider");
 			resourcePath = QCoreApplication::applicationDirPath() + "/../Frameworks/gstreamer-0.10";
 #else
-			pluginFile = findPlugin(ApplicationInfo::libDir() + "/plugins", "gstprovider");
+			foreach(const QString& path, ApplicationInfo::pluginDirs()) {
+				pluginFile = findPlugin(path, "gstprovider");
+				if(!pluginFile.isEmpty())
+					break;
+			}
 #endif
 		}
 
--- psi.orig/src/pluginmanager.cpp
+++ psi/src/pluginmanager.cpp
@@ -59,13 +59,7 @@ public:
  */ 
 static QStringList pluginDirs()
 {
-	QStringList l;
-	l += ApplicationInfo::resourcesDir() + "/plugins";
-	l += ApplicationInfo::homeDir(ApplicationInfo::DataLocation) + "/plugins";
-#if defined(Q_OS_UNIX)
-	l += ApplicationInfo::libDir() + "/plugins";
-#endif
-	return l;
+	return ApplicationInfo::pluginDirs();
 }
 
 /**
