--- psi.orig/src/aboutdlg.cpp
+++ psi/src/aboutdlg.cpp
@@ -71,6 +71,52 @@ AboutDlg::AboutDlg(QWidget* parent)
 			   tr("Developer"));
 	ui_.te_authors->setText( authors );
 
+	// fill in About Psi+ tab...
+	QString psiplus;
+	psiplus += details(QString::fromUtf8("Psi+ Project"),
+			   "", "psi-dev@conference.jabber.ru", "http://psi-dev.googlecode.com/",
+			   tr("Active project members are presents below:"));
+	psiplus += details(QString::fromUtf8("zet"),
+			   "vladimir.shelukhin@gmail.com", "", "",
+			   tr("Founder, Patcher and MS Windows Packager"));
+	psiplus += details(QString::fromUtf8("rion"),
+			   "rion4ik@gmail.com", "", "",
+			   tr("Founder and Lead Patcher"));
+	psiplus += details(QString::fromUtf8("majik"),
+			   "maksim.maj@gmail.com", "", "",
+			   tr("Founder, Designer and Patcher"));
+	psiplus += details(QString::fromUtf8("Dealer_WeARE"),
+			   "wadealer@gmail.com", "", "",
+			   tr("Lead Patcher and Plugins Writer; Mac OS X Packager"));
+	psiplus += details(QString::fromUtf8("liuch"),
+			   "", "", "",
+			   tr("Patcher and Plugins Writer"));
+	psiplus += details(QString::fromUtf8("tehnick"),
+			  "tehnick-8@mail.ru", "", "",
+			  tr("Language coordinator; Debian Packager"));
+	psiplus += details(QString::fromUtf8("KukuRuzo"),
+			   "", "", "",
+			   tr("Patcher and Plugins Writer"));
+	psiplus += details(QString::fromUtf8("taurus"),
+			   "", "", "",
+			   tr("Patcher, Plugins Writer and Fedora Packager"));
+	psiplus += details(QString::fromUtf8("ivan1986"),
+			   "", "", "",
+			   tr("Patcher and Ubuntu Packager"));
+	psiplus += details(QString::fromUtf8("nexor (aka zerkalica)"),
+			   "", "", "",
+			   tr("Ubuntu Packager"));
+	psiplus += details(QString::fromUtf8("Flint (aka Флинт)"),
+			   "", "", "",
+			   tr("Mandriva Linux Packager"));
+	psiplus += details(QString::fromUtf8("ivan101"),
+			   "ivan101@users.sourceforge.net", "", "",
+			   tr("Patcher and Russian localization"));
+	psiplus += details(QString::fromUtf8("Z_God"),
+			   "", "", "",
+			   tr("Psimedia Patcher and Wiki English localization"));
+	ui_.te_psiplus->setText( psiplus );
+
 	// fill in Thanks To tab...
 	QString thanks;
 	thanks += details(QString::fromUtf8("Frederik Schwarzer"),
@@ -182,7 +228,7 @@ QString AboutDlg::details( QString name,
 	if ( !email.isEmpty() )
 		ret += nbsp + "E-mail: " + "<a href=\"mailto:" + email + "\">" + email + "</a><br>\n";
 	if ( !jabber.isEmpty() )
-		ret += nbsp + "XMPP: " + "<a href=\"xmpp:" + jabber + "\">" + jabber + "</a><br>\n";
+		ret += nbsp + "XMPP: " + "<a href=\"xmpp:" + jabber + "?join\">" + jabber + "</a><br>\n";
 	if ( !www.isEmpty() )
 		ret += nbsp + "WWW: " + "<a href=\"" + www + "\">" + www + "</a><br>\n";
 	if ( !desc.isEmpty() )
--- psi.orig/src/about.ui
+++ psi/src/about.ui
@@ -8,12 +8,12 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>504</width>
+    <width>625</width>
     <height>320</height>
    </rect>
   </property>
   <property name="windowTitle" >
-   <string>About Psi</string>
+   <string>About</string>
   </property>
   <layout class="QVBoxLayout" >
    <property name="margin" >
@@ -39,7 +39,7 @@
         <enum>QFrame::Plain</enum>
        </property>
        <property name="psiIconName" stdset="0" >
-        <string>psi/logo_48</string>
+        <string>psi/psiplus_logo</string>
        </property>
       </widget>
      </item>
@@ -221,7 +221,23 @@ Copyright © 2001-2014 The Psi Team.&lt;
        </item>
       </layout>
      </widget>
-    </widget>
+     <widget class="QWidget" name="tab_psiplus" >
+      <attribute name="title" >
+       <string>About Psi+</string>
+      </attribute>
+      <layout class="QVBoxLayout" >
+       <property name="margin" >
+        <number>9</number>
+       </property>
+       <property name="spacing" >
+        <number>6</number>
+       </property>
+       <item>
+        <widget class="PsiTextView" name="te_psiplus" />
+       </item>
+      </layout>
+     </widget>
+	 </widget>
    </item>
    <item>
     <layout class="QHBoxLayout" >
