--- git.orig/src/aboutdlg.cpp
+++ git/src/aboutdlg.cpp
@@ -29,7 +29,7 @@ AboutDlg::AboutDlg(QWidget *parent) : QD
 {
     setAttribute(Qt::WA_DeleteOnClose);
     ui_.setupUi(this);
-    setWindowIcon(IconsetFactory::icon("psi/logo_32").icon());
+    setWindowIcon(IconsetFactory::icon("psi/psiplus_logo").icon());
 
     setModal(false);
 
@@ -58,6 +58,32 @@ AboutDlg::AboutDlg(QWidget *parent) : QD
     authors += details(QString::fromUtf8("Boris Pek"), "tehnick-8@yandex.ru", "", "", tr("Developer and Maintainer"));
     ui_.te_authors->setText(authors);
 
+    // fill in About Psi+ tab...
+    QString psiplus;
+    psiplus += details(QString::fromUtf8("Psi+ Project"), "", "psi-dev@conference.jabber.ru", "https://psi-plus.com/",
+                       tr("Active project members are presents below:"));
+    psiplus += details(QString::fromUtf8("rion"), "rion4ik@gmail.com", "", "", tr("Founder and Lead Patcher"));
+    psiplus += details(QString::fromUtf8("zet"), "vladimir.shelukhin@gmail.com", "", "",
+                       tr("Founder, Patcher and former MS Windows Packager"));
+    psiplus += details(QString::fromUtf8("majik"), "maksim.maj@gmail.com", "", "",
+                       tr("Founder, Patcher and former Designer"));
+    psiplus += details(QString::fromUtf8("Dealer_WeARE"), "wadealer@gmail.com", "", "",
+                       tr("Lead Patcher and Plugins Writer; former macOS Packager"));
+    psiplus += details(QString::fromUtf8("liuch"), "", "", "", tr("Patcher and Plugins Writer"));
+    psiplus += details(QString::fromUtf8("tehnick"), "tehnick-8@yandex.ru", "", "",
+                       tr("Language coordinator; Patcher; Debian and Ubuntu Packager; MS Windows and macOS Packager"));
+    psiplus
+        += details(QString::fromUtf8("KukuRuzo"), "", "", "", tr("Patcher and Plugins Writer; MS Windows Packager"));
+    psiplus
+        += details(QString::fromUtf8("taurus"), "", "", "", tr("Patcher and Plugins Writer; former Fedora Packager"));
+    psiplus += details(QString::fromUtf8("ivan1986"), "", "", "", tr("Patcher and former Ubuntu Packager"));
+    psiplus += details(QString::fromUtf8("nexor (aka zerkalica)"), "", "", "", tr("former Ubuntu Packager"));
+    psiplus += details(QString::fromUtf8("Flint (aka Флинт)"), "", "", "", tr("former Mandriva Linux Packager"));
+    psiplus += details(QString::fromUtf8("ivan101"), "ivan101@users.sourceforge.net", "", "",
+                       tr("Patcher and former Russian translator"));
+    psiplus += details(QString::fromUtf8("Z_God"), "", "", "", tr("Psimedia Patcher and Wiki English localization"));
+    ui_.te_psiplus->setText(psiplus);
+
     // fill in Thanks To tab...
     QString thanks;
     thanks += details(QString::fromUtf8("Boris Pek"), "tehnick-8@yandex.ru", "", "",
@@ -135,7 +161,7 @@ QString AboutDlg::details(QString name,
     if (!email.isEmpty())
         ret += nbsp + "E-mail: " + "<a href=\"mailto:" + email + "\">" + email + "</a><br>\n";
     if (!jabber.isEmpty())
-        ret += nbsp + "XMPP: " + "<a href=\"xmpp:" + jabber + "\">" + jabber + "</a><br>\n";
+        ret += nbsp + "XMPP: " + "<a href=\"xmpp:" + jabber + "?join\">" + jabber + "</a><br>\n";
     if (!www.isEmpty())
         ret += nbsp + "WWW: " + "<a href=\"" + www + "\">" + www + "</a><br>\n";
     if (!desc.isEmpty())
--- git.orig/src/about.ui
+++ git/src/about.ui
@@ -6,12 +6,12 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>504</width>
+    <width>625</width>
     <height>320</height>
    </rect>
   </property>
   <property name="windowTitle">
-   <string>About Psi</string>
+   <string>About</string>
   </property>
   <layout class="QVBoxLayout">
    <property name="spacing">
@@ -49,7 +49,7 @@
      <item>
       <widget class="IconLabel" name="lb_icon" native="true">
        <property name="psiIconName" stdset="0">
-        <string>psi/logo_48</string>
+        <string>psi/psiplus_logo</string>
        </property>
       </widget>
      </item>
@@ -267,7 +267,23 @@
        </item>
       </layout>
      </widget>
-    </widget>
+     <widget class="QWidget" name="tab_psiplus" >
+      <attribute name="title" >
+       <string>About Psi+</string>
+      </attribute>
+      <layout class="QVBoxLayout" >
+       <property name="margin" >
+        <number>9</number>
+       </property>
+       <property name="spacing" >
+        <number>6</number>
+       </property>
+       <item>
+        <widget class="PsiTextView" name="te_psiplus" />
+       </item>
+      </layout>
+     </widget>
+     </widget>
    </item>
    <item>
     <layout class="QHBoxLayout">
