--- git.orig/src/applicationinfo.cpp
+++ git/src/applicationinfo.cpp
@@ -32,11 +32,15 @@
 // place (like an external file loaded through the resources system)
 // Should also be overridable through an optional file.
 
-#define PROG_NAME "Psi"
-#define PROG_VERSION "0.15-dev" " (" __DATE__ ")" //CVS Builds are dated
+#define PROG_NAME "Psi+"
+#ifdef WEBKIT
+#define PROG_VERSION "0.15.xxx-webkit" " (" __DATE__ ")" //CVS Builds are dated
+#else
+#define PROG_VERSION "0.15.xxx" " (" __DATE__ ")" //CVS Builds are dated
+#endif
 //#define PROG_VERSION "0.15";
-#define PROG_CAPS_NODE "http://psi-im.org/caps"
-#define PROG_CAPS_VERSION "caps-b75d8d2b25"
+#define PROG_CAPS_NODE "http://psi-dev.googlecode.com/caps"
+#define PROG_CAPS_VERSION "0.15"
 #define PROG_IPC_NAME "org.psi-im.Psi"	// must not contain '\\' character on Windows
 #define PROG_OPTIONS_NS "http://psi-im.org/options"
 #define PROG_STORAGE_NS "http://psi-im.org/storage"
