Index: psimedia/gstprovider/gstconf.pri
===================================================================
--- psimedia/gstprovider/gstconf.pri	(revision 783)
+++ psimedia/gstprovider/gstconf.pri	(working copy)
@@ -5,29 +5,37 @@

 windows {
 	INCLUDEPATH += \
-		c:/mingw/include/glib-2.0 \
-		c:/mingw/lib/glib-2.0/include \
-		c:/msys/1.0/home/gst/include \
-		c:/msys/1.0/home/gst/include/liboil-0.3 \
-		c:/msys/1.0/home/gst/include/gstreamer-0.10
+		c:/glib/include/glib-2.0 \
+		c:/glib/lib/glib-2.0/include \
+		c:/gstforwin/dxsdk/include \
+		c:/gstforwin/winsdk/include \
+		c:/gstforwin/include \
+		c:/gstforwin/include/liboil-0.3 \
+		c:/gstforwin/include/libxml2 \
+		c:/gstforwin/gstreamer/include/gstreamer-0.10 \
+		c:/gstforwin/gst-plugins-base/include/gstreamer-0.10
 	LIBS += \
-		-Lc:/msys/1.0/home/gst/lib \
-		-lgstreamer-0.10.dll \
-		-lgthread-2.0 \
-		-lglib-2.0 \
-		-lgobject-2.0 \
-		-lgstvideo-0.10.dll \
-		-lgstbase-0.10.dll \
-		-lgstinterfaces-0.10.dll
+		-Lc:/glib/bin \
+		-Lc:/gstforwin/bin \
+		-Lc:/gstforwin/gstreamer/bin \
+		-Lc:/gstforwin/gst-plugins-base/bin \
+		-lgstreamer-0.10-0 \
+		-lgthread-2.0-0 \
+		-lglib-2.0-0 \
+		-lgobject-2.0-0 \
+		-lgstvideo-0.10-0 \
+		-lgstbase-0.10-0 \
+		-lgstinterfaces-0.10-0

 	# qmake mingw seems to have broken prl support, so force these
 	win32-g++|contains($$list($$[QT_VERSION]), 4.0.*|4.1.*|4.2.*|4.3.*) {
 		LIBS *= \
-			-loil-0.3 \
-			-lgstaudio-0.10.dll \
-			-lgstrtp-0.10.dll \
-			-lgstnetbuffer-0.10.dll \
-			-lspeexdsp.dll \
+			-Lc:/gstforwin/winsdk/lib \
+			-loil-0.3-0 \
+			-lgstaudio-0.10-0 \
+			-lgstrtp-0.10-0 \
+			-lgstnetbuffer-0.10-0 \
+			-lspeexdsp-1 \
 			-lsetupapi \
 			-lksuser \
 			-lamstrmid \
