diff -urpN -X psidiff.ignore sources//src/eventdlg.cpp work//src/eventdlg.cpp
--- sources//src/eventdlg.cpp	2010-03-11 21:03:58.000000000 +0500
+++ work//src/eventdlg.cpp	2010-03-11 22:34:50.000000000 +0500
@@ -465,6 +465,8 @@
 AddUrlDlg::AddUrlDlg(QWidget *parent)
 :QDialog(parent)
 {
+	if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+			setAttribute(Qt::WA_MacMetalStyle);
 	setupUi(this);
 #ifndef Q_WS_MAC
 	setWindowIcon(IconsetFactory::icon("psi/www").icon());
@@ -586,6 +588,8 @@
 	: AdvancedWidget<QWidget>(0)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+ 	if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+			setAttribute(Qt::WA_MacMetalStyle);	
 	d = new Private(this);
 	d->composing = true;
 	d->psi = psi;
diff -urpN -X psidiff.ignore sources//src/chatdlg.cpp work//src/chatdlg.cpp
--- sources//src/chatdlg.cpp	2010-03-11 21:04:00.000000000 +0500
+++ work//src/chatdlg.cpp	2010-03-11 22:26:19.000000000 +0500
@@ -93,6 +93,11 @@
 	: TabbableWidget(jid, pa, tabManager)
 	, highlightersInstalled_(false)
 {
+	if (PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool()) {
+			setAttribute(Qt::WA_MacMetalStyle);
+
+	}
+	
 	pending_ = 0;
 	keepOpen_ = false;
 	warnSend_ = false;
diff -urpN -X psidiff.ignore sources//src/infodlg.cpp work//src/infodlg.cpp
--- sources//src/infodlg.cpp	2010-03-11 21:04:03.000000000 +0500
+++ work//src/infodlg.cpp	2010-03-11 22:45:17.000000000 +0500
@@ -120,6 +120,8 @@
 	: QDialog(parent)
 {
 	setAttribute(Qt::WA_DeleteOnClose);
+	if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+			setAttribute(Qt::WA_MacMetalStyle);
 #if QT_VERSION >= 0x040500
 	setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
 #endif
diff -urpN -X psidiff.ignore sources//src/psi_profiles.cpp work//src/psi_profiles.cpp
--- sources//src/psi_profiles.cpp	2010-03-11 21:04:03.000000000 +0500
+++ work//src/psi_profiles.cpp	2010-03-11 22:46:21.000000000 +0500
@@ -884,6 +884,7 @@
 				migrateBoolEntry(tag, "jidComplete", "options.ui.message.use-jid-auto-completion");
 				migrateBoolEntry(tag, "grabUrls", "options.ui.message.auto-grab-urls-from-clipboard");
 				migrateBoolEntry(tag, "smallChats", "options.ui.chat.use-small-chats");
+				migrateBoolEntry(tag, "brushedMetal", "options.ui.mac.use-brushed-metal-windows");
 				migrateBoolEntry(tag, "chatLineEdit", "options.ui.chat.use-expanding-line-edit");
 				migrateBoolEntry(tag, "useTabs", "options.ui.tabs.use-tabs");
 				migrateBoolEntry(tag, "putTabsAtBottom", "options.ui.tabs.put-tabs-at-bottom");
diff -urpN -X psidiff.ignore sources//src/tabs/tabdlg.cpp work//src/tabs/tabdlg.cpp
--- sources//src/tabs/tabdlg.cpp	2010-03-11 21:04:03.000000000 +0500
+++ work//src/tabs/tabdlg.cpp	2010-03-11 22:48:28.000000000 +0500
@@ -131,6 +131,11 @@
 		delegate_->create(this);
 	}
 
+	if (PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool()) {
+			setAttribute(Qt::WA_MacMetalStyle);
+		
+	}
+	
 	// FIXME
 	qRegisterMetaType<TabDlg*>("TabDlg*");
 	qRegisterMetaType<TabbableWidget*>("TabbableWidget*");
diff -urpN -X psidiff.ignore sources//src/whiteboarding/wbdlg.cpp work//src/whiteboarding/wbdlg.cpp
--- sources//src/whiteboarding/wbdlg.cpp	2010-03-11 21:03:55.000000000 +0500
+++ work//src/whiteboarding/wbdlg.cpp	2010-03-11 22:50:08.000000000 +0500
@@ -38,6 +38,9 @@
 //----------------------------------------------------------------------------
 
 WbDlg::WbDlg(SxeSession* session, PsiAccount* pa) {
+	if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+		setAttribute(Qt::WA_MacMetalStyle);
+	
 	groupChat_ = session->groupChat();
 	pending_ = 0;
 	keepOpen_ = false;
