--- psi.orig/src/chatdlg.cpp
+++ psi/src/chatdlg.cpp
@@ -97,6 +97,11 @@ ChatDlg::ChatDlg(const Jid& jid, PsiAcco
     : TabbableWidget(jid, pa, tabManager)
     , highlightersInstalled_(false)
 {
+    if (PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool()) {
+            setAttribute(Qt::WA_MacMetalStyle);
+
+    }
+
     pending_ = 0;
     keepOpen_ = false;
     warnSend_ = false;
--- psi.orig/src/eventdlg.cpp
+++ psi/src/eventdlg.cpp
@@ -458,6 +458,8 @@ void AttachView::addUrlList(const UrlLis
 AddUrlDlg::AddUrlDlg(QWidget *parent)
 :QDialog(parent)
 {
+    if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+            setAttribute(Qt::WA_MacMetalStyle);
     setupUi(this);
 #ifndef Q_OS_MAC
     setWindowIcon(IconsetFactory::icon("psi/www").icon());
@@ -580,6 +582,8 @@ EventDlg::EventDlg(const QString &to, Ps
     : AdvancedWidget<QWidget>(0)
 {
     setAttribute(Qt::WA_DeleteOnClose);
+     if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+            setAttribute(Qt::WA_MacMetalStyle);
     d = new Private(this);
     d->composing = true;
     d->psi = psi;
--- psi.orig/src/infodlg.cpp
+++ psi/src/infodlg.cpp
@@ -127,6 +127,8 @@ InfoDlg::InfoDlg(int type, const Jid &j,
     : QDialog(parent)
 {
     setAttribute(Qt::WA_DeleteOnClose);
+    if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+            setAttribute(Qt::WA_MacMetalStyle);
    setWindowFlags(Qt::Dialog | Qt::WindowTitleHint | Qt::WindowMinimizeButtonHint | Qt::WindowMaximizeButtonHint | Qt::WindowCloseButtonHint | Qt::CustomizeWindowHint);
    ui_.setupUi(this);
    d = new Private;
--- psi.orig/src/psi_profiles.cpp
+++ psi/src/psi_profiles.cpp
@@ -896,6 +896,7 @@ bool OptionsMigration::fromFile(const QS
                 migrateBoolEntry(tag, "jidComplete", "options.ui.message.use-jid-auto-completion");
                 migrateBoolEntry(tag, "grabUrls", "options.ui.message.auto-grab-urls-from-clipboard");
                 migrateBoolEntry(tag, "smallChats", "options.ui.chat.use-small-chats");
+                migrateBoolEntry(tag, "brushedMetal", "options.ui.mac.use-brushed-metal-windows");
                 migrateBoolEntry(tag, "chatLineEdit", "options.ui.chat.use-expanding-line-edit");
                 migrateBoolEntry(tag, "useTabs", "options.ui.tabs.use-tabs");
                 migrateBoolEntry(tag, "putTabsAtBottom", "options.ui.tabs.put-tabs-at-bottom");
--- psi.orig/src/tabs/tabdlg.cpp
+++ psi/src/tabs/tabdlg.cpp
@@ -149,6 +149,11 @@ TabDlg::TabDlg(TabManager* tabManager, c
         delegate_->create(this);
     }

+    if (PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool()) {
+            setAttribute(Qt::WA_MacMetalStyle);
+
+    }
+
     // FIXME
     qRegisterMetaType<TabDlg*>("TabDlg*");
     qRegisterMetaType<TabbableWidget*>("TabbableWidget*");
--- psi.orig/src/whiteboarding/wbdlg.cpp
+++ psi/src/whiteboarding/wbdlg.cpp
@@ -41,6 +41,9 @@ static const QString geometryOption = "o
 //----------------------------------------------------------------------------

 WbDlg::WbDlg(SxeSession* session, PsiAccount* pa) {
+    if ( PsiOptions::instance()->getOption("options.ui.mac.use-brushed-metal-windows").toBool() )
+        setAttribute(Qt::WA_MacMetalStyle);
+
     groupChat_ = session->groupChat();
     pending_ = 0;
     keepOpen_ = false;
