diff --git a/src/historydlg.cpp b/src/historydlg.cpp
index a3631ad..0a1145a 100644
--- a/src/historydlg.cpp
+++ b/src/historydlg.cpp
@@ -615,15 +615,23 @@ void HistoryDlg::startRequest()
 	if(!ui_.busy->isActive()) {
 		ui_.busy->start();
 	}
+
+	// Workaround. On MAC OS X this dialog doesn't get keyboard events
+	// after it was disabled. It's mean you can do only one searching
+	// without reopening this dialog.
+#ifndef Q_OS_MAC
 	setEnabled(false);
+#endif
 }
 
 void HistoryDlg::stopRequest()
 {
 	if(ui_.busy->isActive()) {
-		ui_.busy->stop();		
+		ui_.busy->stop();
 	}
+#ifndef Q_OS_MAC
 	setEnabled(true);
+#endif
 }
 
 EDBHandle* HistoryDlg::getEDBHandle()
