diff --git a/src/systeminfo.cpp b/src/systeminfo.cpp
index 0f3fd31..a407b8f 100644
--- a/src/systeminfo.cpp
+++ b/src/systeminfo.cpp
@@ -198,14 +198,16 @@ SystemInfo::SystemInfo() : QObject(QCoreApplication::instance())
 
 #elif defined(Q_OS_MAC)
 	QSysInfo::MacVersion v = QSysInfo::MacintoshVersion;
-	if(v == QSysInfo::MV_10_3)
-		os_str_ = "Mac OS X 10.3";
-	else if(v == QSysInfo::MV_10_4)
-		os_str_ = "Mac OS X 10.4";
-	else if(v == QSysInfo::MV_10_5)
+	if(v == QSysInfo::MV_10_5)
 		os_str_ = "Mac OS X 10.5";
 	else if(v == QSysInfo::MV_10_6)
 		os_str_ = "Mac OS X 10.6";
+#if QT_VERSION > 0x040700
+	else if(v == QSysInfo::MV_10_7)
+		os_str_ = "Mac OS X 10.7";
+	else if(v == QSysInfo::MV_10_8)
+		os_str_ = "OS X 10.8";
+#endif
 	else
 		os_str_ = "Mac OS X";
 #endif
