diff -urpN git.orig/src/CocoaUtilities/CocoaTrayClick.cpp git/src/CocoaUtilities/CocoaTrayClick.cpp
--- git.orig/src/CocoaUtilities/CocoaTrayClick.cpp	1970-01-01 03:00:00.000000000 +0300
+++ git/src/CocoaUtilities/CocoaTrayClick.cpp	2015-05-04 11:30:57.000000000 +0400
@@ -0,0 +1,91 @@
+/*
+ * CocoaTrayClick
+ * Copyright (C) 2012, 2015  Khryukin Evgeny
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+#include "CocoaTrayClick.h"
+#include <objc/runtime.h>
+#include <objc/message.h>
+#include <QApplication>
+
+#ifdef DEBUG_OUTPUT
+#include <QDebug>
+#endif
+
+bool dockClickHandler(id /*self*/, SEL /*_cmd*/, ...)
+{
+	CocoaTrayClick::instance()->emitTrayClicked();
+	return true;
+}
+
+
+CocoaTrayClick * CocoaTrayClick::instance()
+{
+	if(!instance_)
+		instance_ = new CocoaTrayClick();
+
+	return instance_;
+}
+
+CocoaTrayClick::CocoaTrayClick()
+	: QObject(qApp)
+{
+	Class cls = objc_getClass("NSApplication");
+	objc_object *appInst = objc_msgSend((objc_object*)cls, sel_registerName("sharedApplication"));
+
+	if(appInst != NULL) {
+		objc_object* delegate = objc_msgSend(appInst, sel_registerName("delegate"));
+		Class delClass = (Class)objc_msgSend(delegate,  sel_registerName("class"));
+		SEL shouldHandle = sel_registerName("applicationShouldHandleReopen:hasVisibleWindows:");
+		if (class_getInstanceMethod(delClass, shouldHandle)) {
+			if (class_replaceMethod(delClass, shouldHandle, (IMP)dockClickHandler, "B@:")) {
+#ifdef DEBUG_OUTPUT
+				qDebug() << "Registered dock click handler (replaced original method)";
+#endif
+			}
+			else {
+#ifdef DEBUG_OUTPUT
+				qDebug() << "Failed to replace method for dock click handler";
+#endif
+			}
+		}
+		else {
+			if (class_addMethod(delClass, shouldHandle, (IMP)dockClickHandler,"B@:")) {
+#ifdef DEBUG_OUTPUT
+				qDebug() << "Registered dock click handler";
+#endif
+			}
+			else {
+#ifdef DEBUG_OUTPUT
+				qDebug() << "Failed to register dock click handler";
+#endif
+			}
+		}
+	}
+}
+
+CocoaTrayClick::~CocoaTrayClick()
+{
+}
+
+void CocoaTrayClick::emitTrayClicked()
+{
+	emit trayClicked();
+}
+
+CocoaTrayClick* CocoaTrayClick::instance_ = NULL;
diff -urpN git.orig/src/CocoaUtilities/CocoaTrayClick.mm git/src/CocoaUtilities/CocoaTrayClick.mm
--- git.orig/src/CocoaUtilities/CocoaTrayClick.mm	2014-07-29 14:24:36.000000000 +0400
+++ git/src/CocoaUtilities/CocoaTrayClick.mm	1970-01-01 03:00:00.000000000 +0300
@@ -1,60 +0,0 @@
-/*
- * CocoaTrayClick
- * Copyright (C) 2012  Khryukin Evgeny
- *
- * This program is free software; you can redistribute it and/or
- * modify it under the terms of the GNU General Public License
- * as published by the Free Software Foundation; either version 2
- * of the License, or (at your option) any later version.
- *
- * This program is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
- * GNU General Public License for more details.
- *
- * You should have received a copy of the GNU General Public License
- * along with this library; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
- *
- */
-
-#include "CocoaUtilities/CocoaTrayClick.h"
-#include <objc/runtime.h>
-#include <QApplication>
-#include <QDebug>
-#ifdef HAVE_QT5
-#include <AppKit/NSApplication.h>
-#endif
-
-void dockClickHandler(id /*self*/, SEL /*_cmd*/)
-{
-	CocoaTrayClick::instance()->emitTrayClicked();
-}
-
-
-CocoaTrayClick * CocoaTrayClick::instance()
-{
-	if(!instance_)
-		instance_ = new CocoaTrayClick();
-
-	return instance_;
-}
-
-CocoaTrayClick::CocoaTrayClick()
-	: QObject(qApp)
-{
-	Class cls = [[[NSApplication sharedApplication] delegate] class];
-	if (!class_addMethod(cls, @selector(applicationShouldHandleReopen:hasVisibleWindows:), (IMP) dockClickHandler, "v@:"))
-		qDebug() << "CocoaTrayClick::Private : class_addMethod failed!";
-}
-
-CocoaTrayClick::~CocoaTrayClick()
-{
-}
-
-void CocoaTrayClick::emitTrayClicked()
-{
-	emit trayClicked();
-}
-
-CocoaTrayClick* CocoaTrayClick::instance_ = NULL;
diff -urpN git.orig/src/CocoaUtilities/CocoaUtilities.pri git/src/CocoaUtilities/CocoaUtilities.pri
--- git.orig/src/CocoaUtilities/CocoaUtilities.pri	2014-07-29 14:24:36.000000000 +0400
+++ git/src/CocoaUtilities/CocoaUtilities.pri	2015-05-04 12:48:32.000000000 +0400
@@ -8,5 +8,6 @@ HEADERS += \
 
 OBJECTIVE_SOURCES += \
 	$$PWD/CocoaInitializer.mm \
-	$$PWD/cocoacommon.mm \
-	$$PWD/CocoaTrayClick.mm
+	$$PWD/cocoacommon.mm
+
+SOURCES += $$PWD/CocoaTrayClick.cpp
diff -urpN git.orig/src/src.pri git/src/src.pri
--- git.orig/src/src.pri	2015-05-04 12:46:00.000000000 +0400
+++ git/src/src.pri	2015-05-04 12:57:43.000000000 +0400
@@ -632,6 +632,7 @@ else {
 
 mac {
 	QMAKE_LFLAGS += -framework Carbon -framework IOKit -framework AppKit
+	QMAKE_LFLAGS += -framework CoreFoundation -lobjc #for cocoatrayclick
 }
 
 INCLUDEPATH += $$PWD
