--- psi.orig/src/widgets/psiwindowheader.ui
+++ psi/src/widgets/psiwindowheader.ui
@@ -22,6 +22,9 @@
   <property name="windowTitle">
    <string>PsiWindowHeader</string>
   </property>
+  <property name="layoutDirection">
+   <enum>Qt::RightToLeft</enum>
+  </property>
   <layout class="QVBoxLayout" name="verticalLayout">
    <property name="spacing">
     <number>0</number>
@@ -71,7 +74,7 @@
            </spacer>
           </item>
           <item>
-           <widget class="QToolButton" name="hideButton">
+           <widget class="QToolButton" name="maximizeButton">
             <property name="text">
              <string>...</string>
             </property>
@@ -81,7 +84,7 @@
            </widget>
           </item>
           <item>
-           <widget class="QToolButton" name="maximizeButton">
+           <widget class="QToolButton" name="hideButton">
             <property name="text">
              <string>...</string>
             </property>
--- psi.orig/options/default.xml
+++ psi/options/default.xml
@@ -314,7 +314,7 @@
 			<spell-check comment="Options related to the spell checker">
 				<enabled comment="Whether the spell checker is enabled" type="bool">true</enabled>
 			</spell-check>
-			<disable-send-button type="bool">false</disable-send-button>
+			<disable-send-button type="bool">true</disable-send-button>
 			<systemtray comment="Options related to the system tray">
 				<use-old comment="Use the old system tray code (deprecated)" type="bool">false</use-old>
 				<enable type="bool">true</enable>
