diff -urN src/mac/Makefile work/mac/Makefile
--- src/mac/Makefile	2010-03-30 17:15:28.000000000 +0400
+++ work/mac/Makefile	2011-01-20 07:17:50.000000000 +0300
@@ -23,25 +23,25 @@
 SPARKLEDIR=/Library/Frameworks
 PSI_DIR=..
 
-EXECUTABLE_NAME=psi
-NAME=Psi
+EXECUTABLE_NAME=psi-plus
+NAME=Psi+
 VOLUME_NAME=$(NAME)
-VERSION=0.15-devel_$(shell date +"%Y%m%d")
-#VERSION=0.15
+VERSION=0.15.xxx-macosx
 
-APPFILE_NAME=Psi.app
+APPFILE_NAME=Psi+.app
 FILES=$(APPFILE_NAME)
 DISK_DIR=disk
 ifneq ($(BUILD),no)
 DISK_FILES=$(foreach f,$(FILES), $(addprefix $(DISK_DIR)/,$(f)))
 endif
-MASTER_DMG=$(NAME)-$(VERSION).dmg
+MASTER_DMG=$(EXECUTABLE_NAME)-$(VERSION).dmg
 
 
 # Other variables
 QT_FRAMEWORK_VERSION=4
-QT_FRAMEWORKS=QtCore QtXml QtNetwork QtGui QtSql QtSvg
+QT_FRAMEWORKS=QtCore QtXml QtXmlPatterns QtNetwork QtGui QtSql QtSvg QtDBus phonon
 QT_PLUGINS=imageformats/libqjpeg.dylib imageformats/libqgif.dylib
+PSIPLUS_PLUGINS=`ls ../src/$(EXECUTABLE_NAME).app/Contents/Resources/plugins`
 
 # The rules
 .PHONY: buildall
@@ -68,7 +68,7 @@
 build-psi:
 	@echo
 	@echo --------------------- Building Source Tree --------------------
-	$(MAKE) -C $(PSI_DIR)
+#	$(MAKE) -C $(PSI_DIR)
 	@echo
 else
 build-psi:
@@ -88,7 +88,7 @@
 	mkdir -p "$@/Contents/Frameworks"
 	-ditto -rsrc "$(GROWLDIR)/Growl.framework" "$@/Contents/Frameworks/Growl.framework"
 	-ditto -rsrc "$(SPARKLEDIR)/Sparkle.framework" "$@/Contents/Frameworks/Sparkle.framework"
-	#strip "$@/Contents/MacOS/$(EXECUTABLE_NAME)"
+	strip "$@/Contents/MacOS/$(EXECUTABLE_NAME)"
 	#
 	#@echo
 	#@echo Installing QCA ...
@@ -112,7 +112,7 @@
 	@echo
 	@echo Installing Qt Frameworks ...
 	for f in $(QT_FRAMEWORKS); do \
-		ditto -rsrc "$(QTDIR)/lib/$$f.framework" "$@/Contents/Frameworks/$$f.framework"; \
+		ditto -rsrc "/Library/Frameworks/$$f.framework" "$@/Contents/Frameworks/$$f.framework"; \
 		rm -rf "$@/Contents/Frameworks/$$f.framework/Headers"; \
 		rm -rf "$@/Contents/Frameworks/$$f.framework/$$f""_debug"; \
 		rm -rf "$@/Contents/Frameworks/$$f.framework/$$f""_debug.prl"; \
@@ -121,15 +121,16 @@
 		rm -rf "$@/Contents/Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f""_debug.prl"; \
 		install_name_tool -id "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/Frameworks/$$f.framework/$$f"; \
 		for g in $(QT_FRAMEWORKS); do \
-			install_name_tool -change "$(QTDIR)/lib/$$g.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$g" "@executable_path/../Frameworks/$$g.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$g" "$@/Contents/Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f"; \
+			install_name_tool -change "$$g.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$g" "@executable_path/../Frameworks/$$g.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$g" "$@/Contents/Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f"; \
 		done; \
 		for g in $(QT_PLUGINS); do \
-			install_name_tool -change "$(QTDIR)/lib/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/Plugins/$$g"; \
+			install_name_tool -change "$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/Plugins/$$g"; \
 		done; \
-		\# install_name_tool -change "$(QTDIR)/lib/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/Frameworks/libqca.2.dylib"; \
-		\# install_name_tool -change "$(QTDIR)/lib/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/Resources/crypto/libqca-openssl.dylib"; \
-		install_name_tool -change "$(QTDIR)/lib/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/MacOS/$(EXECUTABLE_NAME)"; \
-	done
+		for p in $(PSIPLUS_PLUGINS); do \
+			install_name_tool -change "$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/Resources/plugins/$$p"; \
+		done; \
+		install_name_tool -change "$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "@executable_path/../Frameworks/$$f.framework/Versions/$(QT_FRAMEWORK_VERSION)/$$f" "$@/Contents/MacOS/$(EXECUTABLE_NAME)"; \
+	done; \
 
 # The rest
 $(DISK_DIR)/%: $(PSI_DIR)/%
@@ -150,9 +151,9 @@
 $(TEMPLATE_DMG).bz2: 
 	@echo
 	@echo --------------------- Generating empty template --------------------
-	mkdir template
-	hdiutil create -size 80m "$(TEMPLATE_DMG)" -srcfolder template -format UDRW -volname "$(VOLUME_NAME)" -quiet
-	rmdir template
+#	mkdir template
+#	hdiutil create -size 200m "$(TEMPLATE_DMG)" -srcfolder template -format UDRW -volname "$(VOLUME_NAME)" -quiet
+#	rmdir template
 	bzip2 "$(TEMPLATE_DMG)"
 	@echo
 
